{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Getting Started\n",
    "\n",
    "## Data Harmonization with `bdikit`\n",
    "\n",
    "Data harmonization is the process of integrating and aligning data from different sources into a consistent format to ensure compatibility and interoperability across data analyses and systems. `bdikit` is a library the helps with key data harmonization steps:\n",
    "- *Schema Mapping*: In this step, data from various sources are mapped to a unified schema or model. This involves identifying equivalent table columns and establishing relationships between disparate datasets.\n",
    "- *Value Mapping (Data Standardization)*: This step involves converting data into a common format or structure, using consistent naming conventions, units, and coding systems to ensure uniformity.\n",
    "\n",
    "In this example, we describe how `bdikit` can be used to harmonize datasets from two papers:\n",
    "- Dou et al. (https://pubmed.ncbi.nlm.nih.gov/37567170/)\n",
    "- Cao et al (https://www.cell.com/cell/fulltext/S0092-8674(21)00997-1)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Loading the data\n",
    "\n",
    "First, import the `bdikit` library and other libraries."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import bdikit as bdi\n",
    "import pandas as pd\n",
    "from IPython.display import display, Markdown"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we load our source data using Pandas and select some columns we are interested in."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>idx</th>\n",
       "      <th>Proteomics_Participant_ID</th>\n",
       "      <th>Case_excluded</th>\n",
       "      <th>Proteomics_TMT_batch</th>\n",
       "      <th>Proteomics_TMT_plex</th>\n",
       "      <th>Proteomics_TMT_channel</th>\n",
       "      <th>Proteomics_Parent_Sample_IDs</th>\n",
       "      <th>Proteomics_Aliquot_ID</th>\n",
       "      <th>Proteomics_Tumor_Normal</th>\n",
       "      <th>Proteomics_OCT</th>\n",
       "      <th>...</th>\n",
       "      <th>RNAseq_R1_sample_type</th>\n",
       "      <th>RNAseq_R1_filename</th>\n",
       "      <th>RNAseq_R1_UUID</th>\n",
       "      <th>RNAseq_R2_sample_type</th>\n",
       "      <th>RNAseq_R2_filename</th>\n",
       "      <th>RNAseq_R2_UUID</th>\n",
       "      <th>miRNAseq_sample_type</th>\n",
       "      <th>miRNAseq_UUID</th>\n",
       "      <th>Methylation_available</th>\n",
       "      <th>Methylation_quality</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>S001</td>\n",
       "      <td>C3L-00006</td>\n",
       "      <td>No</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>128N</td>\n",
       "      <td>C3L-00006-01</td>\n",
       "      <td>CPT0001460012</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>170802_UNC31-K00269_0072_AHK3GVBBXX_TAGCTT_S17...</td>\n",
       "      <td>8a1efc47-1c29-417f-a425-cdbd09565dcb</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>170802_UNC31-K00269_0072_AHK3GVBBXX_TAGCTT_S17...</td>\n",
       "      <td>8c3fe9b7-7acd-4867-8d9c-a8e5d1516eda</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>37bcba98-1094-459e-83ae-c23a602416fb</td>\n",
       "      <td>YES</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>S002</td>\n",
       "      <td>C3L-00008</td>\n",
       "      <td>No</td>\n",
       "      <td>4</td>\n",
       "      <td>16</td>\n",
       "      <td>130N</td>\n",
       "      <td>C3L-00008-01</td>\n",
       "      <td>CPT0001300009</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>170802_UNC31-K00269_0072_AHK3GVBBXX_GGCTAC_S22...</td>\n",
       "      <td>555725e8-cba5-4676-9b0a-80100cbf9f47</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>170802_UNC31-K00269_0072_AHK3GVBBXX_GGCTAC_S22...</td>\n",
       "      <td>15235b12-b67a-4678-acc4-ed03d642bd5e</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>492b50d8-ec35-46e7-a65d-06512aaee394</td>\n",
       "      <td>YES</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>S003</td>\n",
       "      <td>C3L-00032</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>131</td>\n",
       "      <td>C3L-00032-01</td>\n",
       "      <td>CPT0001420009</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>170802_UNC31-K00269_0072_AHK3GVBBXX_GTCCGC_S18...</td>\n",
       "      <td>9ae968f3-691d-4db3-9977-1ab3e5af9085</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>170802_UNC31-K00269_0072_AHK3GVBBXX_GTCCGC_S18...</td>\n",
       "      <td>423b6b09-02aa-4f47-9241-f75c1dad1161</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>1794ff56-db2d-4d1a-8758-cab7fe3d98c1</td>\n",
       "      <td>YES</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>S004</td>\n",
       "      <td>C3L-00084</td>\n",
       "      <td>Yes</td>\n",
       "      <td>3</td>\n",
       "      <td>11</td>\n",
       "      <td>129N</td>\n",
       "      <td>C3L-00084-01</td>\n",
       "      <td>CPT0000820012</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>170818_UNC32-K00270_0050_AHL2FHBBXX_ATCACG_S5_...</td>\n",
       "      <td>b0a7cdf2-2ad8-4442-91b0-548ea4975554</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>170818_UNC32-K00270_0050_AHL2FHBBXX_ATCACG_S5_...</td>\n",
       "      <td>c83987a5-1c13-4af4-b46c-218fe5f60c34</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YES</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>S005</td>\n",
       "      <td>C3L-00090</td>\n",
       "      <td>No</td>\n",
       "      <td>3</td>\n",
       "      <td>12</td>\n",
       "      <td>129C</td>\n",
       "      <td>C3L-00090-01</td>\n",
       "      <td>CPT0001140003</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>170802_UNC31-K00269_0072_AHK3GVBBXX_GAGTGG_S10...</td>\n",
       "      <td>8ce5618d-9ff6-40f9-aeea-8d8e1633ae38</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>170802_UNC31-K00269_0072_AHK3GVBBXX_GAGTGG_S10...</td>\n",
       "      <td>06d3fd4a-a623-4146-8500-4f1f17235253</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>a6524c2d-d7dd-4629-980e-b45dbdc92c49</td>\n",
       "      <td>YES</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>S006</td>\n",
       "      <td>C3L-00098</td>\n",
       "      <td>No</td>\n",
       "      <td>4</td>\n",
       "      <td>14</td>\n",
       "      <td>129N</td>\n",
       "      <td>C3L-00098-02</td>\n",
       "      <td>CPT0000980012</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>170802_UNC31-K00269_0072_AHK3GVBBXX_TTAGGC_S8_...</td>\n",
       "      <td>31252ba9-e052-4b77-809a-f936379ae00c</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>170802_UNC31-K00269_0072_AHK3GVBBXX_TTAGGC_S8_...</td>\n",
       "      <td>23be22ae-de50-4d74-a7c0-c890adbc662a</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YES</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>S007</td>\n",
       "      <td>C3L-00136</td>\n",
       "      <td>No</td>\n",
       "      <td>4</td>\n",
       "      <td>16</td>\n",
       "      <td>129C</td>\n",
       "      <td>C3L-00136-03</td>\n",
       "      <td>CPT0000730011</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>170818_UNC32-K00270_0050_AHL2FHBBXX_GTCCGC_S10...</td>\n",
       "      <td>df0e2942-c702-4135-81a0-fbec4439d753</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>170818_UNC32-K00270_0050_AHL2FHBBXX_GTCCGC_S10...</td>\n",
       "      <td>4e1ad404-4646-4828-91b9-e3c35a4ce505</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YES</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>S008</td>\n",
       "      <td>C3L-00137</td>\n",
       "      <td>No</td>\n",
       "      <td>4</td>\n",
       "      <td>15</td>\n",
       "      <td>130N</td>\n",
       "      <td>C3L-00137-02</td>\n",
       "      <td>CPT0002010011</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>170818_UNC32-K00270_0050_AHL2FHBBXX_GTGAAA_S12...</td>\n",
       "      <td>8fcdd6a1-a7c7-41b5-8b44-e41f2237b236</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>170818_UNC32-K00270_0050_AHL2FHBBXX_GTGAAA_S12...</td>\n",
       "      <td>2bea607d-6eb2-4583-90d7-7823a3d8a572</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YES</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>S009</td>\n",
       "      <td>C3L-00139</td>\n",
       "      <td>No</td>\n",
       "      <td>3</td>\n",
       "      <td>11</td>\n",
       "      <td>130N</td>\n",
       "      <td>C3L-00139-01</td>\n",
       "      <td>CPT0001850012</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>170802_UNC31-K00269_0072_AHK3GVBBXX_CAGATC_S1_...</td>\n",
       "      <td>7785d5a1-a60d-41f9-86f3-e4ebc100704c</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>170802_UNC31-K00269_0072_AHK3GVBBXX_CAGATC_S1_...</td>\n",
       "      <td>90ced367-0342-4739-93b2-4b1a4af800c4</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>a02b2784-9e7f-41b1-8e53-707ae4371c45</td>\n",
       "      <td>YES</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>S010</td>\n",
       "      <td>C3L-00143</td>\n",
       "      <td>No</td>\n",
       "      <td>4</td>\n",
       "      <td>14</td>\n",
       "      <td>130C</td>\n",
       "      <td>C3L-00143-01</td>\n",
       "      <td>CPT0001910016</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>170802_UNC31-K00269_0072_AHK3GVBBXX_ACTTGA_S4_...</td>\n",
       "      <td>6412838b-2f70-4b14-a6ee-3c7baca09fb0</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>170802_UNC31-K00269_0072_AHK3GVBBXX_ACTTGA_S4_...</td>\n",
       "      <td>5d0a26e0-2739-4f38-9350-c685b44911d3</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>872be4b7-1735-48a6-a3a2-7541ec65ea87</td>\n",
       "      <td>YES</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 179 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    idx Proteomics_Participant_ID Case_excluded  Proteomics_TMT_batch  \\\n",
       "0  S001                 C3L-00006            No                     2   \n",
       "1  S002                 C3L-00008            No                     4   \n",
       "2  S003                 C3L-00032            No                     1   \n",
       "3  S004                 C3L-00084           Yes                     3   \n",
       "4  S005                 C3L-00090            No                     3   \n",
       "5  S006                 C3L-00098            No                     4   \n",
       "6  S007                 C3L-00136            No                     4   \n",
       "7  S008                 C3L-00137            No                     4   \n",
       "8  S009                 C3L-00139            No                     3   \n",
       "9  S010                 C3L-00143            No                     4   \n",
       "\n",
       "   Proteomics_TMT_plex Proteomics_TMT_channel Proteomics_Parent_Sample_IDs  \\\n",
       "0                    5                   128N                 C3L-00006-01   \n",
       "1                   16                   130N                 C3L-00008-01   \n",
       "2                    2                    131                 C3L-00032-01   \n",
       "3                   11                   129N                 C3L-00084-01   \n",
       "4                   12                   129C                 C3L-00090-01   \n",
       "5                   14                   129N                 C3L-00098-02   \n",
       "6                   16                   129C                 C3L-00136-03   \n",
       "7                   15                   130N                 C3L-00137-02   \n",
       "8                   11                   130N                 C3L-00139-01   \n",
       "9                   14                   130C                 C3L-00143-01   \n",
       "\n",
       "  Proteomics_Aliquot_ID Proteomics_Tumor_Normal Proteomics_OCT  ...  \\\n",
       "0         CPT0001460012                   Tumor             No  ...   \n",
       "1         CPT0001300009                   Tumor             No  ...   \n",
       "2         CPT0001420009                   Tumor             No  ...   \n",
       "3         CPT0000820012                   Tumor             No  ...   \n",
       "4         CPT0001140003                   Tumor             No  ...   \n",
       "5         CPT0000980012                   Tumor             No  ...   \n",
       "6         CPT0000730011                   Tumor             No  ...   \n",
       "7         CPT0002010011                   Tumor             No  ...   \n",
       "8         CPT0001850012                   Tumor             No  ...   \n",
       "9         CPT0001910016                   Tumor             No  ...   \n",
       "\n",
       "  RNAseq_R1_sample_type                                 RNAseq_R1_filename  \\\n",
       "0                 Tumor  170802_UNC31-K00269_0072_AHK3GVBBXX_TAGCTT_S17...   \n",
       "1                 Tumor  170802_UNC31-K00269_0072_AHK3GVBBXX_GGCTAC_S22...   \n",
       "2                 Tumor  170802_UNC31-K00269_0072_AHK3GVBBXX_GTCCGC_S18...   \n",
       "3                 Tumor  170818_UNC32-K00270_0050_AHL2FHBBXX_ATCACG_S5_...   \n",
       "4                 Tumor  170802_UNC31-K00269_0072_AHK3GVBBXX_GAGTGG_S10...   \n",
       "5                 Tumor  170802_UNC31-K00269_0072_AHK3GVBBXX_TTAGGC_S8_...   \n",
       "6                 Tumor  170818_UNC32-K00270_0050_AHL2FHBBXX_GTCCGC_S10...   \n",
       "7                 Tumor  170818_UNC32-K00270_0050_AHL2FHBBXX_GTGAAA_S12...   \n",
       "8                 Tumor  170802_UNC31-K00269_0072_AHK3GVBBXX_CAGATC_S1_...   \n",
       "9                 Tumor  170802_UNC31-K00269_0072_AHK3GVBBXX_ACTTGA_S4_...   \n",
       "\n",
       "                         RNAseq_R1_UUID RNAseq_R2_sample_type  \\\n",
       "0  8a1efc47-1c29-417f-a425-cdbd09565dcb                 Tumor   \n",
       "1  555725e8-cba5-4676-9b0a-80100cbf9f47                 Tumor   \n",
       "2  9ae968f3-691d-4db3-9977-1ab3e5af9085                 Tumor   \n",
       "3  b0a7cdf2-2ad8-4442-91b0-548ea4975554                 Tumor   \n",
       "4  8ce5618d-9ff6-40f9-aeea-8d8e1633ae38                 Tumor   \n",
       "5  31252ba9-e052-4b77-809a-f936379ae00c                 Tumor   \n",
       "6  df0e2942-c702-4135-81a0-fbec4439d753                 Tumor   \n",
       "7  8fcdd6a1-a7c7-41b5-8b44-e41f2237b236                 Tumor   \n",
       "8  7785d5a1-a60d-41f9-86f3-e4ebc100704c                 Tumor   \n",
       "9  6412838b-2f70-4b14-a6ee-3c7baca09fb0                 Tumor   \n",
       "\n",
       "                                  RNAseq_R2_filename  \\\n",
       "0  170802_UNC31-K00269_0072_AHK3GVBBXX_TAGCTT_S17...   \n",
       "1  170802_UNC31-K00269_0072_AHK3GVBBXX_GGCTAC_S22...   \n",
       "2  170802_UNC31-K00269_0072_AHK3GVBBXX_GTCCGC_S18...   \n",
       "3  170818_UNC32-K00270_0050_AHL2FHBBXX_ATCACG_S5_...   \n",
       "4  170802_UNC31-K00269_0072_AHK3GVBBXX_GAGTGG_S10...   \n",
       "5  170802_UNC31-K00269_0072_AHK3GVBBXX_TTAGGC_S8_...   \n",
       "6  170818_UNC32-K00270_0050_AHL2FHBBXX_GTCCGC_S10...   \n",
       "7  170818_UNC32-K00270_0050_AHL2FHBBXX_GTGAAA_S12...   \n",
       "8  170802_UNC31-K00269_0072_AHK3GVBBXX_CAGATC_S1_...   \n",
       "9  170802_UNC31-K00269_0072_AHK3GVBBXX_ACTTGA_S4_...   \n",
       "\n",
       "                         RNAseq_R2_UUID miRNAseq_sample_type  \\\n",
       "0  8c3fe9b7-7acd-4867-8d9c-a8e5d1516eda                Tumor   \n",
       "1  15235b12-b67a-4678-acc4-ed03d642bd5e                Tumor   \n",
       "2  423b6b09-02aa-4f47-9241-f75c1dad1161                Tumor   \n",
       "3  c83987a5-1c13-4af4-b46c-218fe5f60c34                  NaN   \n",
       "4  06d3fd4a-a623-4146-8500-4f1f17235253                Tumor   \n",
       "5  23be22ae-de50-4d74-a7c0-c890adbc662a                  NaN   \n",
       "6  4e1ad404-4646-4828-91b9-e3c35a4ce505                  NaN   \n",
       "7  2bea607d-6eb2-4583-90d7-7823a3d8a572                  NaN   \n",
       "8  90ced367-0342-4739-93b2-4b1a4af800c4                Tumor   \n",
       "9  5d0a26e0-2739-4f38-9350-c685b44911d3                Tumor   \n",
       "\n",
       "                          miRNAseq_UUID Methylation_available  \\\n",
       "0  37bcba98-1094-459e-83ae-c23a602416fb                   YES   \n",
       "1  492b50d8-ec35-46e7-a65d-06512aaee394                   YES   \n",
       "2  1794ff56-db2d-4d1a-8758-cab7fe3d98c1                   YES   \n",
       "3                                   NaN                   YES   \n",
       "4  a6524c2d-d7dd-4629-980e-b45dbdc92c49                   YES   \n",
       "5                                   NaN                   YES   \n",
       "6                                   NaN                   YES   \n",
       "7                                   NaN                   YES   \n",
       "8  a02b2784-9e7f-41b1-8e53-707ae4371c45                   YES   \n",
       "9  872be4b7-1735-48a6-a3a2-7541ec65ea87                   YES   \n",
       "\n",
       "  Methylation_quality  \n",
       "0                PASS  \n",
       "1                PASS  \n",
       "2                PASS  \n",
       "3                PASS  \n",
       "4                PASS  \n",
       "5                PASS  \n",
       "6                PASS  \n",
       "7                PASS  \n",
       "8                PASS  \n",
       "9                PASS  \n",
       "\n",
       "[10 rows x 179 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_source = pd.read_csv(\"./datasets/Dou-ucec-discovery.csv\")\n",
    "# column_names = [\n",
    "#     \"Country\",\n",
    "#     \"Gender\",\n",
    "#     \"FIGO_stage\",\n",
    "#     \"Path_Stage_Reg_Lymph_Nodes-pN\",\n",
    "#     \"tumor_Stage-Pathological\",\n",
    "#     \"Tumor_Focality\",\n",
    "# ]\n",
    "# df_source = df_source[column_names]\n",
    "df_source.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Our goal is to harmonize the data from our source table (`dou.csv`) with the data from our target table `cao.csv`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Idx</th>\n",
       "      <th>Case_id</th>\n",
       "      <th>Case_excluded</th>\n",
       "      <th>Batch</th>\n",
       "      <th>Plex</th>\n",
       "      <th>ReporterName</th>\n",
       "      <th>Aliquot_ID</th>\n",
       "      <th>Group</th>\n",
       "      <th>Discovery_study</th>\n",
       "      <th>Age</th>\n",
       "      <th>...</th>\n",
       "      <th>Follow-up_additional_surgery_for_new_tumor</th>\n",
       "      <th>Follow-up_additional_treatment_radiation_therapy_for_new_tumor</th>\n",
       "      <th>Follow-up_additional_treatment_pharmaceutical_therapy_for_new_tumor</th>\n",
       "      <th>Follow-up_additional_treatment_immuno_for_new_tumor</th>\n",
       "      <th>Follow-up_days_from_date_of_collection_to_date_of_last_contact</th>\n",
       "      <th>Follow-up_cause_of_death</th>\n",
       "      <th>Follow-up_days_from_date_of_initial_pathologic_diagnosis_to_date_of_death</th>\n",
       "      <th>Follow-up_days_from_date_of_initial_pathologic_diagnosis_to_date_of_additional_surgery_for_new_tumor</th>\n",
       "      <th>Follow-up_procedure_type_of_new_tumor</th>\n",
       "      <th>Follow-up_residual_tumor_after_surgery_for_new_tumor</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C3L-00086</td>\n",
       "      <td>C3L-00086</td>\n",
       "      <td>No</td>\n",
       "      <td>b4</td>\n",
       "      <td>16.0</td>\n",
       "      <td>128N</td>\n",
       "      <td>CPT0092460003</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>No</td>\n",
       "      <td>56</td>\n",
       "      <td>...</td>\n",
       "      <td>n/a|No|No|No|No</td>\n",
       "      <td>n/a|Yes|Yes|Yes|Yes</td>\n",
       "      <td>n/a|Yes|Yes|Yes|Yes</td>\n",
       "      <td>n/a|No|No|No|No</td>\n",
       "      <td>330.0|701.0|1046.0|1436.0|n/a</td>\n",
       "      <td>n/a|n/a|n/a|n/a|Breast Carcinoma</td>\n",
       "      <td>n/a|n/a|n/a|n/a|1578.0</td>\n",
       "      <td>n/a|n/a|n/a|n/a|n/a</td>\n",
       "      <td>n/a|n/a|n/a|n/a|n/a</td>\n",
       "      <td>n/a|n/a|n/a|n/a|n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C3L-00898</td>\n",
       "      <td>C3L-00898</td>\n",
       "      <td>No</td>\n",
       "      <td>b4</td>\n",
       "      <td>14.0</td>\n",
       "      <td>128C</td>\n",
       "      <td>CPT0172200008</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>No</td>\n",
       "      <td>54</td>\n",
       "      <td>...</td>\n",
       "      <td>n/a|n/a|n/a|n/a</td>\n",
       "      <td>n/a|n/a|n/a|n/a</td>\n",
       "      <td>n/a|n/a|n/a|n/a</td>\n",
       "      <td>n/a|n/a|n/a|n/a</td>\n",
       "      <td>396.0|746.0|982.0|1600.0</td>\n",
       "      <td>n/a|n/a|n/a|n/a</td>\n",
       "      <td>n/a|n/a|n/a|n/a</td>\n",
       "      <td>n/a|n/a|n/a|n/a</td>\n",
       "      <td>n/a|n/a|n/a|n/a</td>\n",
       "      <td>n/a|n/a|n/a|n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C3L-00943</td>\n",
       "      <td>C3L-00943</td>\n",
       "      <td>No</td>\n",
       "      <td>b4</td>\n",
       "      <td>15.0</td>\n",
       "      <td>130C</td>\n",
       "      <td>CPT0086090003</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>No</td>\n",
       "      <td>63</td>\n",
       "      <td>...</td>\n",
       "      <td>n/a|n/a|n/a</td>\n",
       "      <td>n/a|n/a|n/a</td>\n",
       "      <td>n/a|n/a|n/a</td>\n",
       "      <td>n/a|n/a|n/a</td>\n",
       "      <td>237.0|693.0|1039.0</td>\n",
       "      <td>n/a|n/a|n/a</td>\n",
       "      <td>n/a|n/a|n/a</td>\n",
       "      <td>n/a|n/a|n/a</td>\n",
       "      <td>n/a|n/a|n/a</td>\n",
       "      <td>n/a|n/a|n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C3L-01064</td>\n",
       "      <td>C3L-01064</td>\n",
       "      <td>No</td>\n",
       "      <td>b3</td>\n",
       "      <td>9.0</td>\n",
       "      <td>129N</td>\n",
       "      <td>CPT0113430004</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>No</td>\n",
       "      <td>54</td>\n",
       "      <td>...</td>\n",
       "      <td>No|No|No|No</td>\n",
       "      <td>No|Yes|No|No</td>\n",
       "      <td>Yes|Yes|Yes|Yes</td>\n",
       "      <td>No|No|No|No</td>\n",
       "      <td>453.0|726.0|1062.0|1447.0</td>\n",
       "      <td>n/a|n/a|n/a|n/a</td>\n",
       "      <td>n/a|n/a|n/a|n/a</td>\n",
       "      <td>n/a|n/a|n/a|n/a</td>\n",
       "      <td>n/a|n/a|n/a|n/a</td>\n",
       "      <td>n/a|n/a|n/a|n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C3L-01277</td>\n",
       "      <td>C3L-01277</td>\n",
       "      <td>No</td>\n",
       "      <td>b4</td>\n",
       "      <td>13.0</td>\n",
       "      <td>130N</td>\n",
       "      <td>CPT0093170003</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>No</td>\n",
       "      <td>61</td>\n",
       "      <td>...</td>\n",
       "      <td>n/a|No|No</td>\n",
       "      <td>n/a|No|Yes</td>\n",
       "      <td>n/a|Yes|No</td>\n",
       "      <td>n/a|No|No</td>\n",
       "      <td>351.0|713.0|967.0</td>\n",
       "      <td>n/a|n/a|n/a</td>\n",
       "      <td>n/a|n/a|n/a</td>\n",
       "      <td>n/a|n/a|n/a</td>\n",
       "      <td>n/a|n/a|n/a</td>\n",
       "      <td>n/a|n/a|n/a</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 213 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Idx    Case_id Case_excluded Batch  Plex ReporterName     Aliquot_ID  \\\n",
       "0  C3L-00086  C3L-00086            No    b4  16.0         128N  CPT0092460003   \n",
       "1  C3L-00898  C3L-00898            No    b4  14.0         128C  CPT0172200008   \n",
       "2  C3L-00943  C3L-00943            No    b4  15.0         130C  CPT0086090003   \n",
       "3  C3L-01064  C3L-01064            No    b3   9.0         129N  CPT0113430004   \n",
       "4  C3L-01277  C3L-01277            No    b4  13.0         130N  CPT0093170003   \n",
       "\n",
       "   Group Discovery_study Age  ... Follow-up_additional_surgery_for_new_tumor  \\\n",
       "0  Tumor              No  56  ...                            n/a|No|No|No|No   \n",
       "1  Tumor              No  54  ...                            n/a|n/a|n/a|n/a   \n",
       "2  Tumor              No  63  ...                                n/a|n/a|n/a   \n",
       "3  Tumor              No  54  ...                                No|No|No|No   \n",
       "4  Tumor              No  61  ...                                  n/a|No|No   \n",
       "\n",
       "  Follow-up_additional_treatment_radiation_therapy_for_new_tumor  \\\n",
       "0                                n/a|Yes|Yes|Yes|Yes               \n",
       "1                                    n/a|n/a|n/a|n/a               \n",
       "2                                        n/a|n/a|n/a               \n",
       "3                                       No|Yes|No|No               \n",
       "4                                         n/a|No|Yes               \n",
       "\n",
       "  Follow-up_additional_treatment_pharmaceutical_therapy_for_new_tumor  \\\n",
       "0                                n/a|Yes|Yes|Yes|Yes                    \n",
       "1                                    n/a|n/a|n/a|n/a                    \n",
       "2                                        n/a|n/a|n/a                    \n",
       "3                                    Yes|Yes|Yes|Yes                    \n",
       "4                                         n/a|Yes|No                    \n",
       "\n",
       "   Follow-up_additional_treatment_immuno_for_new_tumor  \\\n",
       "0                                    n/a|No|No|No|No     \n",
       "1                                    n/a|n/a|n/a|n/a     \n",
       "2                                        n/a|n/a|n/a     \n",
       "3                                        No|No|No|No     \n",
       "4                                          n/a|No|No     \n",
       "\n",
       "   Follow-up_days_from_date_of_collection_to_date_of_last_contact  \\\n",
       "0                      330.0|701.0|1046.0|1436.0|n/a                \n",
       "1                           396.0|746.0|982.0|1600.0                \n",
       "2                                 237.0|693.0|1039.0                \n",
       "3                          453.0|726.0|1062.0|1447.0                \n",
       "4                                  351.0|713.0|967.0                \n",
       "\n",
       "           Follow-up_cause_of_death  \\\n",
       "0  n/a|n/a|n/a|n/a|Breast Carcinoma   \n",
       "1                   n/a|n/a|n/a|n/a   \n",
       "2                       n/a|n/a|n/a   \n",
       "3                   n/a|n/a|n/a|n/a   \n",
       "4                       n/a|n/a|n/a   \n",
       "\n",
       "   Follow-up_days_from_date_of_initial_pathologic_diagnosis_to_date_of_death  \\\n",
       "0                             n/a|n/a|n/a|n/a|1578.0                           \n",
       "1                                    n/a|n/a|n/a|n/a                           \n",
       "2                                        n/a|n/a|n/a                           \n",
       "3                                    n/a|n/a|n/a|n/a                           \n",
       "4                                        n/a|n/a|n/a                           \n",
       "\n",
       "  Follow-up_days_from_date_of_initial_pathologic_diagnosis_to_date_of_additional_surgery_for_new_tumor  \\\n",
       "0                                n/a|n/a|n/a|n/a|n/a                                                     \n",
       "1                                    n/a|n/a|n/a|n/a                                                     \n",
       "2                                        n/a|n/a|n/a                                                     \n",
       "3                                    n/a|n/a|n/a|n/a                                                     \n",
       "4                                        n/a|n/a|n/a                                                     \n",
       "\n",
       "  Follow-up_procedure_type_of_new_tumor  \\\n",
       "0                   n/a|n/a|n/a|n/a|n/a   \n",
       "1                       n/a|n/a|n/a|n/a   \n",
       "2                           n/a|n/a|n/a   \n",
       "3                       n/a|n/a|n/a|n/a   \n",
       "4                           n/a|n/a|n/a   \n",
       "\n",
       "  Follow-up_residual_tumor_after_surgery_for_new_tumor  \n",
       "0                                n/a|n/a|n/a|n/a|n/a    \n",
       "1                                    n/a|n/a|n/a|n/a    \n",
       "2                                        n/a|n/a|n/a    \n",
       "3                                    n/a|n/a|n/a|n/a    \n",
       "4                                        n/a|n/a|n/a    \n",
       "\n",
       "[5 rows x 213 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_target = pd.read_csv(\"./datasets/Dou-ucec-confirmatory.csv\")\n",
    "df_target.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Finding column matches between two tables\n",
    "\n",
    "`bdi-kit` offers a suite of functions to help with data harmonization tasks.\n",
    "\n",
    "For instance, it can help automatically discover one-to-one mappings between the source and target dataset columns.\n",
    "\n",
    "To do so using `bdi-kit`, we can use the `match_schema()` function to match columns of the two schemas as follows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracting features from 179 columns...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "fa1a83a2181547f49d962d9750c579ac",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/179 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracting features from 213 columns...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "25820a1b76c64c9db6127f064cd504e8",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/213 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>idx</td>\n",
       "      <td>xCell_T_cell_CD4+_Th1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Proteomics_Participant_ID</td>\n",
       "      <td>Idx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Case_excluded</td>\n",
       "      <td>Case_excluded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>ABSOLUTE_tumor_purity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Proteomics_TMT_plex</td>\n",
       "      <td>Number_of_para-aortic_lymph_nodes_examined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>RNAseq_R2_UUID</td>\n",
       "      <td>Case_id</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>miRNAseq_sample_type</td>\n",
       "      <td>Mutation_signature_SBS7a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>miRNAseq_UUID</td>\n",
       "      <td>Case_id</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177</th>\n",
       "      <td>Methylation_available</td>\n",
       "      <td>Mutation_signature_SBS42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>Methylation_quality</td>\n",
       "      <td>MSI_status</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>179 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        source                                      target\n",
       "0                          idx                       xCell_T_cell_CD4+_Th1\n",
       "1    Proteomics_Participant_ID                                         Idx\n",
       "2                Case_excluded                               Case_excluded\n",
       "3         Proteomics_TMT_batch                       ABSOLUTE_tumor_purity\n",
       "4          Proteomics_TMT_plex  Number_of_para-aortic_lymph_nodes_examined\n",
       "..                         ...                                         ...\n",
       "174             RNAseq_R2_UUID                                     Case_id\n",
       "175       miRNAseq_sample_type                    Mutation_signature_SBS7a\n",
       "176              miRNAseq_UUID                                     Case_id\n",
       "177      Methylation_available                    Mutation_signature_SBS42\n",
       "178        Methylation_quality                                  MSI_status\n",
       "\n",
       "[179 rows x 2 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bdi.match_schema(df_source, df_target, method=\"ct_learning\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracting features from 179 columns...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "0cd8ea67b72649aa9df0b1a3fcbafd8e",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/179 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracting features from 213 columns...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "ab8e9932498b44918c60ef709c151d16",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/213 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>idx</td>\n",
       "      <td>xCell_T_cell_CD4+_Th1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Proteomics_Participant_ID</td>\n",
       "      <td>Idx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Case_excluded</td>\n",
       "      <td>Case_excluded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>ABSOLUTE_tumor_purity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Proteomics_TMT_plex</td>\n",
       "      <td>Number_of_para-aortic_lymph_nodes_examined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Proteomics_TMT_channel</td>\n",
       "      <td>ReporterName</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Proteomics_Parent_Sample_IDs</td>\n",
       "      <td>Idx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Proteomics_Aliquot_ID</td>\n",
       "      <td>Aliquot_ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Proteomics_Tumor_Normal</td>\n",
       "      <td>Group</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Proteomics_OCT</td>\n",
       "      <td>POLE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Country</td>\n",
       "      <td>Participant_country</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Histologic_Grade_FIGO</td>\n",
       "      <td>Histologic_grade</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Myometrial_invasion_Specify</td>\n",
       "      <td>Myometrial_invasion_present_specify</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>Histologic_Type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Treatment_naive</td>\n",
       "      <td>Follow-up_additional_treatment_radiation_thera...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Tumor_purity</td>\n",
       "      <td>Tumor_necrosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>Pathologic_staging_primary_tumor_pt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Path_Stage_Reg_Lymph_Nodes-pN</td>\n",
       "      <td>Pathologic_staging_regional_lymph_nodes_pn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Clin_Stage_Dist_Mets-cM</td>\n",
       "      <td>Clinical_staging_distant_metastasis_cm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Path_Stage_Dist_Mets-pM</td>\n",
       "      <td>Clinical_staging_distant_metastasis_cm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>tumor_Stage-Pathological</td>\n",
       "      <td>Tumor_stage_pathological</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>FIGO_stage</td>\n",
       "      <td>Pathologic_staging_primary_tumor_pt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>LVSI</td>\n",
       "      <td>Number_of_pelvic_lymph_nodes_examined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>BMI</td>\n",
       "      <td>BMI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Age</td>\n",
       "      <td>Age</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Diabetes</td>\n",
       "      <td>Diabetes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Race</td>\n",
       "      <td>Race</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Ethnicity</td>\n",
       "      <td>Ethnicity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Gender</td>\n",
       "      <td>Sex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Tumor_Site</td>\n",
       "      <td>Tumor_site</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Tumor_Site_Other</td>\n",
       "      <td>Tumor_site_other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Tumor_Focality</td>\n",
       "      <td>Tumor_focality</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Tumor_Size_cm</td>\n",
       "      <td>Tumor_size_cm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Estrogen_Receptor</td>\n",
       "      <td>Ancillary_studies_estrogen_receptor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Estrogen_Receptor_%</td>\n",
       "      <td>Ancillary_studies_estrogen_receptor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Progesterone_Receptor</td>\n",
       "      <td>Ancillary_studies_progesterone_receptor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Progesterone_Receptor_%</td>\n",
       "      <td>Ancillary_studies_progesterone_receptor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>MLH1</td>\n",
       "      <td>Ancillary_studies_mlh1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>MLH2</td>\n",
       "      <td>Ancillary_studies_mlh1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>MSH6</td>\n",
       "      <td>Ancillary_studies_msh2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>PMS2</td>\n",
       "      <td>Ancillary_studies_pms2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>p53</td>\n",
       "      <td>Ancillary_studies_p53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>Other_IHC_specify</td>\n",
       "      <td>Ancillary_studies_other_immunohistochemistry_t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>MLH1_Promoter_Hypermethylation</td>\n",
       "      <td>Ancillary_studies_mlh1_promoter_hypermethylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>Num_full_term_pregnancies</td>\n",
       "      <td>Donor_information_number_of_full_term_pregnancies</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>EPIC_Bcells</td>\n",
       "      <td>Progeny_JAK.STAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>EPIC_CAFs</td>\n",
       "      <td>Aliquot_ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>EPIC_CD4_Tcells</td>\n",
       "      <td>Aliquot_ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>EPIC_CD8_Tcells</td>\n",
       "      <td>xCell_T_cell_CD4+_Th2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>EPIC_Endothelial</td>\n",
       "      <td>Progeny_PI3K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>EPIC_Macrophages</td>\n",
       "      <td>Cibersort_Macrophage_M2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>EPIC_NKcells</td>\n",
       "      <td>xCell_NK_cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>EPIC_otherCells</td>\n",
       "      <td>xCell_Cancer_associated_fibroblast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>CIBERSORT_B _cells _naive</td>\n",
       "      <td>xCell_Cancer_associated_fibroblast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>CIBERSORT_B _cells _memory</td>\n",
       "      <td>xCell_Cancer_associated_fibroblast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>CIBERSORT_Plasma _cells</td>\n",
       "      <td>Progeny_PI3K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>CIBERSORT_T _cells _CD8</td>\n",
       "      <td>xCell_T_cell_CD4+_Th2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>CIBERSORT_T _cells _CD4 _naive</td>\n",
       "      <td>Cibersort_T_cell_CD4+_naive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>CIBERSORT_T _cells _CD4 _memory _resting</td>\n",
       "      <td>CNV_ratio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>CIBERSORT_T _cells _CD4 _memory _activated</td>\n",
       "      <td>xCell_T_cell_CD4+_Th2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>CIBERSORT_T _cells _follicular _helper</td>\n",
       "      <td>xCell_Cancer_associated_fibroblast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>CIBERSORT_T _cells _regulatory _(Tregs)</td>\n",
       "      <td>Progeny_TGFb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>CIBERSORT_T _cells _gamma _delta</td>\n",
       "      <td>Progeny_TGFb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>CIBERSORT_NK _cells _resting</td>\n",
       "      <td>Cibersort_NK_cell_activated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>CIBERSORT_NK _cells _activated</td>\n",
       "      <td>Cibersort_NK_cell_activated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>CIBERSORT_Monocytes</td>\n",
       "      <td>Cibersort_Macrophage_M2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>CIBERSORT_Macrophages _M0</td>\n",
       "      <td>Cibersort_Macrophage_M0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>CIBERSORT_Macrophages _M1</td>\n",
       "      <td>Cibersort_Macrophage_M0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>CIBERSORT_Macrophages _M2</td>\n",
       "      <td>Cibersort_Macrophage_M2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>CIBERSORT_Dendritic _cells _resting</td>\n",
       "      <td>xCell_Cancer_associated_fibroblast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>CIBERSORT_Dendritic _cells _activated</td>\n",
       "      <td>xCell_Cancer_associated_fibroblast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>CIBERSORT_Mast _cells _resting</td>\n",
       "      <td>Cibersort_Mast_cell_activated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>CIBERSORT_Mast _cells _activated</td>\n",
       "      <td>Cibersort_Mast_cell_activated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>CIBERSORT_Eosinophils</td>\n",
       "      <td>xCell_Cancer_associated_fibroblast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>CIBERSORT_Neutrophils</td>\n",
       "      <td>CNV_ratio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>CIBERSORT_Absolute _score</td>\n",
       "      <td>Cibersort_T_cell_follicular_helper</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>ESTIMATE_StromalScore</td>\n",
       "      <td>Estimate_StromalScore</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>ESTIMATE_ImmuneScore</td>\n",
       "      <td>Estimate_ESTIMATEScore</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>ESTIMATE_ESTIMATEScore</td>\n",
       "      <td>Estimate_ESTIMATEScore</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>Stemness_score</td>\n",
       "      <td>Progeny_Androgen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>ER_ESR1</td>\n",
       "      <td>xCell_stroma_score</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>PR_PGR</td>\n",
       "      <td>Progeny_Trail</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>Pathway_activity_EGFR</td>\n",
       "      <td>Progeny_Trail</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>Pathway_activity_Hypoxia</td>\n",
       "      <td>xCell_stroma_score</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>Pathway_activity_JAK.STAT</td>\n",
       "      <td>Mutation_signature_SBS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>Pathway_activity_MAPK</td>\n",
       "      <td>BMI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>Pathway_activity_NFkB</td>\n",
       "      <td>Progeny_NFkB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>Pathway_activity_PI3K</td>\n",
       "      <td>Pathologic_staging_regional_lymph_nodes_pn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>Pathway_activity_TGFb</td>\n",
       "      <td>ABSOLUTE_tumor_purity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>Pathway_activity_TNFa</td>\n",
       "      <td>Pathologic_staging_primary_tumor_pt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>Pathway_activity_Trail</td>\n",
       "      <td>Pathologic_staging_primary_tumor_pt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>Pathway_activity_VEGF</td>\n",
       "      <td>Progeny_VEGF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>Pathway_activity_p53</td>\n",
       "      <td>xCell_stroma_score</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>TP53_ATM</td>\n",
       "      <td>TP53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>TP53_CHEK2</td>\n",
       "      <td>Cibersort_T_cell_CD4+_naive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>TP53_MDM4</td>\n",
       "      <td>TP53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>TP53_RPS6KA3</td>\n",
       "      <td>Ancillary_studies_pms2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>TP53_TP53</td>\n",
       "      <td>TP53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>TP53_pathway</td>\n",
       "      <td>TP53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>PI3K_AKT1</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>PI3K_AKT2</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>PI3K_AKT3</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>PI3K_DEPDC5</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>PI3K_DEPTOR</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>PI3K_INPP4B</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>PI3K_MAPKAP1</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>PI3K_MLST8</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>PI3K_MTOR</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>PI3K_NPRL2</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>PI3K_NPRL3</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>PI3K_PDK1</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>PI3K_PIK3CA</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>PI3K_PIK3CB</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>PI3K_PIK3R1</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>PI3K_PIK3R2</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>PI3K_PPP2R1A</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>PI3K_PTEN</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>PI3K_RHEB</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>PI3K_RICTOR</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>PI3K_RPS6</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>PI3K_RPS6KB1</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>PI3K_RPTOR</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>PI3K_STK11</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>PI3K_TSC1</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>PI3K_TSC2</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>PI3K_pathway</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>HRD_BRCA1</td>\n",
       "      <td>Batch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>HRD_BRCA2</td>\n",
       "      <td>Batch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>HRD_BRCA1_or_BRCA2</td>\n",
       "      <td>Batch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>CNV_DEL</td>\n",
       "      <td>CNV_ratio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>CNV_AMP</td>\n",
       "      <td>CNV_ratio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>CNV_class</td>\n",
       "      <td>CNV_status</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>CNV_idx</td>\n",
       "      <td>CNV_ratio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>CNV_1q_DEL</td>\n",
       "      <td>CNV_ratio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>CNV_3q_DEL</td>\n",
       "      <td>Idx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>CNV_4q_DEL</td>\n",
       "      <td>CNV_ratio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>CNV_1q_AMP</td>\n",
       "      <td>CNV_ratio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>CNV_3q_AMP</td>\n",
       "      <td>CNV_ratio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>CNV_4q_AMP</td>\n",
       "      <td>CNV_ratio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>Purity_Immune</td>\n",
       "      <td>Progeny_Androgen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>Purity_Cancer</td>\n",
       "      <td>Progeny_Androgen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>Purity_Stroma</td>\n",
       "      <td>Progeny_Androgen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>MSI_status</td>\n",
       "      <td>MSI_status</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>POLE_subtype</td>\n",
       "      <td>POLE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>JAK1_MS_INDEL</td>\n",
       "      <td>MSI_status</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>JAK1_Mutation</td>\n",
       "      <td>BMI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>Log2_variant_per_Mbp</td>\n",
       "      <td>Progeny_MAPK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>Log2_SNP_per_Mbp</td>\n",
       "      <td>Progeny_JAK.STAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>Log2_INDEL_per_Mbp</td>\n",
       "      <td>Cibersort_Macrophage_M0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>Log2_variant_total</td>\n",
       "      <td>Estimate_ESTIMATEScore</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>Log2_SNP_total</td>\n",
       "      <td>CNV_ratio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>Log2_INDEL_total</td>\n",
       "      <td>Estimate_ESTIMATEScore</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>Genomics_subtype</td>\n",
       "      <td>Genomic_subtype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>Mutation_signature_C&gt;A</td>\n",
       "      <td>CNV_ratio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>Mutation_signature_C&gt;G</td>\n",
       "      <td>Cibersort_Macrophage_M2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>Mutation_signature_C&gt;T</td>\n",
       "      <td>xCell_Cancer_associated_fibroblast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>Mutation_signature_T&gt;C</td>\n",
       "      <td>Cibersort_Macrophage_M2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>Mutation_signature_T&gt;A</td>\n",
       "      <td>Progeny_Hypoxia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>Mutation_signature_T&gt;G</td>\n",
       "      <td>Progeny_TGFb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>WXS_normal_sample_type</td>\n",
       "      <td>Batch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>WXS_normal_filename</td>\n",
       "      <td>Progeny_WNT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>WXS_normal_UUID</td>\n",
       "      <td>Case_id</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>WXS_tumor_sample_type</td>\n",
       "      <td>Tumor_site</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163</th>\n",
       "      <td>WXS_tumor_filename</td>\n",
       "      <td>Progeny_TGFb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>WXS_tumor_UUID</td>\n",
       "      <td>Progeny_TGFb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>WGS_normal_sample_type</td>\n",
       "      <td>Mutation_signature_SBS7a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>WGS_normal_UUID</td>\n",
       "      <td>Case_id</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>WGS_tumor_sample_type</td>\n",
       "      <td>Tumor_site</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>WGS_tumor_UUID</td>\n",
       "      <td>Progeny_TGFb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>RNAseq_R1_sample_type</td>\n",
       "      <td>ARID1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>RNAseq_R1_filename</td>\n",
       "      <td>Aliquot_ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>RNAseq_R1_UUID</td>\n",
       "      <td>Case_id</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>RNAseq_R2_sample_type</td>\n",
       "      <td>ARID1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>RNAseq_R2_filename</td>\n",
       "      <td>Aliquot_ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>RNAseq_R2_UUID</td>\n",
       "      <td>Case_id</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>miRNAseq_sample_type</td>\n",
       "      <td>Mutation_signature_SBS7a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>miRNAseq_UUID</td>\n",
       "      <td>Case_id</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177</th>\n",
       "      <td>Methylation_available</td>\n",
       "      <td>Mutation_signature_SBS42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>Methylation_quality</td>\n",
       "      <td>MSI_status</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         source  \\\n",
       "0                                           idx   \n",
       "1                     Proteomics_Participant_ID   \n",
       "2                                 Case_excluded   \n",
       "3                          Proteomics_TMT_batch   \n",
       "4                           Proteomics_TMT_plex   \n",
       "5                        Proteomics_TMT_channel   \n",
       "6                  Proteomics_Parent_Sample_IDs   \n",
       "7                         Proteomics_Aliquot_ID   \n",
       "8                       Proteomics_Tumor_Normal   \n",
       "9                                Proteomics_OCT   \n",
       "10                                      Country   \n",
       "11                        Histologic_Grade_FIGO   \n",
       "12                  Myometrial_invasion_Specify   \n",
       "13                              Histologic_type   \n",
       "14                              Treatment_naive   \n",
       "15                                 Tumor_purity   \n",
       "16                  Path_Stage_Primary_Tumor-pT   \n",
       "17                Path_Stage_Reg_Lymph_Nodes-pN   \n",
       "18                      Clin_Stage_Dist_Mets-cM   \n",
       "19                      Path_Stage_Dist_Mets-pM   \n",
       "20                     tumor_Stage-Pathological   \n",
       "21                                   FIGO_stage   \n",
       "22                                         LVSI   \n",
       "23                                          BMI   \n",
       "24                                          Age   \n",
       "25                                     Diabetes   \n",
       "26                                         Race   \n",
       "27                                    Ethnicity   \n",
       "28                                       Gender   \n",
       "29                                   Tumor_Site   \n",
       "30                             Tumor_Site_Other   \n",
       "31                               Tumor_Focality   \n",
       "32                                Tumor_Size_cm   \n",
       "33                            Estrogen_Receptor   \n",
       "34                          Estrogen_Receptor_%   \n",
       "35                        Progesterone_Receptor   \n",
       "36                      Progesterone_Receptor_%   \n",
       "37                                         MLH1   \n",
       "38                                         MLH2   \n",
       "39                                         MSH6   \n",
       "40                                         PMS2   \n",
       "41                                          p53   \n",
       "42                            Other_IHC_specify   \n",
       "43               MLH1_Promoter_Hypermethylation   \n",
       "44                    Num_full_term_pregnancies   \n",
       "45                                  EPIC_Bcells   \n",
       "46                                    EPIC_CAFs   \n",
       "47                              EPIC_CD4_Tcells   \n",
       "48                              EPIC_CD8_Tcells   \n",
       "49                             EPIC_Endothelial   \n",
       "50                             EPIC_Macrophages   \n",
       "51                                 EPIC_NKcells   \n",
       "52                              EPIC_otherCells   \n",
       "53                    CIBERSORT_B _cells _naive   \n",
       "54                   CIBERSORT_B _cells _memory   \n",
       "55                      CIBERSORT_Plasma _cells   \n",
       "56                      CIBERSORT_T _cells _CD8   \n",
       "57               CIBERSORT_T _cells _CD4 _naive   \n",
       "58     CIBERSORT_T _cells _CD4 _memory _resting   \n",
       "59   CIBERSORT_T _cells _CD4 _memory _activated   \n",
       "60       CIBERSORT_T _cells _follicular _helper   \n",
       "61      CIBERSORT_T _cells _regulatory _(Tregs)   \n",
       "62             CIBERSORT_T _cells _gamma _delta   \n",
       "63                 CIBERSORT_NK _cells _resting   \n",
       "64               CIBERSORT_NK _cells _activated   \n",
       "65                          CIBERSORT_Monocytes   \n",
       "66                    CIBERSORT_Macrophages _M0   \n",
       "67                    CIBERSORT_Macrophages _M1   \n",
       "68                    CIBERSORT_Macrophages _M2   \n",
       "69          CIBERSORT_Dendritic _cells _resting   \n",
       "70        CIBERSORT_Dendritic _cells _activated   \n",
       "71               CIBERSORT_Mast _cells _resting   \n",
       "72             CIBERSORT_Mast _cells _activated   \n",
       "73                        CIBERSORT_Eosinophils   \n",
       "74                        CIBERSORT_Neutrophils   \n",
       "75                    CIBERSORT_Absolute _score   \n",
       "76                        ESTIMATE_StromalScore   \n",
       "77                         ESTIMATE_ImmuneScore   \n",
       "78                       ESTIMATE_ESTIMATEScore   \n",
       "79                               Stemness_score   \n",
       "80                                      ER_ESR1   \n",
       "81                                       PR_PGR   \n",
       "82                        Pathway_activity_EGFR   \n",
       "83                     Pathway_activity_Hypoxia   \n",
       "84                    Pathway_activity_JAK.STAT   \n",
       "85                        Pathway_activity_MAPK   \n",
       "86                        Pathway_activity_NFkB   \n",
       "87                        Pathway_activity_PI3K   \n",
       "88                        Pathway_activity_TGFb   \n",
       "89                        Pathway_activity_TNFa   \n",
       "90                       Pathway_activity_Trail   \n",
       "91                        Pathway_activity_VEGF   \n",
       "92                         Pathway_activity_p53   \n",
       "93                                     TP53_ATM   \n",
       "94                                   TP53_CHEK2   \n",
       "95                                    TP53_MDM4   \n",
       "96                                 TP53_RPS6KA3   \n",
       "97                                    TP53_TP53   \n",
       "98                                 TP53_pathway   \n",
       "99                                    PI3K_AKT1   \n",
       "100                                   PI3K_AKT2   \n",
       "101                                   PI3K_AKT3   \n",
       "102                                 PI3K_DEPDC5   \n",
       "103                                 PI3K_DEPTOR   \n",
       "104                                 PI3K_INPP4B   \n",
       "105                                PI3K_MAPKAP1   \n",
       "106                                  PI3K_MLST8   \n",
       "107                                   PI3K_MTOR   \n",
       "108                                  PI3K_NPRL2   \n",
       "109                                  PI3K_NPRL3   \n",
       "110                                   PI3K_PDK1   \n",
       "111                                 PI3K_PIK3CA   \n",
       "112                                 PI3K_PIK3CB   \n",
       "113                                 PI3K_PIK3R1   \n",
       "114                                 PI3K_PIK3R2   \n",
       "115                                PI3K_PPP2R1A   \n",
       "116                                   PI3K_PTEN   \n",
       "117                                   PI3K_RHEB   \n",
       "118                                 PI3K_RICTOR   \n",
       "119                                   PI3K_RPS6   \n",
       "120                                PI3K_RPS6KB1   \n",
       "121                                  PI3K_RPTOR   \n",
       "122                                  PI3K_STK11   \n",
       "123                                   PI3K_TSC1   \n",
       "124                                   PI3K_TSC2   \n",
       "125                                PI3K_pathway   \n",
       "126                                   HRD_BRCA1   \n",
       "127                                   HRD_BRCA2   \n",
       "128                          HRD_BRCA1_or_BRCA2   \n",
       "129                                     CNV_DEL   \n",
       "130                                     CNV_AMP   \n",
       "131                                   CNV_class   \n",
       "132                                     CNV_idx   \n",
       "133                                  CNV_1q_DEL   \n",
       "134                                  CNV_3q_DEL   \n",
       "135                                  CNV_4q_DEL   \n",
       "136                                  CNV_1q_AMP   \n",
       "137                                  CNV_3q_AMP   \n",
       "138                                  CNV_4q_AMP   \n",
       "139                               Purity_Immune   \n",
       "140                               Purity_Cancer   \n",
       "141                               Purity_Stroma   \n",
       "142                                  MSI_status   \n",
       "143                                POLE_subtype   \n",
       "144                               JAK1_MS_INDEL   \n",
       "145                               JAK1_Mutation   \n",
       "146                        Log2_variant_per_Mbp   \n",
       "147                            Log2_SNP_per_Mbp   \n",
       "148                          Log2_INDEL_per_Mbp   \n",
       "149                          Log2_variant_total   \n",
       "150                              Log2_SNP_total   \n",
       "151                            Log2_INDEL_total   \n",
       "152                            Genomics_subtype   \n",
       "153                      Mutation_signature_C>A   \n",
       "154                      Mutation_signature_C>G   \n",
       "155                      Mutation_signature_C>T   \n",
       "156                      Mutation_signature_T>C   \n",
       "157                      Mutation_signature_T>A   \n",
       "158                      Mutation_signature_T>G   \n",
       "159                      WXS_normal_sample_type   \n",
       "160                         WXS_normal_filename   \n",
       "161                             WXS_normal_UUID   \n",
       "162                       WXS_tumor_sample_type   \n",
       "163                          WXS_tumor_filename   \n",
       "164                              WXS_tumor_UUID   \n",
       "165                      WGS_normal_sample_type   \n",
       "166                             WGS_normal_UUID   \n",
       "167                       WGS_tumor_sample_type   \n",
       "168                              WGS_tumor_UUID   \n",
       "169                       RNAseq_R1_sample_type   \n",
       "170                          RNAseq_R1_filename   \n",
       "171                              RNAseq_R1_UUID   \n",
       "172                       RNAseq_R2_sample_type   \n",
       "173                          RNAseq_R2_filename   \n",
       "174                              RNAseq_R2_UUID   \n",
       "175                        miRNAseq_sample_type   \n",
       "176                               miRNAseq_UUID   \n",
       "177                       Methylation_available   \n",
       "178                         Methylation_quality   \n",
       "\n",
       "                                                target  \n",
       "0                                xCell_T_cell_CD4+_Th1  \n",
       "1                                                  Idx  \n",
       "2                                        Case_excluded  \n",
       "3                                ABSOLUTE_tumor_purity  \n",
       "4           Number_of_para-aortic_lymph_nodes_examined  \n",
       "5                                         ReporterName  \n",
       "6                                                  Idx  \n",
       "7                                           Aliquot_ID  \n",
       "8                                                Group  \n",
       "9                                                 POLE  \n",
       "10                                 Participant_country  \n",
       "11                                    Histologic_grade  \n",
       "12                 Myometrial_invasion_present_specify  \n",
       "13                                     Histologic_Type  \n",
       "14   Follow-up_additional_treatment_radiation_thera...  \n",
       "15                                      Tumor_necrosis  \n",
       "16                 Pathologic_staging_primary_tumor_pt  \n",
       "17          Pathologic_staging_regional_lymph_nodes_pn  \n",
       "18              Clinical_staging_distant_metastasis_cm  \n",
       "19              Clinical_staging_distant_metastasis_cm  \n",
       "20                            Tumor_stage_pathological  \n",
       "21                 Pathologic_staging_primary_tumor_pt  \n",
       "22               Number_of_pelvic_lymph_nodes_examined  \n",
       "23                                                 BMI  \n",
       "24                                                 Age  \n",
       "25                                            Diabetes  \n",
       "26                                                Race  \n",
       "27                                           Ethnicity  \n",
       "28                                                 Sex  \n",
       "29                                          Tumor_site  \n",
       "30                                    Tumor_site_other  \n",
       "31                                      Tumor_focality  \n",
       "32                                       Tumor_size_cm  \n",
       "33                 Ancillary_studies_estrogen_receptor  \n",
       "34                 Ancillary_studies_estrogen_receptor  \n",
       "35             Ancillary_studies_progesterone_receptor  \n",
       "36             Ancillary_studies_progesterone_receptor  \n",
       "37                              Ancillary_studies_mlh1  \n",
       "38                              Ancillary_studies_mlh1  \n",
       "39                              Ancillary_studies_msh2  \n",
       "40                              Ancillary_studies_pms2  \n",
       "41                               Ancillary_studies_p53  \n",
       "42   Ancillary_studies_other_immunohistochemistry_t...  \n",
       "43    Ancillary_studies_mlh1_promoter_hypermethylation  \n",
       "44   Donor_information_number_of_full_term_pregnancies  \n",
       "45                                    Progeny_JAK.STAT  \n",
       "46                                          Aliquot_ID  \n",
       "47                                          Aliquot_ID  \n",
       "48                               xCell_T_cell_CD4+_Th2  \n",
       "49                                        Progeny_PI3K  \n",
       "50                             Cibersort_Macrophage_M2  \n",
       "51                                       xCell_NK_cell  \n",
       "52                  xCell_Cancer_associated_fibroblast  \n",
       "53                  xCell_Cancer_associated_fibroblast  \n",
       "54                  xCell_Cancer_associated_fibroblast  \n",
       "55                                        Progeny_PI3K  \n",
       "56                               xCell_T_cell_CD4+_Th2  \n",
       "57                         Cibersort_T_cell_CD4+_naive  \n",
       "58                                           CNV_ratio  \n",
       "59                               xCell_T_cell_CD4+_Th2  \n",
       "60                  xCell_Cancer_associated_fibroblast  \n",
       "61                                        Progeny_TGFb  \n",
       "62                                        Progeny_TGFb  \n",
       "63                         Cibersort_NK_cell_activated  \n",
       "64                         Cibersort_NK_cell_activated  \n",
       "65                             Cibersort_Macrophage_M2  \n",
       "66                             Cibersort_Macrophage_M0  \n",
       "67                             Cibersort_Macrophage_M0  \n",
       "68                             Cibersort_Macrophage_M2  \n",
       "69                  xCell_Cancer_associated_fibroblast  \n",
       "70                  xCell_Cancer_associated_fibroblast  \n",
       "71                       Cibersort_Mast_cell_activated  \n",
       "72                       Cibersort_Mast_cell_activated  \n",
       "73                  xCell_Cancer_associated_fibroblast  \n",
       "74                                           CNV_ratio  \n",
       "75                  Cibersort_T_cell_follicular_helper  \n",
       "76                               Estimate_StromalScore  \n",
       "77                              Estimate_ESTIMATEScore  \n",
       "78                              Estimate_ESTIMATEScore  \n",
       "79                                    Progeny_Androgen  \n",
       "80                                  xCell_stroma_score  \n",
       "81                                       Progeny_Trail  \n",
       "82                                       Progeny_Trail  \n",
       "83                                  xCell_stroma_score  \n",
       "84                             Mutation_signature_SBS1  \n",
       "85                                                 BMI  \n",
       "86                                        Progeny_NFkB  \n",
       "87          Pathologic_staging_regional_lymph_nodes_pn  \n",
       "88                               ABSOLUTE_tumor_purity  \n",
       "89                 Pathologic_staging_primary_tumor_pt  \n",
       "90                 Pathologic_staging_primary_tumor_pt  \n",
       "91                                        Progeny_VEGF  \n",
       "92                                  xCell_stroma_score  \n",
       "93                                                TP53  \n",
       "94                         Cibersort_T_cell_CD4+_naive  \n",
       "95                                                TP53  \n",
       "96                              Ancillary_studies_pms2  \n",
       "97                                                TP53  \n",
       "98                                                TP53  \n",
       "99                                              PIK3CA  \n",
       "100                                             PIK3CA  \n",
       "101                                             PIK3CA  \n",
       "102                                             PIK3CA  \n",
       "103                                             PIK3CA  \n",
       "104                                             PIK3CA  \n",
       "105                                             PIK3CA  \n",
       "106                                             PIK3CA  \n",
       "107                                             PIK3CA  \n",
       "108                                             PIK3CA  \n",
       "109                                             PIK3CA  \n",
       "110                                             PIK3CA  \n",
       "111                                             PIK3CA  \n",
       "112                                             PIK3CA  \n",
       "113                                             PIK3CA  \n",
       "114                                             PIK3CA  \n",
       "115                                             PIK3CA  \n",
       "116                                             PIK3CA  \n",
       "117                                             PIK3CA  \n",
       "118                                             PIK3CA  \n",
       "119                                             PIK3CA  \n",
       "120                                             PIK3CA  \n",
       "121                                             PIK3CA  \n",
       "122                                             PIK3CA  \n",
       "123                                             PIK3CA  \n",
       "124                                             PIK3CA  \n",
       "125                                             PIK3CA  \n",
       "126                                              Batch  \n",
       "127                                              Batch  \n",
       "128                                              Batch  \n",
       "129                                          CNV_ratio  \n",
       "130                                          CNV_ratio  \n",
       "131                                         CNV_status  \n",
       "132                                          CNV_ratio  \n",
       "133                                          CNV_ratio  \n",
       "134                                                Idx  \n",
       "135                                          CNV_ratio  \n",
       "136                                          CNV_ratio  \n",
       "137                                          CNV_ratio  \n",
       "138                                          CNV_ratio  \n",
       "139                                   Progeny_Androgen  \n",
       "140                                   Progeny_Androgen  \n",
       "141                                   Progeny_Androgen  \n",
       "142                                         MSI_status  \n",
       "143                                               POLE  \n",
       "144                                         MSI_status  \n",
       "145                                                BMI  \n",
       "146                                       Progeny_MAPK  \n",
       "147                                   Progeny_JAK.STAT  \n",
       "148                            Cibersort_Macrophage_M0  \n",
       "149                             Estimate_ESTIMATEScore  \n",
       "150                                          CNV_ratio  \n",
       "151                             Estimate_ESTIMATEScore  \n",
       "152                                    Genomic_subtype  \n",
       "153                                          CNV_ratio  \n",
       "154                            Cibersort_Macrophage_M2  \n",
       "155                 xCell_Cancer_associated_fibroblast  \n",
       "156                            Cibersort_Macrophage_M2  \n",
       "157                                    Progeny_Hypoxia  \n",
       "158                                       Progeny_TGFb  \n",
       "159                                              Batch  \n",
       "160                                        Progeny_WNT  \n",
       "161                                            Case_id  \n",
       "162                                         Tumor_site  \n",
       "163                                       Progeny_TGFb  \n",
       "164                                       Progeny_TGFb  \n",
       "165                           Mutation_signature_SBS7a  \n",
       "166                                            Case_id  \n",
       "167                                         Tumor_site  \n",
       "168                                       Progeny_TGFb  \n",
       "169                                             ARID1A  \n",
       "170                                         Aliquot_ID  \n",
       "171                                            Case_id  \n",
       "172                                             ARID1A  \n",
       "173                                         Aliquot_ID  \n",
       "174                                            Case_id  \n",
       "175                           Mutation_signature_SBS7a  \n",
       "176                                            Case_id  \n",
       "177                           Mutation_signature_SBS42  \n",
       "178                                         MSI_status  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.set_option('display.max_rows', None)\n",
    "schema_mapping = bdi.match_schema(df_source, df_target, method=\"ct_learning\")\n",
    "schema_mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracting features from 1 columns...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6ba498c573914394bc4c41cabef033d6",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/1 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracting features from 213 columns...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6b6337e22fd94ddc8fc2c24a775531c1",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/213 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Tumor_purity</td>\n",
       "      <td>Tumor_size_cm</td>\n",
       "      <td>0.108482</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Tumor_purity</td>\n",
       "      <td>ABSOLUTE_tumor_purity</td>\n",
       "      <td>0.099089</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Tumor_purity</td>\n",
       "      <td>Mutation_signature_SBS10b</td>\n",
       "      <td>0.094293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Tumor_purity</td>\n",
       "      <td>Cibersort_T_cell_regulatory_(Tregs)</td>\n",
       "      <td>0.092900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Tumor_purity</td>\n",
       "      <td>Cibersort_Monocyte</td>\n",
       "      <td>0.092766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Tumor_purity</td>\n",
       "      <td>Cibersort_B_cell_naive</td>\n",
       "      <td>0.092648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Tumor_purity</td>\n",
       "      <td>Mutation_signature_SBS1</td>\n",
       "      <td>0.092625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Tumor_purity</td>\n",
       "      <td>Cibersort_T_cell_gamma_delta</td>\n",
       "      <td>0.091828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Tumor_purity</td>\n",
       "      <td>Progeny_TGFb</td>\n",
       "      <td>0.090625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Tumor_purity</td>\n",
       "      <td>Mutation_signature_SBS10a</td>\n",
       "      <td>0.090585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Tumor_purity</td>\n",
       "      <td>Pathologic_staging_primary_tumor_pt</td>\n",
       "      <td>0.089843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Tumor_purity</td>\n",
       "      <td>Mutation_signature_SBS7a</td>\n",
       "      <td>0.089825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Tumor_purity</td>\n",
       "      <td>Mutation_signature_SBS6</td>\n",
       "      <td>0.089275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Tumor_purity</td>\n",
       "      <td>Mutation_signature_SBS42</td>\n",
       "      <td>0.089130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Tumor_purity</td>\n",
       "      <td>Mutation_signature_SBS5</td>\n",
       "      <td>0.088617</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Tumor_purity</td>\n",
       "      <td>Progeny_TNFa</td>\n",
       "      <td>0.088416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Tumor_purity</td>\n",
       "      <td>Mutation_signature_SBS15</td>\n",
       "      <td>0.088232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Tumor_purity</td>\n",
       "      <td>Plex</td>\n",
       "      <td>0.087975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Tumor_purity</td>\n",
       "      <td>BMI</td>\n",
       "      <td>0.087312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Tumor_purity</td>\n",
       "      <td>Follow-up_tumor_status_at_date_of_last_contact...</td>\n",
       "      <td>0.087292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Tumor_purity</td>\n",
       "      <td>Cibersort_T_cell_CD8+</td>\n",
       "      <td>0.087164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Tumor_purity</td>\n",
       "      <td>xCell_Monocyte</td>\n",
       "      <td>0.087080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Tumor_purity</td>\n",
       "      <td>xCell_T_cell_gamma_delta</td>\n",
       "      <td>0.086970</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Tumor_purity</td>\n",
       "      <td>xCell_T_cell_CD8+_naive</td>\n",
       "      <td>0.086835</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Tumor_purity</td>\n",
       "      <td>Tumor_site</td>\n",
       "      <td>0.086777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Tumor_purity</td>\n",
       "      <td>Mutation_signature_SBS20</td>\n",
       "      <td>0.086773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Tumor_purity</td>\n",
       "      <td>Cibersort_Mast_cell_resting</td>\n",
       "      <td>0.086565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Tumor_purity</td>\n",
       "      <td>Cibersort_B_cell_memory</td>\n",
       "      <td>0.086439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Tumor_purity</td>\n",
       "      <td>Tumor_necrosis</td>\n",
       "      <td>0.086347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Tumor_purity</td>\n",
       "      <td>Was_the_participant_exposed_to_secondhand_smoke</td>\n",
       "      <td>0.085964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Tumor_purity</td>\n",
       "      <td>xCell_T_cell_regulatory_(Tregs)</td>\n",
       "      <td>0.085852</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Tumor_purity</td>\n",
       "      <td>Mutation_signature_SBS21</td>\n",
       "      <td>0.085632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Tumor_purity</td>\n",
       "      <td>Cibersort_T_cell_follicular_helper</td>\n",
       "      <td>0.085575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Tumor_purity</td>\n",
       "      <td>Current_medications_medication_name_vitamins_s...</td>\n",
       "      <td>0.085556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Tumor_purity</td>\n",
       "      <td>xCell_T_cell_CD4+_naive</td>\n",
       "      <td>0.085543</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          source                                             target  \\\n",
       "0   Tumor_purity                                      Tumor_size_cm   \n",
       "1   Tumor_purity                              ABSOLUTE_tumor_purity   \n",
       "2   Tumor_purity                          Mutation_signature_SBS10b   \n",
       "3   Tumor_purity                Cibersort_T_cell_regulatory_(Tregs)   \n",
       "4   Tumor_purity                                 Cibersort_Monocyte   \n",
       "5   Tumor_purity                             Cibersort_B_cell_naive   \n",
       "6   Tumor_purity                            Mutation_signature_SBS1   \n",
       "7   Tumor_purity                       Cibersort_T_cell_gamma_delta   \n",
       "8   Tumor_purity                                       Progeny_TGFb   \n",
       "9   Tumor_purity                          Mutation_signature_SBS10a   \n",
       "10  Tumor_purity                Pathologic_staging_primary_tumor_pt   \n",
       "11  Tumor_purity                           Mutation_signature_SBS7a   \n",
       "12  Tumor_purity                            Mutation_signature_SBS6   \n",
       "13  Tumor_purity                           Mutation_signature_SBS42   \n",
       "14  Tumor_purity                            Mutation_signature_SBS5   \n",
       "15  Tumor_purity                                       Progeny_TNFa   \n",
       "16  Tumor_purity                           Mutation_signature_SBS15   \n",
       "17  Tumor_purity                                               Plex   \n",
       "18  Tumor_purity                                                BMI   \n",
       "19  Tumor_purity  Follow-up_tumor_status_at_date_of_last_contact...   \n",
       "20  Tumor_purity                              Cibersort_T_cell_CD8+   \n",
       "21  Tumor_purity                                     xCell_Monocyte   \n",
       "22  Tumor_purity                           xCell_T_cell_gamma_delta   \n",
       "23  Tumor_purity                            xCell_T_cell_CD8+_naive   \n",
       "24  Tumor_purity                                         Tumor_site   \n",
       "25  Tumor_purity                           Mutation_signature_SBS20   \n",
       "26  Tumor_purity                        Cibersort_Mast_cell_resting   \n",
       "27  Tumor_purity                            Cibersort_B_cell_memory   \n",
       "28  Tumor_purity                                     Tumor_necrosis   \n",
       "29  Tumor_purity    Was_the_participant_exposed_to_secondhand_smoke   \n",
       "30  Tumor_purity                    xCell_T_cell_regulatory_(Tregs)   \n",
       "31  Tumor_purity                           Mutation_signature_SBS21   \n",
       "32  Tumor_purity                 Cibersort_T_cell_follicular_helper   \n",
       "33  Tumor_purity  Current_medications_medication_name_vitamins_s...   \n",
       "34  Tumor_purity                            xCell_T_cell_CD4+_naive   \n",
       "\n",
       "    similarity  \n",
       "0     0.108482  \n",
       "1     0.099089  \n",
       "2     0.094293  \n",
       "3     0.092900  \n",
       "4     0.092766  \n",
       "5     0.092648  \n",
       "6     0.092625  \n",
       "7     0.091828  \n",
       "8     0.090625  \n",
       "9     0.090585  \n",
       "10    0.089843  \n",
       "11    0.089825  \n",
       "12    0.089275  \n",
       "13    0.089130  \n",
       "14    0.088617  \n",
       "15    0.088416  \n",
       "16    0.088232  \n",
       "17    0.087975  \n",
       "18    0.087312  \n",
       "19    0.087292  \n",
       "20    0.087164  \n",
       "21    0.087080  \n",
       "22    0.086970  \n",
       "23    0.086835  \n",
       "24    0.086777  \n",
       "25    0.086773  \n",
       "26    0.086565  \n",
       "27    0.086439  \n",
       "28    0.086347  \n",
       "29    0.085964  \n",
       "30    0.085852  \n",
       "31    0.085632  \n",
       "32    0.085575  \n",
       "33    0.085556  \n",
       "34    0.085543  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bdi.top_matches(df_source, columns=['Tumor_purity'], target=df_target, top_k=35)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>value_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Surgery|Surgery</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Surgery</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Other(Mohs treatment)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Radiation,Surgery</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Unknown</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              value_name\n",
       "0                    NaN\n",
       "1        Surgery|Surgery\n",
       "2                Surgery\n",
       "3  Other(Mohs treatment)\n",
       "4      Radiation,Surgery\n",
       "5                Unknown"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bdi.preview_domain(df_target, column='Treatment_naive')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'Ancillary_studies_mlh2'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[26], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# Stemness_score\tProgeny_Androgen\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m \u001b[43mbdi\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmatch_values\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdf_source\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdf_target\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mMLH2\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mAncillary_studies_mlh2\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mtfidf\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m  \n\u001b[1;32m      3\u001b[0m \u001b[38;5;66;03m# bdi.match_values(df_source, df_target, ('Tumor_purity', 'ABSOLUTE_tumor_purity'), method='tfidf')\u001b[39;00m\n",
      "File \u001b[0;32m~/workspace/askem-arpa-h/bdi-kit/bdikit/api.py:317\u001b[0m, in \u001b[0;36mmatch_values\u001b[0;34m(source, target, column_mapping, method, method_args)\u001b[0m\n\u001b[1;32m    315\u001b[0m     method_args \u001b[38;5;241m=\u001b[39m {}\n\u001b[1;32m    316\u001b[0m value_matcher \u001b[38;5;241m=\u001b[39m ValueMatchers\u001b[38;5;241m.\u001b[39mget_instance(method, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mmethod_args)\n\u001b[0;32m--> 317\u001b[0m matches \u001b[38;5;241m=\u001b[39m \u001b[43m_match_values\u001b[49m\u001b[43m(\u001b[49m\u001b[43msource\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtarget_domain\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcolumn_mapping_list\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mvalue_matcher\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    319\u001b[0m result \u001b[38;5;241m=\u001b[39m [\n\u001b[1;32m    320\u001b[0m     _value_matching_result_to_df(matching_result) \u001b[38;5;28;01mfor\u001b[39;00m matching_result \u001b[38;5;129;01min\u001b[39;00m matches\n\u001b[1;32m    321\u001b[0m ]\n\u001b[1;32m    323\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(column_mapping, \u001b[38;5;28mtuple\u001b[39m):\n\u001b[1;32m    324\u001b[0m     \u001b[38;5;66;03m# If only a single mapping is provided (as a tuple), we return the result\u001b[39;00m\n\u001b[1;32m    325\u001b[0m     \u001b[38;5;66;03m# directly as a DataFrame to make it easier to display it in notebooks.\u001b[39;00m\n",
      "File \u001b[0;32m~/workspace/askem-arpa-h/bdi-kit/bdikit/api.py:373\u001b[0m, in \u001b[0;36m_match_values\u001b[0;34m(dataset, target_domain, column_mapping, value_matcher)\u001b[0m\n\u001b[1;32m    370\u001b[0m source_column, target_column \u001b[38;5;241m=\u001b[39m mapping[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msource\u001b[39m\u001b[38;5;124m\"\u001b[39m], mapping[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtarget\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m    372\u001b[0m \u001b[38;5;66;03m# 1. Select candidate columns for value mapping\u001b[39;00m\n\u001b[0;32m--> 373\u001b[0m target_domain_list \u001b[38;5;241m=\u001b[39m \u001b[43mtarget_domain\u001b[49m\u001b[43m[\u001b[49m\u001b[43mtarget_column\u001b[49m\u001b[43m]\u001b[49m\n\u001b[1;32m    374\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m target_domain_list \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(target_domain_list) \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m    375\u001b[0m     \u001b[38;5;28;01mcontinue\u001b[39;00m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'Ancillary_studies_mlh2'"
     ]
    }
   ],
   "source": [
    "# Stemness_score\tProgeny_Androgen\n",
    "bdi.match_values(df_source, df_target, ('MLH2', 'Ancillary_studies_mlh2'), method='tfidf')  \n",
    "# bdi.match_values(df_source, df_target, ('Tumor_purity', 'ABSOLUTE_tumor_purity'), method='tfidf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>b2</td>\n",
       "      <td>0.578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4</td>\n",
       "      <td>b4</td>\n",
       "      <td>0.578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>b1</td>\n",
       "      <td>0.578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>b3</td>\n",
       "      <td>0.578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  source target  similarity\n",
       "0      2     b2       0.578\n",
       "1      4     b4       0.578\n",
       "2      1     b1       0.578\n",
       "3      3     b3       0.578\n",
       "4      5   None         NaN"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bdi.match_values(df_source, df_target, ('Proteomics_TMT_batch', 'Batch'), method='tfidf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracting features from 1 columns...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "8f48aa90b74c4f73856daa6d02be8f16",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/1 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracting features from 213 columns...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "f8c28777ab8e4e48a184e458811702eb",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/213 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MLH2</td>\n",
       "      <td>Ancillary_studies_mlh1</td>\n",
       "      <td>0.103021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MLH2</td>\n",
       "      <td>BMI</td>\n",
       "      <td>0.101541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MLH2</td>\n",
       "      <td>CNV_status</td>\n",
       "      <td>0.100899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MLH2</td>\n",
       "      <td>Mutation_signature_SBS21</td>\n",
       "      <td>0.097867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MLH2</td>\n",
       "      <td>xCell_B_cell</td>\n",
       "      <td>0.097845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MLH2</td>\n",
       "      <td>Mutation_load</td>\n",
       "      <td>0.096779</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MLH2</td>\n",
       "      <td>Mutation_signature_SBS20</td>\n",
       "      <td>0.096626</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>MLH2</td>\n",
       "      <td>xCell_Macrophage</td>\n",
       "      <td>0.096572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MLH2</td>\n",
       "      <td>xCell_Myeloid_dendritic_cell</td>\n",
       "      <td>0.095951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>MLH2</td>\n",
       "      <td>Clinical_staging_distant_metastasis_cm</td>\n",
       "      <td>0.095322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>MLH2</td>\n",
       "      <td>Mutation_signature_SBS54</td>\n",
       "      <td>0.094960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>MLH2</td>\n",
       "      <td>xCell_Monocyte</td>\n",
       "      <td>0.094659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>MLH2</td>\n",
       "      <td>Mutation_signature_SBS42</td>\n",
       "      <td>0.094600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>MLH2</td>\n",
       "      <td>xCell_Endothelial_cell</td>\n",
       "      <td>0.094528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>MLH2</td>\n",
       "      <td>xCell_Class-switched_memory_B_cell</td>\n",
       "      <td>0.094443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>MLH2</td>\n",
       "      <td>Cibersort_Macrophage_M2</td>\n",
       "      <td>0.094272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>MLH2</td>\n",
       "      <td>xCell_Mast_cell</td>\n",
       "      <td>0.094229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>MLH2</td>\n",
       "      <td>xCell_Macrophage_M1</td>\n",
       "      <td>0.093890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>MLH2</td>\n",
       "      <td>Cibersort_Macrophage_M0</td>\n",
       "      <td>0.093889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>MLH2</td>\n",
       "      <td>Tumor_size_cm</td>\n",
       "      <td>0.093696</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>MLH2</td>\n",
       "      <td>xCell_Macrophage_M2</td>\n",
       "      <td>0.093364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>MLH2</td>\n",
       "      <td>Case_id</td>\n",
       "      <td>0.093306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>MLH2</td>\n",
       "      <td>xCell_NK_cell</td>\n",
       "      <td>0.093204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>MLH2</td>\n",
       "      <td>xCell_B_cell_memory</td>\n",
       "      <td>0.092884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>MLH2</td>\n",
       "      <td>Mutation_signature_SBS7a</td>\n",
       "      <td>0.092399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>MLH2</td>\n",
       "      <td>xCell_Plasmacytoid_dendritic_cell</td>\n",
       "      <td>0.091764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>MLH2</td>\n",
       "      <td>xCell_T_cell_CD4+_memory</td>\n",
       "      <td>0.091703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>MLH2</td>\n",
       "      <td>xCell_B_cell_naive</td>\n",
       "      <td>0.091680</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>MLH2</td>\n",
       "      <td>xCell_B_cell_plasma</td>\n",
       "      <td>0.091495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>MLH2</td>\n",
       "      <td>MSI_status</td>\n",
       "      <td>0.091326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>MLH2</td>\n",
       "      <td>Cibersort_Monocyte</td>\n",
       "      <td>0.090943</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>MLH2</td>\n",
       "      <td>Idx</td>\n",
       "      <td>0.090744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>MLH2</td>\n",
       "      <td>xCell_T_cell_CD4+_naive</td>\n",
       "      <td>0.090707</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>MLH2</td>\n",
       "      <td>Mutation_signature_SBS10a</td>\n",
       "      <td>0.090691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>MLH2</td>\n",
       "      <td>Cibersort_Mast_cell_resting</td>\n",
       "      <td>0.090560</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   source                                  target  similarity\n",
       "0    MLH2                  Ancillary_studies_mlh1    0.103021\n",
       "1    MLH2                                     BMI    0.101541\n",
       "2    MLH2                              CNV_status    0.100899\n",
       "3    MLH2                Mutation_signature_SBS21    0.097867\n",
       "4    MLH2                            xCell_B_cell    0.097845\n",
       "5    MLH2                           Mutation_load    0.096779\n",
       "6    MLH2                Mutation_signature_SBS20    0.096626\n",
       "7    MLH2                        xCell_Macrophage    0.096572\n",
       "8    MLH2            xCell_Myeloid_dendritic_cell    0.095951\n",
       "9    MLH2  Clinical_staging_distant_metastasis_cm    0.095322\n",
       "10   MLH2                Mutation_signature_SBS54    0.094960\n",
       "11   MLH2                          xCell_Monocyte    0.094659\n",
       "12   MLH2                Mutation_signature_SBS42    0.094600\n",
       "13   MLH2                  xCell_Endothelial_cell    0.094528\n",
       "14   MLH2      xCell_Class-switched_memory_B_cell    0.094443\n",
       "15   MLH2                 Cibersort_Macrophage_M2    0.094272\n",
       "16   MLH2                         xCell_Mast_cell    0.094229\n",
       "17   MLH2                     xCell_Macrophage_M1    0.093890\n",
       "18   MLH2                 Cibersort_Macrophage_M0    0.093889\n",
       "19   MLH2                           Tumor_size_cm    0.093696\n",
       "20   MLH2                     xCell_Macrophage_M2    0.093364\n",
       "21   MLH2                                 Case_id    0.093306\n",
       "22   MLH2                           xCell_NK_cell    0.093204\n",
       "23   MLH2                     xCell_B_cell_memory    0.092884\n",
       "24   MLH2                Mutation_signature_SBS7a    0.092399\n",
       "25   MLH2       xCell_Plasmacytoid_dendritic_cell    0.091764\n",
       "26   MLH2                xCell_T_cell_CD4+_memory    0.091703\n",
       "27   MLH2                      xCell_B_cell_naive    0.091680\n",
       "28   MLH2                     xCell_B_cell_plasma    0.091495\n",
       "29   MLH2                              MSI_status    0.091326\n",
       "30   MLH2                      Cibersort_Monocyte    0.090943\n",
       "31   MLH2                                     Idx    0.090744\n",
       "32   MLH2                 xCell_T_cell_CD4+_naive    0.090707\n",
       "33   MLH2               Mutation_signature_SBS10a    0.090691\n",
       "34   MLH2             Cibersort_Mast_cell_resting    0.090560"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bdi.top_matches(df_source, columns=['MLH2'], target=df_target, top_k=35)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Finding value matches between two columns\n",
    "\n",
    "Once the matching columns are identified, we can standardize data to ensure that no duplicate values represent the same entity/meaning.\n",
    "\n",
    "To do that, `bdikit` provides the function `match_values()` to find values that should potentially be merged. The library supports multiple methods to perform this task, including syntactic and semantic matching algorithms. In this example, we use the `tfidf` method, which finds values based on the similarity of character n-grams. Please, refer to the [bdikit documentation](https://bdi-kit.readthedocs.io/) to learn more about the methods available."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** Case_excluded<br>**Target column:** Case_excluded<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  source target  similarity\n",
       "0     No     No         1.0\n",
       "1    Yes    Yes         1.0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** Proteomics_TMT_batch<br>**Target column:** ABSOLUTE_tumor_purity<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>0.33</td>\n",
       "      <td>0.622</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4</td>\n",
       "      <td>0.40</td>\n",
       "      <td>0.608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>0.20</td>\n",
       "      <td>0.601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.506</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  source  target  similarity\n",
       "0      3    0.33       0.622\n",
       "1      4    0.40       0.608\n",
       "2      5    0.50       0.606\n",
       "3      2    0.20       0.601\n",
       "4      1    1.00       0.506"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** Proteomics_TMT_plex<br>**Target column:** Number_of_para-aortic_lymph_nodes_examined<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0.626</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>15</td>\n",
       "      <td>5</td>\n",
       "      <td>0.592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>13</td>\n",
       "      <td>3</td>\n",
       "      <td>0.592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>17</td>\n",
       "      <td>7</td>\n",
       "      <td>0.592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>14</td>\n",
       "      <td>4</td>\n",
       "      <td>0.563</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>9</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   source target  similarity\n",
       "0       5      5       1.000\n",
       "1      16     16       1.000\n",
       "2       2      2       1.000\n",
       "3      11     11       1.000\n",
       "4      12     12       1.000\n",
       "5       8      8       1.000\n",
       "6       7      7       1.000\n",
       "7       6      6       1.000\n",
       "8       3      3       1.000\n",
       "9       1      1       1.000\n",
       "10      4      4       1.000\n",
       "11     10      0       0.626\n",
       "12     15      5       0.592\n",
       "13     13      3       0.592\n",
       "14     17      7       0.592\n",
       "15     14      4       0.563\n",
       "16      9   None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** Proteomics_TMT_channel<br>**Target column:** ReporterName<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>128N</td>\n",
       "      <td>128N</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>130N</td>\n",
       "      <td>130N</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>129N</td>\n",
       "      <td>129N</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>129C</td>\n",
       "      <td>129C</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>130C</td>\n",
       "      <td>130C</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>127N</td>\n",
       "      <td>127N</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>127C</td>\n",
       "      <td>127C</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>128C</td>\n",
       "      <td>128C</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>131</td>\n",
       "      <td>131N</td>\n",
       "      <td>0.714</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  source target  similarity\n",
       "0   128N   128N       1.000\n",
       "1   130N   130N       1.000\n",
       "2   129N   129N       1.000\n",
       "3   129C   129C       1.000\n",
       "4   130C   130C       1.000\n",
       "5   127N   127N       1.000\n",
       "6   127C   127C       1.000\n",
       "7   128C   128C       1.000\n",
       "8    131   131N       0.714"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** Proteomics_Tumor_Normal<br>**Target column:** Group<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Adjacent_normal</td>\n",
       "      <td>Adjacent_normal</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Enriched_normal</td>\n",
       "      <td>Enriched_Normal</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Myometrium_normal</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              source           target  similarity\n",
       "0              Tumor            Tumor         1.0\n",
       "1    Adjacent_normal  Adjacent_normal         1.0\n",
       "2    Enriched_normal  Enriched_Normal         1.0\n",
       "3  Myometrium_normal             None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** Proteomics_OCT<br>**Target column:** POLE<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  source target  similarity\n",
       "0     No     No         1.0\n",
       "1    Yes    Yes         1.0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** Country<br>**Target column:** Participant_country<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>United States</td>\n",
       "      <td>United States</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ukraine</td>\n",
       "      <td>Ukraine</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Poland</td>\n",
       "      <td>Poland</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Other_specify</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          source         target  similarity\n",
       "0  United States  United States         1.0\n",
       "1            nan            NaN         1.0\n",
       "2        Ukraine        Ukraine         1.0\n",
       "3         Poland         Poland         1.0\n",
       "4  Other_specify           None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** Histologic_Grade_FIGO<br>**Target column:** Histologic_grade<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FIGO grade 1</td>\n",
       "      <td>Other: High grade</td>\n",
       "      <td>0.426</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FIGO grade 2</td>\n",
       "      <td>Other: High grade</td>\n",
       "      <td>0.426</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FIGO grade 3</td>\n",
       "      <td>Other: High grade</td>\n",
       "      <td>0.426</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         source             target  similarity\n",
       "0           nan                NaN       1.000\n",
       "1  FIGO grade 1  Other: High grade       0.426\n",
       "2  FIGO grade 2  Other: High grade       0.426\n",
       "3  FIGO grade 3  Other: High grade       0.426"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** Myometrial_invasion_Specify<br>**Target column:** Myometrial_invasion_present_specify<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>50 % or more</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Not identified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>under 50 %</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source  target  similarity\n",
       "0             nan     NaN         1.0\n",
       "1    50 % or more     NaN         NaN\n",
       "2  Not identified     NaN         NaN\n",
       "3      under 50 %     NaN         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** Histologic_type<br>**Target column:** Histologic_Type<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>Endometrioid carcinoma</td>\n",
       "      <td>0.855</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Clear cell</td>\n",
       "      <td>Clear cell carcinoma</td>\n",
       "      <td>0.835</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Serous</td>\n",
       "      <td>Serous carcinoma</td>\n",
       "      <td>0.717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>Serous carcinoma</td>\n",
       "      <td>0.618</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source                  target  similarity\n",
       "0             nan                     NaN       1.000\n",
       "1    Endometrioid  Endometrioid carcinoma       0.855\n",
       "2      Clear cell    Clear cell carcinoma       0.835\n",
       "3          Serous        Serous carcinoma       0.717\n",
       "4  Carcinosarcoma        Serous carcinoma       0.618"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** Treatment_naive<br>**Target column:** Follow-up_additional_treatment_radiation_therapy_for_new_tumor<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NO</td>\n",
       "      <td>No|No</td>\n",
       "      <td>0.806</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>YES</td>\n",
       "      <td>Yes|Yes|Yes|Yes</td>\n",
       "      <td>0.787</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  source           target  similarity\n",
       "0    nan              NaN       1.000\n",
       "1     NO            No|No       0.806\n",
       "2    YES  Yes|Yes|Yes|Yes       0.787"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** Tumor_purity<br>**Target column:** Tumor_necrosis<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Low</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   source  target  similarity\n",
       "0     nan     NaN         1.0\n",
       "1  Normal     NaN         NaN\n",
       "2     Low     NaN         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** Path_Stage_Primary_Tumor-pT<br>**Target column:** Pathologic_staging_primary_tumor_pt<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pT3a (FIGO IIIA)</td>\n",
       "      <td>pT3a (FIGO IIIA)</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pT1 (FIGO I)</td>\n",
       "      <td>pT1 (FIGO I)</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pT1b (FIGO IB)</td>\n",
       "      <td>pT1b (FIGO IB)</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>pT2 (FIGO II)</td>\n",
       "      <td>pT2 (FIGO II)</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>pT3b (FIGO IIIB)</td>\n",
       "      <td>pT3b (FIGO IIIB)</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             source            target  similarity\n",
       "0    pT1a (FIGO IA)    pT1a (FIGO IA)         1.0\n",
       "1               nan               NaN         1.0\n",
       "2  pT3a (FIGO IIIA)  pT3a (FIGO IIIA)         1.0\n",
       "3      pT1 (FIGO I)      pT1 (FIGO I)         1.0\n",
       "4    pT1b (FIGO IB)    pT1b (FIGO IB)         1.0\n",
       "5     pT2 (FIGO II)     pT2 (FIGO II)         1.0\n",
       "6  pT3b (FIGO IIIB)  pT3b (FIGO IIIB)         1.0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** Path_Stage_Reg_Lymph_Nodes-pN<br>**Target column:** Pathologic_staging_regional_lymph_nodes_pn<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pN0</td>\n",
       "      <td>pN0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pNX</td>\n",
       "      <td>pNX</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pN2 (FIGO IIIC2)</td>\n",
       "      <td>pN2 (FIGO IIIC2)</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pN1 (FIGO IIIC1)</td>\n",
       "      <td>pN1 (FIGO IIIC1)</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             source            target  similarity\n",
       "0               pN0               pN0         1.0\n",
       "1               pNX               pNX         1.0\n",
       "2               nan               NaN         1.0\n",
       "3  pN2 (FIGO IIIC2)  pN2 (FIGO IIIC2)         1.0\n",
       "4  pN1 (FIGO IIIC1)  pN1 (FIGO IIIC1)         1.0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** Clin_Stage_Dist_Mets-cM<br>**Target column:** Clinical_staging_distant_metastasis_cm<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cM0</td>\n",
       "      <td>cM0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Staging Incomplete</td>\n",
       "      <td>Staging Incomplete</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cM1</td>\n",
       "      <td>cM1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               source              target  similarity\n",
       "0                 cM0                 cM0         1.0\n",
       "1                 nan                 NaN         1.0\n",
       "2  Staging Incomplete  Staging Incomplete         1.0\n",
       "3                 cM1                 cM1         1.0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** Path_Stage_Dist_Mets-pM<br>**Target column:** Clinical_staging_distant_metastasis_cm<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Staging Incomplete</td>\n",
       "      <td>Staging Incomplete</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pM1</td>\n",
       "      <td>cM1</td>\n",
       "      <td>0.445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>No pathologic evidence of distant metastasis</td>\n",
       "      <td>Staging Incomplete</td>\n",
       "      <td>0.423</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         source              target  \\\n",
       "0                            Staging Incomplete  Staging Incomplete   \n",
       "1                                           nan                 NaN   \n",
       "2                                           pM1                 cM1   \n",
       "3  No pathologic evidence of distant metastasis  Staging Incomplete   \n",
       "\n",
       "   similarity  \n",
       "0       1.000  \n",
       "1       1.000  \n",
       "2       0.445  \n",
       "3       0.423  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** tumor_Stage-Pathological<br>**Target column:** Tumor_stage_pathological<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Stage I</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Stage IV</td>\n",
       "      <td>Stage IV</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Stage III</td>\n",
       "      <td>Stage III</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Stage II</td>\n",
       "      <td>Stage II</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      source     target  similarity\n",
       "0    Stage I    Stage I         1.0\n",
       "1   Stage IV   Stage IV         1.0\n",
       "2        nan        NaN         1.0\n",
       "3  Stage III  Stage III         1.0\n",
       "4   Stage II   Stage II         1.0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** FIGO_stage<br>**Target column:** Pathologic_staging_primary_tumor_pt<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>IIIB</td>\n",
       "      <td>pT3b (FIGO IIIB)</td>\n",
       "      <td>0.689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>IIIA</td>\n",
       "      <td>pT3a (FIGO IIIA)</td>\n",
       "      <td>0.682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>IA</td>\n",
       "      <td>pT1a [IA]</td>\n",
       "      <td>0.567</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>II</td>\n",
       "      <td>pT3a (FIGO IIIA)</td>\n",
       "      <td>0.526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>IB</td>\n",
       "      <td>pT1b (FIGO IB)</td>\n",
       "      <td>0.498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>IIIC2</td>\n",
       "      <td>pT2 [II]</td>\n",
       "      <td>0.403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>IIIC1</td>\n",
       "      <td>pT3a (FIGO IIIA)</td>\n",
       "      <td>0.344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>IVB</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  source            target  similarity\n",
       "0    nan               NaN       1.000\n",
       "1   IIIB  pT3b (FIGO IIIB)       0.689\n",
       "2   IIIA  pT3a (FIGO IIIA)       0.682\n",
       "3     IA         pT1a [IA]       0.567\n",
       "4     II  pT3a (FIGO IIIA)       0.526\n",
       "5     IB    pT1b (FIGO IB)       0.498\n",
       "6  IIIC2          pT2 [II]       0.403\n",
       "7  IIIC1  pT3a (FIGO IIIA)       0.344\n",
       "8    IVB              None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** Diabetes<br>**Target column:** Diabetes<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Unknown</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>No</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    source target  similarity\n",
       "0      Yes    Yes         1.0\n",
       "1      nan    NaN         1.0\n",
       "2  Unknown   None         NaN\n",
       "3       No   None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** Race<br>**Target column:** Race<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>White</td>\n",
       "      <td>White</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Asian</td>\n",
       "      <td>Asian</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Not Reported</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Black or African American</td>\n",
       "      <td>Black or African American</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      source                     target  similarity\n",
       "0                      White                      White         1.0\n",
       "1                        nan                        NaN         1.0\n",
       "2                      Asian                      Asian         1.0\n",
       "3               Not Reported               Not Reported         1.0\n",
       "4  Black or African American  Black or African American         1.0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** Ethnicity<br>**Target column:** Ethnicity<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Hispanic or Latino</td>\n",
       "      <td>Hispanic or Latino</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Not reported</td>\n",
       "      <td>Not reported</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   source                  target  similarity\n",
       "0  Not-Hispanic or Latino  Not-Hispanic or Latino         1.0\n",
       "1                     nan                     NaN         1.0\n",
       "2      Hispanic or Latino      Hispanic or Latino         1.0\n",
       "3            Not reported            Not reported         1.0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** Gender<br>**Target column:** Sex<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Female</td>\n",
       "      <td>Female</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   source  target  similarity\n",
       "0  Female  Female         1.0\n",
       "1     nan     NaN         1.0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** Tumor_Site<br>**Target column:** Tumor_site<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Anterior endometrium</td>\n",
       "      <td>Anterior endometrium</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Posterior endometrium</td>\n",
       "      <td>Posterior endometrium</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Other, specify</td>\n",
       "      <td>Other</td>\n",
       "      <td>0.558</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  source                 target  similarity\n",
       "0   Anterior endometrium   Anterior endometrium       1.000\n",
       "1  Posterior endometrium  Posterior endometrium       1.000\n",
       "2                    nan                    NaN       1.000\n",
       "3         Other, specify                  Other       0.558"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** Tumor_Focality<br>**Target column:** Tumor_focality<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Unifocal</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Multifocal</td>\n",
       "      <td>Multifocal</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       source      target  similarity\n",
       "0    Unifocal    Unifocal         1.0\n",
       "1         nan         NaN         1.0\n",
       "2  Multifocal  Multifocal         1.0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** Estrogen_Receptor<br>**Target column:** Ancillary_studies_estrogen_receptor<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Negative</td>\n",
       "      <td>Negative</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Positive</td>\n",
       "      <td>Positive : 5 %</td>\n",
       "      <td>0.941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Unknown</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 source                target  similarity\n",
       "0  Cannot be determined  Cannot be determined       1.000\n",
       "1                   nan                   NaN       1.000\n",
       "2              Negative              Negative       1.000\n",
       "3              Positive        Positive : 5 %       0.941\n",
       "4               Unknown                  None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** Progesterone_Receptor<br>**Target column:** Ancillary_studies_progesterone_receptor<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Negative</td>\n",
       "      <td>Negative</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Positive</td>\n",
       "      <td>Positive : 5 %</td>\n",
       "      <td>0.941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Unknown</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 source                target  similarity\n",
       "0  Cannot be determined  Cannot be determined       1.000\n",
       "1                   nan                   NaN       1.000\n",
       "2              Negative              Negative       1.000\n",
       "3              Positive        Positive : 5 %       0.941\n",
       "4               Unknown                  None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** MLH1<br>**Target column:** Ancillary_studies_mlh1<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Loss of nuclear expression</td>\n",
       "      <td>Loss of nuclear expression</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Unknown</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       source                      target  similarity\n",
       "0   Intact nuclear expression   Intact nuclear expression         1.0\n",
       "1                         nan                         NaN         1.0\n",
       "2  Loss of nuclear expression  Loss of nuclear expression         1.0\n",
       "3        Cannot be determined        Cannot be determined         1.0\n",
       "4                     Unknown                        None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** MLH2<br>**Target column:** Ancillary_studies_mlh1<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Loss of nuclear expression</td>\n",
       "      <td>Loss of nuclear expression</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Unknown</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       source                      target  similarity\n",
       "0   Intact nuclear expression   Intact nuclear expression         1.0\n",
       "1                         nan                         NaN         1.0\n",
       "2        Cannot be determined        Cannot be determined         1.0\n",
       "3  Loss of nuclear expression  Loss of nuclear expression         1.0\n",
       "4                     Unknown                        None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** MSH6<br>**Target column:** Ancillary_studies_msh2<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Loss of nuclear expression</td>\n",
       "      <td>Loss of nuclear expression</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Unknown</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       source                      target  similarity\n",
       "0  Loss of nuclear expression  Loss of nuclear expression         1.0\n",
       "1   Intact nuclear expression   Intact nuclear expression         1.0\n",
       "2                         nan                         NaN         1.0\n",
       "3        Cannot be determined        Cannot be determined         1.0\n",
       "4                     Unknown                        None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** PMS2<br>**Target column:** Ancillary_studies_pms2<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Loss of nuclear expression</td>\n",
       "      <td>Loss of nuclear expression</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Unknown</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       source                      target  similarity\n",
       "0   Intact nuclear expression   Intact nuclear expression         1.0\n",
       "1  Loss of nuclear expression  Loss of nuclear expression         1.0\n",
       "2                         nan                         NaN         1.0\n",
       "3        Cannot be determined        Cannot be determined         1.0\n",
       "4                     Unknown                        None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** p53<br>**Target column:** Ancillary_studies_p53<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Normal</td>\n",
       "      <td>Normal</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Overexpression</td>\n",
       "      <td>Overexpression</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Loss of expression</td>\n",
       "      <td>Loss of expression</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Unknown</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 source                target  similarity\n",
       "0  Cannot be determined  Cannot be determined         1.0\n",
       "1                   nan                   NaN         1.0\n",
       "2                Normal                Normal         1.0\n",
       "3        Overexpression        Overexpression         1.0\n",
       "4    Loss of expression    Loss of expression         1.0\n",
       "5               Unknown                  None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** MLH1_Promoter_Hypermethylation<br>**Target column:** Ancillary_studies_mlh1_promoter_hypermethylation<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Absent</td>\n",
       "      <td>Absent</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Present</td>\n",
       "      <td>Present</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 source                target  similarity\n",
       "0  Cannot be determined  Cannot be determined         1.0\n",
       "1                   nan                   NaN         1.0\n",
       "2                Absent                Absent         1.0\n",
       "3               Present               Present         1.0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** Num_full_term_pregnancies<br>**Target column:** Donor_information_number_of_full_term_pregnancies<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4 or more</td>\n",
       "      <td>4 or more</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      source     target  similarity\n",
       "0          1          1         1.0\n",
       "1  4 or more  4 or more         1.0\n",
       "2        nan        NaN         1.0\n",
       "3          2          2         1.0\n",
       "4          3          3         1.0\n",
       "5    Unknown    Unknown         1.0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** CNV_class<br>**Target column:** CNV_status<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CNV_LOW</td>\n",
       "      <td>CNV_L</td>\n",
       "      <td>0.631</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CNV_HIGH</td>\n",
       "      <td>CNV_H</td>\n",
       "      <td>0.623</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     source target  similarity\n",
       "0       nan    NaN       1.000\n",
       "1   CNV_LOW  CNV_L       0.631\n",
       "2  CNV_HIGH  CNV_H       0.623"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** MSI_status<br>**Target column:** MSI_status<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MSI-H</td>\n",
       "      <td>MSI-H</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MSS</td>\n",
       "      <td>MSS</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  source target  similarity\n",
       "0  MSI-H  MSI-H         1.0\n",
       "1    MSS    MSS         1.0\n",
       "2    nan    NaN         1.0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** POLE_subtype<br>**Target column:** POLE<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  source target  similarity\n",
       "0     No     No         1.0\n",
       "1    Yes    Yes         1.0\n",
       "2    nan    NaN         1.0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** JAK1_MS_INDEL<br>**Target column:** MSI_status<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MS_indel</td>\n",
       "      <td>MSI-H</td>\n",
       "      <td>0.352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WT</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     source target  similarity\n",
       "0       nan    NaN       1.000\n",
       "1  MS_indel  MSI-H       0.352\n",
       "2        WT   None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** JAK1_Mutation<br>**Target column:** BMI<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Frame_Shift_Ins_Nonsense_Mutation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Frame_Shift_Del_Nonsense_Mutation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Frame_Shift_Del_Frame_Shift_Ins</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              source  target  similarity\n",
       "0                                nan     NaN        1.00\n",
       "1                  Nonsense_Mutation     NaN        0.31\n",
       "2                                 WT     NaN         NaN\n",
       "3                  Missense_Mutation     NaN         NaN\n",
       "4  Frame_Shift_Ins_Nonsense_Mutation     NaN         NaN\n",
       "5  Frame_Shift_Del_Nonsense_Mutation     NaN         NaN\n",
       "6                    Frame_Shift_Del     NaN         NaN\n",
       "7    Frame_Shift_Del_Frame_Shift_Ins     NaN         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** Genomics_subtype<br>**Target column:** Genomic_subtype<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MSI-H</td>\n",
       "      <td>MSI-H</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>POLE</td>\n",
       "      <td>POLE</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CNV_low</td>\n",
       "      <td>CNV_L</td>\n",
       "      <td>0.663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CNV_high</td>\n",
       "      <td>CNV_H</td>\n",
       "      <td>0.632</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     source target  similarity\n",
       "0     MSI-H  MSI-H       1.000\n",
       "1       nan    NaN       1.000\n",
       "2      POLE   POLE       1.000\n",
       "3   CNV_low  CNV_L       0.663\n",
       "4  CNV_high  CNV_H       0.632"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** WXS_normal_sample_type<br>**Target column:** Batch<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Blood_normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         source  target  similarity\n",
       "0           nan     NaN         1.0\n",
       "1  Blood_normal     NaN         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** WXS_tumor_sample_type<br>**Target column:** Tumor_site<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Tumor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  source  target  similarity\n",
       "0    nan     NaN         1.0\n",
       "1  Tumor     NaN         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** WGS_normal_sample_type<br>**Target column:** Mutation_signature_SBS7a<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Blood_normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         source  target  similarity\n",
       "0           nan     NaN         1.0\n",
       "1  Blood_normal     NaN         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** WGS_tumor_sample_type<br>**Target column:** Tumor_site<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Tumor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  source  target  similarity\n",
       "0    nan     NaN         1.0\n",
       "1  Tumor     NaN         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** RNAseq_R1_sample_type<br>**Target column:** ARID1A<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Adjacent_normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Tumor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            source  target  similarity\n",
       "0              nan     NaN         1.0\n",
       "1  Adjacent_normal     NaN         NaN\n",
       "2            Tumor     NaN         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** RNAseq_R2_sample_type<br>**Target column:** ARID1A<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Adjacent_normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Tumor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            source  target  similarity\n",
       "0              nan     NaN         1.0\n",
       "1  Adjacent_normal     NaN         NaN\n",
       "2            Tumor     NaN         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** miRNAseq_sample_type<br>**Target column:** Mutation_signature_SBS7a<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Adjacent_normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Tumor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            source  target  similarity\n",
       "0              nan     NaN         1.0\n",
       "1  Adjacent_normal     NaN         NaN\n",
       "2            Tumor     NaN         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** Methylation_available<br>**Target column:** Mutation_signature_SBS42<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>YES</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  source  target  similarity\n",
       "0    nan     NaN         1.0\n",
       "1    YES     NaN         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** Methylation_quality<br>**Target column:** MSI_status<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PASS</td>\n",
       "      <td>MSS</td>\n",
       "      <td>0.414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Failed</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   source target  similarity\n",
       "0     nan    NaN       1.000\n",
       "1    PASS    MSS       0.414\n",
       "2  Failed   None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Matches values from each pair of source-target columns\n",
    "value_matches = bdi.match_values(df_source, df_target, schema_mapping.head(250), method=\"tfidf\")\n",
    "\n",
    "# Print value matches\n",
    "for match in value_matches:\n",
    "    display(\n",
    "        Markdown(\n",
    "            f\"<br>**Source column:** {match.attrs['source']}<br>\"\n",
    "            f\"**Target column:** {match.attrs['target']}<br>\"\n",
    "        )\n",
    "    )\n",
    "    display(match)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['b4', 'b3', 'b1', 'b2', nan]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(df_target[\"Batch\"].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Idx',\n",
       " 'Case_id',\n",
       " 'Case_excluded',\n",
       " 'Batch',\n",
       " 'Plex',\n",
       " 'ReporterName',\n",
       " 'Aliquot_ID',\n",
       " 'Group',\n",
       " 'Discovery_study',\n",
       " 'Age',\n",
       " 'Sex',\n",
       " 'Histologic_Type',\n",
       " 'Histologic_grade',\n",
       " 'Tumor_size_cm',\n",
       " 'Height_at_time_of_surgery_cm',\n",
       " 'Weight_at_time_of_surgery_kg',\n",
       " 'BMI',\n",
       " 'Myometrial_invasion',\n",
       " 'Myometrial_invasion_present_specify',\n",
       " 'AJCC_tnm_cancer_staging_edition_used',\n",
       " 'Pathologic_staging_primary_tumor_pt',\n",
       " 'Pathologic_staging_regional_lymph_nodes_pn',\n",
       " 'Number_of_pelvic_lymph_nodes_examined',\n",
       " 'Tumor_stage_pathological',\n",
       " 'Race',\n",
       " 'CNV_ratio',\n",
       " 'CNV_status',\n",
       " 'POLE',\n",
       " 'MSIsensor_ratio',\n",
       " 'MSI_status',\n",
       " 'Genomic_subtype',\n",
       " 'Mutation_load',\n",
       " 'TP53',\n",
       " 'PTEN',\n",
       " 'CTNNB1',\n",
       " 'ARID1A',\n",
       " 'PIK3CA',\n",
       " 'xCell_Myeloid_dendritic_cell_activated',\n",
       " 'xCell_B_cell',\n",
       " 'xCell_T_cell_CD4+_memory',\n",
       " 'xCell_T_cell_CD4+_naive',\n",
       " 'xCell_T_cell_CD4+_(non-regulatory)',\n",
       " 'xCell_T_cell_CD4+_central_memory',\n",
       " 'xCell_T_cell_CD4+_effector_memory',\n",
       " 'xCell_T_cell_CD8+_naive',\n",
       " 'xCell_T_cell_CD8+',\n",
       " 'xCell_T_cell_CD8+_central_memory',\n",
       " 'xCell_T_cell_CD8+_effector_memory',\n",
       " 'xCell_Class-switched_memory_B_cell',\n",
       " 'xCell_Common_lymphoid_progenitor',\n",
       " 'xCell_Common_myeloid_progenitor',\n",
       " 'xCell_Myeloid_dendritic_cell',\n",
       " 'xCell_Endothelial_cell',\n",
       " 'xCell_Eosinophil',\n",
       " 'xCell_Cancer_associated_fibroblast',\n",
       " 'xCell_Granulocyte-monocyte_progenitor',\n",
       " 'xCell_Hematopoietic_stem_cell',\n",
       " 'xCell_Macrophage',\n",
       " 'xCell_Macrophage_M1',\n",
       " 'xCell_Macrophage_M2',\n",
       " 'xCell_Mast_cell',\n",
       " 'xCell_B_cell_memory',\n",
       " 'xCell_Monocyte',\n",
       " 'xCell_B_cell_naive',\n",
       " 'xCell_Neutrophil',\n",
       " 'xCell_NK_cell',\n",
       " 'xCell_T_cell_NK',\n",
       " 'xCell_Plasmacytoid_dendritic_cell',\n",
       " 'xCell_B_cell_plasma',\n",
       " 'xCell_T_cell_gamma_delta',\n",
       " 'xCell_T_cell_CD4+_Th1',\n",
       " 'xCell_T_cell_CD4+_Th2',\n",
       " 'xCell_T_cell_regulatory_(Tregs)',\n",
       " 'xCell_immune_score',\n",
       " 'xCell_stroma_score',\n",
       " 'xCell_microenvironment_score',\n",
       " 'Cibersort_B_cell_naive',\n",
       " 'Cibersort_B_cell_memory',\n",
       " 'Cibersort_B_cell_plasma',\n",
       " 'Cibersort_T_cell_CD8+',\n",
       " 'Cibersort_T_cell_CD4+_naive',\n",
       " 'Cibersort_T_cell_CD4+_memory_resting',\n",
       " 'Cibersort_T_cell_CD4+_memory_activated',\n",
       " 'Cibersort_T_cell_follicular_helper',\n",
       " 'Cibersort_T_cell_regulatory_(Tregs)',\n",
       " 'Cibersort_T_cell_gamma_delta',\n",
       " 'Cibersort_NK_cell_resting',\n",
       " 'Cibersort_NK_cell_activated',\n",
       " 'Cibersort_Monocyte',\n",
       " 'Cibersort_Macrophage_M0',\n",
       " 'Cibersort_Macrophage_M1',\n",
       " 'Cibersort_Macrophage_M2',\n",
       " 'Cibersort_Myeloid_dendritic_cell_resting',\n",
       " 'Cibersort_Myeloid_dendritic_cell_activated',\n",
       " 'Cibersort_Mast_cell_activated',\n",
       " 'Cibersort_Mast_cell_resting',\n",
       " 'Cibersort_Eosinophil',\n",
       " 'Cibersort_Neutrophil',\n",
       " 'Mutation_signature_SBS1',\n",
       " 'Mutation_signature_SBS5',\n",
       " 'Mutation_signature_SBS6',\n",
       " 'Mutation_signature_SBS7a',\n",
       " 'Mutation_signature_SBS10a',\n",
       " 'Mutation_signature_SBS10b',\n",
       " 'Mutation_signature_SBS15',\n",
       " 'Mutation_signature_SBS20',\n",
       " 'Mutation_signature_SBS21',\n",
       " 'Mutation_signature_SBS42',\n",
       " 'Mutation_signature_SBS54',\n",
       " 'ABSOLUTE_tumor_purity',\n",
       " 'Diabetes',\n",
       " 'Metformin_treatment',\n",
       " 'Progeny_Androgen',\n",
       " 'Progeny_EGFR',\n",
       " 'Progeny_Estrogen',\n",
       " 'Progeny_Hypoxia',\n",
       " 'Progeny_JAK.STAT',\n",
       " 'Progeny_MAPK',\n",
       " 'Progeny_NFkB',\n",
       " 'Progeny_p53',\n",
       " 'Progeny_PI3K',\n",
       " 'Progeny_TGFb',\n",
       " 'Progeny_TNFa',\n",
       " 'Progeny_Trail',\n",
       " 'Progeny_VEGF',\n",
       " 'Progeny_WNT',\n",
       " 'Estimate_StromalScore',\n",
       " 'Estimate_ImmuneScore',\n",
       " 'Estimate_ESTIMATEScore',\n",
       " 'Peritoneal_ascitic_fluid',\n",
       " 'Tumor_necrosis',\n",
       " 'Margin_status',\n",
       " 'Ethnicity',\n",
       " 'Ethnicity_race_ancestry_identified',\n",
       " 'Participant_country',\n",
       " 'Tumor_site',\n",
       " 'Tumor_site_other',\n",
       " 'Tumor_focality',\n",
       " 'Number_of_pelvic_lymph_nodes_positive_for_tumor_by_ihc_staining_only',\n",
       " 'Number_of_pelvic_lymph_nodes_positive_for_tumor_by_he',\n",
       " 'Number_of_para-aortic_lymph_nodes_examined',\n",
       " 'Number_of_para-aortic_lymph_nodes_positive_for_tumor_by_ihc_staining_only',\n",
       " 'Number_of_para-aortic_lymph_nodes_positive_for_tumor_by_he',\n",
       " 'Number_of_other_lymph_nodes_examined',\n",
       " 'Number_of_other_lymph_nodes_positive_for_tumor_by_ihc_staining_only',\n",
       " 'Number_of_other_lymph_nodes_positive_for_tumor_by_he',\n",
       " 'Clinical_staging_distant_metastasis_cm',\n",
       " 'Residual_tumor',\n",
       " 'Ancillary_studies_estrogen_receptor',\n",
       " 'Ancillary_studies_progesterone_receptor',\n",
       " 'Ancillary_studies_mlh1',\n",
       " 'Ancillary_studies_msh2',\n",
       " 'Ancillary_studies_msh6',\n",
       " 'Ancillary_studies_pms2',\n",
       " 'Ancillary_studies_p53',\n",
       " 'Ancillary_studies_other_immunohistochemistry_performed',\n",
       " 'Ancillary_studies_mlh1_promoter_hypermethylation',\n",
       " 'Ancillary_studies_other_testing_performed',\n",
       " 'Donor_information_menopause_status',\n",
       " 'Donor_information_has_the_patient_ever_taken_menopausal_hormone_therapy',\n",
       " 'Donor_information_number_of_full_term_pregnancies',\n",
       " 'Ancillary_studies_other_immunohistochemistry_type_and_result',\n",
       " 'baseline_ancillary_studies_other_testing_type_and_result',\n",
       " 'History_of_cancer',\n",
       " 'Alcohol_consumption',\n",
       " 'Tobacco_smoking_history',\n",
       " 'Age_at_which_the_participant_started_smoking',\n",
       " 'Age_at_which_the_participant_stopped_smoking',\n",
       " 'On_the_days_participant_smoked_how_many_cigarettes_did_he_she_usually_smoke',\n",
       " 'Number_of_pack_years_smoked',\n",
       " 'Was_the_participant_exposed_to_secondhand_smoke',\n",
       " 'Personal_medical_history_history_source',\n",
       " 'medical_history_current_medications',\n",
       " 'Current_medications_history_source',\n",
       " 'Current_medications_medication_name_vitamins_supplements',\n",
       " 'Cancer_history_cancer_type',\n",
       " 'Cancer_history_history_source',\n",
       " 'Cancer_history_history_of_any_treatment',\n",
       " 'Cancer_history_medical_record_documentation_of_this_history_of_cancer_and_treatment',\n",
       " 'Procurement_blood_collection_minimum_required_blood_collected',\n",
       " 'Procurement_blood_collection_number_of_blood_tubes_collected',\n",
       " 'Follow-up_follow_up_period',\n",
       " 'Follow-up_is_this_patient_lost_to_follow-up',\n",
       " 'Follow-up_vital_status_at_date_of_last_contact',\n",
       " 'Follow-up_days_from_date_of_initial_pathologic_diagnosis_to_date_of_last_contact',\n",
       " 'Follow-up_adjuvant_post-operative_radiation_therapy',\n",
       " 'Follow-up_adjuvant_post-operative_pharmaceutical_therapy',\n",
       " 'Follow-up_adjuvant_post-operative_immunological_therapy',\n",
       " 'Follow-up_tumor_status_at_date_of_last_contact_or_death',\n",
       " 'Follow-up_has_the_patient_ever_taken_menopausal_hormone_therapy',\n",
       " 'Follow-up_has_the_patient_ever_taken_oral_contraceptives',\n",
       " 'Follow-up_has_the_patient_ever_taken_tamoxifen',\n",
       " 'Follow-up_hypertension',\n",
       " 'Follow-up_has_the_patient_ever_been_diagnosed_with_diabetes_by_a_physician',\n",
       " 'Follow-up_number_of_full_term_pregnancies',\n",
       " 'Follow-up_has_the_patient_had_colorectal_cancer',\n",
       " 'Follow-up_measure_of_success_of_outcome_at_the_completion_of_initial_first_course_treatment',\n",
       " 'Follow-up_measure_of_success_of_outcome_at_completion_of_this_follow-up_form',\n",
       " 'Follow-up_measure_of_success_of_outcome_at_date_of_last_contact_or_death',\n",
       " 'follow-up_new_tumor_after_initial_treatment',\n",
       " 'Follow-up_type_of_new_tumor',\n",
       " 'Follow-up_site_of_new_tumor',\n",
       " 'Follow-up_days_from_date_of_initial_pathologic_diagnosis_to_date_of_new_tumor_after_initial_treatment',\n",
       " 'Follow-up_additional_surgery_for_new_tumor',\n",
       " 'Follow-up_additional_treatment_radiation_therapy_for_new_tumor',\n",
       " 'Follow-up_additional_treatment_pharmaceutical_therapy_for_new_tumor',\n",
       " 'Follow-up_additional_treatment_immuno_for_new_tumor',\n",
       " 'Follow-up_days_from_date_of_collection_to_date_of_last_contact',\n",
       " 'Follow-up_cause_of_death',\n",
       " 'Follow-up_days_from_date_of_initial_pathologic_diagnosis_to_date_of_death',\n",
       " 'Follow-up_days_from_date_of_initial_pathologic_diagnosis_to_date_of_additional_surgery_for_new_tumor',\n",
       " 'Follow-up_procedure_type_of_new_tumor',\n",
       " 'Follow-up_residual_tumor_after_surgery_for_new_tumor']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(df_target.columns)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Generating a harmonized table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>participant_country</th>\n",
       "      <th>sex</th>\n",
       "      <th>tumor_stage_pathological</th>\n",
       "      <th>pathologic_staging_regional_lymph_nodes_pn</th>\n",
       "      <th>tumor_focality</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>United States</td>\n",
       "      <td>Female</td>\n",
       "      <td>Stage IA</td>\n",
       "      <td>pN0</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>United States</td>\n",
       "      <td>Female</td>\n",
       "      <td>Stage IV</td>\n",
       "      <td>pNX</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>United States</td>\n",
       "      <td>Female</td>\n",
       "      <td>Stage IA</td>\n",
       "      <td>pN0</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>United States</td>\n",
       "      <td>Female</td>\n",
       "      <td>Stage IA</td>\n",
       "      <td>pNX</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>None</td>\n",
       "      <td>Female</td>\n",
       "      <td>Stage IA</td>\n",
       "      <td>pNX</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>None</td>\n",
       "      <td>Female</td>\n",
       "      <td>Stage III</td>\n",
       "      <td>pN0</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>United States</td>\n",
       "      <td>Female</td>\n",
       "      <td>Stage III</td>\n",
       "      <td>pN0</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>None</td>\n",
       "      <td>Female</td>\n",
       "      <td>Stage IA</td>\n",
       "      <td>pN0</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>104 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    participant_country     sex tumor_stage_pathological  \\\n",
       "0         United States  Female                 Stage IA   \n",
       "1         United States  Female                 Stage IV   \n",
       "2         United States  Female                 Stage IA   \n",
       "3                   NaN     NaN                      NaN   \n",
       "4         United States  Female                 Stage IA   \n",
       "..                  ...     ...                      ...   \n",
       "99                 None  Female                 Stage IA   \n",
       "100                None  Female                Stage III   \n",
       "101       United States  Female                Stage III   \n",
       "102                None  Female                 Stage IA   \n",
       "103                None     NaN                      NaN   \n",
       "\n",
       "    pathologic_staging_regional_lymph_nodes_pn tumor_focality  \n",
       "0                                          pN0       Unifocal  \n",
       "1                                          pNX       Unifocal  \n",
       "2                                          pN0       Unifocal  \n",
       "3                                          NaN            NaN  \n",
       "4                                          pNX       Unifocal  \n",
       "..                                         ...            ...  \n",
       "99                                         pNX       Unifocal  \n",
       "100                                        pN0       Unifocal  \n",
       "101                                        pN0       Unifocal  \n",
       "102                                        pN0       Unifocal  \n",
       "103                                        NaN            NaN  \n",
       "\n",
       "[104 rows x 5 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_mapped = bdi.materialize_mapping(df_source, value_matches)\n",
    "df_mapped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Country</th>\n",
       "      <th>Gender</th>\n",
       "      <th>FIGO_stage</th>\n",
       "      <th>Path_Stage_Reg_Lymph_Nodes-pN</th>\n",
       "      <th>tumor_Stage-Pathological</th>\n",
       "      <th>Tumor_Focality</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>United States</td>\n",
       "      <td>Female</td>\n",
       "      <td>IA</td>\n",
       "      <td>pN0</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>United States</td>\n",
       "      <td>Female</td>\n",
       "      <td>IA</td>\n",
       "      <td>pNX</td>\n",
       "      <td>Stage IV</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>United States</td>\n",
       "      <td>Female</td>\n",
       "      <td>IA</td>\n",
       "      <td>pN0</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>United States</td>\n",
       "      <td>Female</td>\n",
       "      <td>IA</td>\n",
       "      <td>pNX</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>Ukraine</td>\n",
       "      <td>Female</td>\n",
       "      <td>IA</td>\n",
       "      <td>pNX</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>Ukraine</td>\n",
       "      <td>Female</td>\n",
       "      <td>II</td>\n",
       "      <td>pN0</td>\n",
       "      <td>Stage II</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>United States</td>\n",
       "      <td>Female</td>\n",
       "      <td>II</td>\n",
       "      <td>pN0</td>\n",
       "      <td>Stage II</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>Ukraine</td>\n",
       "      <td>Female</td>\n",
       "      <td>IA</td>\n",
       "      <td>pN0</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>Ukraine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>104 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Country  Gender FIGO_stage Path_Stage_Reg_Lymph_Nodes-pN  \\\n",
       "0    United States  Female         IA                           pN0   \n",
       "1    United States  Female         IA                           pNX   \n",
       "2    United States  Female         IA                           pN0   \n",
       "3              NaN     NaN        NaN                           NaN   \n",
       "4    United States  Female         IA                           pNX   \n",
       "..             ...     ...        ...                           ...   \n",
       "99         Ukraine  Female         IA                           pNX   \n",
       "100        Ukraine  Female         II                           pN0   \n",
       "101  United States  Female         II                           pN0   \n",
       "102        Ukraine  Female         IA                           pN0   \n",
       "103        Ukraine     NaN        NaN                           NaN   \n",
       "\n",
       "    tumor_Stage-Pathological Tumor_Focality  \n",
       "0                    Stage I       Unifocal  \n",
       "1                   Stage IV       Unifocal  \n",
       "2                    Stage I       Unifocal  \n",
       "3                        NaN            NaN  \n",
       "4                    Stage I       Unifocal  \n",
       "..                       ...            ...  \n",
       "99                   Stage I       Unifocal  \n",
       "100                 Stage II       Unifocal  \n",
       "101                 Stage II       Unifocal  \n",
       "102                  Stage I       Unifocal  \n",
       "103                      NaN            NaN  \n",
       "\n",
       "[104 rows x 6 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "source_column_names = list(map(lambda m: m.attrs['source'], value_matches))\n",
    "target_column_names = list(map(lambda m: m.attrs['target'], value_matches))\n",
    "df_source[source_column_names]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>participant_country</th>\n",
       "      <th>sex</th>\n",
       "      <th>tumor_stage_pathological</th>\n",
       "      <th>pathologic_staging_regional_lymph_nodes_pn</th>\n",
       "      <th>tumor_stage_pathological</th>\n",
       "      <th>tumor_focality</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>United States</td>\n",
       "      <td>Female</td>\n",
       "      <td>Stage IA</td>\n",
       "      <td>pN0</td>\n",
       "      <td>Stage IA</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>United States</td>\n",
       "      <td>Female</td>\n",
       "      <td>Stage IV</td>\n",
       "      <td>pNX</td>\n",
       "      <td>Stage IV</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>United States</td>\n",
       "      <td>Female</td>\n",
       "      <td>Stage IA</td>\n",
       "      <td>pN0</td>\n",
       "      <td>Stage IA</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>United States</td>\n",
       "      <td>Female</td>\n",
       "      <td>Stage IA</td>\n",
       "      <td>pNX</td>\n",
       "      <td>Stage IA</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>Poland</td>\n",
       "      <td>Male</td>\n",
       "      <td>Stage III</td>\n",
       "      <td>pN2</td>\n",
       "      <td>Stage III</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>China</td>\n",
       "      <td>Female</td>\n",
       "      <td>Stage III</td>\n",
       "      <td>pN2</td>\n",
       "      <td>Stage III</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>China</td>\n",
       "      <td>Male</td>\n",
       "      <td>Stage III</td>\n",
       "      <td>pN2</td>\n",
       "      <td>Stage III</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>Poland</td>\n",
       "      <td>Female</td>\n",
       "      <td>Stage III</td>\n",
       "      <td>pN2</td>\n",
       "      <td>Stage III</td>\n",
       "      <td>Multifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>Poland</td>\n",
       "      <td>Female</td>\n",
       "      <td>Stage III</td>\n",
       "      <td>pN2</td>\n",
       "      <td>Stage III</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>244 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    participant_country     sex tumor_stage_pathological  \\\n",
       "0         United States  Female                 Stage IA   \n",
       "1         United States  Female                 Stage IV   \n",
       "2         United States  Female                 Stage IA   \n",
       "3                   NaN     NaN                      NaN   \n",
       "4         United States  Female                 Stage IA   \n",
       "..                  ...     ...                      ...   \n",
       "135              Poland    Male                Stage III   \n",
       "136               China  Female                Stage III   \n",
       "137               China    Male                Stage III   \n",
       "138              Poland  Female                Stage III   \n",
       "139              Poland  Female                Stage III   \n",
       "\n",
       "    pathologic_staging_regional_lymph_nodes_pn tumor_stage_pathological  \\\n",
       "0                                          pN0                 Stage IA   \n",
       "1                                          pNX                 Stage IV   \n",
       "2                                          pN0                 Stage IA   \n",
       "3                                          NaN                      NaN   \n",
       "4                                          pNX                 Stage IA   \n",
       "..                                         ...                      ...   \n",
       "135                                        pN2                Stage III   \n",
       "136                                        pN2                Stage III   \n",
       "137                                        pN2                Stage III   \n",
       "138                                        pN2                Stage III   \n",
       "139                                        pN2                Stage III   \n",
       "\n",
       "    tumor_focality  \n",
       "0         Unifocal  \n",
       "1         Unifocal  \n",
       "2         Unifocal  \n",
       "3              NaN  \n",
       "4         Unifocal  \n",
       "..             ...  \n",
       "135       Unifocal  \n",
       "136       Unifocal  \n",
       "137       Unifocal  \n",
       "138     Multifocal  \n",
       "139       Unifocal  \n",
       "\n",
       "[244 rows x 6 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.concat([df_mapped[target_column_names], df_target[target_column_names]])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
