{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/javascript": "(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  var force = true;\n  var py_version = '3.4.3'.replace('rc', '-rc.').replace('.dev', '-dev.');\n  var reloading = false;\n  var Bokeh = root.Bokeh;\n\n  if (typeof (root._bokeh_timeout) === \"undefined\" || force) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks;\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, js_modules, js_exports, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n    if (js_modules == null) js_modules = [];\n    if (js_exports == null) js_exports = {};\n\n    root._bokeh_onload_callbacks.push(callback);\n\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls.length === 0 && js_modules.length === 0 && Object.keys(js_exports).length === 0) {\n      run_callbacks();\n      return null;\n    }\n    if (!reloading) {\n      console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    }\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n    window._bokeh_on_load = on_load\n\n    function on_error() {\n      console.error(\"failed to load \" + url);\n    }\n\n    var skip = [];\n    if (window.requirejs) {\n      window.requirejs.config({'packages': {}, 'paths': {'tabulator': 'https://cdn.jsdelivr.net/npm/tabulator-tables@5.5.0/dist/js/tabulator.min', 'moment': 'https://cdn.jsdelivr.net/npm/luxon/build/global/luxon.min'}, 'shim': {}});\n      require([\"tabulator\"], function(Tabulator) {\n\twindow.Tabulator = Tabulator\n\ton_load()\n      })\n      require([\"moment\"], function(moment) {\n\twindow.moment = moment\n\ton_load()\n      })\n      root._bokeh_is_loading = css_urls.length + 2;\n    } else {\n      root._bokeh_is_loading = css_urls.length + js_urls.length + js_modules.length + Object.keys(js_exports).length;\n    }\n\n    var existing_stylesheets = []\n    var links = document.getElementsByTagName('link')\n    for (var i = 0; i < links.length; i++) {\n      var link = links[i]\n      if (link.href != null) {\n\texisting_stylesheets.push(link.href)\n      }\n    }\n    for (var i = 0; i < css_urls.length; i++) {\n      var url = css_urls[i];\n      if (existing_stylesheets.indexOf(url) !== -1) {\n\ton_load()\n\tcontinue;\n      }\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error;\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }    if (((window.Tabulator !== undefined) && (!(window.Tabulator instanceof HTMLElement))) || window.requirejs) {\n      var urls = ['https://cdn.holoviz.org/panel/1.4.5/dist/bundled/datatabulator/tabulator-tables@5.5.0/dist/js/tabulator.min.js'];\n      for (var i = 0; i < urls.length; i++) {\n        skip.push(urls[i])\n      }\n    }    if (((window.moment !== undefined) && (!(window.moment instanceof HTMLElement))) || window.requirejs) {\n      var urls = ['https://cdn.holoviz.org/panel/1.4.5/dist/bundled/datatabulator/luxon/build/global/luxon.min.js'];\n      for (var i = 0; i < urls.length; i++) {\n        skip.push(urls[i])\n      }\n    }    var existing_scripts = []\n    var scripts = document.getElementsByTagName('script')\n    for (var i = 0; i < scripts.length; i++) {\n      var script = scripts[i]\n      if (script.src != null) {\n\texisting_scripts.push(script.src)\n      }\n    }\n    for (var i = 0; i < js_urls.length; i++) {\n      var url = js_urls[i];\n      if (skip.indexOf(url) !== -1 || existing_scripts.indexOf(url) !== -1) {\n\tif (!window.requirejs) {\n\t  on_load();\n\t}\n\tcontinue;\n      }\n      var element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error;\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n    for (var i = 0; i < js_modules.length; i++) {\n      var url = js_modules[i];\n      if (skip.indexOf(url) !== -1 || existing_scripts.indexOf(url) !== -1) {\n\tif (!window.requirejs) {\n\t  on_load();\n\t}\n\tcontinue;\n      }\n      var element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error;\n      element.async = false;\n      element.src = url;\n      element.type = \"module\";\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n    for (const name in js_exports) {\n      var url = js_exports[name];\n      if (skip.indexOf(url) >= 0 || root[name] != null) {\n\tif (!window.requirejs) {\n\t  on_load();\n\t}\n\tcontinue;\n      }\n      var element = document.createElement('script');\n      element.onerror = on_error;\n      element.async = false;\n      element.type = \"module\";\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      element.textContent = `\n      import ${name} from \"${url}\"\n      window.${name} = ${name}\n      window._bokeh_on_load()\n      `\n      document.head.appendChild(element);\n    }\n    if (!js_urls.length && !js_modules.length) {\n      on_load()\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  var js_urls = [\"https://cdn.holoviz.org/panel/1.4.5/dist/bundled/datatabulator/tabulator-tables@5.5.0/dist/js/tabulator.min.js\", \"https://cdn.holoviz.org/panel/1.4.5/dist/bundled/datatabulator/luxon/build/global/luxon.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-3.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.4.3.min.js\", \"https://cdn.holoviz.org/panel/1.4.5/dist/panel.min.js\"];\n  var js_modules = [];\n  var js_exports = {};\n  var css_urls = [\"https://cdn.holoviz.org/panel/1.4.5/dist/bundled/datatabulator/tabulator-tables@5.5.0/dist/css/tabulator_simple.min.css?v=1.4.5\"];\n  var inline_js = [    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\nfunction(Bokeh) {} // ensure no trailing comma for IE\n  ];\n\n  function run_inline_js() {\n    if ((root.Bokeh !== undefined) || (force === true)) {\n      for (var i = 0; i < inline_js.length; i++) {\n\ttry {\n          inline_js[i].call(root, root.Bokeh);\n\t} catch(e) {\n\t  if (!reloading) {\n\t    throw e;\n\t  }\n\t}\n      }\n      // Cache old bokeh versions\n      if (Bokeh != undefined && !reloading) {\n\tvar NewBokeh = root.Bokeh;\n\tif (Bokeh.versions === undefined) {\n\t  Bokeh.versions = new Map();\n\t}\n\tif (NewBokeh.version !== Bokeh.version) {\n\t  Bokeh.versions.set(NewBokeh.version, NewBokeh)\n\t}\n\troot.Bokeh = Bokeh;\n      }} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    }\n    root._bokeh_is_initializing = false\n  }\n\n  function load_or_wait() {\n    // Implement a backoff loop that tries to ensure we do not load multiple\n    // versions of Bokeh and its dependencies at the same time.\n    // In recent versions we use the root._bokeh_is_initializing flag\n    // to determine whether there is an ongoing attempt to initialize\n    // bokeh, however for backward compatibility we also try to ensure\n    // that we do not start loading a newer (Panel>=1.0 and Bokeh>3) version\n    // before older versions are fully initialized.\n    if (root._bokeh_is_initializing && Date.now() > root._bokeh_timeout) {\n      root._bokeh_is_initializing = false;\n      root._bokeh_onload_callbacks = undefined;\n      console.log(\"Bokeh: BokehJS was loaded multiple times but one version failed to initialize.\");\n      load_or_wait();\n    } else if (root._bokeh_is_initializing || (typeof root._bokeh_is_initializing === \"undefined\" && root._bokeh_onload_callbacks !== undefined)) {\n      setTimeout(load_or_wait, 100);\n    } else {\n      root._bokeh_is_initializing = true\n      root._bokeh_onload_callbacks = []\n      var bokeh_loaded = Bokeh != null && (Bokeh.version === py_version || (Bokeh.versions !== undefined && Bokeh.versions.has(py_version)));\n      if (!reloading && !bokeh_loaded) {\n\troot.Bokeh = undefined;\n      }\n      load_libs(css_urls, js_urls, js_modules, js_exports, function() {\n\tconsole.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n\trun_inline_js();\n      });\n    }\n  }\n  // Give older versions of the autoload script a head-start to ensure\n  // they initialize before we start loading newer version.\n  setTimeout(load_or_wait, 100)\n}(window));",
      "application/vnd.holoviews_load.v0+json": ""
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": "\nif ((window.PyViz === undefined) || (window.PyViz instanceof HTMLElement)) {\n  window.PyViz = {comms: {}, comm_status:{}, kernels:{}, receivers: {}, plot_index: []}\n}\n\n\n    function JupyterCommManager() {\n    }\n\n    JupyterCommManager.prototype.register_target = function(plot_id, comm_id, msg_handler) {\n      if (window.comm_manager || ((window.Jupyter !== undefined) && (Jupyter.notebook.kernel != null))) {\n        var comm_manager = window.comm_manager || Jupyter.notebook.kernel.comm_manager;\n        comm_manager.register_target(comm_id, function(comm) {\n          comm.on_msg(msg_handler);\n        });\n      } else if ((plot_id in window.PyViz.kernels) && (window.PyViz.kernels[plot_id])) {\n        window.PyViz.kernels[plot_id].registerCommTarget(comm_id, function(comm) {\n          comm.onMsg = msg_handler;\n        });\n      } else if (typeof google != 'undefined' && google.colab.kernel != null) {\n        google.colab.kernel.comms.registerTarget(comm_id, (comm) => {\n          var messages = comm.messages[Symbol.asyncIterator]();\n          function processIteratorResult(result) {\n            var message = result.value;\n            console.log(message)\n            var content = {data: message.data, comm_id};\n            var buffers = []\n            for (var buffer of message.buffers || []) {\n              buffers.push(new DataView(buffer))\n            }\n            var metadata = message.metadata || {};\n            var msg = {content, buffers, metadata}\n            msg_handler(msg);\n            return messages.next().then(processIteratorResult);\n          }\n          return messages.next().then(processIteratorResult);\n        })\n      }\n    }\n\n    JupyterCommManager.prototype.get_client_comm = function(plot_id, comm_id, msg_handler) {\n      if (comm_id in window.PyViz.comms) {\n        return window.PyViz.comms[comm_id];\n      } else if (window.comm_manager || ((window.Jupyter !== undefined) && (Jupyter.notebook.kernel != null))) {\n        var comm_manager = window.comm_manager || Jupyter.notebook.kernel.comm_manager;\n        var comm = comm_manager.new_comm(comm_id, {}, {}, {}, comm_id);\n        if (msg_handler) {\n          comm.on_msg(msg_handler);\n        }\n      } else if ((plot_id in window.PyViz.kernels) && (window.PyViz.kernels[plot_id])) {\n        var comm = window.PyViz.kernels[plot_id].connectToComm(comm_id);\n        comm.open();\n        if (msg_handler) {\n          comm.onMsg = msg_handler;\n        }\n      } else if (typeof google != 'undefined' && google.colab.kernel != null) {\n        var comm_promise = google.colab.kernel.comms.open(comm_id)\n        comm_promise.then((comm) => {\n          window.PyViz.comms[comm_id] = comm;\n          if (msg_handler) {\n            var messages = comm.messages[Symbol.asyncIterator]();\n            function processIteratorResult(result) {\n              var message = result.value;\n              var content = {data: message.data};\n              var metadata = message.metadata || {comm_id};\n              var msg = {content, metadata}\n              msg_handler(msg);\n              return messages.next().then(processIteratorResult);\n            }\n            return messages.next().then(processIteratorResult);\n          }\n        }) \n        var sendClosure = (data, metadata, buffers, disposeOnDone) => {\n          return comm_promise.then((comm) => {\n            comm.send(data, metadata, buffers, disposeOnDone);\n          });\n        };\n        var comm = {\n          send: sendClosure\n        };\n      }\n      window.PyViz.comms[comm_id] = comm;\n      return comm;\n    }\n    window.PyViz.comm_manager = new JupyterCommManager();\n    \n\n\nvar JS_MIME_TYPE = 'application/javascript';\nvar HTML_MIME_TYPE = 'text/html';\nvar EXEC_MIME_TYPE = 'application/vnd.holoviews_exec.v0+json';\nvar CLASS_NAME = 'output';\n\n/**\n * Render data to the DOM node\n */\nfunction render(props, node) {\n  var div = document.createElement(\"div\");\n  var script = document.createElement(\"script\");\n  node.appendChild(div);\n  node.appendChild(script);\n}\n\n/**\n * Handle when a new output is added\n */\nfunction handle_add_output(event, handle) {\n  var output_area = handle.output_area;\n  var output = handle.output;\n  if ((output.data == undefined) || (!output.data.hasOwnProperty(EXEC_MIME_TYPE))) {\n    return\n  }\n  var id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n  var toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n  if (id !== undefined) {\n    var nchildren = toinsert.length;\n    var html_node = toinsert[nchildren-1].children[0];\n    html_node.innerHTML = output.data[HTML_MIME_TYPE];\n    var scripts = [];\n    var nodelist = html_node.querySelectorAll(\"script\");\n    for (var i in nodelist) {\n      if (nodelist.hasOwnProperty(i)) {\n        scripts.push(nodelist[i])\n      }\n    }\n\n    scripts.forEach( function (oldScript) {\n      var newScript = document.createElement(\"script\");\n      var attrs = [];\n      var nodemap = oldScript.attributes;\n      for (var j in nodemap) {\n        if (nodemap.hasOwnProperty(j)) {\n          attrs.push(nodemap[j])\n        }\n      }\n      attrs.forEach(function(attr) { newScript.setAttribute(attr.name, attr.value) });\n      newScript.appendChild(document.createTextNode(oldScript.innerHTML));\n      oldScript.parentNode.replaceChild(newScript, oldScript);\n    });\n    if (JS_MIME_TYPE in output.data) {\n      toinsert[nchildren-1].children[1].textContent = output.data[JS_MIME_TYPE];\n    }\n    output_area._hv_plot_id = id;\n    if ((window.Bokeh !== undefined) && (id in Bokeh.index)) {\n      window.PyViz.plot_index[id] = Bokeh.index[id];\n    } else {\n      window.PyViz.plot_index[id] = null;\n    }\n  } else if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n    var bk_div = document.createElement(\"div\");\n    bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n    var script_attrs = bk_div.children[0].attributes;\n    for (var i = 0; i < script_attrs.length; i++) {\n      toinsert[toinsert.length - 1].childNodes[1].setAttribute(script_attrs[i].name, script_attrs[i].value);\n    }\n    // store reference to server id on output_area\n    output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n  }\n}\n\n/**\n * Handle when an output is cleared or removed\n */\nfunction handle_clear_output(event, handle) {\n  var id = handle.cell.output_area._hv_plot_id;\n  var server_id = handle.cell.output_area._bokeh_server_id;\n  if (((id === undefined) || !(id in PyViz.plot_index)) && (server_id !== undefined)) { return; }\n  var comm = window.PyViz.comm_manager.get_client_comm(\"hv-extension-comm\", \"hv-extension-comm\", function () {});\n  if (server_id !== null) {\n    comm.send({event_type: 'server_delete', 'id': server_id});\n    return;\n  } else if (comm !== null) {\n    comm.send({event_type: 'delete', 'id': id});\n  }\n  delete PyViz.plot_index[id];\n  if ((window.Bokeh !== undefined) & (id in window.Bokeh.index)) {\n    var doc = window.Bokeh.index[id].model.document\n    doc.clear();\n    const i = window.Bokeh.documents.indexOf(doc);\n    if (i > -1) {\n      window.Bokeh.documents.splice(i, 1);\n    }\n  }\n}\n\n/**\n * Handle kernel restart event\n */\nfunction handle_kernel_cleanup(event, handle) {\n  delete PyViz.comms[\"hv-extension-comm\"];\n  window.PyViz.plot_index = {}\n}\n\n/**\n * Handle update_display_data messages\n */\nfunction handle_update_output(event, handle) {\n  handle_clear_output(event, {cell: {output_area: handle.output_area}})\n  handle_add_output(event, handle)\n}\n\nfunction register_renderer(events, OutputArea) {\n  function append_mime(data, metadata, element) {\n    // create a DOM node to render to\n    var toinsert = this.create_output_subarea(\n    metadata,\n    CLASS_NAME,\n    EXEC_MIME_TYPE\n    );\n    this.keyboard_manager.register_events(toinsert);\n    // Render to node\n    var props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n    render(props, toinsert[0]);\n    element.append(toinsert);\n    return toinsert\n  }\n\n  events.on('output_added.OutputArea', handle_add_output);\n  events.on('output_updated.OutputArea', handle_update_output);\n  events.on('clear_output.CodeCell', handle_clear_output);\n  events.on('delete.Cell', handle_clear_output);\n  events.on('kernel_ready.Kernel', handle_kernel_cleanup);\n\n  OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n    safe: true,\n    index: 0\n  });\n}\n\nif (window.Jupyter !== undefined) {\n  try {\n    var events = require('base/js/events');\n    var OutputArea = require('notebook/js/outputarea').OutputArea;\n    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n      register_renderer(events, OutputArea);\n    }\n  } catch(err) {\n  }\n}\n",
      "application/vnd.holoviews_load.v0+json": ""
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>*[data-root-id],\n",
       "*[data-root-id] > * {\n",
       "  box-sizing: border-box;\n",
       "  font-family: var(--jp-ui-font-family);\n",
       "  font-size: var(--jp-ui-font-size1);\n",
       "  color: var(--vscode-editor-foreground, var(--jp-ui-font-color1));\n",
       "}\n",
       "\n",
       "/* Override VSCode background color */\n",
       ".cell-output-ipywidget-background:has(\n",
       "    > .cell-output-ipywidget-background > .lm-Widget > *[data-root-id]\n",
       "  ),\n",
       ".cell-output-ipywidget-background:has(> .lm-Widget > *[data-root-id]) {\n",
       "  background-color: transparent !important;\n",
       "}\n",
       "</style>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.holoviews_exec.v0+json": "",
      "text/html": [
       "<div id='3f8fdd9c-53b4-4204-9ce7-100e5a7d3b23'>\n",
       "  <div id=\"f38846ea-c9c5-475c-a48f-2147e5b407f2\" data-root-id=\"3f8fdd9c-53b4-4204-9ce7-100e5a7d3b23\" style=\"display: contents;\"></div>\n",
       "</div>\n",
       "<script type=\"application/javascript\">(function(root) {\n",
       "  var docs_json = {\"48235a87-c3b0-4747-9c5c-2d0052dca732\":{\"version\":\"3.4.3\",\"title\":\"Bokeh Application\",\"roots\":[{\"type\":\"object\",\"name\":\"panel.models.browser.BrowserInfo\",\"id\":\"3f8fdd9c-53b4-4204-9ce7-100e5a7d3b23\"},{\"type\":\"object\",\"name\":\"panel.models.comm_manager.CommManager\",\"id\":\"3758eade-28d9-4ebb-b538-a05356c163de\",\"attributes\":{\"plot_id\":\"3f8fdd9c-53b4-4204-9ce7-100e5a7d3b23\",\"comm_id\":\"b23a01f8e61346dabfd8e176e1a288f5\",\"client_comm_id\":\"81f0bcba8eda4a65852bd15a42457c6f\"}}],\"defs\":[{\"type\":\"model\",\"name\":\"ReactiveHTML1\"},{\"type\":\"model\",\"name\":\"FlexBox1\",\"properties\":[{\"name\":\"align_content\",\"kind\":\"Any\",\"default\":\"flex-start\"},{\"name\":\"align_items\",\"kind\":\"Any\",\"default\":\"flex-start\"},{\"name\":\"flex_direction\",\"kind\":\"Any\",\"default\":\"row\"},{\"name\":\"flex_wrap\",\"kind\":\"Any\",\"default\":\"wrap\"},{\"name\":\"gap\",\"kind\":\"Any\",\"default\":\"\"},{\"name\":\"justify_content\",\"kind\":\"Any\",\"default\":\"flex-start\"}]},{\"type\":\"model\",\"name\":\"FloatPanel1\",\"properties\":[{\"name\":\"config\",\"kind\":\"Any\",\"default\":{\"type\":\"map\"}},{\"name\":\"contained\",\"kind\":\"Any\",\"default\":true},{\"name\":\"position\",\"kind\":\"Any\",\"default\":\"right-top\"},{\"name\":\"offsetx\",\"kind\":\"Any\",\"default\":null},{\"name\":\"offsety\",\"kind\":\"Any\",\"default\":null},{\"name\":\"theme\",\"kind\":\"Any\",\"default\":\"primary\"},{\"name\":\"status\",\"kind\":\"Any\",\"default\":\"normalized\"}]},{\"type\":\"model\",\"name\":\"GridStack1\",\"properties\":[{\"name\":\"mode\",\"kind\":\"Any\",\"default\":\"warn\"},{\"name\":\"ncols\",\"kind\":\"Any\",\"default\":null},{\"name\":\"nrows\",\"kind\":\"Any\",\"default\":null},{\"name\":\"allow_resize\",\"kind\":\"Any\",\"default\":true},{\"name\":\"allow_drag\",\"kind\":\"Any\",\"default\":true},{\"name\":\"state\",\"kind\":\"Any\",\"default\":[]}]},{\"type\":\"model\",\"name\":\"drag1\",\"properties\":[{\"name\":\"slider_width\",\"kind\":\"Any\",\"default\":5},{\"name\":\"slider_color\",\"kind\":\"Any\",\"default\":\"black\"},{\"name\":\"value\",\"kind\":\"Any\",\"default\":50}]},{\"type\":\"model\",\"name\":\"click1\",\"properties\":[{\"name\":\"terminal_output\",\"kind\":\"Any\",\"default\":\"\"},{\"name\":\"debug_name\",\"kind\":\"Any\",\"default\":\"\"},{\"name\":\"clears\",\"kind\":\"Any\",\"default\":0}]},{\"type\":\"model\",\"name\":\"FastWrapper1\",\"properties\":[{\"name\":\"object\",\"kind\":\"Any\",\"default\":null},{\"name\":\"style\",\"kind\":\"Any\",\"default\":null}]},{\"type\":\"model\",\"name\":\"NotificationAreaBase1\",\"properties\":[{\"name\":\"js_events\",\"kind\":\"Any\",\"default\":{\"type\":\"map\"}},{\"name\":\"position\",\"kind\":\"Any\",\"default\":\"bottom-right\"},{\"name\":\"_clear\",\"kind\":\"Any\",\"default\":0}]},{\"type\":\"model\",\"name\":\"NotificationArea1\",\"properties\":[{\"name\":\"js_events\",\"kind\":\"Any\",\"default\":{\"type\":\"map\"}},{\"name\":\"notifications\",\"kind\":\"Any\",\"default\":[]},{\"name\":\"position\",\"kind\":\"Any\",\"default\":\"bottom-right\"},{\"name\":\"_clear\",\"kind\":\"Any\",\"default\":0},{\"name\":\"types\",\"kind\":\"Any\",\"default\":[{\"type\":\"map\",\"entries\":[[\"type\",\"warning\"],[\"background\",\"#ffc107\"],[\"icon\",{\"type\":\"map\",\"entries\":[[\"className\",\"fas fa-exclamation-triangle\"],[\"tagName\",\"i\"],[\"color\",\"white\"]]}]]},{\"type\":\"map\",\"entries\":[[\"type\",\"info\"],[\"background\",\"#007bff\"],[\"icon\",{\"type\":\"map\",\"entries\":[[\"className\",\"fas fa-info-circle\"],[\"tagName\",\"i\"],[\"color\",\"white\"]]}]]}]}]},{\"type\":\"model\",\"name\":\"Notification\",\"properties\":[{\"name\":\"background\",\"kind\":\"Any\",\"default\":null},{\"name\":\"duration\",\"kind\":\"Any\",\"default\":3000},{\"name\":\"icon\",\"kind\":\"Any\",\"default\":null},{\"name\":\"message\",\"kind\":\"Any\",\"default\":\"\"},{\"name\":\"notification_type\",\"kind\":\"Any\",\"default\":null},{\"name\":\"_destroyed\",\"kind\":\"Any\",\"default\":false}]},{\"type\":\"model\",\"name\":\"TemplateActions1\",\"properties\":[{\"name\":\"open_modal\",\"kind\":\"Any\",\"default\":0},{\"name\":\"close_modal\",\"kind\":\"Any\",\"default\":0}]},{\"type\":\"model\",\"name\":\"BootstrapTemplateActions1\",\"properties\":[{\"name\":\"open_modal\",\"kind\":\"Any\",\"default\":0},{\"name\":\"close_modal\",\"kind\":\"Any\",\"default\":0}]},{\"type\":\"model\",\"name\":\"TemplateEditor1\",\"properties\":[{\"name\":\"layout\",\"kind\":\"Any\",\"default\":[]}]},{\"type\":\"model\",\"name\":\"MaterialTemplateActions1\",\"properties\":[{\"name\":\"open_modal\",\"kind\":\"Any\",\"default\":0},{\"name\":\"close_modal\",\"kind\":\"Any\",\"default\":0}]},{\"type\":\"model\",\"name\":\"copy_to_clipboard1\",\"properties\":[{\"name\":\"fill\",\"kind\":\"Any\",\"default\":\"none\"},{\"name\":\"value\",\"kind\":\"Any\",\"default\":null}]}]}};\n",
       "  var render_items = [{\"docid\":\"48235a87-c3b0-4747-9c5c-2d0052dca732\",\"roots\":{\"3f8fdd9c-53b4-4204-9ce7-100e5a7d3b23\":\"f38846ea-c9c5-475c-a48f-2147e5b407f2\"},\"root_ids\":[\"3f8fdd9c-53b4-4204-9ce7-100e5a7d3b23\"]}];\n",
       "  var docs = Object.values(docs_json)\n",
       "  if (!docs) {\n",
       "    return\n",
       "  }\n",
       "  const py_version = docs[0].version.replace('rc', '-rc.').replace('.dev', '-dev.')\n",
       "  async function embed_document(root) {\n",
       "    var Bokeh = get_bokeh(root)\n",
       "    await Bokeh.embed.embed_items_notebook(docs_json, render_items);\n",
       "    for (const render_item of render_items) {\n",
       "      for (const root_id of render_item.root_ids) {\n",
       "\tconst id_el = document.getElementById(root_id)\n",
       "\tif (id_el.children.length && id_el.children[0].hasAttribute('data-root-id')) {\n",
       "\t  const root_el = id_el.children[0]\n",
       "\t  root_el.id = root_el.id + '-rendered'\n",
       "\t  for (const child of root_el.children) {\n",
       "            // Ensure JupyterLab does not capture keyboard shortcuts\n",
       "            // see: https://jupyterlab.readthedocs.io/en/4.1.x/extension/notebook.html#keyboard-interaction-model\n",
       "\t    child.setAttribute('data-lm-suppress-shortcuts', 'true')\n",
       "\t  }\n",
       "\t}\n",
       "      }\n",
       "    }\n",
       "  }\n",
       "  function get_bokeh(root) {\n",
       "    if (root.Bokeh === undefined) {\n",
       "      return null\n",
       "    } else if (root.Bokeh.version !== py_version) {\n",
       "      if (root.Bokeh.versions === undefined || !root.Bokeh.versions.has(py_version)) {\n",
       "\treturn null\n",
       "      }\n",
       "      return root.Bokeh.versions.get(py_version);\n",
       "    } else if (root.Bokeh.version === py_version) {\n",
       "      return root.Bokeh\n",
       "    }\n",
       "    return null\n",
       "  }\n",
       "  function is_loaded(root) {\n",
       "    var Bokeh = get_bokeh(root)\n",
       "    return (Bokeh != null && Bokeh.Panel !== undefined && ( root.Tabulator !== undefined) && ( root.Tabulator !== undefined))\n",
       "  }\n",
       "  if (is_loaded(root)) {\n",
       "    embed_document(root);\n",
       "  } else {\n",
       "    var attempts = 0;\n",
       "    var timer = setInterval(function(root) {\n",
       "      if (is_loaded(root)) {\n",
       "        clearInterval(timer);\n",
       "        embed_document(root);\n",
       "      } else if (document.readyState == \"complete\") {\n",
       "        attempts++;\n",
       "        if (attempts > 200) {\n",
       "          clearInterval(timer);\n",
       "\t  var Bokeh = get_bokeh(root)\n",
       "\t  if (Bokeh == null || Bokeh.Panel == null) {\n",
       "            console.warn(\"Panel: ERROR: Unable to run Panel code because Bokeh or Panel library is missing\");\n",
       "\t  } else {\n",
       "\t    console.warn(\"Panel: WARNING: Attempting to render but not all required libraries could be resolved.\")\n",
       "\t    embed_document(root)\n",
       "\t  }\n",
       "        }\n",
       "      }\n",
       "    }, 25, root)\n",
       "  }\n",
       "})(window);</script>"
      ]
     },
     "metadata": {
      "application/vnd.holoviews_exec.v0+json": {
       "id": "3f8fdd9c-53b4-4204-9ce7-100e5a7d3b23"
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import bdikit as bdi\n",
    "import pandas as pd\n",
    "from IPython.display import display, Markdown\n",
    "pd.set_option('display.max_rows', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_source = pd.read_csv(\"./datasets/Dou-ucec-discovery.csv\")\n",
    "df_target = pd.read_csv(\"./datasets/Dou-ucec-confirmatory.csv\")\n",
    "df_gt = pd.read_csv(\"./datasets/Dou-ucec-ground-truth.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_ground_truth = df_gt[~df_gt['target'].isna()]\n",
    "gt_set = df_ground_truth.set_index('source')['target'].to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def accuracy(gt_set, schema_mapping):\n",
    "    schema_dict = schema_mapping.set_index('source')['target'].to_dict()\n",
    "    correct_count = 0\n",
    "    total = 0\n",
    "    for source_column, target_column in  schema_dict.items():\n",
    "        if source_column in gt_set:\n",
    "            total += 1\n",
    "            correct_target_columns = set(gt_set[source_column].split(\",\"))\n",
    "            if target_column in correct_target_columns:\n",
    "                correct_count += 1\n",
    "                # print(f\"OK: {source_column} -> out={target_column} gt={correct_target_columns}\")\n",
    "            else:\n",
    "                print(f\"ER: {source_column} -> out={target_column} gt={correct_target_columns}\")\n",
    "    return correct_count/ float(total)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_source.columns = df_source.columns.str.replace(' ', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracting features from 179 columns...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "8a055f6f3aac4175b83cda61c3bd7ea6",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/179 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracting features from 213 columns...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "9761eef33a4547ffa457c7c0e1fcc6d8",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/213 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ER: Proteomics_Participant_ID -> out=Idx gt={'Case_id'}\n",
      "ER: Proteomics_TMT_batch -> out=Metformin_treatment gt={'Batch'}\n",
      "ER: Proteomics_TMT_plex -> out=Number_of_para-aortic_lymph_nodes_positive_for_tumor_by_he gt={'Plex'}\n",
      "ER: Proteomics_TMT_channel -> out=Number_of_para-aortic_lymph_nodes_positive_for_tumor_by_he gt={'ReporterName'}\n",
      "ER: Treatment_naive -> out=Follow-up_additional_treatment_radiation_therapy_for_new_tumor gt={'Cancer_history_history_of_any_treatment'}\n",
      "ER: Tumor_purity -> out=Tumor_necrosis gt={'ABSOLUTE_tumor_purity'}\n",
      "ER: MSH6 -> out=Ancillary_studies_msh2 gt={'Ancillary_studies_msh6'}\n",
      "ER: CIBERSORT_Monocytes -> out=Cibersort_Macrophage_M2 gt={'Cibersort_Monocyte'}\n",
      "ER: CIBERSORT_Eosinophils -> out=xCell_Cancer_associated_fibroblast gt={'Cibersort_Eosinophil'}\n",
      "ER: CIBERSORT_Neutrophils -> out=CNV_ratio gt={'Cibersort_Neutrophil'}\n",
      "ER: ESTIMATE_ImmuneScore -> out=Estimate_ESTIMATEScore gt={'Estimate_ImmuneScore'}\n",
      "Recall: 75.556\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>idx</td>\n",
       "      <td>xCell_T_cell_CD4+_Th1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Proteomics_Participant_ID</td>\n",
       "      <td>Idx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Case_excluded</td>\n",
       "      <td>Case_excluded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>Metformin_treatment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Proteomics_TMT_plex</td>\n",
       "      <td>Number_of_para-aortic_lymph_nodes_positive_for...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Proteomics_TMT_channel</td>\n",
       "      <td>Number_of_para-aortic_lymph_nodes_positive_for...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Proteomics_Parent_Sample_IDs</td>\n",
       "      <td>Idx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Proteomics_Aliquot_ID</td>\n",
       "      <td>Aliquot_ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Proteomics_Tumor_Normal</td>\n",
       "      <td>Group</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Proteomics_OCT</td>\n",
       "      <td>POLE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Country</td>\n",
       "      <td>Participant_country</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Histologic_Grade_FIGO</td>\n",
       "      <td>Histologic_grade</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Myometrial_invasion_Specify</td>\n",
       "      <td>Myometrial_invasion_present_specify</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>Histologic_Type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Treatment_naive</td>\n",
       "      <td>Follow-up_additional_treatment_radiation_thera...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Tumor_purity</td>\n",
       "      <td>Tumor_necrosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>Pathologic_staging_primary_tumor_pt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Path_Stage_Reg_Lymph_Nodes-pN</td>\n",
       "      <td>Pathologic_staging_regional_lymph_nodes_pn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Clin_Stage_Dist_Mets-cM</td>\n",
       "      <td>Clinical_staging_distant_metastasis_cm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Path_Stage_Dist_Mets-pM</td>\n",
       "      <td>Clinical_staging_distant_metastasis_cm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>tumor_Stage-Pathological</td>\n",
       "      <td>Tumor_stage_pathological</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>FIGO_stage</td>\n",
       "      <td>Tumor_stage_pathological</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>LVSI</td>\n",
       "      <td>Number_of_other_lymph_nodes_examined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>BMI</td>\n",
       "      <td>BMI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Age</td>\n",
       "      <td>Age</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Diabetes</td>\n",
       "      <td>Diabetes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Race</td>\n",
       "      <td>Race</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Ethnicity</td>\n",
       "      <td>Ethnicity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Gender</td>\n",
       "      <td>Sex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Tumor_Site</td>\n",
       "      <td>Tumor_site</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Tumor_Site_Other</td>\n",
       "      <td>Tumor_site_other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Tumor_Focality</td>\n",
       "      <td>Tumor_focality</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Tumor_Size_cm</td>\n",
       "      <td>Tumor_size_cm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Estrogen_Receptor</td>\n",
       "      <td>Ancillary_studies_estrogen_receptor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Estrogen_Receptor_%</td>\n",
       "      <td>Ancillary_studies_estrogen_receptor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Progesterone_Receptor</td>\n",
       "      <td>Ancillary_studies_progesterone_receptor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Progesterone_Receptor_%</td>\n",
       "      <td>Ancillary_studies_progesterone_receptor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>MLH1</td>\n",
       "      <td>Ancillary_studies_mlh1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>MLH2</td>\n",
       "      <td>Ancillary_studies_mlh1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>MSH6</td>\n",
       "      <td>Ancillary_studies_msh2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>PMS2</td>\n",
       "      <td>Ancillary_studies_pms2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>p53</td>\n",
       "      <td>Ancillary_studies_p53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>Other_IHC_specify</td>\n",
       "      <td>Ancillary_studies_other_immunohistochemistry_t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>MLH1_Promoter_Hypermethylation</td>\n",
       "      <td>Ancillary_studies_mlh1_promoter_hypermethylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>Num_full_term_pregnancies</td>\n",
       "      <td>Donor_information_number_of_full_term_pregnancies</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>EPIC_Bcells</td>\n",
       "      <td>Progeny_Androgen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>EPIC_CAFs</td>\n",
       "      <td>CNV_ratio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>EPIC_CD4_Tcells</td>\n",
       "      <td>Aliquot_ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>EPIC_CD8_Tcells</td>\n",
       "      <td>xCell_T_cell_CD4+_Th2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>EPIC_Endothelial</td>\n",
       "      <td>Progeny_Androgen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>EPIC_Macrophages</td>\n",
       "      <td>Cibersort_Macrophage_M2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>EPIC_NKcells</td>\n",
       "      <td>xCell_NK_cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>EPIC_otherCells</td>\n",
       "      <td>Estimate_ESTIMATEScore</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>CIBERSORT_B_cells_naive</td>\n",
       "      <td>Cibersort_T_cell_CD4+_memory_resting</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>CIBERSORT_B_cells_memory</td>\n",
       "      <td>xCell_Cancer_associated_fibroblast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>CIBERSORT_Plasma_cells</td>\n",
       "      <td>Progeny_PI3K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>CIBERSORT_T_cells_CD8</td>\n",
       "      <td>Cibersort_T_cell_CD4+_memory_resting</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>CIBERSORT_T_cells_CD4_naive</td>\n",
       "      <td>Cibersort_T_cell_CD4+_naive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>CIBERSORT_T_cells_CD4_memory_resting</td>\n",
       "      <td>CNV_ratio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>CIBERSORT_T_cells_CD4_memory_activated</td>\n",
       "      <td>xCell_T_cell_CD4+_Th2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>CIBERSORT_T_cells_follicular_helper</td>\n",
       "      <td>Cibersort_T_cell_CD4+_memory_resting</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>CIBERSORT_T_cells_regulatory_(Tregs)</td>\n",
       "      <td>Cibersort_T_cell_CD4+_memory_resting</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>CIBERSORT_T_cells_gamma_delta</td>\n",
       "      <td>xCell_T_cell_CD4+_Th2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>CIBERSORT_NK_cells_resting</td>\n",
       "      <td>Cibersort_NK_cell_activated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>CIBERSORT_NK_cells_activated</td>\n",
       "      <td>Cibersort_NK_cell_activated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>CIBERSORT_Monocytes</td>\n",
       "      <td>Cibersort_Macrophage_M2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>CIBERSORT_Macrophages_M0</td>\n",
       "      <td>Cibersort_Macrophage_M0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>CIBERSORT_Macrophages_M1</td>\n",
       "      <td>Cibersort_Macrophage_M1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>CIBERSORT_Macrophages_M2</td>\n",
       "      <td>Cibersort_Macrophage_M2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>CIBERSORT_Dendritic_cells_resting</td>\n",
       "      <td>xCell_Cancer_associated_fibroblast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>CIBERSORT_Dendritic_cells_activated</td>\n",
       "      <td>Cibersort_T_cell_CD4+_memory_resting</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>CIBERSORT_Mast_cells_resting</td>\n",
       "      <td>Cibersort_Mast_cell_activated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>CIBERSORT_Mast_cells_activated</td>\n",
       "      <td>Cibersort_Mast_cell_activated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>CIBERSORT_Eosinophils</td>\n",
       "      <td>xCell_Cancer_associated_fibroblast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>CIBERSORT_Neutrophils</td>\n",
       "      <td>CNV_ratio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>CIBERSORT_Absolute_score</td>\n",
       "      <td>Cibersort_T_cell_follicular_helper</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>ESTIMATE_StromalScore</td>\n",
       "      <td>Estimate_StromalScore</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>ESTIMATE_ImmuneScore</td>\n",
       "      <td>Estimate_ESTIMATEScore</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>ESTIMATE_ESTIMATEScore</td>\n",
       "      <td>Estimate_ESTIMATEScore</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>Stemness_score</td>\n",
       "      <td>Progeny_Androgen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>ER_ESR1</td>\n",
       "      <td>xCell_stroma_score</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>PR_PGR</td>\n",
       "      <td>Progeny_EGFR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>Pathway_activity_EGFR</td>\n",
       "      <td>Ancillary_studies_progesterone_receptor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>Pathway_activity_Hypoxia</td>\n",
       "      <td>xCell_immune_score</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>Pathway_activity_JAK.STAT</td>\n",
       "      <td>Mutation_signature_SBS5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>Pathway_activity_MAPK</td>\n",
       "      <td>Progeny_MAPK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>Pathway_activity_NFkB</td>\n",
       "      <td>Pathologic_staging_regional_lymph_nodes_pn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>Pathway_activity_PI3K</td>\n",
       "      <td>Pathologic_staging_regional_lymph_nodes_pn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>Pathway_activity_TGFb</td>\n",
       "      <td>ABSOLUTE_tumor_purity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>Pathway_activity_TNFa</td>\n",
       "      <td>Pathologic_staging_primary_tumor_pt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>Pathway_activity_Trail</td>\n",
       "      <td>Pathologic_staging_primary_tumor_pt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>Pathway_activity_VEGF</td>\n",
       "      <td>Progeny_VEGF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>Pathway_activity_p53</td>\n",
       "      <td>Pathologic_staging_regional_lymph_nodes_pn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>TP53_ATM</td>\n",
       "      <td>TP53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>TP53_CHEK2</td>\n",
       "      <td>Ancillary_studies_pms2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>TP53_MDM4</td>\n",
       "      <td>TP53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>TP53_RPS6KA3</td>\n",
       "      <td>Ancillary_studies_pms2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>TP53_TP53</td>\n",
       "      <td>TP53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>TP53_pathway</td>\n",
       "      <td>TP53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>PI3K_AKT1</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>PI3K_AKT2</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>PI3K_AKT3</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>PI3K_DEPDC5</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>PI3K_DEPTOR</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>PI3K_INPP4B</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>PI3K_MAPKAP1</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>PI3K_MLST8</td>\n",
       "      <td>Ancillary_studies_pms2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>PI3K_MTOR</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>PI3K_NPRL2</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>PI3K_NPRL3</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>PI3K_PDK1</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>PI3K_PIK3CA</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>PI3K_PIK3CB</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>PI3K_PIK3R1</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>PI3K_PIK3R2</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>PI3K_PPP2R1A</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>PI3K_PTEN</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>PI3K_RHEB</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>PI3K_RICTOR</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>PI3K_RPS6</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>PI3K_RPS6KB1</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>PI3K_RPTOR</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>PI3K_STK11</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>PI3K_TSC1</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>PI3K_TSC2</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>PI3K_pathway</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>HRD_BRCA1</td>\n",
       "      <td>ARID1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>HRD_BRCA2</td>\n",
       "      <td>ARID1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>HRD_BRCA1_or_BRCA2</td>\n",
       "      <td>Batch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>CNV_DEL</td>\n",
       "      <td>CNV_ratio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>CNV_AMP</td>\n",
       "      <td>CNV_ratio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>CNV_class</td>\n",
       "      <td>CNV_status</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>CNV_idx</td>\n",
       "      <td>CNV_ratio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>CNV_1q_DEL</td>\n",
       "      <td>CNV_ratio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>CNV_3q_DEL</td>\n",
       "      <td>CNV_ratio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>CNV_4q_DEL</td>\n",
       "      <td>CNV_ratio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>CNV_1q_AMP</td>\n",
       "      <td>CNV_ratio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>CNV_3q_AMP</td>\n",
       "      <td>CNV_ratio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>CNV_4q_AMP</td>\n",
       "      <td>CNV_ratio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>Purity_Immune</td>\n",
       "      <td>Progeny_Androgen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>Purity_Cancer</td>\n",
       "      <td>Progeny_Hypoxia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>Purity_Stroma</td>\n",
       "      <td>Progeny_Androgen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>MSI_status</td>\n",
       "      <td>MSI_status</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>POLE_subtype</td>\n",
       "      <td>POLE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>JAK1_MS_INDEL</td>\n",
       "      <td>MSI_status</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>JAK1_Mutation</td>\n",
       "      <td>Clinical_staging_distant_metastasis_cm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>Log2_variant_per_Mbp</td>\n",
       "      <td>Progeny_MAPK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>Log2_SNP_per_Mbp</td>\n",
       "      <td>Progeny_JAK.STAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>Log2_INDEL_per_Mbp</td>\n",
       "      <td>Cibersort_Macrophage_M0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>Log2_variant_total</td>\n",
       "      <td>Estimate_ESTIMATEScore</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>Log2_SNP_total</td>\n",
       "      <td>CNV_ratio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>Log2_INDEL_total</td>\n",
       "      <td>Estimate_ESTIMATEScore</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>Genomics_subtype</td>\n",
       "      <td>Genomic_subtype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>Mutation_signature_C&gt;A</td>\n",
       "      <td>CNV_ratio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>Mutation_signature_C&gt;G</td>\n",
       "      <td>Cibersort_Macrophage_M2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>Mutation_signature_C&gt;T</td>\n",
       "      <td>Estimate_ImmuneScore</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>Mutation_signature_T&gt;C</td>\n",
       "      <td>Cibersort_Macrophage_M2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>Mutation_signature_T&gt;A</td>\n",
       "      <td>Progeny_Hypoxia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>Mutation_signature_T&gt;G</td>\n",
       "      <td>Progeny_TGFb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>WXS_normal_sample_type</td>\n",
       "      <td>Ancillary_studies_other_immunohistochemistry_p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>WXS_normal_filename</td>\n",
       "      <td>Progeny_WNT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>WXS_normal_UUID</td>\n",
       "      <td>Progeny_WNT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>WXS_tumor_sample_type</td>\n",
       "      <td>Tumor_necrosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163</th>\n",
       "      <td>WXS_tumor_filename</td>\n",
       "      <td>Progeny_TGFb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>WXS_tumor_UUID</td>\n",
       "      <td>Progeny_WNT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>WGS_normal_sample_type</td>\n",
       "      <td>Group</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>WGS_normal_UUID</td>\n",
       "      <td>Progeny_WNT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>WGS_tumor_sample_type</td>\n",
       "      <td>Tumor_site</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>WGS_tumor_UUID</td>\n",
       "      <td>Progeny_TGFb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>RNAseq_R1_sample_type</td>\n",
       "      <td>ARID1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>RNAseq_R1_filename</td>\n",
       "      <td>Aliquot_ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>RNAseq_R1_UUID</td>\n",
       "      <td>Aliquot_ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>RNAseq_R2_sample_type</td>\n",
       "      <td>ARID1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>RNAseq_R2_filename</td>\n",
       "      <td>Aliquot_ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>RNAseq_R2_UUID</td>\n",
       "      <td>Aliquot_ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>miRNAseq_sample_type</td>\n",
       "      <td>Mutation_signature_SBS7a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>miRNAseq_UUID</td>\n",
       "      <td>Batch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177</th>\n",
       "      <td>Methylation_available</td>\n",
       "      <td>Group</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>Methylation_quality</td>\n",
       "      <td>MSI_status</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     source  \\\n",
       "0                                       idx   \n",
       "1                 Proteomics_Participant_ID   \n",
       "2                             Case_excluded   \n",
       "3                      Proteomics_TMT_batch   \n",
       "4                       Proteomics_TMT_plex   \n",
       "5                    Proteomics_TMT_channel   \n",
       "6              Proteomics_Parent_Sample_IDs   \n",
       "7                     Proteomics_Aliquot_ID   \n",
       "8                   Proteomics_Tumor_Normal   \n",
       "9                            Proteomics_OCT   \n",
       "10                                  Country   \n",
       "11                    Histologic_Grade_FIGO   \n",
       "12              Myometrial_invasion_Specify   \n",
       "13                          Histologic_type   \n",
       "14                          Treatment_naive   \n",
       "15                             Tumor_purity   \n",
       "16              Path_Stage_Primary_Tumor-pT   \n",
       "17            Path_Stage_Reg_Lymph_Nodes-pN   \n",
       "18                  Clin_Stage_Dist_Mets-cM   \n",
       "19                  Path_Stage_Dist_Mets-pM   \n",
       "20                 tumor_Stage-Pathological   \n",
       "21                               FIGO_stage   \n",
       "22                                     LVSI   \n",
       "23                                      BMI   \n",
       "24                                      Age   \n",
       "25                                 Diabetes   \n",
       "26                                     Race   \n",
       "27                                Ethnicity   \n",
       "28                                   Gender   \n",
       "29                               Tumor_Site   \n",
       "30                         Tumor_Site_Other   \n",
       "31                           Tumor_Focality   \n",
       "32                            Tumor_Size_cm   \n",
       "33                        Estrogen_Receptor   \n",
       "34                      Estrogen_Receptor_%   \n",
       "35                    Progesterone_Receptor   \n",
       "36                  Progesterone_Receptor_%   \n",
       "37                                     MLH1   \n",
       "38                                     MLH2   \n",
       "39                                     MSH6   \n",
       "40                                     PMS2   \n",
       "41                                      p53   \n",
       "42                        Other_IHC_specify   \n",
       "43           MLH1_Promoter_Hypermethylation   \n",
       "44                Num_full_term_pregnancies   \n",
       "45                              EPIC_Bcells   \n",
       "46                                EPIC_CAFs   \n",
       "47                          EPIC_CD4_Tcells   \n",
       "48                          EPIC_CD8_Tcells   \n",
       "49                         EPIC_Endothelial   \n",
       "50                         EPIC_Macrophages   \n",
       "51                             EPIC_NKcells   \n",
       "52                          EPIC_otherCells   \n",
       "53                  CIBERSORT_B_cells_naive   \n",
       "54                 CIBERSORT_B_cells_memory   \n",
       "55                   CIBERSORT_Plasma_cells   \n",
       "56                    CIBERSORT_T_cells_CD8   \n",
       "57              CIBERSORT_T_cells_CD4_naive   \n",
       "58     CIBERSORT_T_cells_CD4_memory_resting   \n",
       "59   CIBERSORT_T_cells_CD4_memory_activated   \n",
       "60      CIBERSORT_T_cells_follicular_helper   \n",
       "61     CIBERSORT_T_cells_regulatory_(Tregs)   \n",
       "62            CIBERSORT_T_cells_gamma_delta   \n",
       "63               CIBERSORT_NK_cells_resting   \n",
       "64             CIBERSORT_NK_cells_activated   \n",
       "65                      CIBERSORT_Monocytes   \n",
       "66                 CIBERSORT_Macrophages_M0   \n",
       "67                 CIBERSORT_Macrophages_M1   \n",
       "68                 CIBERSORT_Macrophages_M2   \n",
       "69        CIBERSORT_Dendritic_cells_resting   \n",
       "70      CIBERSORT_Dendritic_cells_activated   \n",
       "71             CIBERSORT_Mast_cells_resting   \n",
       "72           CIBERSORT_Mast_cells_activated   \n",
       "73                    CIBERSORT_Eosinophils   \n",
       "74                    CIBERSORT_Neutrophils   \n",
       "75                 CIBERSORT_Absolute_score   \n",
       "76                    ESTIMATE_StromalScore   \n",
       "77                     ESTIMATE_ImmuneScore   \n",
       "78                   ESTIMATE_ESTIMATEScore   \n",
       "79                           Stemness_score   \n",
       "80                                  ER_ESR1   \n",
       "81                                   PR_PGR   \n",
       "82                    Pathway_activity_EGFR   \n",
       "83                 Pathway_activity_Hypoxia   \n",
       "84                Pathway_activity_JAK.STAT   \n",
       "85                    Pathway_activity_MAPK   \n",
       "86                    Pathway_activity_NFkB   \n",
       "87                    Pathway_activity_PI3K   \n",
       "88                    Pathway_activity_TGFb   \n",
       "89                    Pathway_activity_TNFa   \n",
       "90                   Pathway_activity_Trail   \n",
       "91                    Pathway_activity_VEGF   \n",
       "92                     Pathway_activity_p53   \n",
       "93                                 TP53_ATM   \n",
       "94                               TP53_CHEK2   \n",
       "95                                TP53_MDM4   \n",
       "96                             TP53_RPS6KA3   \n",
       "97                                TP53_TP53   \n",
       "98                             TP53_pathway   \n",
       "99                                PI3K_AKT1   \n",
       "100                               PI3K_AKT2   \n",
       "101                               PI3K_AKT3   \n",
       "102                             PI3K_DEPDC5   \n",
       "103                             PI3K_DEPTOR   \n",
       "104                             PI3K_INPP4B   \n",
       "105                            PI3K_MAPKAP1   \n",
       "106                              PI3K_MLST8   \n",
       "107                               PI3K_MTOR   \n",
       "108                              PI3K_NPRL2   \n",
       "109                              PI3K_NPRL3   \n",
       "110                               PI3K_PDK1   \n",
       "111                             PI3K_PIK3CA   \n",
       "112                             PI3K_PIK3CB   \n",
       "113                             PI3K_PIK3R1   \n",
       "114                             PI3K_PIK3R2   \n",
       "115                            PI3K_PPP2R1A   \n",
       "116                               PI3K_PTEN   \n",
       "117                               PI3K_RHEB   \n",
       "118                             PI3K_RICTOR   \n",
       "119                               PI3K_RPS6   \n",
       "120                            PI3K_RPS6KB1   \n",
       "121                              PI3K_RPTOR   \n",
       "122                              PI3K_STK11   \n",
       "123                               PI3K_TSC1   \n",
       "124                               PI3K_TSC2   \n",
       "125                            PI3K_pathway   \n",
       "126                               HRD_BRCA1   \n",
       "127                               HRD_BRCA2   \n",
       "128                      HRD_BRCA1_or_BRCA2   \n",
       "129                                 CNV_DEL   \n",
       "130                                 CNV_AMP   \n",
       "131                               CNV_class   \n",
       "132                                 CNV_idx   \n",
       "133                              CNV_1q_DEL   \n",
       "134                              CNV_3q_DEL   \n",
       "135                              CNV_4q_DEL   \n",
       "136                              CNV_1q_AMP   \n",
       "137                              CNV_3q_AMP   \n",
       "138                              CNV_4q_AMP   \n",
       "139                           Purity_Immune   \n",
       "140                           Purity_Cancer   \n",
       "141                           Purity_Stroma   \n",
       "142                              MSI_status   \n",
       "143                            POLE_subtype   \n",
       "144                           JAK1_MS_INDEL   \n",
       "145                           JAK1_Mutation   \n",
       "146                    Log2_variant_per_Mbp   \n",
       "147                        Log2_SNP_per_Mbp   \n",
       "148                      Log2_INDEL_per_Mbp   \n",
       "149                      Log2_variant_total   \n",
       "150                          Log2_SNP_total   \n",
       "151                        Log2_INDEL_total   \n",
       "152                        Genomics_subtype   \n",
       "153                  Mutation_signature_C>A   \n",
       "154                  Mutation_signature_C>G   \n",
       "155                  Mutation_signature_C>T   \n",
       "156                  Mutation_signature_T>C   \n",
       "157                  Mutation_signature_T>A   \n",
       "158                  Mutation_signature_T>G   \n",
       "159                  WXS_normal_sample_type   \n",
       "160                     WXS_normal_filename   \n",
       "161                         WXS_normal_UUID   \n",
       "162                   WXS_tumor_sample_type   \n",
       "163                      WXS_tumor_filename   \n",
       "164                          WXS_tumor_UUID   \n",
       "165                  WGS_normal_sample_type   \n",
       "166                         WGS_normal_UUID   \n",
       "167                   WGS_tumor_sample_type   \n",
       "168                          WGS_tumor_UUID   \n",
       "169                   RNAseq_R1_sample_type   \n",
       "170                      RNAseq_R1_filename   \n",
       "171                          RNAseq_R1_UUID   \n",
       "172                   RNAseq_R2_sample_type   \n",
       "173                      RNAseq_R2_filename   \n",
       "174                          RNAseq_R2_UUID   \n",
       "175                    miRNAseq_sample_type   \n",
       "176                           miRNAseq_UUID   \n",
       "177                   Methylation_available   \n",
       "178                     Methylation_quality   \n",
       "\n",
       "                                                target  \n",
       "0                                xCell_T_cell_CD4+_Th1  \n",
       "1                                                  Idx  \n",
       "2                                        Case_excluded  \n",
       "3                                  Metformin_treatment  \n",
       "4    Number_of_para-aortic_lymph_nodes_positive_for...  \n",
       "5    Number_of_para-aortic_lymph_nodes_positive_for...  \n",
       "6                                                  Idx  \n",
       "7                                           Aliquot_ID  \n",
       "8                                                Group  \n",
       "9                                                 POLE  \n",
       "10                                 Participant_country  \n",
       "11                                    Histologic_grade  \n",
       "12                 Myometrial_invasion_present_specify  \n",
       "13                                     Histologic_Type  \n",
       "14   Follow-up_additional_treatment_radiation_thera...  \n",
       "15                                      Tumor_necrosis  \n",
       "16                 Pathologic_staging_primary_tumor_pt  \n",
       "17          Pathologic_staging_regional_lymph_nodes_pn  \n",
       "18              Clinical_staging_distant_metastasis_cm  \n",
       "19              Clinical_staging_distant_metastasis_cm  \n",
       "20                            Tumor_stage_pathological  \n",
       "21                            Tumor_stage_pathological  \n",
       "22                Number_of_other_lymph_nodes_examined  \n",
       "23                                                 BMI  \n",
       "24                                                 Age  \n",
       "25                                            Diabetes  \n",
       "26                                                Race  \n",
       "27                                           Ethnicity  \n",
       "28                                                 Sex  \n",
       "29                                          Tumor_site  \n",
       "30                                    Tumor_site_other  \n",
       "31                                      Tumor_focality  \n",
       "32                                       Tumor_size_cm  \n",
       "33                 Ancillary_studies_estrogen_receptor  \n",
       "34                 Ancillary_studies_estrogen_receptor  \n",
       "35             Ancillary_studies_progesterone_receptor  \n",
       "36             Ancillary_studies_progesterone_receptor  \n",
       "37                              Ancillary_studies_mlh1  \n",
       "38                              Ancillary_studies_mlh1  \n",
       "39                              Ancillary_studies_msh2  \n",
       "40                              Ancillary_studies_pms2  \n",
       "41                               Ancillary_studies_p53  \n",
       "42   Ancillary_studies_other_immunohistochemistry_t...  \n",
       "43    Ancillary_studies_mlh1_promoter_hypermethylation  \n",
       "44   Donor_information_number_of_full_term_pregnancies  \n",
       "45                                    Progeny_Androgen  \n",
       "46                                           CNV_ratio  \n",
       "47                                          Aliquot_ID  \n",
       "48                               xCell_T_cell_CD4+_Th2  \n",
       "49                                    Progeny_Androgen  \n",
       "50                             Cibersort_Macrophage_M2  \n",
       "51                                       xCell_NK_cell  \n",
       "52                              Estimate_ESTIMATEScore  \n",
       "53                Cibersort_T_cell_CD4+_memory_resting  \n",
       "54                  xCell_Cancer_associated_fibroblast  \n",
       "55                                        Progeny_PI3K  \n",
       "56                Cibersort_T_cell_CD4+_memory_resting  \n",
       "57                         Cibersort_T_cell_CD4+_naive  \n",
       "58                                           CNV_ratio  \n",
       "59                               xCell_T_cell_CD4+_Th2  \n",
       "60                Cibersort_T_cell_CD4+_memory_resting  \n",
       "61                Cibersort_T_cell_CD4+_memory_resting  \n",
       "62                               xCell_T_cell_CD4+_Th2  \n",
       "63                         Cibersort_NK_cell_activated  \n",
       "64                         Cibersort_NK_cell_activated  \n",
       "65                             Cibersort_Macrophage_M2  \n",
       "66                             Cibersort_Macrophage_M0  \n",
       "67                             Cibersort_Macrophage_M1  \n",
       "68                             Cibersort_Macrophage_M2  \n",
       "69                  xCell_Cancer_associated_fibroblast  \n",
       "70                Cibersort_T_cell_CD4+_memory_resting  \n",
       "71                       Cibersort_Mast_cell_activated  \n",
       "72                       Cibersort_Mast_cell_activated  \n",
       "73                  xCell_Cancer_associated_fibroblast  \n",
       "74                                           CNV_ratio  \n",
       "75                  Cibersort_T_cell_follicular_helper  \n",
       "76                               Estimate_StromalScore  \n",
       "77                              Estimate_ESTIMATEScore  \n",
       "78                              Estimate_ESTIMATEScore  \n",
       "79                                    Progeny_Androgen  \n",
       "80                                  xCell_stroma_score  \n",
       "81                                        Progeny_EGFR  \n",
       "82             Ancillary_studies_progesterone_receptor  \n",
       "83                                  xCell_immune_score  \n",
       "84                             Mutation_signature_SBS5  \n",
       "85                                        Progeny_MAPK  \n",
       "86          Pathologic_staging_regional_lymph_nodes_pn  \n",
       "87          Pathologic_staging_regional_lymph_nodes_pn  \n",
       "88                               ABSOLUTE_tumor_purity  \n",
       "89                 Pathologic_staging_primary_tumor_pt  \n",
       "90                 Pathologic_staging_primary_tumor_pt  \n",
       "91                                        Progeny_VEGF  \n",
       "92          Pathologic_staging_regional_lymph_nodes_pn  \n",
       "93                                                TP53  \n",
       "94                              Ancillary_studies_pms2  \n",
       "95                                                TP53  \n",
       "96                              Ancillary_studies_pms2  \n",
       "97                                                TP53  \n",
       "98                                                TP53  \n",
       "99                                              PIK3CA  \n",
       "100                                             PIK3CA  \n",
       "101                                             PIK3CA  \n",
       "102                                             PIK3CA  \n",
       "103                                             PIK3CA  \n",
       "104                                             PIK3CA  \n",
       "105                                             PIK3CA  \n",
       "106                             Ancillary_studies_pms2  \n",
       "107                                             PIK3CA  \n",
       "108                                             PIK3CA  \n",
       "109                                             PIK3CA  \n",
       "110                                             PIK3CA  \n",
       "111                                             PIK3CA  \n",
       "112                                             PIK3CA  \n",
       "113                                             PIK3CA  \n",
       "114                                             PIK3CA  \n",
       "115                                             PIK3CA  \n",
       "116                                             PIK3CA  \n",
       "117                                             PIK3CA  \n",
       "118                                             PIK3CA  \n",
       "119                                             PIK3CA  \n",
       "120                                             PIK3CA  \n",
       "121                                             PIK3CA  \n",
       "122                                             PIK3CA  \n",
       "123                                             PIK3CA  \n",
       "124                                             PIK3CA  \n",
       "125                                             PIK3CA  \n",
       "126                                             ARID1A  \n",
       "127                                             ARID1A  \n",
       "128                                              Batch  \n",
       "129                                          CNV_ratio  \n",
       "130                                          CNV_ratio  \n",
       "131                                         CNV_status  \n",
       "132                                          CNV_ratio  \n",
       "133                                          CNV_ratio  \n",
       "134                                          CNV_ratio  \n",
       "135                                          CNV_ratio  \n",
       "136                                          CNV_ratio  \n",
       "137                                          CNV_ratio  \n",
       "138                                          CNV_ratio  \n",
       "139                                   Progeny_Androgen  \n",
       "140                                    Progeny_Hypoxia  \n",
       "141                                   Progeny_Androgen  \n",
       "142                                         MSI_status  \n",
       "143                                               POLE  \n",
       "144                                         MSI_status  \n",
       "145             Clinical_staging_distant_metastasis_cm  \n",
       "146                                       Progeny_MAPK  \n",
       "147                                   Progeny_JAK.STAT  \n",
       "148                            Cibersort_Macrophage_M0  \n",
       "149                             Estimate_ESTIMATEScore  \n",
       "150                                          CNV_ratio  \n",
       "151                             Estimate_ESTIMATEScore  \n",
       "152                                    Genomic_subtype  \n",
       "153                                          CNV_ratio  \n",
       "154                            Cibersort_Macrophage_M2  \n",
       "155                               Estimate_ImmuneScore  \n",
       "156                            Cibersort_Macrophage_M2  \n",
       "157                                    Progeny_Hypoxia  \n",
       "158                                       Progeny_TGFb  \n",
       "159  Ancillary_studies_other_immunohistochemistry_p...  \n",
       "160                                        Progeny_WNT  \n",
       "161                                        Progeny_WNT  \n",
       "162                                     Tumor_necrosis  \n",
       "163                                       Progeny_TGFb  \n",
       "164                                        Progeny_WNT  \n",
       "165                                              Group  \n",
       "166                                        Progeny_WNT  \n",
       "167                                         Tumor_site  \n",
       "168                                       Progeny_TGFb  \n",
       "169                                             ARID1A  \n",
       "170                                         Aliquot_ID  \n",
       "171                                         Aliquot_ID  \n",
       "172                                             ARID1A  \n",
       "173                                         Aliquot_ID  \n",
       "174                                         Aliquot_ID  \n",
       "175                           Mutation_signature_SBS7a  \n",
       "176                                              Batch  \n",
       "177                                              Group  \n",
       "178                                         MSI_status  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "schema_mapping = bdi.match_schema(df_source, df_target, method='ct_learning')\n",
    "print(f\"Recall: {100*accuracy(gt_set, schema_mapping):.3f}\")\n",
    "schema_mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracting features from 179 columns...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "fe32ed0bab28433ea1383cb720417793",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/179 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracting features from 213 columns...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "791829c5722e45f9b247e06cee321d87",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/213 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ER: Proteomics_TMT_batch -> out=Metformin_treatment gt={'Batch'}\n",
      "ER: Proteomics_TMT_plex -> out=Number_of_para-aortic_lymph_nodes_positive_for_tumor_by_he gt={'Plex'}\n",
      "ER: Treatment_naive -> out=Follow-up_additional_treatment_radiation_therapy_for_new_tumor gt={'Cancer_history_history_of_any_treatment'}\n",
      "ER: Tumor_purity -> out=Tumor_necrosis gt={'ABSOLUTE_tumor_purity'}\n",
      "ER: MSH6 -> out=Ancillary_studies_msh2 gt={'Ancillary_studies_msh6'}\n",
      "ER: CIBERSORT_Monocytes -> out=Cibersort_Macrophage_M2 gt={'Cibersort_Monocyte'}\n",
      "ER: CIBERSORT_Eosinophils -> out=xCell_Cancer_associated_fibroblast gt={'Cibersort_Eosinophil'}\n",
      "ER: CIBERSORT_Neutrophils -> out=CNV_ratio gt={'Cibersort_Neutrophil'}\n",
      "ER: ESTIMATE_ImmuneScore -> out=Estimate_ESTIMATEScore gt={'Estimate_ImmuneScore'}\n",
      "Recall: 80.000\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>idx</td>\n",
       "      <td>Number_of_para-aortic_lymph_nodes_examined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Proteomics_Participant_ID</td>\n",
       "      <td>Case_id</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Case_excluded</td>\n",
       "      <td>Case_excluded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>Metformin_treatment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Proteomics_TMT_plex</td>\n",
       "      <td>Number_of_para-aortic_lymph_nodes_positive_for...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Proteomics_TMT_channel</td>\n",
       "      <td>ReporterName</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Proteomics_Parent_Sample_IDs</td>\n",
       "      <td>Case_id</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Proteomics_Aliquot_ID</td>\n",
       "      <td>Aliquot_ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Proteomics_Tumor_Normal</td>\n",
       "      <td>Group</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Proteomics_OCT</td>\n",
       "      <td>POLE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Country</td>\n",
       "      <td>Participant_country</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Histologic_Grade_FIGO</td>\n",
       "      <td>Histologic_grade</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Myometrial_invasion_Specify</td>\n",
       "      <td>Myometrial_invasion_present_specify</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>Histologic_Type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Treatment_naive</td>\n",
       "      <td>Follow-up_additional_treatment_radiation_thera...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Tumor_purity</td>\n",
       "      <td>Tumor_necrosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>Pathologic_staging_primary_tumor_pt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Path_Stage_Reg_Lymph_Nodes-pN</td>\n",
       "      <td>Pathologic_staging_regional_lymph_nodes_pn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Clin_Stage_Dist_Mets-cM</td>\n",
       "      <td>Clinical_staging_distant_metastasis_cm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Path_Stage_Dist_Mets-pM</td>\n",
       "      <td>Clinical_staging_distant_metastasis_cm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>tumor_Stage-Pathological</td>\n",
       "      <td>Tumor_stage_pathological</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>FIGO_stage</td>\n",
       "      <td>Tumor_stage_pathological</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>LVSI</td>\n",
       "      <td>Number_of_other_lymph_nodes_examined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>BMI</td>\n",
       "      <td>BMI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Age</td>\n",
       "      <td>Age</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Diabetes</td>\n",
       "      <td>Diabetes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Race</td>\n",
       "      <td>Race</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Ethnicity</td>\n",
       "      <td>Ethnicity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Gender</td>\n",
       "      <td>Sex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Tumor_Site</td>\n",
       "      <td>Tumor_site</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Tumor_Site_Other</td>\n",
       "      <td>Tumor_site_other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Tumor_Focality</td>\n",
       "      <td>Tumor_focality</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Tumor_Size_cm</td>\n",
       "      <td>Tumor_size_cm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Estrogen_Receptor</td>\n",
       "      <td>Ancillary_studies_estrogen_receptor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Estrogen_Receptor_%</td>\n",
       "      <td>Ancillary_studies_estrogen_receptor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Progesterone_Receptor</td>\n",
       "      <td>Ancillary_studies_progesterone_receptor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Progesterone_Receptor_%</td>\n",
       "      <td>Ancillary_studies_progesterone_receptor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>MLH1</td>\n",
       "      <td>Ancillary_studies_mlh1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>MLH2</td>\n",
       "      <td>Ancillary_studies_mlh1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>MSH6</td>\n",
       "      <td>Ancillary_studies_msh2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>PMS2</td>\n",
       "      <td>Ancillary_studies_pms2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>p53</td>\n",
       "      <td>Ancillary_studies_p53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>Other_IHC_specify</td>\n",
       "      <td>Ancillary_studies_other_immunohistochemistry_t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>MLH1_Promoter_Hypermethylation</td>\n",
       "      <td>Ancillary_studies_mlh1_promoter_hypermethylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>Num_full_term_pregnancies</td>\n",
       "      <td>Donor_information_number_of_full_term_pregnancies</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>EPIC_Bcells</td>\n",
       "      <td>Progeny_Androgen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>EPIC_CAFs</td>\n",
       "      <td>CNV_ratio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>EPIC_CD4_Tcells</td>\n",
       "      <td>Aliquot_ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>EPIC_CD8_Tcells</td>\n",
       "      <td>xCell_T_cell_CD4+_Th2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>EPIC_Endothelial</td>\n",
       "      <td>Progeny_Androgen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>EPIC_Macrophages</td>\n",
       "      <td>Cibersort_Macrophage_M2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>EPIC_NKcells</td>\n",
       "      <td>xCell_NK_cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>EPIC_otherCells</td>\n",
       "      <td>Estimate_ESTIMATEScore</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>CIBERSORT_B_cells_naive</td>\n",
       "      <td>Cibersort_T_cell_CD4+_memory_resting</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>CIBERSORT_B_cells_memory</td>\n",
       "      <td>xCell_Cancer_associated_fibroblast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>CIBERSORT_Plasma_cells</td>\n",
       "      <td>Progeny_PI3K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>CIBERSORT_T_cells_CD8</td>\n",
       "      <td>Cibersort_T_cell_CD4+_memory_resting</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>CIBERSORT_T_cells_CD4_naive</td>\n",
       "      <td>Cibersort_T_cell_CD4+_naive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>CIBERSORT_T_cells_CD4_memory_resting</td>\n",
       "      <td>CNV_ratio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>CIBERSORT_T_cells_CD4_memory_activated</td>\n",
       "      <td>xCell_T_cell_CD4+_Th2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>CIBERSORT_T_cells_follicular_helper</td>\n",
       "      <td>Cibersort_T_cell_CD4+_memory_resting</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>CIBERSORT_T_cells_regulatory_(Tregs)</td>\n",
       "      <td>Cibersort_T_cell_CD4+_memory_resting</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>CIBERSORT_T_cells_gamma_delta</td>\n",
       "      <td>xCell_T_cell_CD4+_Th2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>CIBERSORT_NK_cells_resting</td>\n",
       "      <td>Cibersort_NK_cell_activated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>CIBERSORT_NK_cells_activated</td>\n",
       "      <td>Cibersort_NK_cell_activated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>CIBERSORT_Monocytes</td>\n",
       "      <td>Cibersort_Macrophage_M2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>CIBERSORT_Macrophages_M0</td>\n",
       "      <td>Cibersort_Macrophage_M0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>CIBERSORT_Macrophages_M1</td>\n",
       "      <td>Cibersort_Macrophage_M1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>CIBERSORT_Macrophages_M2</td>\n",
       "      <td>Cibersort_Macrophage_M2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>CIBERSORT_Dendritic_cells_resting</td>\n",
       "      <td>xCell_Cancer_associated_fibroblast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>CIBERSORT_Dendritic_cells_activated</td>\n",
       "      <td>Cibersort_T_cell_CD4+_memory_resting</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>CIBERSORT_Mast_cells_resting</td>\n",
       "      <td>Cibersort_Mast_cell_activated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>CIBERSORT_Mast_cells_activated</td>\n",
       "      <td>Cibersort_Mast_cell_activated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>CIBERSORT_Eosinophils</td>\n",
       "      <td>xCell_Cancer_associated_fibroblast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>CIBERSORT_Neutrophils</td>\n",
       "      <td>CNV_ratio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>CIBERSORT_Absolute_score</td>\n",
       "      <td>Cibersort_T_cell_follicular_helper</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>ESTIMATE_StromalScore</td>\n",
       "      <td>Estimate_StromalScore</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>ESTIMATE_ImmuneScore</td>\n",
       "      <td>Estimate_ESTIMATEScore</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>ESTIMATE_ESTIMATEScore</td>\n",
       "      <td>Estimate_ESTIMATEScore</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>Stemness_score</td>\n",
       "      <td>Progeny_Androgen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>ER_ESR1</td>\n",
       "      <td>xCell_stroma_score</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>PR_PGR</td>\n",
       "      <td>Progeny_EGFR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>Pathway_activity_EGFR</td>\n",
       "      <td>Ancillary_studies_progesterone_receptor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>Pathway_activity_Hypoxia</td>\n",
       "      <td>xCell_immune_score</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>Pathway_activity_JAK.STAT</td>\n",
       "      <td>Mutation_signature_SBS5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>Pathway_activity_MAPK</td>\n",
       "      <td>Progeny_MAPK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>Pathway_activity_NFkB</td>\n",
       "      <td>Pathologic_staging_regional_lymph_nodes_pn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>Pathway_activity_PI3K</td>\n",
       "      <td>Pathologic_staging_regional_lymph_nodes_pn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>Pathway_activity_TGFb</td>\n",
       "      <td>ABSOLUTE_tumor_purity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>Pathway_activity_TNFa</td>\n",
       "      <td>Pathologic_staging_primary_tumor_pt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>Pathway_activity_Trail</td>\n",
       "      <td>Pathologic_staging_primary_tumor_pt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>Pathway_activity_VEGF</td>\n",
       "      <td>Progeny_VEGF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>Pathway_activity_p53</td>\n",
       "      <td>Pathologic_staging_regional_lymph_nodes_pn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>TP53_ATM</td>\n",
       "      <td>TP53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>TP53_CHEK2</td>\n",
       "      <td>Ancillary_studies_pms2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>TP53_MDM4</td>\n",
       "      <td>TP53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>TP53_RPS6KA3</td>\n",
       "      <td>Ancillary_studies_pms2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>TP53_TP53</td>\n",
       "      <td>TP53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>TP53_pathway</td>\n",
       "      <td>TP53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>PI3K_AKT1</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>PI3K_AKT2</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>PI3K_AKT3</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>PI3K_DEPDC5</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>PI3K_DEPTOR</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>PI3K_INPP4B</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>PI3K_MAPKAP1</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>PI3K_MLST8</td>\n",
       "      <td>Ancillary_studies_pms2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>PI3K_MTOR</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>PI3K_NPRL2</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>PI3K_NPRL3</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>PI3K_PDK1</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>PI3K_PIK3CA</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>PI3K_PIK3CB</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>PI3K_PIK3R1</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>PI3K_PIK3R2</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>PI3K_PPP2R1A</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>PI3K_PTEN</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>PI3K_RHEB</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>PI3K_RICTOR</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>PI3K_RPS6</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>PI3K_RPS6KB1</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>PI3K_RPTOR</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>PI3K_STK11</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>PI3K_TSC1</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>PI3K_TSC2</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>PI3K_pathway</td>\n",
       "      <td>PIK3CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>HRD_BRCA1</td>\n",
       "      <td>ARID1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>HRD_BRCA2</td>\n",
       "      <td>ARID1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>HRD_BRCA1_or_BRCA2</td>\n",
       "      <td>Batch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>CNV_DEL</td>\n",
       "      <td>CNV_ratio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>CNV_AMP</td>\n",
       "      <td>CNV_ratio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>CNV_class</td>\n",
       "      <td>CNV_status</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>CNV_idx</td>\n",
       "      <td>CNV_ratio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>CNV_1q_DEL</td>\n",
       "      <td>CNV_ratio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>CNV_3q_DEL</td>\n",
       "      <td>CNV_ratio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>CNV_4q_DEL</td>\n",
       "      <td>CNV_ratio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>CNV_1q_AMP</td>\n",
       "      <td>CNV_ratio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>CNV_3q_AMP</td>\n",
       "      <td>CNV_ratio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>CNV_4q_AMP</td>\n",
       "      <td>CNV_ratio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>Purity_Immune</td>\n",
       "      <td>Progeny_Androgen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>Purity_Cancer</td>\n",
       "      <td>Progeny_Hypoxia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>Purity_Stroma</td>\n",
       "      <td>Progeny_Androgen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>MSI_status</td>\n",
       "      <td>MSI_status</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>POLE_subtype</td>\n",
       "      <td>POLE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>JAK1_MS_INDEL</td>\n",
       "      <td>MSI_status</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>JAK1_Mutation</td>\n",
       "      <td>Clinical_staging_distant_metastasis_cm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>Log2_variant_per_Mbp</td>\n",
       "      <td>Progeny_MAPK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>Log2_SNP_per_Mbp</td>\n",
       "      <td>Progeny_JAK.STAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>Log2_INDEL_per_Mbp</td>\n",
       "      <td>Cibersort_Macrophage_M0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>Log2_variant_total</td>\n",
       "      <td>Estimate_ESTIMATEScore</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>Log2_SNP_total</td>\n",
       "      <td>CNV_ratio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>Log2_INDEL_total</td>\n",
       "      <td>Estimate_ESTIMATEScore</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>Genomics_subtype</td>\n",
       "      <td>Genomic_subtype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>Mutation_signature_C&gt;A</td>\n",
       "      <td>CNV_ratio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>Mutation_signature_C&gt;G</td>\n",
       "      <td>Cibersort_Macrophage_M2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>Mutation_signature_C&gt;T</td>\n",
       "      <td>Estimate_ImmuneScore</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>Mutation_signature_T&gt;C</td>\n",
       "      <td>Cibersort_Macrophage_M2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>Mutation_signature_T&gt;A</td>\n",
       "      <td>Progeny_Hypoxia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>Mutation_signature_T&gt;G</td>\n",
       "      <td>Progeny_TGFb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>WXS_normal_sample_type</td>\n",
       "      <td>Ancillary_studies_other_immunohistochemistry_p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>WXS_normal_filename</td>\n",
       "      <td>Progeny_WNT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>WXS_normal_UUID</td>\n",
       "      <td>Progeny_WNT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>WXS_tumor_sample_type</td>\n",
       "      <td>Tumor_necrosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163</th>\n",
       "      <td>WXS_tumor_filename</td>\n",
       "      <td>Progeny_TGFb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>WXS_tumor_UUID</td>\n",
       "      <td>Progeny_WNT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>WGS_normal_sample_type</td>\n",
       "      <td>Group</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>WGS_normal_UUID</td>\n",
       "      <td>Progeny_WNT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>WGS_tumor_sample_type</td>\n",
       "      <td>Tumor_site</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>WGS_tumor_UUID</td>\n",
       "      <td>Progeny_TGFb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>RNAseq_R1_sample_type</td>\n",
       "      <td>ARID1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>RNAseq_R1_filename</td>\n",
       "      <td>Aliquot_ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>RNAseq_R1_UUID</td>\n",
       "      <td>Aliquot_ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>RNAseq_R2_sample_type</td>\n",
       "      <td>ARID1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>RNAseq_R2_filename</td>\n",
       "      <td>Aliquot_ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>RNAseq_R2_UUID</td>\n",
       "      <td>Aliquot_ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>miRNAseq_sample_type</td>\n",
       "      <td>Mutation_signature_SBS7a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>miRNAseq_UUID</td>\n",
       "      <td>Batch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177</th>\n",
       "      <td>Methylation_available</td>\n",
       "      <td>Group</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>Methylation_quality</td>\n",
       "      <td>MSI_status</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     source  \\\n",
       "0                                       idx   \n",
       "1                 Proteomics_Participant_ID   \n",
       "2                             Case_excluded   \n",
       "3                      Proteomics_TMT_batch   \n",
       "4                       Proteomics_TMT_plex   \n",
       "5                    Proteomics_TMT_channel   \n",
       "6              Proteomics_Parent_Sample_IDs   \n",
       "7                     Proteomics_Aliquot_ID   \n",
       "8                   Proteomics_Tumor_Normal   \n",
       "9                            Proteomics_OCT   \n",
       "10                                  Country   \n",
       "11                    Histologic_Grade_FIGO   \n",
       "12              Myometrial_invasion_Specify   \n",
       "13                          Histologic_type   \n",
       "14                          Treatment_naive   \n",
       "15                             Tumor_purity   \n",
       "16              Path_Stage_Primary_Tumor-pT   \n",
       "17            Path_Stage_Reg_Lymph_Nodes-pN   \n",
       "18                  Clin_Stage_Dist_Mets-cM   \n",
       "19                  Path_Stage_Dist_Mets-pM   \n",
       "20                 tumor_Stage-Pathological   \n",
       "21                               FIGO_stage   \n",
       "22                                     LVSI   \n",
       "23                                      BMI   \n",
       "24                                      Age   \n",
       "25                                 Diabetes   \n",
       "26                                     Race   \n",
       "27                                Ethnicity   \n",
       "28                                   Gender   \n",
       "29                               Tumor_Site   \n",
       "30                         Tumor_Site_Other   \n",
       "31                           Tumor_Focality   \n",
       "32                            Tumor_Size_cm   \n",
       "33                        Estrogen_Receptor   \n",
       "34                      Estrogen_Receptor_%   \n",
       "35                    Progesterone_Receptor   \n",
       "36                  Progesterone_Receptor_%   \n",
       "37                                     MLH1   \n",
       "38                                     MLH2   \n",
       "39                                     MSH6   \n",
       "40                                     PMS2   \n",
       "41                                      p53   \n",
       "42                        Other_IHC_specify   \n",
       "43           MLH1_Promoter_Hypermethylation   \n",
       "44                Num_full_term_pregnancies   \n",
       "45                              EPIC_Bcells   \n",
       "46                                EPIC_CAFs   \n",
       "47                          EPIC_CD4_Tcells   \n",
       "48                          EPIC_CD8_Tcells   \n",
       "49                         EPIC_Endothelial   \n",
       "50                         EPIC_Macrophages   \n",
       "51                             EPIC_NKcells   \n",
       "52                          EPIC_otherCells   \n",
       "53                  CIBERSORT_B_cells_naive   \n",
       "54                 CIBERSORT_B_cells_memory   \n",
       "55                   CIBERSORT_Plasma_cells   \n",
       "56                    CIBERSORT_T_cells_CD8   \n",
       "57              CIBERSORT_T_cells_CD4_naive   \n",
       "58     CIBERSORT_T_cells_CD4_memory_resting   \n",
       "59   CIBERSORT_T_cells_CD4_memory_activated   \n",
       "60      CIBERSORT_T_cells_follicular_helper   \n",
       "61     CIBERSORT_T_cells_regulatory_(Tregs)   \n",
       "62            CIBERSORT_T_cells_gamma_delta   \n",
       "63               CIBERSORT_NK_cells_resting   \n",
       "64             CIBERSORT_NK_cells_activated   \n",
       "65                      CIBERSORT_Monocytes   \n",
       "66                 CIBERSORT_Macrophages_M0   \n",
       "67                 CIBERSORT_Macrophages_M1   \n",
       "68                 CIBERSORT_Macrophages_M2   \n",
       "69        CIBERSORT_Dendritic_cells_resting   \n",
       "70      CIBERSORT_Dendritic_cells_activated   \n",
       "71             CIBERSORT_Mast_cells_resting   \n",
       "72           CIBERSORT_Mast_cells_activated   \n",
       "73                    CIBERSORT_Eosinophils   \n",
       "74                    CIBERSORT_Neutrophils   \n",
       "75                 CIBERSORT_Absolute_score   \n",
       "76                    ESTIMATE_StromalScore   \n",
       "77                     ESTIMATE_ImmuneScore   \n",
       "78                   ESTIMATE_ESTIMATEScore   \n",
       "79                           Stemness_score   \n",
       "80                                  ER_ESR1   \n",
       "81                                   PR_PGR   \n",
       "82                    Pathway_activity_EGFR   \n",
       "83                 Pathway_activity_Hypoxia   \n",
       "84                Pathway_activity_JAK.STAT   \n",
       "85                    Pathway_activity_MAPK   \n",
       "86                    Pathway_activity_NFkB   \n",
       "87                    Pathway_activity_PI3K   \n",
       "88                    Pathway_activity_TGFb   \n",
       "89                    Pathway_activity_TNFa   \n",
       "90                   Pathway_activity_Trail   \n",
       "91                    Pathway_activity_VEGF   \n",
       "92                     Pathway_activity_p53   \n",
       "93                                 TP53_ATM   \n",
       "94                               TP53_CHEK2   \n",
       "95                                TP53_MDM4   \n",
       "96                             TP53_RPS6KA3   \n",
       "97                                TP53_TP53   \n",
       "98                             TP53_pathway   \n",
       "99                                PI3K_AKT1   \n",
       "100                               PI3K_AKT2   \n",
       "101                               PI3K_AKT3   \n",
       "102                             PI3K_DEPDC5   \n",
       "103                             PI3K_DEPTOR   \n",
       "104                             PI3K_INPP4B   \n",
       "105                            PI3K_MAPKAP1   \n",
       "106                              PI3K_MLST8   \n",
       "107                               PI3K_MTOR   \n",
       "108                              PI3K_NPRL2   \n",
       "109                              PI3K_NPRL3   \n",
       "110                               PI3K_PDK1   \n",
       "111                             PI3K_PIK3CA   \n",
       "112                             PI3K_PIK3CB   \n",
       "113                             PI3K_PIK3R1   \n",
       "114                             PI3K_PIK3R2   \n",
       "115                            PI3K_PPP2R1A   \n",
       "116                               PI3K_PTEN   \n",
       "117                               PI3K_RHEB   \n",
       "118                             PI3K_RICTOR   \n",
       "119                               PI3K_RPS6   \n",
       "120                            PI3K_RPS6KB1   \n",
       "121                              PI3K_RPTOR   \n",
       "122                              PI3K_STK11   \n",
       "123                               PI3K_TSC1   \n",
       "124                               PI3K_TSC2   \n",
       "125                            PI3K_pathway   \n",
       "126                               HRD_BRCA1   \n",
       "127                               HRD_BRCA2   \n",
       "128                      HRD_BRCA1_or_BRCA2   \n",
       "129                                 CNV_DEL   \n",
       "130                                 CNV_AMP   \n",
       "131                               CNV_class   \n",
       "132                                 CNV_idx   \n",
       "133                              CNV_1q_DEL   \n",
       "134                              CNV_3q_DEL   \n",
       "135                              CNV_4q_DEL   \n",
       "136                              CNV_1q_AMP   \n",
       "137                              CNV_3q_AMP   \n",
       "138                              CNV_4q_AMP   \n",
       "139                           Purity_Immune   \n",
       "140                           Purity_Cancer   \n",
       "141                           Purity_Stroma   \n",
       "142                              MSI_status   \n",
       "143                            POLE_subtype   \n",
       "144                           JAK1_MS_INDEL   \n",
       "145                           JAK1_Mutation   \n",
       "146                    Log2_variant_per_Mbp   \n",
       "147                        Log2_SNP_per_Mbp   \n",
       "148                      Log2_INDEL_per_Mbp   \n",
       "149                      Log2_variant_total   \n",
       "150                          Log2_SNP_total   \n",
       "151                        Log2_INDEL_total   \n",
       "152                        Genomics_subtype   \n",
       "153                  Mutation_signature_C>A   \n",
       "154                  Mutation_signature_C>G   \n",
       "155                  Mutation_signature_C>T   \n",
       "156                  Mutation_signature_T>C   \n",
       "157                  Mutation_signature_T>A   \n",
       "158                  Mutation_signature_T>G   \n",
       "159                  WXS_normal_sample_type   \n",
       "160                     WXS_normal_filename   \n",
       "161                         WXS_normal_UUID   \n",
       "162                   WXS_tumor_sample_type   \n",
       "163                      WXS_tumor_filename   \n",
       "164                          WXS_tumor_UUID   \n",
       "165                  WGS_normal_sample_type   \n",
       "166                         WGS_normal_UUID   \n",
       "167                   WGS_tumor_sample_type   \n",
       "168                          WGS_tumor_UUID   \n",
       "169                   RNAseq_R1_sample_type   \n",
       "170                      RNAseq_R1_filename   \n",
       "171                          RNAseq_R1_UUID   \n",
       "172                   RNAseq_R2_sample_type   \n",
       "173                      RNAseq_R2_filename   \n",
       "174                          RNAseq_R2_UUID   \n",
       "175                    miRNAseq_sample_type   \n",
       "176                           miRNAseq_UUID   \n",
       "177                   Methylation_available   \n",
       "178                     Methylation_quality   \n",
       "\n",
       "                                                target  \n",
       "0           Number_of_para-aortic_lymph_nodes_examined  \n",
       "1                                              Case_id  \n",
       "2                                        Case_excluded  \n",
       "3                                  Metformin_treatment  \n",
       "4    Number_of_para-aortic_lymph_nodes_positive_for...  \n",
       "5                                         ReporterName  \n",
       "6                                              Case_id  \n",
       "7                                           Aliquot_ID  \n",
       "8                                                Group  \n",
       "9                                                 POLE  \n",
       "10                                 Participant_country  \n",
       "11                                    Histologic_grade  \n",
       "12                 Myometrial_invasion_present_specify  \n",
       "13                                     Histologic_Type  \n",
       "14   Follow-up_additional_treatment_radiation_thera...  \n",
       "15                                      Tumor_necrosis  \n",
       "16                 Pathologic_staging_primary_tumor_pt  \n",
       "17          Pathologic_staging_regional_lymph_nodes_pn  \n",
       "18              Clinical_staging_distant_metastasis_cm  \n",
       "19              Clinical_staging_distant_metastasis_cm  \n",
       "20                            Tumor_stage_pathological  \n",
       "21                            Tumor_stage_pathological  \n",
       "22                Number_of_other_lymph_nodes_examined  \n",
       "23                                                 BMI  \n",
       "24                                                 Age  \n",
       "25                                            Diabetes  \n",
       "26                                                Race  \n",
       "27                                           Ethnicity  \n",
       "28                                                 Sex  \n",
       "29                                          Tumor_site  \n",
       "30                                    Tumor_site_other  \n",
       "31                                      Tumor_focality  \n",
       "32                                       Tumor_size_cm  \n",
       "33                 Ancillary_studies_estrogen_receptor  \n",
       "34                 Ancillary_studies_estrogen_receptor  \n",
       "35             Ancillary_studies_progesterone_receptor  \n",
       "36             Ancillary_studies_progesterone_receptor  \n",
       "37                              Ancillary_studies_mlh1  \n",
       "38                              Ancillary_studies_mlh1  \n",
       "39                              Ancillary_studies_msh2  \n",
       "40                              Ancillary_studies_pms2  \n",
       "41                               Ancillary_studies_p53  \n",
       "42   Ancillary_studies_other_immunohistochemistry_t...  \n",
       "43    Ancillary_studies_mlh1_promoter_hypermethylation  \n",
       "44   Donor_information_number_of_full_term_pregnancies  \n",
       "45                                    Progeny_Androgen  \n",
       "46                                           CNV_ratio  \n",
       "47                                          Aliquot_ID  \n",
       "48                               xCell_T_cell_CD4+_Th2  \n",
       "49                                    Progeny_Androgen  \n",
       "50                             Cibersort_Macrophage_M2  \n",
       "51                                       xCell_NK_cell  \n",
       "52                              Estimate_ESTIMATEScore  \n",
       "53                Cibersort_T_cell_CD4+_memory_resting  \n",
       "54                  xCell_Cancer_associated_fibroblast  \n",
       "55                                        Progeny_PI3K  \n",
       "56                Cibersort_T_cell_CD4+_memory_resting  \n",
       "57                         Cibersort_T_cell_CD4+_naive  \n",
       "58                                           CNV_ratio  \n",
       "59                               xCell_T_cell_CD4+_Th2  \n",
       "60                Cibersort_T_cell_CD4+_memory_resting  \n",
       "61                Cibersort_T_cell_CD4+_memory_resting  \n",
       "62                               xCell_T_cell_CD4+_Th2  \n",
       "63                         Cibersort_NK_cell_activated  \n",
       "64                         Cibersort_NK_cell_activated  \n",
       "65                             Cibersort_Macrophage_M2  \n",
       "66                             Cibersort_Macrophage_M0  \n",
       "67                             Cibersort_Macrophage_M1  \n",
       "68                             Cibersort_Macrophage_M2  \n",
       "69                  xCell_Cancer_associated_fibroblast  \n",
       "70                Cibersort_T_cell_CD4+_memory_resting  \n",
       "71                       Cibersort_Mast_cell_activated  \n",
       "72                       Cibersort_Mast_cell_activated  \n",
       "73                  xCell_Cancer_associated_fibroblast  \n",
       "74                                           CNV_ratio  \n",
       "75                  Cibersort_T_cell_follicular_helper  \n",
       "76                               Estimate_StromalScore  \n",
       "77                              Estimate_ESTIMATEScore  \n",
       "78                              Estimate_ESTIMATEScore  \n",
       "79                                    Progeny_Androgen  \n",
       "80                                  xCell_stroma_score  \n",
       "81                                        Progeny_EGFR  \n",
       "82             Ancillary_studies_progesterone_receptor  \n",
       "83                                  xCell_immune_score  \n",
       "84                             Mutation_signature_SBS5  \n",
       "85                                        Progeny_MAPK  \n",
       "86          Pathologic_staging_regional_lymph_nodes_pn  \n",
       "87          Pathologic_staging_regional_lymph_nodes_pn  \n",
       "88                               ABSOLUTE_tumor_purity  \n",
       "89                 Pathologic_staging_primary_tumor_pt  \n",
       "90                 Pathologic_staging_primary_tumor_pt  \n",
       "91                                        Progeny_VEGF  \n",
       "92          Pathologic_staging_regional_lymph_nodes_pn  \n",
       "93                                                TP53  \n",
       "94                              Ancillary_studies_pms2  \n",
       "95                                                TP53  \n",
       "96                              Ancillary_studies_pms2  \n",
       "97                                                TP53  \n",
       "98                                                TP53  \n",
       "99                                              PIK3CA  \n",
       "100                                             PIK3CA  \n",
       "101                                             PIK3CA  \n",
       "102                                             PIK3CA  \n",
       "103                                             PIK3CA  \n",
       "104                                             PIK3CA  \n",
       "105                                             PIK3CA  \n",
       "106                             Ancillary_studies_pms2  \n",
       "107                                             PIK3CA  \n",
       "108                                             PIK3CA  \n",
       "109                                             PIK3CA  \n",
       "110                                             PIK3CA  \n",
       "111                                             PIK3CA  \n",
       "112                                             PIK3CA  \n",
       "113                                             PIK3CA  \n",
       "114                                             PIK3CA  \n",
       "115                                             PIK3CA  \n",
       "116                                             PIK3CA  \n",
       "117                                             PIK3CA  \n",
       "118                                             PIK3CA  \n",
       "119                                             PIK3CA  \n",
       "120                                             PIK3CA  \n",
       "121                                             PIK3CA  \n",
       "122                                             PIK3CA  \n",
       "123                                             PIK3CA  \n",
       "124                                             PIK3CA  \n",
       "125                                             PIK3CA  \n",
       "126                                             ARID1A  \n",
       "127                                             ARID1A  \n",
       "128                                              Batch  \n",
       "129                                          CNV_ratio  \n",
       "130                                          CNV_ratio  \n",
       "131                                         CNV_status  \n",
       "132                                          CNV_ratio  \n",
       "133                                          CNV_ratio  \n",
       "134                                          CNV_ratio  \n",
       "135                                          CNV_ratio  \n",
       "136                                          CNV_ratio  \n",
       "137                                          CNV_ratio  \n",
       "138                                          CNV_ratio  \n",
       "139                                   Progeny_Androgen  \n",
       "140                                    Progeny_Hypoxia  \n",
       "141                                   Progeny_Androgen  \n",
       "142                                         MSI_status  \n",
       "143                                               POLE  \n",
       "144                                         MSI_status  \n",
       "145             Clinical_staging_distant_metastasis_cm  \n",
       "146                                       Progeny_MAPK  \n",
       "147                                   Progeny_JAK.STAT  \n",
       "148                            Cibersort_Macrophage_M0  \n",
       "149                             Estimate_ESTIMATEScore  \n",
       "150                                          CNV_ratio  \n",
       "151                             Estimate_ESTIMATEScore  \n",
       "152                                    Genomic_subtype  \n",
       "153                                          CNV_ratio  \n",
       "154                            Cibersort_Macrophage_M2  \n",
       "155                               Estimate_ImmuneScore  \n",
       "156                            Cibersort_Macrophage_M2  \n",
       "157                                    Progeny_Hypoxia  \n",
       "158                                       Progeny_TGFb  \n",
       "159  Ancillary_studies_other_immunohistochemistry_p...  \n",
       "160                                        Progeny_WNT  \n",
       "161                                        Progeny_WNT  \n",
       "162                                     Tumor_necrosis  \n",
       "163                                       Progeny_TGFb  \n",
       "164                                        Progeny_WNT  \n",
       "165                                              Group  \n",
       "166                                        Progeny_WNT  \n",
       "167                                         Tumor_site  \n",
       "168                                       Progeny_TGFb  \n",
       "169                                             ARID1A  \n",
       "170                                         Aliquot_ID  \n",
       "171                                         Aliquot_ID  \n",
       "172                                             ARID1A  \n",
       "173                                         Aliquot_ID  \n",
       "174                                         Aliquot_ID  \n",
       "175                           Mutation_signature_SBS7a  \n",
       "176                                              Batch  \n",
       "177                                              Group  \n",
       "178                                         MSI_status  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "schema_mapping = bdi.match_schema(df_source, df_target, method='max_val_sim')\n",
    "print(f\"Recall: {100*accuracy(gt_set, schema_mapping):.3f}\")\n",
    "schema_mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracting features from 1 columns...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "4a7415c43f464277a03431032c884d82",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/1 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracting features from 213 columns...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6787b5314beb4557bef5b1d63bad3b0b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/213 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Proteomics_Participant_ID</td>\n",
       "      <td>Idx</td>\n",
       "      <td>0.173261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Proteomics_Participant_ID</td>\n",
       "      <td>Case_id</td>\n",
       "      <td>0.158389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Proteomics_Participant_ID</td>\n",
       "      <td>Aliquot_ID</td>\n",
       "      <td>0.123715</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Proteomics_Participant_ID</td>\n",
       "      <td>xCell_T_cell_CD8+_central_memory</td>\n",
       "      <td>0.103956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Proteomics_Participant_ID</td>\n",
       "      <td>Progeny_p53</td>\n",
       "      <td>0.102078</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      source                            target  similarity\n",
       "0  Proteomics_Participant_ID                               Idx    0.173261\n",
       "1  Proteomics_Participant_ID                           Case_id    0.158389\n",
       "2  Proteomics_Participant_ID                        Aliquot_ID    0.123715\n",
       "3  Proteomics_Participant_ID  xCell_T_cell_CD8+_central_memory    0.103956\n",
       "4  Proteomics_Participant_ID                       Progeny_p53    0.102078"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bdi.top_matches(df_source, columns=[\"Proteomics_Participant_ID\"], target=df_target, top_k=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>value_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.201397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.059600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.032400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.055500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.107774</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   value_name\n",
       "0    0.201397\n",
       "1    0.059600\n",
       "2    0.032400\n",
       "3    0.055500\n",
       "4    0.107774"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bdi.preview_domain(df_target, \"Cibersort_T_cell_CD8+\").head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>value_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   value_name\n",
       "0         1.0\n",
       "1         0.0\n",
       "2         NaN"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bdi.preview_domain(df_source, \"TP53_TP53\").head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>value_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Mutated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  value_name\n",
       "0         WT\n",
       "1    Mutated\n",
       "2        NaN"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bdi.preview_domain(df_target, \"TP53\").head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [source, target, similarity]\n",
       "Index: []"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bdi.match_values(df_source, df_target, (\"TP53_TP53\", \"TP53\"), method='tfidf')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
