{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/javascript": "(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  var force = true;\n  var py_version = '3.4.3'.replace('rc', '-rc.').replace('.dev', '-dev.');\n  var reloading = false;\n  var Bokeh = root.Bokeh;\n\n  if (typeof (root._bokeh_timeout) === \"undefined\" || force) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks;\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, js_modules, js_exports, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n    if (js_modules == null) js_modules = [];\n    if (js_exports == null) js_exports = {};\n\n    root._bokeh_onload_callbacks.push(callback);\n\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls.length === 0 && js_modules.length === 0 && Object.keys(js_exports).length === 0) {\n      run_callbacks();\n      return null;\n    }\n    if (!reloading) {\n      console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    }\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n    window._bokeh_on_load = on_load\n\n    function on_error() {\n      console.error(\"failed to load \" + url);\n    }\n\n    var skip = [];\n    if (window.requirejs) {\n      window.requirejs.config({'packages': {}, 'paths': {'tabulator': 'https://cdn.jsdelivr.net/npm/tabulator-tables@5.5.0/dist/js/tabulator.min', 'moment': 'https://cdn.jsdelivr.net/npm/luxon/build/global/luxon.min'}, 'shim': {}});\n      require([\"tabulator\"], function(Tabulator) {\n\twindow.Tabulator = Tabulator\n\ton_load()\n      })\n      require([\"moment\"], function(moment) {\n\twindow.moment = moment\n\ton_load()\n      })\n      root._bokeh_is_loading = css_urls.length + 2;\n    } else {\n      root._bokeh_is_loading = css_urls.length + js_urls.length + js_modules.length + Object.keys(js_exports).length;\n    }\n\n    var existing_stylesheets = []\n    var links = document.getElementsByTagName('link')\n    for (var i = 0; i < links.length; i++) {\n      var link = links[i]\n      if (link.href != null) {\n\texisting_stylesheets.push(link.href)\n      }\n    }\n    for (var i = 0; i < css_urls.length; i++) {\n      var url = css_urls[i];\n      if (existing_stylesheets.indexOf(url) !== -1) {\n\ton_load()\n\tcontinue;\n      }\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error;\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }    if (((window.Tabulator !== undefined) && (!(window.Tabulator instanceof HTMLElement))) || window.requirejs) {\n      var urls = ['https://cdn.holoviz.org/panel/1.4.5/dist/bundled/datatabulator/tabulator-tables@5.5.0/dist/js/tabulator.min.js'];\n      for (var i = 0; i < urls.length; i++) {\n        skip.push(urls[i])\n      }\n    }    if (((window.moment !== undefined) && (!(window.moment instanceof HTMLElement))) || window.requirejs) {\n      var urls = ['https://cdn.holoviz.org/panel/1.4.5/dist/bundled/datatabulator/luxon/build/global/luxon.min.js'];\n      for (var i = 0; i < urls.length; i++) {\n        skip.push(urls[i])\n      }\n    }    var existing_scripts = []\n    var scripts = document.getElementsByTagName('script')\n    for (var i = 0; i < scripts.length; i++) {\n      var script = scripts[i]\n      if (script.src != null) {\n\texisting_scripts.push(script.src)\n      }\n    }\n    for (var i = 0; i < js_urls.length; i++) {\n      var url = js_urls[i];\n      if (skip.indexOf(url) !== -1 || existing_scripts.indexOf(url) !== -1) {\n\tif (!window.requirejs) {\n\t  on_load();\n\t}\n\tcontinue;\n      }\n      var element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error;\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n    for (var i = 0; i < js_modules.length; i++) {\n      var url = js_modules[i];\n      if (skip.indexOf(url) !== -1 || existing_scripts.indexOf(url) !== -1) {\n\tif (!window.requirejs) {\n\t  on_load();\n\t}\n\tcontinue;\n      }\n      var element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error;\n      element.async = false;\n      element.src = url;\n      element.type = \"module\";\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n    for (const name in js_exports) {\n      var url = js_exports[name];\n      if (skip.indexOf(url) >= 0 || root[name] != null) {\n\tif (!window.requirejs) {\n\t  on_load();\n\t}\n\tcontinue;\n      }\n      var element = document.createElement('script');\n      element.onerror = on_error;\n      element.async = false;\n      element.type = \"module\";\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      element.textContent = `\n      import ${name} from \"${url}\"\n      window.${name} = ${name}\n      window._bokeh_on_load()\n      `\n      document.head.appendChild(element);\n    }\n    if (!js_urls.length && !js_modules.length) {\n      on_load()\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  var js_urls = [\"https://cdn.holoviz.org/panel/1.4.5/dist/bundled/datatabulator/tabulator-tables@5.5.0/dist/js/tabulator.min.js\", \"https://cdn.holoviz.org/panel/1.4.5/dist/bundled/datatabulator/luxon/build/global/luxon.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-3.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.4.3.min.js\", \"https://cdn.holoviz.org/panel/1.4.5/dist/panel.min.js\"];\n  var js_modules = [];\n  var js_exports = {};\n  var css_urls = [\"https://cdn.holoviz.org/panel/1.4.5/dist/bundled/datatabulator/tabulator-tables@5.5.0/dist/css/tabulator_simple.min.css?v=1.4.5\"];\n  var inline_js = [    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\nfunction(Bokeh) {} // ensure no trailing comma for IE\n  ];\n\n  function run_inline_js() {\n    if ((root.Bokeh !== undefined) || (force === true)) {\n      for (var i = 0; i < inline_js.length; i++) {\n\ttry {\n          inline_js[i].call(root, root.Bokeh);\n\t} catch(e) {\n\t  if (!reloading) {\n\t    throw e;\n\t  }\n\t}\n      }\n      // Cache old bokeh versions\n      if (Bokeh != undefined && !reloading) {\n\tvar NewBokeh = root.Bokeh;\n\tif (Bokeh.versions === undefined) {\n\t  Bokeh.versions = new Map();\n\t}\n\tif (NewBokeh.version !== Bokeh.version) {\n\t  Bokeh.versions.set(NewBokeh.version, NewBokeh)\n\t}\n\troot.Bokeh = Bokeh;\n      }} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    }\n    root._bokeh_is_initializing = false\n  }\n\n  function load_or_wait() {\n    // Implement a backoff loop that tries to ensure we do not load multiple\n    // versions of Bokeh and its dependencies at the same time.\n    // In recent versions we use the root._bokeh_is_initializing flag\n    // to determine whether there is an ongoing attempt to initialize\n    // bokeh, however for backward compatibility we also try to ensure\n    // that we do not start loading a newer (Panel>=1.0 and Bokeh>3) version\n    // before older versions are fully initialized.\n    if (root._bokeh_is_initializing && Date.now() > root._bokeh_timeout) {\n      root._bokeh_is_initializing = false;\n      root._bokeh_onload_callbacks = undefined;\n      console.log(\"Bokeh: BokehJS was loaded multiple times but one version failed to initialize.\");\n      load_or_wait();\n    } else if (root._bokeh_is_initializing || (typeof root._bokeh_is_initializing === \"undefined\" && root._bokeh_onload_callbacks !== undefined)) {\n      setTimeout(load_or_wait, 100);\n    } else {\n      root._bokeh_is_initializing = true\n      root._bokeh_onload_callbacks = []\n      var bokeh_loaded = Bokeh != null && (Bokeh.version === py_version || (Bokeh.versions !== undefined && Bokeh.versions.has(py_version)));\n      if (!reloading && !bokeh_loaded) {\n\troot.Bokeh = undefined;\n      }\n      load_libs(css_urls, js_urls, js_modules, js_exports, function() {\n\tconsole.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n\trun_inline_js();\n      });\n    }\n  }\n  // Give older versions of the autoload script a head-start to ensure\n  // they initialize before we start loading newer version.\n  setTimeout(load_or_wait, 100)\n}(window));",
      "application/vnd.holoviews_load.v0+json": ""
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": "\nif ((window.PyViz === undefined) || (window.PyViz instanceof HTMLElement)) {\n  window.PyViz = {comms: {}, comm_status:{}, kernels:{}, receivers: {}, plot_index: []}\n}\n\n\n    function JupyterCommManager() {\n    }\n\n    JupyterCommManager.prototype.register_target = function(plot_id, comm_id, msg_handler) {\n      if (window.comm_manager || ((window.Jupyter !== undefined) && (Jupyter.notebook.kernel != null))) {\n        var comm_manager = window.comm_manager || Jupyter.notebook.kernel.comm_manager;\n        comm_manager.register_target(comm_id, function(comm) {\n          comm.on_msg(msg_handler);\n        });\n      } else if ((plot_id in window.PyViz.kernels) && (window.PyViz.kernels[plot_id])) {\n        window.PyViz.kernels[plot_id].registerCommTarget(comm_id, function(comm) {\n          comm.onMsg = msg_handler;\n        });\n      } else if (typeof google != 'undefined' && google.colab.kernel != null) {\n        google.colab.kernel.comms.registerTarget(comm_id, (comm) => {\n          var messages = comm.messages[Symbol.asyncIterator]();\n          function processIteratorResult(result) {\n            var message = result.value;\n            console.log(message)\n            var content = {data: message.data, comm_id};\n            var buffers = []\n            for (var buffer of message.buffers || []) {\n              buffers.push(new DataView(buffer))\n            }\n            var metadata = message.metadata || {};\n            var msg = {content, buffers, metadata}\n            msg_handler(msg);\n            return messages.next().then(processIteratorResult);\n          }\n          return messages.next().then(processIteratorResult);\n        })\n      }\n    }\n\n    JupyterCommManager.prototype.get_client_comm = function(plot_id, comm_id, msg_handler) {\n      if (comm_id in window.PyViz.comms) {\n        return window.PyViz.comms[comm_id];\n      } else if (window.comm_manager || ((window.Jupyter !== undefined) && (Jupyter.notebook.kernel != null))) {\n        var comm_manager = window.comm_manager || Jupyter.notebook.kernel.comm_manager;\n        var comm = comm_manager.new_comm(comm_id, {}, {}, {}, comm_id);\n        if (msg_handler) {\n          comm.on_msg(msg_handler);\n        }\n      } else if ((plot_id in window.PyViz.kernels) && (window.PyViz.kernels[plot_id])) {\n        var comm = window.PyViz.kernels[plot_id].connectToComm(comm_id);\n        comm.open();\n        if (msg_handler) {\n          comm.onMsg = msg_handler;\n        }\n      } else if (typeof google != 'undefined' && google.colab.kernel != null) {\n        var comm_promise = google.colab.kernel.comms.open(comm_id)\n        comm_promise.then((comm) => {\n          window.PyViz.comms[comm_id] = comm;\n          if (msg_handler) {\n            var messages = comm.messages[Symbol.asyncIterator]();\n            function processIteratorResult(result) {\n              var message = result.value;\n              var content = {data: message.data};\n              var metadata = message.metadata || {comm_id};\n              var msg = {content, metadata}\n              msg_handler(msg);\n              return messages.next().then(processIteratorResult);\n            }\n            return messages.next().then(processIteratorResult);\n          }\n        }) \n        var sendClosure = (data, metadata, buffers, disposeOnDone) => {\n          return comm_promise.then((comm) => {\n            comm.send(data, metadata, buffers, disposeOnDone);\n          });\n        };\n        var comm = {\n          send: sendClosure\n        };\n      }\n      window.PyViz.comms[comm_id] = comm;\n      return comm;\n    }\n    window.PyViz.comm_manager = new JupyterCommManager();\n    \n\n\nvar JS_MIME_TYPE = 'application/javascript';\nvar HTML_MIME_TYPE = 'text/html';\nvar EXEC_MIME_TYPE = 'application/vnd.holoviews_exec.v0+json';\nvar CLASS_NAME = 'output';\n\n/**\n * Render data to the DOM node\n */\nfunction render(props, node) {\n  var div = document.createElement(\"div\");\n  var script = document.createElement(\"script\");\n  node.appendChild(div);\n  node.appendChild(script);\n}\n\n/**\n * Handle when a new output is added\n */\nfunction handle_add_output(event, handle) {\n  var output_area = handle.output_area;\n  var output = handle.output;\n  if ((output.data == undefined) || (!output.data.hasOwnProperty(EXEC_MIME_TYPE))) {\n    return\n  }\n  var id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n  var toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n  if (id !== undefined) {\n    var nchildren = toinsert.length;\n    var html_node = toinsert[nchildren-1].children[0];\n    html_node.innerHTML = output.data[HTML_MIME_TYPE];\n    var scripts = [];\n    var nodelist = html_node.querySelectorAll(\"script\");\n    for (var i in nodelist) {\n      if (nodelist.hasOwnProperty(i)) {\n        scripts.push(nodelist[i])\n      }\n    }\n\n    scripts.forEach( function (oldScript) {\n      var newScript = document.createElement(\"script\");\n      var attrs = [];\n      var nodemap = oldScript.attributes;\n      for (var j in nodemap) {\n        if (nodemap.hasOwnProperty(j)) {\n          attrs.push(nodemap[j])\n        }\n      }\n      attrs.forEach(function(attr) { newScript.setAttribute(attr.name, attr.value) });\n      newScript.appendChild(document.createTextNode(oldScript.innerHTML));\n      oldScript.parentNode.replaceChild(newScript, oldScript);\n    });\n    if (JS_MIME_TYPE in output.data) {\n      toinsert[nchildren-1].children[1].textContent = output.data[JS_MIME_TYPE];\n    }\n    output_area._hv_plot_id = id;\n    if ((window.Bokeh !== undefined) && (id in Bokeh.index)) {\n      window.PyViz.plot_index[id] = Bokeh.index[id];\n    } else {\n      window.PyViz.plot_index[id] = null;\n    }\n  } else if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n    var bk_div = document.createElement(\"div\");\n    bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n    var script_attrs = bk_div.children[0].attributes;\n    for (var i = 0; i < script_attrs.length; i++) {\n      toinsert[toinsert.length - 1].childNodes[1].setAttribute(script_attrs[i].name, script_attrs[i].value);\n    }\n    // store reference to server id on output_area\n    output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n  }\n}\n\n/**\n * Handle when an output is cleared or removed\n */\nfunction handle_clear_output(event, handle) {\n  var id = handle.cell.output_area._hv_plot_id;\n  var server_id = handle.cell.output_area._bokeh_server_id;\n  if (((id === undefined) || !(id in PyViz.plot_index)) && (server_id !== undefined)) { return; }\n  var comm = window.PyViz.comm_manager.get_client_comm(\"hv-extension-comm\", \"hv-extension-comm\", function () {});\n  if (server_id !== null) {\n    comm.send({event_type: 'server_delete', 'id': server_id});\n    return;\n  } else if (comm !== null) {\n    comm.send({event_type: 'delete', 'id': id});\n  }\n  delete PyViz.plot_index[id];\n  if ((window.Bokeh !== undefined) & (id in window.Bokeh.index)) {\n    var doc = window.Bokeh.index[id].model.document\n    doc.clear();\n    const i = window.Bokeh.documents.indexOf(doc);\n    if (i > -1) {\n      window.Bokeh.documents.splice(i, 1);\n    }\n  }\n}\n\n/**\n * Handle kernel restart event\n */\nfunction handle_kernel_cleanup(event, handle) {\n  delete PyViz.comms[\"hv-extension-comm\"];\n  window.PyViz.plot_index = {}\n}\n\n/**\n * Handle update_display_data messages\n */\nfunction handle_update_output(event, handle) {\n  handle_clear_output(event, {cell: {output_area: handle.output_area}})\n  handle_add_output(event, handle)\n}\n\nfunction register_renderer(events, OutputArea) {\n  function append_mime(data, metadata, element) {\n    // create a DOM node to render to\n    var toinsert = this.create_output_subarea(\n    metadata,\n    CLASS_NAME,\n    EXEC_MIME_TYPE\n    );\n    this.keyboard_manager.register_events(toinsert);\n    // Render to node\n    var props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n    render(props, toinsert[0]);\n    element.append(toinsert);\n    return toinsert\n  }\n\n  events.on('output_added.OutputArea', handle_add_output);\n  events.on('output_updated.OutputArea', handle_update_output);\n  events.on('clear_output.CodeCell', handle_clear_output);\n  events.on('delete.Cell', handle_clear_output);\n  events.on('kernel_ready.Kernel', handle_kernel_cleanup);\n\n  OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n    safe: true,\n    index: 0\n  });\n}\n\nif (window.Jupyter !== undefined) {\n  try {\n    var events = require('base/js/events');\n    var OutputArea = require('notebook/js/outputarea').OutputArea;\n    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n      register_renderer(events, OutputArea);\n    }\n  } catch(err) {\n  }\n}\n",
      "application/vnd.holoviews_load.v0+json": ""
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>*[data-root-id],\n",
       "*[data-root-id] > * {\n",
       "  box-sizing: border-box;\n",
       "  font-family: var(--jp-ui-font-family);\n",
       "  font-size: var(--jp-ui-font-size1);\n",
       "  color: var(--vscode-editor-foreground, var(--jp-ui-font-color1));\n",
       "}\n",
       "\n",
       "/* Override VSCode background color */\n",
       ".cell-output-ipywidget-background:has(\n",
       "    > .cell-output-ipywidget-background > .lm-Widget > *[data-root-id]\n",
       "  ),\n",
       ".cell-output-ipywidget-background:has(> .lm-Widget > *[data-root-id]) {\n",
       "  background-color: transparent !important;\n",
       "}\n",
       "</style>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.holoviews_exec.v0+json": "",
      "text/html": [
       "<div id='799de2b7-3e72-460d-b909-ff6dcc166011'>\n",
       "  <div id=\"dfcf3f5f-6a23-459d-ad46-dcfde3ef8851\" data-root-id=\"799de2b7-3e72-460d-b909-ff6dcc166011\" style=\"display: contents;\"></div>\n",
       "</div>\n",
       "<script type=\"application/javascript\">(function(root) {\n",
       "  var docs_json = {\"40de9fc3-a77c-48cd-a774-2bb190a86e83\":{\"version\":\"3.4.3\",\"title\":\"Bokeh Application\",\"roots\":[{\"type\":\"object\",\"name\":\"panel.models.browser.BrowserInfo\",\"id\":\"799de2b7-3e72-460d-b909-ff6dcc166011\"},{\"type\":\"object\",\"name\":\"panel.models.comm_manager.CommManager\",\"id\":\"a85cb3be-fad6-49f7-b8bc-c5a78bcd6e66\",\"attributes\":{\"plot_id\":\"799de2b7-3e72-460d-b909-ff6dcc166011\",\"comm_id\":\"7820191b7de243b0bade8a631dd1b5ab\",\"client_comm_id\":\"8c883eb5b6b44d5da5e62de3eb44e2d6\"}}],\"defs\":[{\"type\":\"model\",\"name\":\"ReactiveHTML1\"},{\"type\":\"model\",\"name\":\"FlexBox1\",\"properties\":[{\"name\":\"align_content\",\"kind\":\"Any\",\"default\":\"flex-start\"},{\"name\":\"align_items\",\"kind\":\"Any\",\"default\":\"flex-start\"},{\"name\":\"flex_direction\",\"kind\":\"Any\",\"default\":\"row\"},{\"name\":\"flex_wrap\",\"kind\":\"Any\",\"default\":\"wrap\"},{\"name\":\"gap\",\"kind\":\"Any\",\"default\":\"\"},{\"name\":\"justify_content\",\"kind\":\"Any\",\"default\":\"flex-start\"}]},{\"type\":\"model\",\"name\":\"FloatPanel1\",\"properties\":[{\"name\":\"config\",\"kind\":\"Any\",\"default\":{\"type\":\"map\"}},{\"name\":\"contained\",\"kind\":\"Any\",\"default\":true},{\"name\":\"position\",\"kind\":\"Any\",\"default\":\"right-top\"},{\"name\":\"offsetx\",\"kind\":\"Any\",\"default\":null},{\"name\":\"offsety\",\"kind\":\"Any\",\"default\":null},{\"name\":\"theme\",\"kind\":\"Any\",\"default\":\"primary\"},{\"name\":\"status\",\"kind\":\"Any\",\"default\":\"normalized\"}]},{\"type\":\"model\",\"name\":\"GridStack1\",\"properties\":[{\"name\":\"mode\",\"kind\":\"Any\",\"default\":\"warn\"},{\"name\":\"ncols\",\"kind\":\"Any\",\"default\":null},{\"name\":\"nrows\",\"kind\":\"Any\",\"default\":null},{\"name\":\"allow_resize\",\"kind\":\"Any\",\"default\":true},{\"name\":\"allow_drag\",\"kind\":\"Any\",\"default\":true},{\"name\":\"state\",\"kind\":\"Any\",\"default\":[]}]},{\"type\":\"model\",\"name\":\"drag1\",\"properties\":[{\"name\":\"slider_width\",\"kind\":\"Any\",\"default\":5},{\"name\":\"slider_color\",\"kind\":\"Any\",\"default\":\"black\"},{\"name\":\"value\",\"kind\":\"Any\",\"default\":50}]},{\"type\":\"model\",\"name\":\"click1\",\"properties\":[{\"name\":\"terminal_output\",\"kind\":\"Any\",\"default\":\"\"},{\"name\":\"debug_name\",\"kind\":\"Any\",\"default\":\"\"},{\"name\":\"clears\",\"kind\":\"Any\",\"default\":0}]},{\"type\":\"model\",\"name\":\"FastWrapper1\",\"properties\":[{\"name\":\"object\",\"kind\":\"Any\",\"default\":null},{\"name\":\"style\",\"kind\":\"Any\",\"default\":null}]},{\"type\":\"model\",\"name\":\"NotificationAreaBase1\",\"properties\":[{\"name\":\"js_events\",\"kind\":\"Any\",\"default\":{\"type\":\"map\"}},{\"name\":\"position\",\"kind\":\"Any\",\"default\":\"bottom-right\"},{\"name\":\"_clear\",\"kind\":\"Any\",\"default\":0}]},{\"type\":\"model\",\"name\":\"NotificationArea1\",\"properties\":[{\"name\":\"js_events\",\"kind\":\"Any\",\"default\":{\"type\":\"map\"}},{\"name\":\"notifications\",\"kind\":\"Any\",\"default\":[]},{\"name\":\"position\",\"kind\":\"Any\",\"default\":\"bottom-right\"},{\"name\":\"_clear\",\"kind\":\"Any\",\"default\":0},{\"name\":\"types\",\"kind\":\"Any\",\"default\":[{\"type\":\"map\",\"entries\":[[\"type\",\"warning\"],[\"background\",\"#ffc107\"],[\"icon\",{\"type\":\"map\",\"entries\":[[\"className\",\"fas fa-exclamation-triangle\"],[\"tagName\",\"i\"],[\"color\",\"white\"]]}]]},{\"type\":\"map\",\"entries\":[[\"type\",\"info\"],[\"background\",\"#007bff\"],[\"icon\",{\"type\":\"map\",\"entries\":[[\"className\",\"fas fa-info-circle\"],[\"tagName\",\"i\"],[\"color\",\"white\"]]}]]}]}]},{\"type\":\"model\",\"name\":\"Notification\",\"properties\":[{\"name\":\"background\",\"kind\":\"Any\",\"default\":null},{\"name\":\"duration\",\"kind\":\"Any\",\"default\":3000},{\"name\":\"icon\",\"kind\":\"Any\",\"default\":null},{\"name\":\"message\",\"kind\":\"Any\",\"default\":\"\"},{\"name\":\"notification_type\",\"kind\":\"Any\",\"default\":null},{\"name\":\"_destroyed\",\"kind\":\"Any\",\"default\":false}]},{\"type\":\"model\",\"name\":\"TemplateActions1\",\"properties\":[{\"name\":\"open_modal\",\"kind\":\"Any\",\"default\":0},{\"name\":\"close_modal\",\"kind\":\"Any\",\"default\":0}]},{\"type\":\"model\",\"name\":\"BootstrapTemplateActions1\",\"properties\":[{\"name\":\"open_modal\",\"kind\":\"Any\",\"default\":0},{\"name\":\"close_modal\",\"kind\":\"Any\",\"default\":0}]},{\"type\":\"model\",\"name\":\"TemplateEditor1\",\"properties\":[{\"name\":\"layout\",\"kind\":\"Any\",\"default\":[]}]},{\"type\":\"model\",\"name\":\"MaterialTemplateActions1\",\"properties\":[{\"name\":\"open_modal\",\"kind\":\"Any\",\"default\":0},{\"name\":\"close_modal\",\"kind\":\"Any\",\"default\":0}]},{\"type\":\"model\",\"name\":\"copy_to_clipboard1\",\"properties\":[{\"name\":\"fill\",\"kind\":\"Any\",\"default\":\"none\"},{\"name\":\"value\",\"kind\":\"Any\",\"default\":null}]}]}};\n",
       "  var render_items = [{\"docid\":\"40de9fc3-a77c-48cd-a774-2bb190a86e83\",\"roots\":{\"799de2b7-3e72-460d-b909-ff6dcc166011\":\"dfcf3f5f-6a23-459d-ad46-dcfde3ef8851\"},\"root_ids\":[\"799de2b7-3e72-460d-b909-ff6dcc166011\"]}];\n",
       "  var docs = Object.values(docs_json)\n",
       "  if (!docs) {\n",
       "    return\n",
       "  }\n",
       "  const py_version = docs[0].version.replace('rc', '-rc.').replace('.dev', '-dev.')\n",
       "  async function embed_document(root) {\n",
       "    var Bokeh = get_bokeh(root)\n",
       "    await Bokeh.embed.embed_items_notebook(docs_json, render_items);\n",
       "    for (const render_item of render_items) {\n",
       "      for (const root_id of render_item.root_ids) {\n",
       "\tconst id_el = document.getElementById(root_id)\n",
       "\tif (id_el.children.length && id_el.children[0].hasAttribute('data-root-id')) {\n",
       "\t  const root_el = id_el.children[0]\n",
       "\t  root_el.id = root_el.id + '-rendered'\n",
       "\t  for (const child of root_el.children) {\n",
       "            // Ensure JupyterLab does not capture keyboard shortcuts\n",
       "            // see: https://jupyterlab.readthedocs.io/en/4.1.x/extension/notebook.html#keyboard-interaction-model\n",
       "\t    child.setAttribute('data-lm-suppress-shortcuts', 'true')\n",
       "\t  }\n",
       "\t}\n",
       "      }\n",
       "    }\n",
       "  }\n",
       "  function get_bokeh(root) {\n",
       "    if (root.Bokeh === undefined) {\n",
       "      return null\n",
       "    } else if (root.Bokeh.version !== py_version) {\n",
       "      if (root.Bokeh.versions === undefined || !root.Bokeh.versions.has(py_version)) {\n",
       "\treturn null\n",
       "      }\n",
       "      return root.Bokeh.versions.get(py_version);\n",
       "    } else if (root.Bokeh.version === py_version) {\n",
       "      return root.Bokeh\n",
       "    }\n",
       "    return null\n",
       "  }\n",
       "  function is_loaded(root) {\n",
       "    var Bokeh = get_bokeh(root)\n",
       "    return (Bokeh != null && Bokeh.Panel !== undefined && ( root.Tabulator !== undefined) && ( root.Tabulator !== undefined))\n",
       "  }\n",
       "  if (is_loaded(root)) {\n",
       "    embed_document(root);\n",
       "  } else {\n",
       "    var attempts = 0;\n",
       "    var timer = setInterval(function(root) {\n",
       "      if (is_loaded(root)) {\n",
       "        clearInterval(timer);\n",
       "        embed_document(root);\n",
       "      } else if (document.readyState == \"complete\") {\n",
       "        attempts++;\n",
       "        if (attempts > 200) {\n",
       "          clearInterval(timer);\n",
       "\t  var Bokeh = get_bokeh(root)\n",
       "\t  if (Bokeh == null || Bokeh.Panel == null) {\n",
       "            console.warn(\"Panel: ERROR: Unable to run Panel code because Bokeh or Panel library is missing\");\n",
       "\t  } else {\n",
       "\t    console.warn(\"Panel: WARNING: Attempting to render but not all required libraries could be resolved.\")\n",
       "\t    embed_document(root)\n",
       "\t  }\n",
       "        }\n",
       "      }\n",
       "    }, 25, root)\n",
       "  }\n",
       "})(window);</script>"
      ]
     },
     "metadata": {
      "application/vnd.holoviews_exec.v0+json": {
       "id": "799de2b7-3e72-460d-b909-ff6dcc166011"
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import bdikit as bdi\n",
    "import pandas as pd\n",
    "from IPython.display import display, Markdown\n",
    "pd.set_option('display.max_rows', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_source = pd.read_csv(\"./datasets/Dou-ucec-discovery.csv\")\n",
    "df_target = pd.read_csv(\"./datasets/Dou-ucec-confirmatory.csv\")\n",
    "df_gt = pd.read_csv(\"./datasets/Dou-ucec-ground-truth.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Proteomics_Participant_ID',\n",
       " 'Case_excluded',\n",
       " 'Proteomics_TMT_batch',\n",
       " 'Proteomics_TMT_plex',\n",
       " 'Proteomics_TMT_channel',\n",
       " 'Proteomics_Aliquot_ID',\n",
       " 'Proteomics_Tumor_Normal',\n",
       " 'Country',\n",
       " 'Histologic_Grade_FIGO',\n",
       " 'Myometrial_invasion_Specify',\n",
       " 'Histologic_type',\n",
       " 'Treatment_naive',\n",
       " 'Tumor_purity',\n",
       " 'Path_Stage_Primary_Tumor-pT',\n",
       " 'Path_Stage_Reg_Lymph_Nodes-pN',\n",
       " 'Clin_Stage_Dist_Mets-cM',\n",
       " 'tumor_Stage-Pathological',\n",
       " 'BMI',\n",
       " 'Age',\n",
       " 'Diabetes',\n",
       " 'Race',\n",
       " 'Ethnicity',\n",
       " 'Gender',\n",
       " 'Tumor_Site',\n",
       " 'Tumor_Site_Other',\n",
       " 'Tumor_Focality',\n",
       " 'Tumor_Size_cm',\n",
       " 'Estrogen_Receptor',\n",
       " 'Progesterone_Receptor',\n",
       " 'MLH1',\n",
       " 'MSH6',\n",
       " 'PMS2',\n",
       " 'p53',\n",
       " 'MLH1_Promoter_Hypermethylation',\n",
       " 'Num_full_term_pregnancies',\n",
       " 'CIBERSORT_B _cells _naive',\n",
       " 'CIBERSORT_B _cells _memory',\n",
       " 'CIBERSORT_Plasma _cells',\n",
       " 'CIBERSORT_T _cells _CD8',\n",
       " 'CIBERSORT_T _cells _CD4 _naive',\n",
       " 'CIBERSORT_T _cells _CD4 _memory _resting',\n",
       " 'CIBERSORT_T _cells _CD4 _memory _activated',\n",
       " 'CIBERSORT_T _cells _follicular _helper',\n",
       " 'CIBERSORT_T _cells _regulatory _(Tregs)',\n",
       " 'CIBERSORT_T _cells _gamma _delta',\n",
       " 'CIBERSORT_NK _cells _resting',\n",
       " 'CIBERSORT_NK _cells _activated',\n",
       " 'CIBERSORT_Monocytes',\n",
       " 'CIBERSORT_Macrophages _M0',\n",
       " 'CIBERSORT_Macrophages _M1',\n",
       " 'CIBERSORT_Macrophages _M2',\n",
       " 'CIBERSORT_Dendritic _cells _resting',\n",
       " 'CIBERSORT_Dendritic _cells _activated',\n",
       " 'CIBERSORT_Mast _cells _resting',\n",
       " 'CIBERSORT_Mast _cells _activated',\n",
       " 'CIBERSORT_Eosinophils',\n",
       " 'CIBERSORT_Neutrophils',\n",
       " 'ESTIMATE_StromalScore',\n",
       " 'ESTIMATE_ImmuneScore',\n",
       " 'ESTIMATE_ESTIMATEScore',\n",
       " 'CNV_class',\n",
       " 'MSI_status',\n",
       " 'POLE_subtype',\n",
       " 'Genomics_subtype']"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ground_truth = df_gt[~df_gt['target'].isna()]\n",
    "gt_set = df_ground_truth.set_index('source')['target'].to_dict()\n",
    "list(gt_set.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "def accuracy(gt_set, schema_mapping):\n",
    "    schema_dict = schema_mapping.set_index('source')['target'].to_dict()\n",
    "    correct_count = 0\n",
    "    total = 0\n",
    "    correct = []\n",
    "    wrong = []\n",
    "    for source_column, target_column in  schema_dict.items():\n",
    "        if source_column in gt_set:\n",
    "            total += 1\n",
    "            correct_target_columns = set(gt_set[source_column].split(\",\"))\n",
    "            if target_column in correct_target_columns:\n",
    "                correct_count += 1\n",
    "                correct.append((source_column, target_column))\n",
    "                # print(f\"OK: {source_column} -> out={target_column} gt={correct_target_columns}\")\n",
    "            else:\n",
    "                wrong.append((source_column, target_column))\n",
    "                # print(f\"ER: {source_column} -> out={target_column} gt={correct_target_columns}\")\n",
    "    return correct_count/float(total), correct, wrong"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_source.columns = df_source.columns.str.replace(' ', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracting features from 179 columns...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "7a4deece3d74488cb9acdcabe8bc7e2c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/179 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracting features from 213 columns...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "39d6d0936a12489cb204cff083e33ab5",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/213 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "75.556\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>idx</td>\n",
       "      <td>xCell_T_cell_CD4+_Th1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Proteomics_Participant_ID</td>\n",
       "      <td>Idx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Case_excluded</td>\n",
       "      <td>Case_excluded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>Metformin_treatment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Proteomics_TMT_plex</td>\n",
       "      <td>Number_of_para-aortic_lymph_nodes_positive_for...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Proteomics_TMT_channel</td>\n",
       "      <td>Number_of_para-aortic_lymph_nodes_positive_for...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Proteomics_Parent_Sample_IDs</td>\n",
       "      <td>Idx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Proteomics_Aliquot_ID</td>\n",
       "      <td>Aliquot_ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Proteomics_Tumor_Normal</td>\n",
       "      <td>Group</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Proteomics_OCT</td>\n",
       "      <td>POLE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         source  \\\n",
       "0                           idx   \n",
       "1     Proteomics_Participant_ID   \n",
       "2                 Case_excluded   \n",
       "3          Proteomics_TMT_batch   \n",
       "4           Proteomics_TMT_plex   \n",
       "5        Proteomics_TMT_channel   \n",
       "6  Proteomics_Parent_Sample_IDs   \n",
       "7         Proteomics_Aliquot_ID   \n",
       "8       Proteomics_Tumor_Normal   \n",
       "9                Proteomics_OCT   \n",
       "\n",
       "                                              target  \n",
       "0                              xCell_T_cell_CD4+_Th1  \n",
       "1                                                Idx  \n",
       "2                                      Case_excluded  \n",
       "3                                Metformin_treatment  \n",
       "4  Number_of_para-aortic_lymph_nodes_positive_for...  \n",
       "5  Number_of_para-aortic_lymph_nodes_positive_for...  \n",
       "6                                                Idx  \n",
       "7                                         Aliquot_ID  \n",
       "8                                              Group  \n",
       "9                                               POLE  "
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "schema_mapping = bdi.match_schema(df_source, df_target, method='ct_learning')\n",
    "acc, correct, wrong = accuracy(gt_set, schema_mapping)\n",
    "print(f\"{100*acc:.3f}\")\n",
    "schema_mapping.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracting features from 179 columns...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "87138683756c4c38a531f209380362d9",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/179 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracting features from 213 columns...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "67da5138df604f0082bd6eb3f3891abd",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/213 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "80.000\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>idx</td>\n",
       "      <td>Number_of_para-aortic_lymph_nodes_examined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Proteomics_Participant_ID</td>\n",
       "      <td>Case_id</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Case_excluded</td>\n",
       "      <td>Case_excluded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>Metformin_treatment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Proteomics_TMT_plex</td>\n",
       "      <td>Number_of_para-aortic_lymph_nodes_positive_for...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Proteomics_TMT_channel</td>\n",
       "      <td>ReporterName</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Proteomics_Parent_Sample_IDs</td>\n",
       "      <td>Case_id</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Proteomics_Aliquot_ID</td>\n",
       "      <td>Aliquot_ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Proteomics_Tumor_Normal</td>\n",
       "      <td>Group</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Proteomics_OCT</td>\n",
       "      <td>POLE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         source  \\\n",
       "0                           idx   \n",
       "1     Proteomics_Participant_ID   \n",
       "2                 Case_excluded   \n",
       "3          Proteomics_TMT_batch   \n",
       "4           Proteomics_TMT_plex   \n",
       "5        Proteomics_TMT_channel   \n",
       "6  Proteomics_Parent_Sample_IDs   \n",
       "7         Proteomics_Aliquot_ID   \n",
       "8       Proteomics_Tumor_Normal   \n",
       "9                Proteomics_OCT   \n",
       "\n",
       "                                              target  \n",
       "0         Number_of_para-aortic_lymph_nodes_examined  \n",
       "1                                            Case_id  \n",
       "2                                      Case_excluded  \n",
       "3                                Metformin_treatment  \n",
       "4  Number_of_para-aortic_lymph_nodes_positive_for...  \n",
       "5                                       ReporterName  \n",
       "6                                            Case_id  \n",
       "7                                         Aliquot_ID  \n",
       "8                                              Group  \n",
       "9                                               POLE  "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "schema_mapping = bdi.match_schema(df_source, df_target, method='max_val_sim')\n",
    "acc, correct, wrong = accuracy(gt_set, schema_mapping)\n",
    "print(f\"{100*acc:.3f}\")\n",
    "schema_mapping.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Proteomics_Participant_ID',\n",
       " 'Case_excluded',\n",
       " 'Proteomics_TMT_channel',\n",
       " 'Proteomics_Aliquot_ID',\n",
       " 'Proteomics_Tumor_Normal',\n",
       " 'Country',\n",
       " 'Histologic_Grade_FIGO',\n",
       " 'Myometrial_invasion_Specify',\n",
       " 'Histologic_type',\n",
       " 'Path_Stage_Primary_Tumor-pT',\n",
       " 'Path_Stage_Reg_Lymph_Nodes-pN',\n",
       " 'Clin_Stage_Dist_Mets-cM',\n",
       " 'tumor_Stage-Pathological',\n",
       " 'BMI',\n",
       " 'Age',\n",
       " 'Diabetes',\n",
       " 'Race',\n",
       " 'Ethnicity',\n",
       " 'Gender',\n",
       " 'Tumor_Site',\n",
       " 'Tumor_Site_Other',\n",
       " 'Tumor_Focality',\n",
       " 'Tumor_Size_cm',\n",
       " 'Estrogen_Receptor',\n",
       " 'Progesterone_Receptor',\n",
       " 'MLH1',\n",
       " 'PMS2',\n",
       " 'p53',\n",
       " 'MLH1_Promoter_Hypermethylation',\n",
       " 'Num_full_term_pregnancies',\n",
       " 'ESTIMATE_StromalScore',\n",
       " 'ESTIMATE_ESTIMATEScore',\n",
       " 'CNV_class',\n",
       " 'MSI_status',\n",
       " 'POLE_subtype',\n",
       " 'Genomics_subtype']"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(map(lambda x: x[0], correct))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['G1 Well differentiated', 'G2 Moderately differentiated',\n",
       "       'G3 Poorly differentiated', 'Other: High grade',\n",
       "       'Other: Not specified.', nan], dtype=object)"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_target['Histologic_grade'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OK: ESTIMATE_StromalScore -> out=Estimate_StromalScore gt={'Estimate_StromalScore'}\n",
      "OK: ESTIMATE_ImmuneScore -> out=Estimate_ImmuneScore gt={'Estimate_ImmuneScore'}\n",
      "OK: Tumor_Site_Other -> out=Tumor_site_other gt={'Tumor_site_other'}\n",
      "OK: ESTIMATE_ESTIMATEScore -> out=Estimate_ESTIMATEScore gt={'Estimate_ESTIMATEScore'}\n",
      "OK: Histologic_type -> out=Histologic_Type gt={'Histologic_Type'}\n",
      "OK: Tumor_Focality -> out=Tumor_focality gt={'Tumor_focality'}\n",
      "OK: Case_excluded -> out=Case_excluded gt={'Case_excluded'}\n",
      "OK: Tumor_Size_cm -> out=Tumor_size_cm gt={'Tumor_size_cm'}\n",
      "OK: Tumor_Site -> out=Tumor_site gt={'Tumor_site'}\n",
      "OK: MSI_status -> out=MSI_status gt={'MSI_status'}\n",
      "OK: Ethnicity -> out=Ethnicity gt={'Ethnicity'}\n",
      "OK: Diabetes -> out=Diabetes gt={'Diabetes'}\n",
      "OK: Race -> out=Race gt={'Race'}\n",
      "OK: BMI -> out=BMI gt={'BMI'}\n",
      "OK: Age -> out=Age gt={'Age'}\n",
      "OK: CIBERSORT_Eosinophils -> out=Cibersort_Eosinophil gt={'Cibersort_Eosinophil'}\n",
      "OK: CIBERSORT_Neutrophils -> out=Cibersort_Neutrophil gt={'Cibersort_Neutrophil'}\n",
      "OK: tumor_Stage-Pathological -> out=Tumor_stage_pathological gt={'Tumor_stage_pathological'}\n",
      "OK: CIBERSORT_Monocytes -> out=Cibersort_Monocyte gt={'Cibersort_Monocyte'}\n",
      "OK: Genomics_subtype -> out=Genomic_subtype gt={'Genomic_subtype'}\n",
      "OK: Myometrial_invasion_Specify -> out=Myometrial_invasion_present_specify gt={'Myometrial_invasion', 'Myometrial_invasion_present_specify'}\n",
      "OK: Histologic_Grade_FIGO -> out=Histologic_grade gt={'Histologic_grade'}\n",
      "Recall: 48.889\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>Pathway_activity_Hypoxia</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>CNV_AMP</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>BMI</td>\n",
       "      <td>BMI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>Log2_SNP_per_Mbp</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>PI3K_DEPTOR</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>CNV_4q_DEL</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>Estrogen_Receptor</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>miRNAseq_UUID</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>JAK1_Mutation</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>CNV_class</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       source target\n",
       "90   Pathway_activity_Hypoxia       \n",
       "136                   CNV_AMP       \n",
       "14                        BMI    BMI\n",
       "152          Log2_SNP_per_Mbp       \n",
       "109               PI3K_DEPTOR       \n",
       "141                CNV_4q_DEL       \n",
       "64          Estrogen_Receptor       \n",
       "176             miRNAseq_UUID       \n",
       "150             JAK1_Mutation       \n",
       "137                 CNV_class       "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "schema_mapping = bdi.match_schema(df_source, df_target, method='coma')\n",
    "print(f\"Recall: {100*accuracy(gt_set, schema_mapping):.3f}\")\n",
    "schema_mapping.sample(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "unsupported format string passed to tuple.__format__",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[60], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m schema_mapping \u001b[38;5;241m=\u001b[39m bdi\u001b[38;5;241m.\u001b[39mmatch_schema(df_source, df_target, method\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124msimilarity_flooding\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m----> 2\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mRecall: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;241m100\u001b[39m\u001b[38;5;241m*\u001b[39maccuracy(gt_set,\u001b[38;5;250m \u001b[39mschema_mapping)\u001b[38;5;132;01m:\u001b[39;00m\u001b[38;5;124m.3f\u001b[39m\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m      3\u001b[0m schema_mapping\u001b[38;5;241m.\u001b[39msample(\u001b[38;5;241m10\u001b[39m)\n",
      "\u001b[0;31mTypeError\u001b[0m: unsupported format string passed to tuple.__format__"
     ]
    }
   ],
   "source": [
    "schema_mapping = bdi.match_schema(df_source, df_target, method='similarity_flooding')\n",
    "print(f\"Recall: {100*accuracy(gt_set, schema_mapping):.3f}\")\n",
    "schema_mapping.sample(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OK: Case_excluded -> out=Case_excluded gt={'Case_excluded'}\n",
      "OK: Ethnicity -> out=Ethnicity gt={'Ethnicity'}\n",
      "OK: Gender -> out=Sex gt={'Sex'}\n",
      "OK: Tumor_Focality -> out=Tumor_focality gt={'Tumor_focality'}\n",
      "OK: MLH1_Promoter_Hypermethylation -> out=Ancillary_studies_mlh1_promoter_hypermethylation gt={'Ancillary_studies_mlh1_promoter_hypermethylation'}\n",
      "OK: Num_full_term_pregnancies -> out=Donor_information_number_of_full_term_pregnancies gt={'Donor_information_number_of_full_term_pregnancies'}\n",
      "OK: MSI_status -> out=MSI_status gt={'MSI_status'}\n",
      "OK: p53 -> out=Ancillary_studies_p53 gt={'Ancillary_studies_p53'}\n",
      "OK: Proteomics_Tumor_Normal -> out=Group gt={'Group'}\n",
      "OK: MLH1 -> out=Ancillary_studies_mlh1 gt={'Ancillary_studies_mlh1'}\n",
      "OK: Proteomics_TMT_channel -> out=ReporterName gt={'ReporterName'}\n",
      "OK: Race -> out=Race gt={'Race'}\n",
      "OK: Clin_Stage_Dist_Mets-cM -> out=Clinical_staging_distant_metastasis_cm gt={'Clinical_staging_distant_metastasis_cm'}\n",
      "OK: Tumor_Size_cm -> out=Tumor_size_cm gt={'Tumor_size_cm'}\n",
      "OK: BMI -> out=BMI gt={'BMI'}\n",
      "OK: Tumor_Site -> out=Tumor_site gt={'Tumor_site'}\n",
      "OK: Path_Stage_Primary_Tumor-pT -> out=Pathologic_staging_primary_tumor_pt gt={'Pathologic_staging_primary_tumor_pt'}\n",
      "OK: Path_Stage_Reg_Lymph_Nodes-pN -> out=Pathologic_staging_regional_lymph_nodes_pn gt={'Pathologic_staging_regional_lymph_nodes_pn'}\n",
      "OK: Tumor_Site_Other -> out=Tumor_site_other gt={'Tumor_site_other'}\n",
      "OK: Genomics_subtype -> out=Genomic_subtype gt={'Genomic_subtype'}\n",
      "OK: Country -> out=Participant_country gt={'Participant_country'}\n",
      "OK: Myometrial_invasion_Specify -> out=Myometrial_invasion gt={'Myometrial_invasion', 'Myometrial_invasion_present_specify'}\n",
      "OK: Proteomics_Aliquot_ID -> out=Aliquot_ID gt={'Aliquot_ID'}\n",
      "OK: Proteomics_Participant_ID -> out=Case_id gt={'Case_id'}\n",
      "Recall: 53.333\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>Purity_Cancer</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>Pathway_activity_Hypoxia</td>\n",
       "      <td>Progeny_NFkB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>CIBERSORT_Macrophages_M2</td>\n",
       "      <td>Cibersort_Monocyte</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177</th>\n",
       "      <td>Methylation_available</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>ESTIMATE_StromalScore</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>Progesterone_Receptor</td>\n",
       "      <td>History_of_cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Tumor_Size_cm</td>\n",
       "      <td>Tumor_size_cm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>PI3K_PIK3R1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>PI3K_RPTOR</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>Purity_Stroma</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       source              target\n",
       "150             Purity_Cancer                    \n",
       "73   Pathway_activity_Hypoxia        Progeny_NFkB\n",
       "72   CIBERSORT_Macrophages_M2  Cibersort_Monocyte\n",
       "177     Methylation_available                    \n",
       "107     ESTIMATE_StromalScore                    \n",
       "43      Progesterone_Receptor   History_of_cancer\n",
       "22              Tumor_Size_cm       Tumor_size_cm\n",
       "124               PI3K_PIK3R1                    \n",
       "132                PI3K_RPTOR                    \n",
       "151             Purity_Stroma                    "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "schema_mapping = bdi.match_schema(df_source, df_target, method='jaccard_distance')\n",
    "print(f\"Recall: {100*accuracy(gt_set, schema_mapping):.3f}\")\n",
    "schema_mapping.sample(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracting features from 1 columns...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "3983fe75eae542a4ab225aaffdd73fd0",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/1 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracting features from 213 columns...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "b48ab36d07674b7ebe80ea5588639bfb",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/213 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "channel_matches = bdi.top_matches(df_source, columns=[\"Proteomics_TMT_channel\"], target=df_target, top_k=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** Proteomics_TMT_channel<br>**Target column:** Pathologic_staging_primary_tumor_pt<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "411064f32fcf4d688308c7c89c70eecc",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "BokehModel(combine_events=True, render_bundle={'docs_json': {'5bd1165c-8f4a-4301-9b5f-a6f2b4f0778d': {'version…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** Proteomics_TMT_channel<br>**Target column:** Number_of_pelvic_lymph_nodes_examined<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c5bc9f7bc78e46cf8dcf4f409ba70809",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "BokehModel(combine_events=True, render_bundle={'docs_json': {'44dd946d-f996-4dca-bc64-19306286c802': {'version…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** Proteomics_TMT_channel<br>**Target column:** ReporterName<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "7a6346af979a47448efafef518e68ec0",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "BokehModel(combine_events=True, render_bundle={'docs_json': {'b913b222-2543-4e26-b937-55941670bf40': {'version…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** Proteomics_TMT_channel<br>**Target column:** Number_of_para-aortic_lymph_nodes_examined<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "a494851b7a014511af2654bcf79f03f0",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "BokehModel(combine_events=True, render_bundle={'docs_json': {'ce1bd0c8-6918-481f-a64b-f8da82701b9e': {'version…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "<br>**Source column:** Proteomics_TMT_channel<br>**Target column:** Number_of_para-aortic_lymph_nodes_positive_for_tumor_by_he<br>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "b347ba74138d48a399e8254b313bd1fb",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "BokehModel(combine_events=True, render_bundle={'docs_json': {'4393b060-5945-49eb-be42-609a15af0837': {'version…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "channel_value_matches = bdi.match_values(df_source, df_target, channel_matches, method='tfidf')\n",
    "bdi.view_value_matches(channel_value_matches)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracting features from 1 columns...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "a0d9ec6051d44cb8b52f9c48a754b96b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/1 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracting features from 213 columns...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "3dcb032a60e44ec89e942b0b19cda27f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/213 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Proteomics_Participant_ID</td>\n",
       "      <td>Idx</td>\n",
       "      <td>0.824948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Proteomics_Participant_ID</td>\n",
       "      <td>Case_id</td>\n",
       "      <td>0.777057</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Proteomics_Participant_ID</td>\n",
       "      <td>Aliquot_ID</td>\n",
       "      <td>0.616938</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      source      target  similarity\n",
       "0  Proteomics_Participant_ID         Idx    0.824948\n",
       "1  Proteomics_Participant_ID     Case_id    0.777057\n",
       "2  Proteomics_Participant_ID  Aliquot_ID    0.616938"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "part_id_matches = bdi.top_matches(df_source, columns=[\"Proteomics_Participant_ID\"], target=df_target, top_k=3)\n",
    "part_id_matches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>value_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>C3L-00771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>C3L-00136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>C3N-00337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>NX9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>C3N-00847</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    value_name\n",
       "27   C3L-00771\n",
       "6    C3L-00136\n",
       "69   C3N-00337\n",
       "113        NX9\n",
       "82   C3N-00847"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bdi.preview_domain(df_source, 'Proteomics_Participant_ID').sample(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>value_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C3L-00086</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C3L-00898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C3L-00943</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C3L-01064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C3L-01277</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  value_name\n",
       "0  C3L-00086\n",
       "1  C3L-00898\n",
       "2  C3L-00943\n",
       "3  C3L-01064\n",
       "4  C3L-01277"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bdi.preview_domain(df_target, 'Idx').head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>value_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C3L-00086</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C3L-00898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C3L-00943</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C3L-01064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C3L-01277</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  value_name\n",
       "0  C3L-00086\n",
       "1  C3L-00898\n",
       "2  C3L-00943\n",
       "3  C3L-01064\n",
       "4  C3L-01277"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bdi.preview_domain(df_target, 'Case_id').head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Idx</th>\n",
       "      <th>Case_id</th>\n",
       "      <th>Case_excluded</th>\n",
       "      <th>Batch</th>\n",
       "      <th>Plex</th>\n",
       "      <th>ReporterName</th>\n",
       "      <th>Aliquot_ID</th>\n",
       "      <th>Group</th>\n",
       "      <th>Discovery_study</th>\n",
       "      <th>Age</th>\n",
       "      <th>...</th>\n",
       "      <th>Follow-up_additional_surgery_for_new_tumor</th>\n",
       "      <th>Follow-up_additional_treatment_radiation_therapy_for_new_tumor</th>\n",
       "      <th>Follow-up_additional_treatment_pharmaceutical_therapy_for_new_tumor</th>\n",
       "      <th>Follow-up_additional_treatment_immuno_for_new_tumor</th>\n",
       "      <th>Follow-up_days_from_date_of_collection_to_date_of_last_contact</th>\n",
       "      <th>Follow-up_cause_of_death</th>\n",
       "      <th>Follow-up_days_from_date_of_initial_pathologic_diagnosis_to_date_of_death</th>\n",
       "      <th>Follow-up_days_from_date_of_initial_pathologic_diagnosis_to_date_of_additional_surgery_for_new_tumor</th>\n",
       "      <th>Follow-up_procedure_type_of_new_tumor</th>\n",
       "      <th>Follow-up_residual_tumor_after_surgery_for_new_tumor</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>C3L-05571-b1</td>\n",
       "      <td>C3L-05571</td>\n",
       "      <td>Yes</td>\n",
       "      <td>b1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>131C</td>\n",
       "      <td>CPT0275270003</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>C3N-02978-b1</td>\n",
       "      <td>C3N-02978</td>\n",
       "      <td>Yes</td>\n",
       "      <td>b3</td>\n",
       "      <td>12.0</td>\n",
       "      <td>129C</td>\n",
       "      <td>CPT0244470003</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>C3L-00935-A</td>\n",
       "      <td>C3L-00935</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Adjacent_normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>C3L-01277-A</td>\n",
       "      <td>C3L-01277</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Adjacent_normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>C3L-01639-A</td>\n",
       "      <td>C3L-01639</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Adjacent_normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>C3L-01967-A</td>\n",
       "      <td>C3L-01967</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Adjacent_normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>C3L-02121-A</td>\n",
       "      <td>C3L-02121</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Adjacent_normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>C3L-02216-A</td>\n",
       "      <td>C3L-02216</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Adjacent_normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>C3L-02347-A</td>\n",
       "      <td>C3L-02347</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Adjacent_normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>C3L-02353-A</td>\n",
       "      <td>C3L-02353</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Adjacent_normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>C3L-02408-A</td>\n",
       "      <td>C3L-02408</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Adjacent_normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>C3L-02409-A</td>\n",
       "      <td>C3L-02409</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Adjacent_normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>C3L-02412-A</td>\n",
       "      <td>C3L-02412</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Adjacent_normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>C3L-02468-A</td>\n",
       "      <td>C3L-02468</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Adjacent_normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>C3L-02557-A</td>\n",
       "      <td>C3L-02557</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Adjacent_normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>C3L-02602-A</td>\n",
       "      <td>C3L-02602</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Adjacent_normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>C3N-00520-A</td>\n",
       "      <td>C3N-00520</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Adjacent_normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>C3N-00859-A</td>\n",
       "      <td>C3N-00859</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Adjacent_normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>C3N-00872-A</td>\n",
       "      <td>C3N-00872</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Adjacent_normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>C3N-01344-A</td>\n",
       "      <td>C3N-01344</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Adjacent_normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>C3N-01525-A</td>\n",
       "      <td>C3N-01525</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Adjacent_normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>C3N-01526-A</td>\n",
       "      <td>C3N-01526</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Adjacent_normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>C3N-01761-A</td>\n",
       "      <td>C3N-01761</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Adjacent_normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>C3N-01765-A</td>\n",
       "      <td>C3N-01765</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Adjacent_normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>C3N-01830-A</td>\n",
       "      <td>C3N-01830</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Adjacent_normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163</th>\n",
       "      <td>C3N-01847-A</td>\n",
       "      <td>C3N-01847</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Adjacent_normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>C3N-01848-A</td>\n",
       "      <td>C3N-01848</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Adjacent_normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>C3N-01875-A</td>\n",
       "      <td>C3N-01875</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Adjacent_normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>C3N-01876-A</td>\n",
       "      <td>C3N-01876</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Adjacent_normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>C3N-02249-A</td>\n",
       "      <td>C3N-02249</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Adjacent_normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>C3N-02253-A</td>\n",
       "      <td>C3N-02253</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Adjacent_normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>C3N-02436-A</td>\n",
       "      <td>C3N-02436</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Adjacent_normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32 rows × 213 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Idx    Case_id Case_excluded Batch  Plex ReporterName  \\\n",
       "60   C3L-05571-b1  C3L-05571           Yes    b1   4.0         131C   \n",
       "133  C3N-02978-b1  C3N-02978           Yes    b3  12.0         129C   \n",
       "140   C3L-00935-A  C3L-00935            No   NaN   NaN          NaN   \n",
       "141   C3L-01277-A  C3L-01277            No   NaN   NaN          NaN   \n",
       "142   C3L-01639-A  C3L-01639            No   NaN   NaN          NaN   \n",
       "143   C3L-01967-A  C3L-01967            No   NaN   NaN          NaN   \n",
       "144   C3L-02121-A  C3L-02121            No   NaN   NaN          NaN   \n",
       "145   C3L-02216-A  C3L-02216            No   NaN   NaN          NaN   \n",
       "146   C3L-02347-A  C3L-02347            No   NaN   NaN          NaN   \n",
       "147   C3L-02353-A  C3L-02353            No   NaN   NaN          NaN   \n",
       "148   C3L-02408-A  C3L-02408            No   NaN   NaN          NaN   \n",
       "149   C3L-02409-A  C3L-02409            No   NaN   NaN          NaN   \n",
       "150   C3L-02412-A  C3L-02412            No   NaN   NaN          NaN   \n",
       "151   C3L-02468-A  C3L-02468            No   NaN   NaN          NaN   \n",
       "152   C3L-02557-A  C3L-02557            No   NaN   NaN          NaN   \n",
       "153   C3L-02602-A  C3L-02602            No   NaN   NaN          NaN   \n",
       "154   C3N-00520-A  C3N-00520            No   NaN   NaN          NaN   \n",
       "155   C3N-00859-A  C3N-00859            No   NaN   NaN          NaN   \n",
       "156   C3N-00872-A  C3N-00872            No   NaN   NaN          NaN   \n",
       "157   C3N-01344-A  C3N-01344            No   NaN   NaN          NaN   \n",
       "158   C3N-01525-A  C3N-01525            No   NaN   NaN          NaN   \n",
       "159   C3N-01526-A  C3N-01526            No   NaN   NaN          NaN   \n",
       "160   C3N-01761-A  C3N-01761            No   NaN   NaN          NaN   \n",
       "161   C3N-01765-A  C3N-01765            No   NaN   NaN          NaN   \n",
       "162   C3N-01830-A  C3N-01830            No   NaN   NaN          NaN   \n",
       "163   C3N-01847-A  C3N-01847            No   NaN   NaN          NaN   \n",
       "164   C3N-01848-A  C3N-01848            No   NaN   NaN          NaN   \n",
       "165   C3N-01875-A  C3N-01875            No   NaN   NaN          NaN   \n",
       "166   C3N-01876-A  C3N-01876            No   NaN   NaN          NaN   \n",
       "167   C3N-02249-A  C3N-02249            No   NaN   NaN          NaN   \n",
       "168   C3N-02253-A  C3N-02253            No   NaN   NaN          NaN   \n",
       "169   C3N-02436-A  C3N-02436            No   NaN   NaN          NaN   \n",
       "\n",
       "        Aliquot_ID            Group Discovery_study  Age  ...  \\\n",
       "60   CPT0275270003            Tumor             NaN  NaN  ...   \n",
       "133  CPT0244470003            Tumor             NaN  NaN  ...   \n",
       "140            NaN  Adjacent_normal             NaN  NaN  ...   \n",
       "141            NaN  Adjacent_normal             NaN  NaN  ...   \n",
       "142            NaN  Adjacent_normal             NaN  NaN  ...   \n",
       "143            NaN  Adjacent_normal             NaN  NaN  ...   \n",
       "144            NaN  Adjacent_normal             NaN  NaN  ...   \n",
       "145            NaN  Adjacent_normal             NaN  NaN  ...   \n",
       "146            NaN  Adjacent_normal             NaN  NaN  ...   \n",
       "147            NaN  Adjacent_normal             NaN  NaN  ...   \n",
       "148            NaN  Adjacent_normal             NaN  NaN  ...   \n",
       "149            NaN  Adjacent_normal             NaN  NaN  ...   \n",
       "150            NaN  Adjacent_normal             NaN  NaN  ...   \n",
       "151            NaN  Adjacent_normal             NaN  NaN  ...   \n",
       "152            NaN  Adjacent_normal             NaN  NaN  ...   \n",
       "153            NaN  Adjacent_normal             NaN  NaN  ...   \n",
       "154            NaN  Adjacent_normal             NaN  NaN  ...   \n",
       "155            NaN  Adjacent_normal             NaN  NaN  ...   \n",
       "156            NaN  Adjacent_normal             NaN  NaN  ...   \n",
       "157            NaN  Adjacent_normal             NaN  NaN  ...   \n",
       "158            NaN  Adjacent_normal             NaN  NaN  ...   \n",
       "159            NaN  Adjacent_normal             NaN  NaN  ...   \n",
       "160            NaN  Adjacent_normal             NaN  NaN  ...   \n",
       "161            NaN  Adjacent_normal             NaN  NaN  ...   \n",
       "162            NaN  Adjacent_normal             NaN  NaN  ...   \n",
       "163            NaN  Adjacent_normal             NaN  NaN  ...   \n",
       "164            NaN  Adjacent_normal             NaN  NaN  ...   \n",
       "165            NaN  Adjacent_normal             NaN  NaN  ...   \n",
       "166            NaN  Adjacent_normal             NaN  NaN  ...   \n",
       "167            NaN  Adjacent_normal             NaN  NaN  ...   \n",
       "168            NaN  Adjacent_normal             NaN  NaN  ...   \n",
       "169            NaN  Adjacent_normal             NaN  NaN  ...   \n",
       "\n",
       "    Follow-up_additional_surgery_for_new_tumor  \\\n",
       "60                                         NaN   \n",
       "133                                        NaN   \n",
       "140                                        NaN   \n",
       "141                                        NaN   \n",
       "142                                        NaN   \n",
       "143                                        NaN   \n",
       "144                                        NaN   \n",
       "145                                        NaN   \n",
       "146                                        NaN   \n",
       "147                                        NaN   \n",
       "148                                        NaN   \n",
       "149                                        NaN   \n",
       "150                                        NaN   \n",
       "151                                        NaN   \n",
       "152                                        NaN   \n",
       "153                                        NaN   \n",
       "154                                        NaN   \n",
       "155                                        NaN   \n",
       "156                                        NaN   \n",
       "157                                        NaN   \n",
       "158                                        NaN   \n",
       "159                                        NaN   \n",
       "160                                        NaN   \n",
       "161                                        NaN   \n",
       "162                                        NaN   \n",
       "163                                        NaN   \n",
       "164                                        NaN   \n",
       "165                                        NaN   \n",
       "166                                        NaN   \n",
       "167                                        NaN   \n",
       "168                                        NaN   \n",
       "169                                        NaN   \n",
       "\n",
       "    Follow-up_additional_treatment_radiation_therapy_for_new_tumor  \\\n",
       "60                                                 NaN               \n",
       "133                                                NaN               \n",
       "140                                                NaN               \n",
       "141                                                NaN               \n",
       "142                                                NaN               \n",
       "143                                                NaN               \n",
       "144                                                NaN               \n",
       "145                                                NaN               \n",
       "146                                                NaN               \n",
       "147                                                NaN               \n",
       "148                                                NaN               \n",
       "149                                                NaN               \n",
       "150                                                NaN               \n",
       "151                                                NaN               \n",
       "152                                                NaN               \n",
       "153                                                NaN               \n",
       "154                                                NaN               \n",
       "155                                                NaN               \n",
       "156                                                NaN               \n",
       "157                                                NaN               \n",
       "158                                                NaN               \n",
       "159                                                NaN               \n",
       "160                                                NaN               \n",
       "161                                                NaN               \n",
       "162                                                NaN               \n",
       "163                                                NaN               \n",
       "164                                                NaN               \n",
       "165                                                NaN               \n",
       "166                                                NaN               \n",
       "167                                                NaN               \n",
       "168                                                NaN               \n",
       "169                                                NaN               \n",
       "\n",
       "    Follow-up_additional_treatment_pharmaceutical_therapy_for_new_tumor  \\\n",
       "60                                                 NaN                    \n",
       "133                                                NaN                    \n",
       "140                                                NaN                    \n",
       "141                                                NaN                    \n",
       "142                                                NaN                    \n",
       "143                                                NaN                    \n",
       "144                                                NaN                    \n",
       "145                                                NaN                    \n",
       "146                                                NaN                    \n",
       "147                                                NaN                    \n",
       "148                                                NaN                    \n",
       "149                                                NaN                    \n",
       "150                                                NaN                    \n",
       "151                                                NaN                    \n",
       "152                                                NaN                    \n",
       "153                                                NaN                    \n",
       "154                                                NaN                    \n",
       "155                                                NaN                    \n",
       "156                                                NaN                    \n",
       "157                                                NaN                    \n",
       "158                                                NaN                    \n",
       "159                                                NaN                    \n",
       "160                                                NaN                    \n",
       "161                                                NaN                    \n",
       "162                                                NaN                    \n",
       "163                                                NaN                    \n",
       "164                                                NaN                    \n",
       "165                                                NaN                    \n",
       "166                                                NaN                    \n",
       "167                                                NaN                    \n",
       "168                                                NaN                    \n",
       "169                                                NaN                    \n",
       "\n",
       "     Follow-up_additional_treatment_immuno_for_new_tumor  \\\n",
       "60                                                 NaN     \n",
       "133                                                NaN     \n",
       "140                                                NaN     \n",
       "141                                                NaN     \n",
       "142                                                NaN     \n",
       "143                                                NaN     \n",
       "144                                                NaN     \n",
       "145                                                NaN     \n",
       "146                                                NaN     \n",
       "147                                                NaN     \n",
       "148                                                NaN     \n",
       "149                                                NaN     \n",
       "150                                                NaN     \n",
       "151                                                NaN     \n",
       "152                                                NaN     \n",
       "153                                                NaN     \n",
       "154                                                NaN     \n",
       "155                                                NaN     \n",
       "156                                                NaN     \n",
       "157                                                NaN     \n",
       "158                                                NaN     \n",
       "159                                                NaN     \n",
       "160                                                NaN     \n",
       "161                                                NaN     \n",
       "162                                                NaN     \n",
       "163                                                NaN     \n",
       "164                                                NaN     \n",
       "165                                                NaN     \n",
       "166                                                NaN     \n",
       "167                                                NaN     \n",
       "168                                                NaN     \n",
       "169                                                NaN     \n",
       "\n",
       "     Follow-up_days_from_date_of_collection_to_date_of_last_contact  \\\n",
       "60                                                 NaN                \n",
       "133                                                NaN                \n",
       "140                                                NaN                \n",
       "141                                                NaN                \n",
       "142                                                NaN                \n",
       "143                                                NaN                \n",
       "144                                                NaN                \n",
       "145                                                NaN                \n",
       "146                                                NaN                \n",
       "147                                                NaN                \n",
       "148                                                NaN                \n",
       "149                                                NaN                \n",
       "150                                                NaN                \n",
       "151                                                NaN                \n",
       "152                                                NaN                \n",
       "153                                                NaN                \n",
       "154                                                NaN                \n",
       "155                                                NaN                \n",
       "156                                                NaN                \n",
       "157                                                NaN                \n",
       "158                                                NaN                \n",
       "159                                                NaN                \n",
       "160                                                NaN                \n",
       "161                                                NaN                \n",
       "162                                                NaN                \n",
       "163                                                NaN                \n",
       "164                                                NaN                \n",
       "165                                                NaN                \n",
       "166                                                NaN                \n",
       "167                                                NaN                \n",
       "168                                                NaN                \n",
       "169                                                NaN                \n",
       "\n",
       "     Follow-up_cause_of_death  \\\n",
       "60                        NaN   \n",
       "133                       NaN   \n",
       "140                       NaN   \n",
       "141                       NaN   \n",
       "142                       NaN   \n",
       "143                       NaN   \n",
       "144                       NaN   \n",
       "145                       NaN   \n",
       "146                       NaN   \n",
       "147                       NaN   \n",
       "148                       NaN   \n",
       "149                       NaN   \n",
       "150                       NaN   \n",
       "151                       NaN   \n",
       "152                       NaN   \n",
       "153                       NaN   \n",
       "154                       NaN   \n",
       "155                       NaN   \n",
       "156                       NaN   \n",
       "157                       NaN   \n",
       "158                       NaN   \n",
       "159                       NaN   \n",
       "160                       NaN   \n",
       "161                       NaN   \n",
       "162                       NaN   \n",
       "163                       NaN   \n",
       "164                       NaN   \n",
       "165                       NaN   \n",
       "166                       NaN   \n",
       "167                       NaN   \n",
       "168                       NaN   \n",
       "169                       NaN   \n",
       "\n",
       "     Follow-up_days_from_date_of_initial_pathologic_diagnosis_to_date_of_death  \\\n",
       "60                                                 NaN                           \n",
       "133                                                NaN                           \n",
       "140                                                NaN                           \n",
       "141                                                NaN                           \n",
       "142                                                NaN                           \n",
       "143                                                NaN                           \n",
       "144                                                NaN                           \n",
       "145                                                NaN                           \n",
       "146                                                NaN                           \n",
       "147                                                NaN                           \n",
       "148                                                NaN                           \n",
       "149                                                NaN                           \n",
       "150                                                NaN                           \n",
       "151                                                NaN                           \n",
       "152                                                NaN                           \n",
       "153                                                NaN                           \n",
       "154                                                NaN                           \n",
       "155                                                NaN                           \n",
       "156                                                NaN                           \n",
       "157                                                NaN                           \n",
       "158                                                NaN                           \n",
       "159                                                NaN                           \n",
       "160                                                NaN                           \n",
       "161                                                NaN                           \n",
       "162                                                NaN                           \n",
       "163                                                NaN                           \n",
       "164                                                NaN                           \n",
       "165                                                NaN                           \n",
       "166                                                NaN                           \n",
       "167                                                NaN                           \n",
       "168                                                NaN                           \n",
       "169                                                NaN                           \n",
       "\n",
       "    Follow-up_days_from_date_of_initial_pathologic_diagnosis_to_date_of_additional_surgery_for_new_tumor  \\\n",
       "60                                                 NaN                                                     \n",
       "133                                                NaN                                                     \n",
       "140                                                NaN                                                     \n",
       "141                                                NaN                                                     \n",
       "142                                                NaN                                                     \n",
       "143                                                NaN                                                     \n",
       "144                                                NaN                                                     \n",
       "145                                                NaN                                                     \n",
       "146                                                NaN                                                     \n",
       "147                                                NaN                                                     \n",
       "148                                                NaN                                                     \n",
       "149                                                NaN                                                     \n",
       "150                                                NaN                                                     \n",
       "151                                                NaN                                                     \n",
       "152                                                NaN                                                     \n",
       "153                                                NaN                                                     \n",
       "154                                                NaN                                                     \n",
       "155                                                NaN                                                     \n",
       "156                                                NaN                                                     \n",
       "157                                                NaN                                                     \n",
       "158                                                NaN                                                     \n",
       "159                                                NaN                                                     \n",
       "160                                                NaN                                                     \n",
       "161                                                NaN                                                     \n",
       "162                                                NaN                                                     \n",
       "163                                                NaN                                                     \n",
       "164                                                NaN                                                     \n",
       "165                                                NaN                                                     \n",
       "166                                                NaN                                                     \n",
       "167                                                NaN                                                     \n",
       "168                                                NaN                                                     \n",
       "169                                                NaN                                                     \n",
       "\n",
       "    Follow-up_procedure_type_of_new_tumor  \\\n",
       "60                                    NaN   \n",
       "133                                   NaN   \n",
       "140                                   NaN   \n",
       "141                                   NaN   \n",
       "142                                   NaN   \n",
       "143                                   NaN   \n",
       "144                                   NaN   \n",
       "145                                   NaN   \n",
       "146                                   NaN   \n",
       "147                                   NaN   \n",
       "148                                   NaN   \n",
       "149                                   NaN   \n",
       "150                                   NaN   \n",
       "151                                   NaN   \n",
       "152                                   NaN   \n",
       "153                                   NaN   \n",
       "154                                   NaN   \n",
       "155                                   NaN   \n",
       "156                                   NaN   \n",
       "157                                   NaN   \n",
       "158                                   NaN   \n",
       "159                                   NaN   \n",
       "160                                   NaN   \n",
       "161                                   NaN   \n",
       "162                                   NaN   \n",
       "163                                   NaN   \n",
       "164                                   NaN   \n",
       "165                                   NaN   \n",
       "166                                   NaN   \n",
       "167                                   NaN   \n",
       "168                                   NaN   \n",
       "169                                   NaN   \n",
       "\n",
       "    Follow-up_residual_tumor_after_surgery_for_new_tumor  \n",
       "60                                                 NaN    \n",
       "133                                                NaN    \n",
       "140                                                NaN    \n",
       "141                                                NaN    \n",
       "142                                                NaN    \n",
       "143                                                NaN    \n",
       "144                                                NaN    \n",
       "145                                                NaN    \n",
       "146                                                NaN    \n",
       "147                                                NaN    \n",
       "148                                                NaN    \n",
       "149                                                NaN    \n",
       "150                                                NaN    \n",
       "151                                                NaN    \n",
       "152                                                NaN    \n",
       "153                                                NaN    \n",
       "154                                                NaN    \n",
       "155                                                NaN    \n",
       "156                                                NaN    \n",
       "157                                                NaN    \n",
       "158                                                NaN    \n",
       "159                                                NaN    \n",
       "160                                                NaN    \n",
       "161                                                NaN    \n",
       "162                                                NaN    \n",
       "163                                                NaN    \n",
       "164                                                NaN    \n",
       "165                                                NaN    \n",
       "166                                                NaN    \n",
       "167                                                NaN    \n",
       "168                                                NaN    \n",
       "169                                                NaN    \n",
       "\n",
       "[32 rows x 213 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_target[(df_target['Idx'] != df_target['Case_id'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>value_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.201397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.059600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.032400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.055500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.107774</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   value_name\n",
       "0    0.201397\n",
       "1    0.059600\n",
       "2    0.032400\n",
       "3    0.055500\n",
       "4    0.107774"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bdi.preview_domain(df_target, \"Cibersort_T_cell_CD8+\").head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>value_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>YES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  value_name\n",
       "0        YES\n",
       "1         NO\n",
       "2        NaN"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bdi.preview_domain(df_source, \"Treatment_naive\").head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>value_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>b4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>b3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>b1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>b2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  value_name\n",
       "0         b4\n",
       "1         b3\n",
       "2         b1\n",
       "3         b2\n",
       "4        NaN"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bdi.preview_domain(df_target, \"Batch\").head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [source, target, similarity]\n",
       "Index: []"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bdi.match_values(df_source, df_target, (\"TP53_TP53\", \"TP53\"), method='tfidf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracting features from 1 columns...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "ac0de2bc1db24d669041f96fba73ca3c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/1 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracting features from 213 columns...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e2b62b64f9d041feb545c4ffc05a1a91",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/213 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>Pathologic_staging_primary_tumor_pt</td>\n",
       "      <td>0.408878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>Cibersort_T_cell_regulatory_(Tregs)</td>\n",
       "      <td>0.373619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>ABSOLUTE_tumor_purity</td>\n",
       "      <td>0.363810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>Cibersort_T_cell_gamma_delta</td>\n",
       "      <td>0.354802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>Metformin_treatment</td>\n",
       "      <td>0.350562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>Number_of_para-aortic_lymph_nodes_positive_for...</td>\n",
       "      <td>0.335916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>Tumor_stage_pathological</td>\n",
       "      <td>0.308310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>Number_of_pelvic_lymph_nodes_positive_for_tumo...</td>\n",
       "      <td>0.301036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>TP53</td>\n",
       "      <td>0.284595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>baseline_ancillary_studies_other_testing_type_...</td>\n",
       "      <td>0.270495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>PTEN</td>\n",
       "      <td>0.269900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>Residual_tumor</td>\n",
       "      <td>0.262319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>xCell_Plasmacytoid_dendritic_cell</td>\n",
       "      <td>0.244273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>Number_of_para-aortic_lymph_nodes_examined</td>\n",
       "      <td>0.235909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>Number_of_other_lymph_nodes_positive_for_tumor...</td>\n",
       "      <td>0.228211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>Ancillary_studies_pms2</td>\n",
       "      <td>0.227522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>xCell_T_cell_gamma_delta</td>\n",
       "      <td>0.220487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>xCell_T_cell_regulatory_(Tregs)</td>\n",
       "      <td>0.218774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>Group</td>\n",
       "      <td>0.217741</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>Number_of_pelvic_lymph_nodes_examined</td>\n",
       "      <td>0.214605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>Progeny_TNFa</td>\n",
       "      <td>0.213904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>Tumor_necrosis</td>\n",
       "      <td>0.213028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>Progeny_TGFb</td>\n",
       "      <td>0.209516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>Follow-up_procedure_type_of_new_tumor</td>\n",
       "      <td>0.206324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>Procurement_blood_collection_number_of_blood_t...</td>\n",
       "      <td>0.200936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>Ancillary_studies_other_immunohistochemistry_t...</td>\n",
       "      <td>0.200408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>Tumor_size_cm</td>\n",
       "      <td>0.193912</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>PIK3CA</td>\n",
       "      <td>0.193866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>Cibersort_T_cell_CD8+</td>\n",
       "      <td>0.193583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>Cibersort_T_cell_CD4+_naive</td>\n",
       "      <td>0.192633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>Number_of_para-aortic_lymph_nodes_positive_for...</td>\n",
       "      <td>0.187850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>Mutation_signature_SBS7a</td>\n",
       "      <td>0.183603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>Cibersort_T_cell_follicular_helper</td>\n",
       "      <td>0.182404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>Ancillary_studies_mlh1_promoter_hypermethylation</td>\n",
       "      <td>0.177488</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>Mutation_signature_SBS10a</td>\n",
       "      <td>0.173041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>Mutation_signature_SBS5</td>\n",
       "      <td>0.172607</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>MSIsensor_ratio</td>\n",
       "      <td>0.164316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>xCell_Common_lymphoid_progenitor</td>\n",
       "      <td>0.161361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>Myometrial_invasion</td>\n",
       "      <td>0.158397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>Cibersort_B_cell_plasma</td>\n",
       "      <td>0.158243</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>Current_medications_medication_name_vitamins_s...</td>\n",
       "      <td>0.156453</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>Number_of_pelvic_lymph_nodes_positive_for_tumo...</td>\n",
       "      <td>0.155242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>xCell_Macrophage_M1</td>\n",
       "      <td>0.153883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>xCell_T_cell_CD4+_Th1</td>\n",
       "      <td>0.153741</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>Proteomics_TMT_batch</td>\n",
       "      <td>Aliquot_ID</td>\n",
       "      <td>0.152065</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  source                                             target  \\\n",
       "0   Proteomics_TMT_batch                Pathologic_staging_primary_tumor_pt   \n",
       "1   Proteomics_TMT_batch                Cibersort_T_cell_regulatory_(Tregs)   \n",
       "2   Proteomics_TMT_batch                              ABSOLUTE_tumor_purity   \n",
       "3   Proteomics_TMT_batch                       Cibersort_T_cell_gamma_delta   \n",
       "4   Proteomics_TMT_batch                                Metformin_treatment   \n",
       "5   Proteomics_TMT_batch  Number_of_para-aortic_lymph_nodes_positive_for...   \n",
       "6   Proteomics_TMT_batch                           Tumor_stage_pathological   \n",
       "7   Proteomics_TMT_batch  Number_of_pelvic_lymph_nodes_positive_for_tumo...   \n",
       "8   Proteomics_TMT_batch                                               TP53   \n",
       "9   Proteomics_TMT_batch  baseline_ancillary_studies_other_testing_type_...   \n",
       "10  Proteomics_TMT_batch                                               PTEN   \n",
       "11  Proteomics_TMT_batch                                     Residual_tumor   \n",
       "12  Proteomics_TMT_batch                  xCell_Plasmacytoid_dendritic_cell   \n",
       "13  Proteomics_TMT_batch         Number_of_para-aortic_lymph_nodes_examined   \n",
       "14  Proteomics_TMT_batch  Number_of_other_lymph_nodes_positive_for_tumor...   \n",
       "15  Proteomics_TMT_batch                             Ancillary_studies_pms2   \n",
       "16  Proteomics_TMT_batch                           xCell_T_cell_gamma_delta   \n",
       "17  Proteomics_TMT_batch                    xCell_T_cell_regulatory_(Tregs)   \n",
       "18  Proteomics_TMT_batch                                              Group   \n",
       "19  Proteomics_TMT_batch              Number_of_pelvic_lymph_nodes_examined   \n",
       "20  Proteomics_TMT_batch                                       Progeny_TNFa   \n",
       "21  Proteomics_TMT_batch                                     Tumor_necrosis   \n",
       "22  Proteomics_TMT_batch                                       Progeny_TGFb   \n",
       "23  Proteomics_TMT_batch              Follow-up_procedure_type_of_new_tumor   \n",
       "24  Proteomics_TMT_batch  Procurement_blood_collection_number_of_blood_t...   \n",
       "25  Proteomics_TMT_batch  Ancillary_studies_other_immunohistochemistry_t...   \n",
       "26  Proteomics_TMT_batch                                      Tumor_size_cm   \n",
       "27  Proteomics_TMT_batch                                             PIK3CA   \n",
       "28  Proteomics_TMT_batch                              Cibersort_T_cell_CD8+   \n",
       "29  Proteomics_TMT_batch                        Cibersort_T_cell_CD4+_naive   \n",
       "30  Proteomics_TMT_batch  Number_of_para-aortic_lymph_nodes_positive_for...   \n",
       "31  Proteomics_TMT_batch                           Mutation_signature_SBS7a   \n",
       "32  Proteomics_TMT_batch                 Cibersort_T_cell_follicular_helper   \n",
       "33  Proteomics_TMT_batch   Ancillary_studies_mlh1_promoter_hypermethylation   \n",
       "34  Proteomics_TMT_batch                          Mutation_signature_SBS10a   \n",
       "35  Proteomics_TMT_batch                            Mutation_signature_SBS5   \n",
       "36  Proteomics_TMT_batch                                    MSIsensor_ratio   \n",
       "37  Proteomics_TMT_batch                   xCell_Common_lymphoid_progenitor   \n",
       "38  Proteomics_TMT_batch                                Myometrial_invasion   \n",
       "39  Proteomics_TMT_batch                            Cibersort_B_cell_plasma   \n",
       "40  Proteomics_TMT_batch  Current_medications_medication_name_vitamins_s...   \n",
       "41  Proteomics_TMT_batch  Number_of_pelvic_lymph_nodes_positive_for_tumo...   \n",
       "42  Proteomics_TMT_batch                                xCell_Macrophage_M1   \n",
       "43  Proteomics_TMT_batch                              xCell_T_cell_CD4+_Th1   \n",
       "44  Proteomics_TMT_batch                                         Aliquot_ID   \n",
       "\n",
       "    similarity  \n",
       "0     0.408878  \n",
       "1     0.373619  \n",
       "2     0.363810  \n",
       "3     0.354802  \n",
       "4     0.350562  \n",
       "5     0.335916  \n",
       "6     0.308310  \n",
       "7     0.301036  \n",
       "8     0.284595  \n",
       "9     0.270495  \n",
       "10    0.269900  \n",
       "11    0.262319  \n",
       "12    0.244273  \n",
       "13    0.235909  \n",
       "14    0.228211  \n",
       "15    0.227522  \n",
       "16    0.220487  \n",
       "17    0.218774  \n",
       "18    0.217741  \n",
       "19    0.214605  \n",
       "20    0.213904  \n",
       "21    0.213028  \n",
       "22    0.209516  \n",
       "23    0.206324  \n",
       "24    0.200936  \n",
       "25    0.200408  \n",
       "26    0.193912  \n",
       "27    0.193866  \n",
       "28    0.193583  \n",
       "29    0.192633  \n",
       "30    0.187850  \n",
       "31    0.183603  \n",
       "32    0.182404  \n",
       "33    0.177488  \n",
       "34    0.173041  \n",
       "35    0.172607  \n",
       "36    0.164316  \n",
       "37    0.161361  \n",
       "38    0.158397  \n",
       "39    0.158243  \n",
       "40    0.156453  \n",
       "41    0.155242  \n",
       "42    0.153883  \n",
       "43    0.153741  \n",
       "44    0.152065  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bdi.top_matches(df_source, columns=[\"Proteomics_TMT_batch\"], target=df_target, top_k=45)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
