{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Getting Started"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, import the `bdikit` library."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import bdikit as bdi\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this example, we are mapping data from Dou et al. (https://pubmed.ncbi.nlm.nih.gov/37567170/) to the GDC format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>case_id</th>\n",
       "      <th>age</th>\n",
       "      <th>gender</th>\n",
       "      <th>country</th>\n",
       "      <th>smoking_history</th>\n",
       "      <th>smoke_age_start</th>\n",
       "      <th>smoke_age_stop</th>\n",
       "      <th>num_smoke_per_day</th>\n",
       "      <th>num_pack_years_sm</th>\n",
       "      <th>smoking_second_hand</th>\n",
       "      <th>...</th>\n",
       "      <th>tumor_pathology_review</th>\n",
       "      <th>ESTIMATE_stromal_score</th>\n",
       "      <th>ESTIMATE_immune_score</th>\n",
       "      <th>CD3_IHC_count</th>\n",
       "      <th>stemness_score</th>\n",
       "      <th>mutation_count</th>\n",
       "      <th>neoAntigen_count</th>\n",
       "      <th>chr_instability_idx</th>\n",
       "      <th>integrated_subtype</th>\n",
       "      <th>transcriptomic_subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C3L-00977</td>\n",
       "      <td>56</td>\n",
       "      <td>Male</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Current reformed smoker, years unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yes</td>\n",
       "      <td>...</td>\n",
       "      <td>SCC 80%</td>\n",
       "      <td>6825.995755</td>\n",
       "      <td>7989.115925</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.953243</td>\n",
       "      <td>106</td>\n",
       "      <td>0</td>\n",
       "      <td>2.003654</td>\n",
       "      <td>Basal</td>\n",
       "      <td>Mesenchymal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C3L-00987</td>\n",
       "      <td>61</td>\n",
       "      <td>Male</td>\n",
       "      <td>Ukraine</td>\n",
       "      <td>Current smoker: Includes daily and non-daily s...</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20</td>\n",
       "      <td>43.0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>...</td>\n",
       "      <td>SCC 90%;SCC 90%;SCC 80%;SCC 70%</td>\n",
       "      <td>5999.793467</td>\n",
       "      <td>4772.409716</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.825330</td>\n",
       "      <td>83</td>\n",
       "      <td>0</td>\n",
       "      <td>5.205612</td>\n",
       "      <td>CIN</td>\n",
       "      <td>Classical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C3L-00994</td>\n",
       "      <td>50</td>\n",
       "      <td>Male</td>\n",
       "      <td>Ukraine</td>\n",
       "      <td>Current reformed smoker within past 15 years</td>\n",
       "      <td>16</td>\n",
       "      <td>50</td>\n",
       "      <td>6</td>\n",
       "      <td>10.2</td>\n",
       "      <td>Exposure to secondhand smoke history not avail...</td>\n",
       "      <td>...</td>\n",
       "      <td>SCC 40%;SCC 70%;SCC 40%;SCC 75%</td>\n",
       "      <td>8924.036564</td>\n",
       "      <td>8176.233903</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.664581</td>\n",
       "      <td>67</td>\n",
       "      <td>0</td>\n",
       "      <td>1.684475</td>\n",
       "      <td>Immune</td>\n",
       "      <td>Mesenchymal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C3L-00995</td>\n",
       "      <td>56</td>\n",
       "      <td>Male</td>\n",
       "      <td>Ukraine</td>\n",
       "      <td>Current reformed smoker, more than 15 years</td>\n",
       "      <td>25</td>\n",
       "      <td>56</td>\n",
       "      <td>20</td>\n",
       "      <td>31.0</td>\n",
       "      <td>Exposure to secondhand smoke history not avail...</td>\n",
       "      <td>...</td>\n",
       "      <td>SCC 90%;SCC 80%;SCC 40%;SCC 70%</td>\n",
       "      <td>8723.429667</td>\n",
       "      <td>8342.246345</td>\n",
       "      <td>70.0</td>\n",
       "      <td>0.539918</td>\n",
       "      <td>64</td>\n",
       "      <td>0</td>\n",
       "      <td>1.340483</td>\n",
       "      <td>Immune</td>\n",
       "      <td>Mesenchymal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C3L-00997</td>\n",
       "      <td>47</td>\n",
       "      <td>Male</td>\n",
       "      <td>Ukraine</td>\n",
       "      <td>Current reformed smoker, more than 15 years</td>\n",
       "      <td>12</td>\n",
       "      <td>27</td>\n",
       "      <td>20</td>\n",
       "      <td>15.0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>...</td>\n",
       "      <td>SCC 70%;SCC 70%</td>\n",
       "      <td>7025.911695</td>\n",
       "      <td>7445.251991</td>\n",
       "      <td>60.0</td>\n",
       "      <td>0.843765</td>\n",
       "      <td>129</td>\n",
       "      <td>0</td>\n",
       "      <td>3.906370</td>\n",
       "      <td>CIN</td>\n",
       "      <td>Classical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>C3L-00999</td>\n",
       "      <td>56</td>\n",
       "      <td>Male</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Current smoker: Includes daily and non-daily s...</td>\n",
       "      <td>20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20</td>\n",
       "      <td>36.0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>...</td>\n",
       "      <td>SCC 80%;SCC 70%</td>\n",
       "      <td>8510.704551</td>\n",
       "      <td>8210.549555</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0.548977</td>\n",
       "      <td>159</td>\n",
       "      <td>0</td>\n",
       "      <td>1.148834</td>\n",
       "      <td>Immune</td>\n",
       "      <td>Atypical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>C3L-01138</td>\n",
       "      <td>62</td>\n",
       "      <td>Male</td>\n",
       "      <td>Ukraine</td>\n",
       "      <td>Current smoker: Includes daily and non-daily s...</td>\n",
       "      <td>8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20</td>\n",
       "      <td>54.0</td>\n",
       "      <td>Exposure to secondhand smoke history not avail...</td>\n",
       "      <td>...</td>\n",
       "      <td>SCC 70%</td>\n",
       "      <td>6283.423855</td>\n",
       "      <td>6407.893478</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.890790</td>\n",
       "      <td>187</td>\n",
       "      <td>1</td>\n",
       "      <td>3.924982</td>\n",
       "      <td>Basal</td>\n",
       "      <td>Classical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>C3L-01237</td>\n",
       "      <td>57</td>\n",
       "      <td>Male</td>\n",
       "      <td>Ukraine</td>\n",
       "      <td>Current reformed smoker, more than 15 years</td>\n",
       "      <td>17</td>\n",
       "      <td>20</td>\n",
       "      <td>40</td>\n",
       "      <td>6.0</td>\n",
       "      <td>Exposure to secondhand smoke history not avail...</td>\n",
       "      <td>...</td>\n",
       "      <td>SCC 80%</td>\n",
       "      <td>7175.667725</td>\n",
       "      <td>5720.287055</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0.738328</td>\n",
       "      <td>141</td>\n",
       "      <td>1</td>\n",
       "      <td>4.286490</td>\n",
       "      <td>CIN</td>\n",
       "      <td>Classical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>C3L-02617</td>\n",
       "      <td>64</td>\n",
       "      <td>Male</td>\n",
       "      <td>Bulgaria</td>\n",
       "      <td>Current smoker: Includes daily and non-daily s...</td>\n",
       "      <td>20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>30</td>\n",
       "      <td>66.0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>...</td>\n",
       "      <td>SCC 90%;SCC 80%;SCC 80%;SCC 80%;SCC 80%</td>\n",
       "      <td>6346.815584</td>\n",
       "      <td>6958.573390</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.906466</td>\n",
       "      <td>470</td>\n",
       "      <td>1</td>\n",
       "      <td>4.744818</td>\n",
       "      <td>CIN</td>\n",
       "      <td>Classical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>C3L-02621</td>\n",
       "      <td>68</td>\n",
       "      <td>Male</td>\n",
       "      <td>Bulgaria</td>\n",
       "      <td>Current smoker: Includes daily and non-daily s...</td>\n",
       "      <td>38</td>\n",
       "      <td>NaN</td>\n",
       "      <td>30</td>\n",
       "      <td>45.0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>...</td>\n",
       "      <td>SCC 70%;SCC 70%;SCC 70%;SCC 70%;SCC 65%</td>\n",
       "      <td>6806.384264</td>\n",
       "      <td>8039.787184</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.754950</td>\n",
       "      <td>358</td>\n",
       "      <td>0</td>\n",
       "      <td>1.774521</td>\n",
       "      <td>CIN</td>\n",
       "      <td>Atypical</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     case_id  age gender   country  \\\n",
       "0  C3L-00977   56   Male    Russia   \n",
       "1  C3L-00987   61   Male   Ukraine   \n",
       "2  C3L-00994   50   Male   Ukraine   \n",
       "3  C3L-00995   56   Male   Ukraine   \n",
       "4  C3L-00997   47   Male   Ukraine   \n",
       "5  C3L-00999   56   Male    Russia   \n",
       "6  C3L-01138   62   Male   Ukraine   \n",
       "7  C3L-01237   57   Male   Ukraine   \n",
       "8  C3L-02617   64   Male  Bulgaria   \n",
       "9  C3L-02621   68   Male  Bulgaria   \n",
       "\n",
       "                                     smoking_history smoke_age_start  \\\n",
       "0             Current reformed smoker, years unknown             NaN   \n",
       "1  Current smoker: Includes daily and non-daily s...              18   \n",
       "2       Current reformed smoker within past 15 years              16   \n",
       "3        Current reformed smoker, more than 15 years              25   \n",
       "4        Current reformed smoker, more than 15 years              12   \n",
       "5  Current smoker: Includes daily and non-daily s...              20   \n",
       "6  Current smoker: Includes daily and non-daily s...               8   \n",
       "7        Current reformed smoker, more than 15 years              17   \n",
       "8  Current smoker: Includes daily and non-daily s...              20   \n",
       "9  Current smoker: Includes daily and non-daily s...              38   \n",
       "\n",
       "  smoke_age_stop num_smoke_per_day  num_pack_years_sm  \\\n",
       "0            NaN                20                NaN   \n",
       "1            NaN                20               43.0   \n",
       "2             50                 6               10.2   \n",
       "3             56                20               31.0   \n",
       "4             27                20               15.0   \n",
       "5            NaN                20               36.0   \n",
       "6            NaN                20               54.0   \n",
       "7             20                40                6.0   \n",
       "8            NaN                30               66.0   \n",
       "9            NaN                30               45.0   \n",
       "\n",
       "                                 smoking_second_hand  ...  \\\n",
       "0                                                Yes  ...   \n",
       "1                                                Yes  ...   \n",
       "2  Exposure to secondhand smoke history not avail...  ...   \n",
       "3  Exposure to secondhand smoke history not avail...  ...   \n",
       "4                                                Yes  ...   \n",
       "5                                                Yes  ...   \n",
       "6  Exposure to secondhand smoke history not avail...  ...   \n",
       "7  Exposure to secondhand smoke history not avail...  ...   \n",
       "8                                                Yes  ...   \n",
       "9                                                Yes  ...   \n",
       "\n",
       "                    tumor_pathology_review ESTIMATE_stromal_score  \\\n",
       "0                                  SCC 80%            6825.995755   \n",
       "1          SCC 90%;SCC 90%;SCC 80%;SCC 70%            5999.793467   \n",
       "2          SCC 40%;SCC 70%;SCC 40%;SCC 75%            8924.036564   \n",
       "3          SCC 90%;SCC 80%;SCC 40%;SCC 70%            8723.429667   \n",
       "4                          SCC 70%;SCC 70%            7025.911695   \n",
       "5                          SCC 80%;SCC 70%            8510.704551   \n",
       "6                                  SCC 70%            6283.423855   \n",
       "7                                  SCC 80%            7175.667725   \n",
       "8  SCC 90%;SCC 80%;SCC 80%;SCC 80%;SCC 80%            6346.815584   \n",
       "9  SCC 70%;SCC 70%;SCC 70%;SCC 70%;SCC 65%            6806.384264   \n",
       "\n",
       "  ESTIMATE_immune_score CD3_IHC_count stemness_score mutation_count  \\\n",
       "0           7989.115925           1.0       0.953243            106   \n",
       "1           4772.409716           0.0       0.825330             83   \n",
       "2           8176.233903           5.0       0.664581             67   \n",
       "3           8342.246345          70.0       0.539918             64   \n",
       "4           7445.251991          60.0       0.843765            129   \n",
       "5           8210.549555          30.0       0.548977            159   \n",
       "6           6407.893478          10.0       0.890790            187   \n",
       "7           5720.287055          30.0       0.738328            141   \n",
       "8           6958.573390           NaN       0.906466            470   \n",
       "9           8039.787184           NaN       0.754950            358   \n",
       "\n",
       "   neoAntigen_count chr_instability_idx integrated_subtype  \\\n",
       "0                 0            2.003654              Basal   \n",
       "1                 0            5.205612                CIN   \n",
       "2                 0            1.684475             Immune   \n",
       "3                 0            1.340483             Immune   \n",
       "4                 0            3.906370                CIN   \n",
       "5                 0            1.148834             Immune   \n",
       "6                 1            3.924982              Basal   \n",
       "7                 1            4.286490                CIN   \n",
       "8                 1            4.744818                CIN   \n",
       "9                 0            1.774521                CIN   \n",
       "\n",
       "  transcriptomic_subtype  \n",
       "0            Mesenchymal  \n",
       "1              Classical  \n",
       "2            Mesenchymal  \n",
       "3            Mesenchymal  \n",
       "4              Classical  \n",
       "5               Atypical  \n",
       "6              Classical  \n",
       "7              Classical  \n",
       "8              Classical  \n",
       "9               Atypical  \n",
       "\n",
       "[10 rows x 37 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset = pd.read_csv(\"./datasets/Huang.csv\")\n",
    "\n",
    "# columns = [\n",
    "#     \"Country\",\n",
    "#     \"Histologic_type\",\n",
    "#     \"FIGO_stage\",\n",
    "#     \"BMI\",\n",
    "#     \"Age\",\n",
    "#     \"Race\",\n",
    "#     \"Ethnicity\",\n",
    "#     \"Gender\",\n",
    "#     \"Tumor_Focality\",\n",
    "#     \"Tumor_Size_cm\",\n",
    "# ]\n",
    "\n",
    "# dataset[columns].head(10)\n",
    "dataset.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Matching the table schema to GDC standard vocabulary\n",
    "\n",
    "`bdi-kit` offers a suite of functions to help with data harmonization tasks.\n",
    "For instance, it can help with automatic discovery of one-to-one mappings between the columns in the input (source) dataset and a target dataset schema. The target schema can be either another table or a standard data vocabulary such as the GDC (Genomic Data Commons).\n",
    "\n",
    "To achieve this using `bdi-kit`, we can use the `match_schema()` function to match columns to the GDC vocabulary schema as follows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracting features from 37 columns...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "ac73cc03d9f24fbba8dec5864b71faab",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/37 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table features loaded for 734 columns\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>case_id</td>\n",
       "      <td>catalog_reference</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>age</td>\n",
       "      <td>age_at_onset</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gender</td>\n",
       "      <td>gender</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>country</td>\n",
       "      <td>country_of_birth</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>smoking_history</td>\n",
       "      <td>tobacco_smoking_status</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>smoke_age_start</td>\n",
       "      <td>years_smoked</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>smoke_age_stop</td>\n",
       "      <td>years_smoked</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>num_smoke_per_day</td>\n",
       "      <td>years_smoked</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>num_pack_years_sm</td>\n",
       "      <td>pack_years_smoked</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>smoking_second_hand</td>\n",
       "      <td>relative_smoker</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>smoking_inferred</td>\n",
       "      <td>relative_smoker</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>alcohol_consum</td>\n",
       "      <td>alcohol_intensity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>num_yrs_alc_con</td>\n",
       "      <td>undescended_testis_corrected_age</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>tumor_site_original</td>\n",
       "      <td>primary_site</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>tumor_site_curated</td>\n",
       "      <td>tumor_depth_descriptor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>tumor_focality</td>\n",
       "      <td>tumor_focality</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>tumor_size_cm</td>\n",
       "      <td>tumor_width_measurement</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>histologic_type</td>\n",
       "      <td>roots</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>histologic_grade</td>\n",
       "      <td>who_nte_grade</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>tumor_necrosis</td>\n",
       "      <td>necrosis_present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>patho_staging_pt</td>\n",
       "      <td>ensat_pathologic_t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>patho_staging_pn</td>\n",
       "      <td>ajcc_pathologic_n</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>clinic_staging_dist_metas</td>\n",
       "      <td>uicc_clinical_m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>tumor_stage</td>\n",
       "      <td>ensat_pathologic_stage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>P16</td>\n",
       "      <td>scan_tracer_used</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>HPV_inference</td>\n",
       "      <td>vascular_invasion_present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>NAT_pathology_review</td>\n",
       "      <td>staining_intensity_scale</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>tumor_pathology_review</td>\n",
       "      <td>best_overall_response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>ESTIMATE_stromal_score</td>\n",
       "      <td>transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>ESTIMATE_immune_score</td>\n",
       "      <td>transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>CD3_IHC_count</td>\n",
       "      <td>cd4_count</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>stemness_score</td>\n",
       "      <td>transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>mutation_count</td>\n",
       "      <td>mutation_codon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>neoAntigen_count</td>\n",
       "      <td>ensat_pathologic_n</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>chr_instability_idx</td>\n",
       "      <td>transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>integrated_subtype</td>\n",
       "      <td>roots</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>transcriptomic_subtype</td>\n",
       "      <td>fab_morphology_code</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       source                            target\n",
       "0                     case_id                 catalog_reference\n",
       "1                         age                      age_at_onset\n",
       "2                      gender                            gender\n",
       "3                     country                  country_of_birth\n",
       "4             smoking_history            tobacco_smoking_status\n",
       "5             smoke_age_start                      years_smoked\n",
       "6              smoke_age_stop                      years_smoked\n",
       "7           num_smoke_per_day                      years_smoked\n",
       "8           num_pack_years_sm                 pack_years_smoked\n",
       "9         smoking_second_hand                   relative_smoker\n",
       "10           smoking_inferred                   relative_smoker\n",
       "11             alcohol_consum                 alcohol_intensity\n",
       "12            num_yrs_alc_con  undescended_testis_corrected_age\n",
       "13        tumor_site_original                      primary_site\n",
       "14         tumor_site_curated            tumor_depth_descriptor\n",
       "15             tumor_focality                    tumor_focality\n",
       "16              tumor_size_cm           tumor_width_measurement\n",
       "17            histologic_type                             roots\n",
       "18           histologic_grade                     who_nte_grade\n",
       "19             tumor_necrosis                  necrosis_present\n",
       "20           patho_staging_pt                ensat_pathologic_t\n",
       "21           patho_staging_pn                 ajcc_pathologic_n\n",
       "22  clinic_staging_dist_metas                   uicc_clinical_m\n",
       "23                tumor_stage            ensat_pathologic_stage\n",
       "24                        P16                  scan_tracer_used\n",
       "25              HPV_inference         vascular_invasion_present\n",
       "26       NAT_pathology_review          staining_intensity_scale\n",
       "27     tumor_pathology_review             best_overall_response\n",
       "28     ESTIMATE_stromal_score                        transcript\n",
       "29      ESTIMATE_immune_score                        transcript\n",
       "30              CD3_IHC_count                         cd4_count\n",
       "31             stemness_score                        transcript\n",
       "32             mutation_count                    mutation_codon\n",
       "33           neoAntigen_count                ensat_pathologic_n\n",
       "34        chr_instability_idx                        transcript\n",
       "35         integrated_subtype                             roots\n",
       "36     transcriptomic_subtype               fab_morphology_code"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "column_mappings = bdi.match_schema(dataset, target=\"gdc\", method=\"ct_learning\")\n",
    "column_mappings"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generating a harmonized table\n",
    "\n",
    "After discovering a schema mapping, we can generate a new table (DataFrame) using the new column names from the GDC standard vocabulary.\n",
    "\n",
    "To do so using `bdi-kit`, we can use the function `materialize_mapping()` as follows. Note that the column headers have been renamed to the target schema."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>country_of_birth</th>\n",
       "      <th>dysplasia_type</th>\n",
       "      <th>figo_stage</th>\n",
       "      <th>hpv_positive_type</th>\n",
       "      <th>weight</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>gender</th>\n",
       "      <th>tumor_focality</th>\n",
       "      <th>tumor_depth</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>United States</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>IA</td>\n",
       "      <td>38.88</td>\n",
       "      <td>64.0</td>\n",
       "      <td>White</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>Female</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>2.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>United States</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>IA</td>\n",
       "      <td>39.76</td>\n",
       "      <td>58.0</td>\n",
       "      <td>White</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>Female</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>3.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>United States</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>IA</td>\n",
       "      <td>51.19</td>\n",
       "      <td>50.0</td>\n",
       "      <td>White</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>Female</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>4.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>United States</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>IA</td>\n",
       "      <td>32.69</td>\n",
       "      <td>75.0</td>\n",
       "      <td>White</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>Female</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>3.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>Ukraine</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>IA</td>\n",
       "      <td>29.40</td>\n",
       "      <td>75.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Female</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>4.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>Ukraine</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>II</td>\n",
       "      <td>35.42</td>\n",
       "      <td>74.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Female</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>1.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>United States</td>\n",
       "      <td>Serous</td>\n",
       "      <td>II</td>\n",
       "      <td>24.32</td>\n",
       "      <td>85.0</td>\n",
       "      <td>Black or African American</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>Female</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>3.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>Ukraine</td>\n",
       "      <td>Serous</td>\n",
       "      <td>IA</td>\n",
       "      <td>34.06</td>\n",
       "      <td>70.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Female</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>Ukraine</td>\n",
       "      <td>Serous</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>104 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    country_of_birth  dysplasia_type figo_stage  hpv_positive_type  weight  \\\n",
       "0      United States    Endometrioid         IA              38.88    64.0   \n",
       "1      United States    Endometrioid         IA              39.76    58.0   \n",
       "2      United States    Endometrioid         IA              51.19    50.0   \n",
       "3                NaN  Carcinosarcoma        NaN                NaN     NaN   \n",
       "4      United States    Endometrioid         IA              32.69    75.0   \n",
       "..               ...             ...        ...                ...     ...   \n",
       "99           Ukraine    Endometrioid         IA              29.40    75.0   \n",
       "100          Ukraine    Endometrioid         II              35.42    74.0   \n",
       "101    United States          Serous         II              24.32    85.0   \n",
       "102          Ukraine          Serous         IA              34.06    70.0   \n",
       "103          Ukraine          Serous        NaN                NaN     NaN   \n",
       "\n",
       "                          race               ethnicity  gender tumor_focality  \\\n",
       "0                        White  Not-Hispanic or Latino  Female       Unifocal   \n",
       "1                        White  Not-Hispanic or Latino  Female       Unifocal   \n",
       "2                        White  Not-Hispanic or Latino  Female       Unifocal   \n",
       "3                          NaN                     NaN     NaN            NaN   \n",
       "4                        White  Not-Hispanic or Latino  Female       Unifocal   \n",
       "..                         ...                     ...     ...            ...   \n",
       "99                         NaN                     NaN  Female       Unifocal   \n",
       "100                        NaN                     NaN  Female       Unifocal   \n",
       "101  Black or African American  Not-Hispanic or Latino  Female       Unifocal   \n",
       "102                        NaN                     NaN  Female       Unifocal   \n",
       "103                        NaN                     NaN     NaN            NaN   \n",
       "\n",
       "     tumor_depth  \n",
       "0            2.9  \n",
       "1            3.5  \n",
       "2            4.5  \n",
       "3            NaN  \n",
       "4            3.5  \n",
       "..           ...  \n",
       "99           4.2  \n",
       "100          1.5  \n",
       "101          3.8  \n",
       "102          5.0  \n",
       "103          NaN  \n",
       "\n",
       "[104 rows x 10 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bdi.materialize_mapping(dataset, column_mappings)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generating a harmonized table with value mappings\n",
    "\n",
    "`bdi-kit` can also help with translation of the values from the source table to the target standard format.\n",
    "\n",
    "To this end, `bdi-kit` provides the function `match_values()` that automatically creates value mappings for each string column.\n",
    "The output of `match_values()` can be fed to `materialize_mapping()` which materialized the final target using both schema and value mappings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>country_of_birth</th>\n",
       "      <th>dysplasia_type</th>\n",
       "      <th>figo_stage</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>gender</th>\n",
       "      <th>tumor_focality</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>United States</td>\n",
       "      <td>None</td>\n",
       "      <td>Stage IA</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>female</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>United States</td>\n",
       "      <td>None</td>\n",
       "      <td>Stage IA</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>female</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>United States</td>\n",
       "      <td>None</td>\n",
       "      <td>Stage IA</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>female</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Esophageal Mucosa Columnar Dysplasia</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>United States</td>\n",
       "      <td>None</td>\n",
       "      <td>Stage IA</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>female</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>Ukraine</td>\n",
       "      <td>None</td>\n",
       "      <td>Stage IA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>female</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>Ukraine</td>\n",
       "      <td>None</td>\n",
       "      <td>Stage III</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>female</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>United States</td>\n",
       "      <td>None</td>\n",
       "      <td>Stage III</td>\n",
       "      <td>black or african american</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>female</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>Ukraine</td>\n",
       "      <td>None</td>\n",
       "      <td>Stage IA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>female</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>Ukraine</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>104 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    country_of_birth                        dysplasia_type figo_stage  \\\n",
       "0      United States                                  None   Stage IA   \n",
       "1      United States                                  None   Stage IA   \n",
       "2      United States                                  None   Stage IA   \n",
       "3                NaN  Esophageal Mucosa Columnar Dysplasia        NaN   \n",
       "4      United States                                  None   Stage IA   \n",
       "..               ...                                   ...        ...   \n",
       "99           Ukraine                                  None   Stage IA   \n",
       "100          Ukraine                                  None  Stage III   \n",
       "101    United States                                  None  Stage III   \n",
       "102          Ukraine                                  None   Stage IA   \n",
       "103          Ukraine                                  None        NaN   \n",
       "\n",
       "                          race               ethnicity  gender tumor_focality  \n",
       "0                        white  not hispanic or latino  female       Unifocal  \n",
       "1                        white  not hispanic or latino  female       Unifocal  \n",
       "2                        white  not hispanic or latino  female       Unifocal  \n",
       "3                          NaN                     NaN     NaN            NaN  \n",
       "4                        white  not hispanic or latino  female       Unifocal  \n",
       "..                         ...                     ...     ...            ...  \n",
       "99                         NaN                     NaN  female       Unifocal  \n",
       "100                        NaN                     NaN  female       Unifocal  \n",
       "101  black or african american  not hispanic or latino  female       Unifocal  \n",
       "102                        NaN                     NaN  female       Unifocal  \n",
       "103                        NaN                     NaN     NaN            NaN  \n",
       "\n",
       "[104 rows x 7 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "value_mappings = bdi.match_values(dataset, column_mapping=column_mappings, target=\"gdc\", method=\"tfidf\")\n",
    "bdi.materialize_mapping(dataset, value_mappings)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Verifying the schema mappings\n",
    "\n",
    "Sometimes the mappings generated automatically may be incorrect or you may to want verify them individually.\n",
    "To verify the suggested column mappings, `bdi-kit` offers additional APIs to visualize the data and make any modifications when necessary. \n",
    "\n",
    "For this example, we will use the column `Histologic_type`. We can start by exploring the columns most similar to `Histologic_type`. \n",
    "\n",
    "For this, we can use the `top_matches()` function. Here, we notice that `primary_diagnosis` could be a potential target column.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Some weights of RobertaModel were not initialized from the model checkpoint at roberta-base and are newly initialized: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']\n",
      "You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.\n",
      "100%|██████████| 1/1 [00:00<00:00, 14.70it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table features extracted from 1 columns\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 734/734 [00:58<00:00, 12.48it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table features extracted from 734 columns\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>described_cases</td>\n",
       "      <td>0.589956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>slide_images</td>\n",
       "      <td>0.587552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>history_of_tumor_type</td>\n",
       "      <td>0.574640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>primary_diagnosis</td>\n",
       "      <td>0.573583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>additional_pathology_findings</td>\n",
       "      <td>0.562278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>pathology_details</td>\n",
       "      <td>0.562007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>pathology_reports</td>\n",
       "      <td>0.547307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>relationship_primary_diagnosis</td>\n",
       "      <td>0.524285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>diagnoses</td>\n",
       "      <td>0.519854</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>family_histories</td>\n",
       "      <td>0.516649</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            source                          target  similarity\n",
       "0  Histologic_type                 described_cases    0.589956\n",
       "1  Histologic_type                    slide_images    0.587552\n",
       "2  Histologic_type           history_of_tumor_type    0.574640\n",
       "3  Histologic_type               primary_diagnosis    0.573583\n",
       "4  Histologic_type   additional_pathology_findings    0.562278\n",
       "5  Histologic_type               pathology_details    0.562007\n",
       "6  Histologic_type               pathology_reports    0.547307\n",
       "7  Histologic_type  relationship_primary_diagnosis    0.524285\n",
       "8  Histologic_type                       diagnoses    0.519854\n",
       "9  Histologic_type                family_histories    0.516649"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hist_type_matches = bdi.top_matches(dataset, columns=[\"Histologic_type\"], target=\"gdc\")\n",
    "hist_type_matches"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Viewing the column domains\n",
    "\n",
    "To verify that `primary_diagnosis` is a good target column, we view and compare the domains of each column using the `preview_domain()` function. For the source table, it returns the list of unique values in the source column. For the GDC target, it returns the list of unique valid values that a column can have.\n",
    "\n",
    "Here we see that the values seem to be related."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>value_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Endometrioid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Carcinosarcoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Serous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Clear cell</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       value_name\n",
       "0    Endometrioid\n",
       "1  Carcinosarcoma\n",
       "2          Serous\n",
       "3      Clear cell"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bdi.preview_domain(dataset, \"Histologic_type\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>value_name</th>\n",
       "      <th>value_description</th>\n",
       "      <th>column_description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Abdominal desmoid</td>\n",
       "      <td>An insidious poorly circumscribed neoplasm ari...</td>\n",
       "      <td>Text term used to describe the patient's histo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Abdominal fibromatosis</td>\n",
       "      <td>An insidious poorly circumscribed neoplasm ari...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Achromic nevus</td>\n",
       "      <td>A benign nevus characterized by the absence of...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Acidophil adenocarcinoma</td>\n",
       "      <td>A malignant epithelial neoplasm of the anterio...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Acidophil adenoma</td>\n",
       "      <td>An epithelial neoplasm of the anterior pituita...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2620</th>\n",
       "      <td>Wolffian duct tumor</td>\n",
       "      <td>An epithelial neoplasm of the female reproduct...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2621</th>\n",
       "      <td>Xanthofibroma</td>\n",
       "      <td>A benign neoplasm composed of fibroblastic spi...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2622</th>\n",
       "      <td>Yolk sac tumor</td>\n",
       "      <td>A non-seminomatous malignant germ cell tumor c...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2623</th>\n",
       "      <td>Unknown</td>\n",
       "      <td>Not known, not observed, not recorded, or refu...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2624</th>\n",
       "      <td>Not Reported</td>\n",
       "      <td>Not provided or available.</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2625 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    value_name  \\\n",
       "0            Abdominal desmoid   \n",
       "1       Abdominal fibromatosis   \n",
       "2               Achromic nevus   \n",
       "3     Acidophil adenocarcinoma   \n",
       "4            Acidophil adenoma   \n",
       "...                        ...   \n",
       "2620       Wolffian duct tumor   \n",
       "2621             Xanthofibroma   \n",
       "2622            Yolk sac tumor   \n",
       "2623                   Unknown   \n",
       "2624              Not Reported   \n",
       "\n",
       "                                      value_description  \\\n",
       "0     An insidious poorly circumscribed neoplasm ari...   \n",
       "1     An insidious poorly circumscribed neoplasm ari...   \n",
       "2     A benign nevus characterized by the absence of...   \n",
       "3     A malignant epithelial neoplasm of the anterio...   \n",
       "4     An epithelial neoplasm of the anterior pituita...   \n",
       "...                                                 ...   \n",
       "2620  An epithelial neoplasm of the female reproduct...   \n",
       "2621  A benign neoplasm composed of fibroblastic spi...   \n",
       "2622  A non-seminomatous malignant germ cell tumor c...   \n",
       "2623  Not known, not observed, not recorded, or refu...   \n",
       "2624                         Not provided or available.   \n",
       "\n",
       "                                     column_description  \n",
       "0     Text term used to describe the patient's histo...  \n",
       "1                                                        \n",
       "2                                                        \n",
       "3                                                        \n",
       "4                                                        \n",
       "...                                                 ...  \n",
       "2620                                                     \n",
       "2621                                                     \n",
       "2622                                                     \n",
       "2623                                                     \n",
       "2624                                                     \n",
       "\n",
       "[2625 rows x 3 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bdi.preview_domain(\"gdc\", \"primary_diagnosis\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since `primary_diagnosis` looks like a correct match for `Histologic_type`, we can modify the `column_mappings` variable directly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Country</td>\n",
       "      <td>country_of_birth</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>primary_diagnosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FIGO_stage</td>\n",
       "      <td>figo_stage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BMI</td>\n",
       "      <td>hpv_positive_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Age</td>\n",
       "      <td>weight</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Race</td>\n",
       "      <td>race</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Ethnicity</td>\n",
       "      <td>ethnicity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Gender</td>\n",
       "      <td>gender</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Tumor_Focality</td>\n",
       "      <td>tumor_focality</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Tumor_Size_cm</td>\n",
       "      <td>tumor_depth</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            source             target\n",
       "0          Country   country_of_birth\n",
       "1  Histologic_type  primary_diagnosis\n",
       "2       FIGO_stage         figo_stage\n",
       "3              BMI  hpv_positive_type\n",
       "4              Age             weight\n",
       "5             Race               race\n",
       "6        Ethnicity          ethnicity\n",
       "7           Gender             gender\n",
       "8   Tumor_Focality     tumor_focality\n",
       "9    Tumor_Size_cm        tumor_depth"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "column_mappings.loc[column_mappings[\"source\"] == \"Histologic_type\", \"target\"] = \"primary_diagnosis\"\n",
    "column_mappings"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Finding correct value mappings\n",
    "\n",
    "After finding the correct column, we need to find appropriate value mappings. \n",
    "Using `match_values()`, we can inspect what the possible value mappings for this would look like after the harmonization.\n",
    "\n",
    "`bdi-kit` implements multiple methods for value mapping discovery, including:\n",
    "\n",
    " - `edit_distance` - Computes value similarities using Levenstein's edit distance measure.\n",
    " - `tfidf` - A method based on tf-idf importance weighting computed over charcter n-grams.\n",
    " - `embeddings` - Uses BERT word embeddings to compute \"semantic similarity\" between the values.\n",
    "\n",
    "To specify a value mapping approach, we can pass the `method` parameter."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>Carcinosarcoma, NOS</td>\n",
       "      <td>0.848485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Clear cell</td>\n",
       "      <td>Clear cell adenoma</td>\n",
       "      <td>0.714286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>Stromal endometriosis</td>\n",
       "      <td>0.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Serous</td>\n",
       "      <td>Neuronevus</td>\n",
       "      <td>0.625000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source                 target  similarity\n",
       "0  Carcinosarcoma    Carcinosarcoma, NOS    0.848485\n",
       "1      Clear cell     Clear cell adenoma    0.714286\n",
       "2    Endometrioid  Stromal endometriosis    0.666667\n",
       "3          Serous             Neuronevus    0.625000"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bdi.match_values(\n",
    "    dataset, column_mapping=(\"Histologic_type\", \"primary_diagnosis\"), target=\"gdc\", method=\"edit_distance\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>Carcinosarcoma, NOS</td>\n",
       "      <td>0.969</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>Endometrioid adenoma, NOS</td>\n",
       "      <td>0.897</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Clear cell</td>\n",
       "      <td>Clear cell adenoma</td>\n",
       "      <td>0.853</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Serous</td>\n",
       "      <td>Serous carcinoma, NOS</td>\n",
       "      <td>0.755</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source                     target  similarity\n",
       "0  Carcinosarcoma        Carcinosarcoma, NOS       0.969\n",
       "1    Endometrioid  Endometrioid adenoma, NOS       0.897\n",
       "2      Clear cell         Clear cell adenoma       0.853\n",
       "3          Serous      Serous carcinoma, NOS       0.755"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bdi.match_values(\n",
    "    dataset, column_mapping=(\"Histologic_type\", \"primary_diagnosis\"), target=\"gdc\", method=\"tfidf\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>Carcinofibroma</td>\n",
       "      <td>0.919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>Endometrioid cystadenocarcinoma</td>\n",
       "      <td>0.810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Clear cell</td>\n",
       "      <td>Clear cell carcinoma</td>\n",
       "      <td>0.760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Serous</td>\n",
       "      <td>Serous cystoma</td>\n",
       "      <td>0.661</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source                           target  similarity\n",
       "0  Carcinosarcoma                   Carcinofibroma       0.919\n",
       "1    Endometrioid  Endometrioid cystadenocarcinoma       0.810\n",
       "2      Clear cell             Clear cell carcinoma       0.760\n",
       "3          Serous                   Serous cystoma       0.661"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bdi.match_values(\n",
    "    dataset, column_mapping=(\"Histologic_type\", \"primary_diagnosis\"), target=\"gdc\", method=\"embedding\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>Carcinosarcoma, NOS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Clear cell</td>\n",
       "      <td>Clear cell adenocarcinoma, NOS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>Endometrioid carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Serous</td>\n",
       "      <td>Serous cystadenocarcinoma</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source                          target\n",
       "0  Carcinosarcoma             Carcinosarcoma, NOS\n",
       "1      Clear cell  Clear cell adenocarcinoma, NOS\n",
       "2    Endometrioid          Endometrioid carcinoma\n",
       "3          Serous       Serous cystadenocarcinoma"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hist_type_vmap = pd.DataFrame(\n",
    "    columns=[\"source\", \"target\"],\n",
    "    data=[\n",
    "        (\"Carcinosarcoma\", \"Carcinosarcoma, NOS\"),\n",
    "        (\"Clear cell\", \"Clear cell adenocarcinoma, NOS\"),\n",
    "        (\"Endometrioid\", \"Endometrioid carcinoma\"),\n",
    "        (\"Serous\", \"Serous cystadenocarcinoma\"),\n",
    "    ],\n",
    ")\n",
    "hist_type_vmap"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Verifying multiple value mappings at once\n",
    "\n",
    "Besides verifying value mappings individually, you can also do it for all column mappings at once."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gender => gender\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Male</td>\n",
       "      <td>male</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Female</td>\n",
       "      <td>female</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   source  target  similarity\n",
       "0    Male    male         1.0\n",
       "1  Female  female         1.0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "country => country_of_birth\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Russia</td>\n",
       "      <td>Russia</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Ukraine</td>\n",
       "      <td>Ukraine</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Bulgaria</td>\n",
       "      <td>Bulgaria</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Poland</td>\n",
       "      <td>Poland</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>United States</td>\n",
       "      <td>United States</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>China</td>\n",
       "      <td>China</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Armenia</td>\n",
       "      <td>Armenia</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>TSS did not collect</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                source         target  similarity\n",
       "0               Russia         Russia         1.0\n",
       "1              Ukraine        Ukraine         1.0\n",
       "2             Bulgaria       Bulgaria         1.0\n",
       "3               Poland         Poland         1.0\n",
       "4        United States  United States         1.0\n",
       "5                China          China         1.0\n",
       "6              Armenia        Armenia         1.0\n",
       "7  TSS did not collect           None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "smoking_history => tobacco_smoking_status\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Current reformed smoker, more than 15 years</td>\n",
       "      <td>Current Reformed Smoker for &gt; 15 yrs</td>\n",
       "      <td>0.823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Current reformed smoker, years unknown</td>\n",
       "      <td>Current Reformed Smoker for &gt; 15 yrs</td>\n",
       "      <td>0.756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Current reformed smoker within past 15 years</td>\n",
       "      <td>Current Reformed Smoker for &gt; 15 yrs</td>\n",
       "      <td>0.719</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Lifelong non-smoker: Less than 100 cigarettes ...</td>\n",
       "      <td>Lifelong Non-Smoker</td>\n",
       "      <td>0.697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Smoking history not available</td>\n",
       "      <td>Smoking history not documented</td>\n",
       "      <td>0.564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Current smoker: Includes daily and non-daily s...</td>\n",
       "      <td>Current Smoker</td>\n",
       "      <td>0.506</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              source  \\\n",
       "0        Current reformed smoker, more than 15 years   \n",
       "1             Current reformed smoker, years unknown   \n",
       "2       Current reformed smoker within past 15 years   \n",
       "3  Lifelong non-smoker: Less than 100 cigarettes ...   \n",
       "4                      Smoking history not available   \n",
       "5  Current smoker: Includes daily and non-daily s...   \n",
       "\n",
       "                                 target  similarity  \n",
       "0  Current Reformed Smoker for > 15 yrs       0.823  \n",
       "1  Current Reformed Smoker for > 15 yrs       0.756  \n",
       "2  Current Reformed Smoker for > 15 yrs       0.719  \n",
       "3                   Lifelong Non-Smoker       0.697  \n",
       "4        Smoking history not documented       0.564  \n",
       "5                        Current Smoker       0.506  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "smoking_second_hand => relative_smoker\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>No or minimal exposure to secondhand smoke</td>\n",
       "      <td>No</td>\n",
       "      <td>0.410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Exposure to secondhand smoke history not avail...</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>0.352</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              source        target  similarity\n",
       "0                                                Yes           Yes       1.000\n",
       "1         No or minimal exposure to secondhand smoke            No       0.410\n",
       "2  Exposure to secondhand smoke history not avail...  Not Reported       0.352"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "smoking_inferred => relative_smoker\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>weak_evidence</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>strong_evidence</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            source target similarity\n",
       "0    weak_evidence   None       None\n",
       "1  strong_evidence   None       None"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "alcohol_consum => alcohol_intensity\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Lifelong non-drinker</td>\n",
       "      <td>Lifelong Non-Drinker</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Consumed alcohol in the past, but currently a ...</td>\n",
       "      <td>Non-Drinker</td>\n",
       "      <td>0.437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Alcohol consumption more than 2 drinks per day...</td>\n",
       "      <td>Non-Drinker</td>\n",
       "      <td>0.370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Alcohol consumption equal to or less than 2 dr...</td>\n",
       "      <td>Occasional Drinker</td>\n",
       "      <td>0.360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Alcohol consumption history not available</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>nan</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              source                target  \\\n",
       "0                               Lifelong non-drinker  Lifelong Non-Drinker   \n",
       "1  Consumed alcohol in the past, but currently a ...           Non-Drinker   \n",
       "2  Alcohol consumption more than 2 drinks per day...           Non-Drinker   \n",
       "3  Alcohol consumption equal to or less than 2 dr...    Occasional Drinker   \n",
       "4          Alcohol consumption history not available                  None   \n",
       "5                                                nan                  None   \n",
       "\n",
       "   similarity  \n",
       "0       1.000  \n",
       "1       0.437  \n",
       "2       0.370  \n",
       "3       0.360  \n",
       "4         NaN  \n",
       "5         NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "tumor_site_original => primary_site\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Floor of mouth</td>\n",
       "      <td>Floor of mouth</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Larynx</td>\n",
       "      <td>Larynx</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Oropharynx</td>\n",
       "      <td>Oropharynx</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Base of tongue</td>\n",
       "      <td>Base of tongue</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Lip</td>\n",
       "      <td>Lip</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Tonsil</td>\n",
       "      <td>Tonsil</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Tongue</td>\n",
       "      <td>Base of tongue</td>\n",
       "      <td>0.733</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Oral cavity</td>\n",
       "      <td>Other and ill-defined sites in lip, oral cavit...</td>\n",
       "      <td>0.542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Other</td>\n",
       "      <td>Other and unspecified parts of mouth</td>\n",
       "      <td>0.428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Alveolar ridge</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Buccal mucosa</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            source                                             target  \\\n",
       "0   Floor of mouth                                     Floor of mouth   \n",
       "1           Larynx                                             Larynx   \n",
       "2       Oropharynx                                         Oropharynx   \n",
       "3   Base of tongue                                     Base of tongue   \n",
       "4              Lip                                                Lip   \n",
       "5           Tonsil                                             Tonsil   \n",
       "6           Tongue                                     Base of tongue   \n",
       "7      Oral cavity  Other and ill-defined sites in lip, oral cavit...   \n",
       "8            Other               Other and unspecified parts of mouth   \n",
       "9   Alveolar ridge                                               None   \n",
       "10   Buccal mucosa                                               None   \n",
       "\n",
       "    similarity  \n",
       "0        1.000  \n",
       "1        1.000  \n",
       "2        1.000  \n",
       "3        1.000  \n",
       "4        1.000  \n",
       "5        1.000  \n",
       "6        0.733  \n",
       "7        0.542  \n",
       "8        0.428  \n",
       "9          NaN  \n",
       "10         NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "tumor_site_curated => tumor_depth_descriptor\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Lip</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Oropharynx</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Oral cavity</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Hypopharynx</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Larynx</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        source target similarity\n",
       "0          Lip   None       None\n",
       "1   Oropharynx   None       None\n",
       "2  Oral cavity   None       None\n",
       "3  Hypopharynx   None       None\n",
       "4       Larynx   None       None"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "tumor_focality => tumor_focality\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Unifocal</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Multifocal</td>\n",
       "      <td>Multifocal</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       source      target  similarity\n",
       "0    Unifocal    Unifocal         1.0\n",
       "1  Multifocal  Multifocal         1.0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "histologic_grade => who_nte_grade\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>G3 Poorly differentiated</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>0.431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G2 Moderately differentiated</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>0.358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>G1 Well differentiated</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>0.310</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         source        target  similarity\n",
       "0      G3 Poorly differentiated  Not Reported       0.431\n",
       "1  G2 Moderately differentiated  Not Reported       0.358\n",
       "2        G1 Well differentiated  Not Reported       0.310"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "tumor_necrosis => necrosis_present\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Not identified</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Present</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source target similarity\n",
       "0  Not identified   None       None\n",
       "1         Present   None       None"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "patho_staging_pt => ensat_pathologic_t\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T2</td>\n",
       "      <td>T2</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pT3</td>\n",
       "      <td>T3</td>\n",
       "      <td>0.703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pT1</td>\n",
       "      <td>T1</td>\n",
       "      <td>0.693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pT2</td>\n",
       "      <td>T2</td>\n",
       "      <td>0.661</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pT4</td>\n",
       "      <td>T4</td>\n",
       "      <td>0.658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>T4a</td>\n",
       "      <td>T4</td>\n",
       "      <td>0.567</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>pT4a</td>\n",
       "      <td>T4</td>\n",
       "      <td>0.476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>pT1b</td>\n",
       "      <td>T1</td>\n",
       "      <td>0.475</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  source target  similarity\n",
       "0     T2     T2       1.000\n",
       "1    pT3     T3       0.703\n",
       "2    pT1     T1       0.693\n",
       "3    pT2     T2       0.661\n",
       "4    pT4     T4       0.658\n",
       "5    T4a     T4       0.567\n",
       "6   pT4a     T4       0.476\n",
       "7   pT1b     T1       0.475"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "patho_staging_pn => ajcc_pathologic_n\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>N3b</td>\n",
       "      <td>N3b</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>N1</td>\n",
       "      <td>N1</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pN2c</td>\n",
       "      <td>N2c</td>\n",
       "      <td>0.828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pN3a</td>\n",
       "      <td>N3a</td>\n",
       "      <td>0.828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pN2b</td>\n",
       "      <td>N2b</td>\n",
       "      <td>0.819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>pN3b</td>\n",
       "      <td>N3b</td>\n",
       "      <td>0.806</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>pNx</td>\n",
       "      <td>NX</td>\n",
       "      <td>0.706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>pN2</td>\n",
       "      <td>N2</td>\n",
       "      <td>0.617</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>pN3</td>\n",
       "      <td>N3</td>\n",
       "      <td>0.617</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>pN0</td>\n",
       "      <td>N0</td>\n",
       "      <td>0.611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>PN1</td>\n",
       "      <td>N1</td>\n",
       "      <td>0.534</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   source target  similarity\n",
       "0     N3b    N3b       1.000\n",
       "1      N1     N1       1.000\n",
       "2    pN2c    N2c       0.828\n",
       "3    pN3a    N3a       0.828\n",
       "4    pN2b    N2b       0.819\n",
       "5    pN3b    N3b       0.806\n",
       "6     pNx     NX       0.706\n",
       "7     pN2     N2       0.617\n",
       "8     pN3     N3       0.617\n",
       "9     pN0     N0       0.611\n",
       "10    PN1     N1       0.534"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "clinic_staging_dist_metas => uicc_clinical_m\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>m0</td>\n",
       "      <td>M0</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cM0</td>\n",
       "      <td>cM0 (i+)</td>\n",
       "      <td>0.891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>cMX</td>\n",
       "      <td>MX</td>\n",
       "      <td>0.674</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Staging Incomplete</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>nan</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               source    target  similarity\n",
       "0                  m0        M0       1.000\n",
       "1                 cM0  cM0 (i+)       0.891\n",
       "2                 cMX        MX       0.674\n",
       "3  Staging Incomplete      None         NaN\n",
       "4                 nan      None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "tumor_stage => ensat_pathologic_stage\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Stage III</td>\n",
       "      <td>Stage III</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Stage II</td>\n",
       "      <td>Stage II</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Stage IV</td>\n",
       "      <td>Stage IV</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Stage I</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      source     target  similarity\n",
       "0  Stage III  Stage III         1.0\n",
       "1   Stage II   Stage II         1.0\n",
       "2   Stage IV   Stage IV         1.0\n",
       "3    Stage I    Stage I         1.0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "P16 => scan_tracer_used\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Not Evaluated</td>\n",
       "      <td>Acetate</td>\n",
       "      <td>0.326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Positive (&gt;70% nuclear and cytoplasmic staining)</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Unknown</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Negative</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>nan</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             source   target  similarity\n",
       "0                                     Not Evaluated  Acetate       0.326\n",
       "1  Positive (>70% nuclear and cytoplasmic staining)     None         NaN\n",
       "2                                           Unknown     None         NaN\n",
       "3                                          Negative     None         NaN\n",
       "4                                               nan     None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "HPV_inference => vascular_invasion_present\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NO</td>\n",
       "      <td>No</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>YES</td>\n",
       "      <td>Yes</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  source target  similarity\n",
       "0     NO     No         1.0\n",
       "1    YES    Yes         1.0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "NAT_pathology_review => staining_intensity_scale\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Normal 60%%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Normal 15%;Normal 10%%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>nan</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Normal 35%%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Normal Cartilage 0%%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Normal 25%%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Soft Tissue  No EPI%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Normal 20%%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Muscle 0%%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Normal 50%%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Normal 40%;Normal 20%;Normal 15%%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>N 35%%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>N 5%%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Normal EPI 10%;Normal EPI 10%%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Normal 70$ EPI%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Normal Thyroid;Normal Thyroid;Normal No EPI%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Normal 40%%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Normal 5%%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Normal 1%%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Normal 20%;SCC 75%%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>SCC 30%;SCC 20%%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>N 3%%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>N 25%%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Normal Cartilage%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Normal 0% EPI%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>N 15%%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Soft Tissue  No EPI;Normal No EPI%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Normal 20%;Normal No EPI;Normal 5%%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>N 0%%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>N 20%%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Normal 20% EPI%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Normal 15%;Normal 5%%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Normal 30%%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Normal 10%%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Hyperplasia 70%%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>N lymphoid B%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Normal no EPI%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>SCC 75%%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Normal 10%;Normal 15%;SCC 90%%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>Normal 0%%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>N 10%%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>Normal 15%%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>N 30%%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>SCC 70%;Normal 10%;Normal 30%%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>Normal 10%;Normal 10%;Normal 10%%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>SCC 70%%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>Normal%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          source target similarity\n",
       "0                                    Normal 60%%   None       None\n",
       "1                         Normal 15%;Normal 10%%   None       None\n",
       "2                                            nan   None       None\n",
       "3                                    Normal 35%%   None       None\n",
       "4                           Normal Cartilage 0%%   None       None\n",
       "5                                    Normal 25%%   None       None\n",
       "6                           Soft Tissue  No EPI%   None       None\n",
       "7                                    Normal 20%%   None       None\n",
       "8                                     Muscle 0%%   None       None\n",
       "9                                    Normal 50%%   None       None\n",
       "10             Normal 40%;Normal 20%;Normal 15%%   None       None\n",
       "11                                        N 35%%   None       None\n",
       "12                                         N 5%%   None       None\n",
       "13                Normal EPI 10%;Normal EPI 10%%   None       None\n",
       "14                               Normal 70$ EPI%   None       None\n",
       "15  Normal Thyroid;Normal Thyroid;Normal No EPI%   None       None\n",
       "16                                   Normal 40%%   None       None\n",
       "17                                    Normal 5%%   None       None\n",
       "18                                    Normal 1%%   None       None\n",
       "19                           Normal 20%;SCC 75%%   None       None\n",
       "20                              SCC 30%;SCC 20%%   None       None\n",
       "21                                         N 3%%   None       None\n",
       "22                                        N 25%%   None       None\n",
       "23                             Normal Cartilage%   None       None\n",
       "24                                Normal 0% EPI%   None       None\n",
       "25                                        N 15%%   None       None\n",
       "26            Soft Tissue  No EPI;Normal No EPI%   None       None\n",
       "27           Normal 20%;Normal No EPI;Normal 5%%   None       None\n",
       "28                                         N 0%%   None       None\n",
       "29                                        N 20%%   None       None\n",
       "30                               Normal 20% EPI%   None       None\n",
       "31                         Normal 15%;Normal 5%%   None       None\n",
       "32                                   Normal 30%%   None       None\n",
       "33                                   Normal 10%%   None       None\n",
       "34                              Hyperplasia 70%%   None       None\n",
       "35                                 N lymphoid B%   None       None\n",
       "36                                Normal no EPI%   None       None\n",
       "37                                      SCC 75%%   None       None\n",
       "38                Normal 10%;Normal 15%;SCC 90%%   None       None\n",
       "39                                    Normal 0%%   None       None\n",
       "40                                        N 10%%   None       None\n",
       "41                                   Normal 15%%   None       None\n",
       "42                                        N 30%%   None       None\n",
       "43                SCC 70%;Normal 10%;Normal 30%%   None       None\n",
       "44             Normal 10%;Normal 10%;Normal 10%%   None       None\n",
       "45                                      SCC 70%%   None       None\n",
       "46                                       Normal%   None       None"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "neoAntigen_count => ensat_pathologic_n\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>N0</td>\n",
       "      <td>0.54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>N1</td>\n",
       "      <td>0.54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>3</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  source target  similarity\n",
       "0      0     N0        0.54\n",
       "1      1     N1        0.54\n",
       "2      5   None         NaN\n",
       "3      4   None         NaN\n",
       "4      2   None         NaN\n",
       "5      3   None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "transcriptomic_subtype => fab_morphology_code\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Classical</td>\n",
       "      <td>Not Classified</td>\n",
       "      <td>0.53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Mesenchymal</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Atypical</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Basal</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        source          target  similarity\n",
       "0    Classical  Not Classified        0.53\n",
       "1  Mesenchymal            None         NaN\n",
       "2     Atypical            None         NaN\n",
       "3        Basal            None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "mappings = bdi.match_values(\n",
    "    dataset,\n",
    "    column_mapping=column_mappings,\n",
    "    target=\"gdc\",\n",
    "    method=\"tfidf\",\n",
    ")\n",
    "\n",
    "for mapping in mappings:\n",
    "    print(f\"{mapping.attrs['source']} => {mapping.attrs['target']}\")\n",
    "    display(mapping)\n",
    "    print(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>value_name</th>\n",
       "      <th>value_description</th>\n",
       "      <th>column_description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>G1</td>\n",
       "      <td>A morphologic qualifier indicating that a canc...</td>\n",
       "      <td>The WHO (World Health Organization) grading cl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G2</td>\n",
       "      <td>A morphologic qualifier indicating that a canc...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>G3</td>\n",
       "      <td>A morphologic qualifier indicating that a canc...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GX</td>\n",
       "      <td>A morphologic qualifier indicating that the gr...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Unknown</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Not Reported</td>\n",
       "      <td>Not provided or available.</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     value_name                                  value_description  \\\n",
       "0            G1  A morphologic qualifier indicating that a canc...   \n",
       "1            G2  A morphologic qualifier indicating that a canc...   \n",
       "2            G3  A morphologic qualifier indicating that a canc...   \n",
       "3            GX  A morphologic qualifier indicating that the gr...   \n",
       "4       Unknown                                                      \n",
       "5  Not Reported                         Not provided or available.   \n",
       "\n",
       "                                  column_description  \n",
       "0  The WHO (World Health Organization) grading cl...  \n",
       "1                                                     \n",
       "2                                                     \n",
       "3                                                     \n",
       "4                                                     \n",
       "5                                                     "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bdi.preview_domain(\"gdc\", \"who_nte_grade\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fixing remaining value mappings\n",
    "\n",
    "We need fix a few value mappings:\n",
    "- Race\n",
    "- Ethnicity\n",
    "- Tumor_Site\n",
    "\n",
    "For race, we need to fix: `nan` -> `american indian or alaska native`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>White</td>\n",
       "      <td>white</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Asian</td>\n",
       "      <td>asian</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Not Reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Black or African American</td>\n",
       "      <td>black or african american</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>nan</td>\n",
       "      <td>american indian or alaska native</td>\n",
       "      <td>0.359</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      source                            target  similarity\n",
       "0                      White                             white       1.000\n",
       "1                      Asian                             asian       1.000\n",
       "2               Not Reported                      not reported       1.000\n",
       "3  Black or African American         black or african american       1.000\n",
       "4                        nan  american indian or alaska native       0.359"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "race_vmap = bdi.match_values(\n",
    "    dataset,\n",
    "    column_mapping=(\"Race\", \"race\"),\n",
    "    target=\"gdc\",\n",
    "    method=\"tfidf\",\n",
    ")\n",
    "race_vmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>White</td>\n",
       "      <td>white</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Asian</td>\n",
       "      <td>asian</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Not Reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Black or African American</td>\n",
       "      <td>black or african american</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      source                     target  similarity\n",
       "0                      White                      white         1.0\n",
       "1                      Asian                      asian         1.0\n",
       "2               Not Reported               not reported         1.0\n",
       "3  Black or African American  black or african american         1.0"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "race_vmap = race_vmap[race_vmap[\"similarity\"] >= 1.0]\n",
    "race_vmap"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For `Ethnicity`, we need to fix: `Not reported` -> `not hispanic or latino`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Hispanic or Latino</td>\n",
       "      <td>hispanic or latino</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>0.935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Not reported</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>0.268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>nan</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   source                  target  similarity\n",
       "0      Hispanic or Latino      hispanic or latino       1.000\n",
       "1  Not-Hispanic or Latino  not hispanic or latino       0.935\n",
       "2            Not reported  not hispanic or latino       0.268\n",
       "3                     nan                    None         NaN"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ethinicity_vmap = bdi.match_values(\n",
    "    dataset,\n",
    "    column_mapping=(\"Ethnicity\", \"ethnicity\"),\n",
    "    target=\"gdc\",\n",
    "    method=\"tfidf\",\n",
    ")\n",
    "ethinicity_vmap\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Hispanic or Latino</td>\n",
       "      <td>hispanic or latino</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>0.935</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   source                  target  similarity\n",
       "0      Hispanic or Latino      hispanic or latino       1.000\n",
       "1  Not-Hispanic or Latino  not hispanic or latino       0.935"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ethinicity_vmap = ethinicity_vmap[ethinicity_vmap[\"similarity\"] > 0.9]\n",
    "ethinicity_vmap"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For `Tumor_Site`, given that this dataset is about endometrial cancer, all values must be mapped to \"Endometrium\". So instead of fixing each mapping individually, we will write a custom function that returns \"Endometrium\" regardless of the input value. Later, we will show how to use this function to transform the dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Anterior endometrium</td>\n",
       "      <td>Endometrium</td>\n",
       "      <td>0.852</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Posterior endometrium</td>\n",
       "      <td>Endometrium</td>\n",
       "      <td>0.823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Other, specify</td>\n",
       "      <td>Other specified parts of pancreas</td>\n",
       "      <td>0.543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>nan</td>\n",
       "      <td>Anal canal</td>\n",
       "      <td>0.301</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  source                             target  similarity\n",
       "0   Anterior endometrium                        Endometrium       0.852\n",
       "1  Posterior endometrium                        Endometrium       0.823\n",
       "2         Other, specify  Other specified parts of pancreas       0.543\n",
       "3                    nan                         Anal canal       0.301"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bdi.match_values(\n",
    "    dataset, column_mapping=(\"Tumor_Site\", \"tissue_or_organ_of_origin\"), target=\"gdc\", method=\"tfidf\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Custom mapping function that will be used to map the values of the 'Tumor_Site' column\n",
    "def map_tumor_site(source_value):\n",
    "    return \"Endometrium\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Combining custom user mappings with suggested mappings\n",
    "\n",
    "Before generating a final harmonized dataset, we can combine the automatically generated value mappings with the fixed mappings provided by the user. To do so, we use `bdi.merge_mappings()` function, which take a list of mappings (e.g., generated automatically) and a list of \"user-defined mapping overrides\" that will be combined with the first list of mappings and will take precedence whenever they conflict.\n",
    "\n",
    "In our example below, all mappings specified in the variable `user_mappings` will override the mappings in `value_mappings` generated by the `bdi.match_values()` function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "from math import ceil\n",
    "\n",
    "user_mappings = [\n",
    "    {\n",
    "        # When no mapping is need, specifying the source and target is enough\n",
    "        \"source\": \"BMI\",\n",
    "        \"target\": \"bmi\",\n",
    "    },\n",
    "    {\n",
    "        \"source\": \"Tumor_Size_cm\",\n",
    "        \"target\": \"tumor_largest_dimension_diameter\",\n",
    "    },\n",
    "    {\n",
    "        # mapper can be a custom Python function\n",
    "        \"source\": \"Tumor_Site\",\n",
    "        \"target\": \"tissue_or_organ_of_origin\",\n",
    "        \"mapper\": map_tumor_site,\n",
    "    },\n",
    "    {\n",
    "        # Lambda functions can also be used as mappers\n",
    "        \"source\": \"Age\",\n",
    "        \"target\": \"days_to_birth\",\n",
    "        \"mapper\": lambda age: -age * 365.25,\n",
    "    },\n",
    "    {\n",
    "        \"source\": \"Age\",\n",
    "        \"target\": \"age_at_diagnosis\",\n",
    "        \"mapper\": lambda age: float(\"nan\") if pd.isnull(age) else ceil(age*365.25),\n",
    "    },\n",
    "    {\n",
    "        # We can also use a data frame to specify value mappings using the `matches` attribute\n",
    "        \"source\": \"Histologic_type\",\n",
    "        \"target\": \"primary_diagnosis\",\n",
    "        \"matches\": hist_type_vmap\n",
    "    },\n",
    "    # For dataframes that contain the 'source' and 'target' columns as attributes,\n",
    "    # such as the ones returned by the match_values() function, we can directly\n",
    "    # use them as mappings\n",
    "    ethinicity_vmap,\n",
    "    race_vmap,\n",
    "]\n",
    "\n",
    "\n",
    "harmonization_spec = bdi.merge_mappings(value_mappings, user_mappings)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, we generate the harmonized dataset, with the user-defined value mappings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tissue_or_organ_of_origin</th>\n",
       "      <th>bmi</th>\n",
       "      <th>days_to_birth</th>\n",
       "      <th>age_at_diagnosis</th>\n",
       "      <th>tumor_largest_dimension_diameter</th>\n",
       "      <th>country_of_birth</th>\n",
       "      <th>primary_diagnosis</th>\n",
       "      <th>figo_stage</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>gender</th>\n",
       "      <th>tumor_focality</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Endometrium</td>\n",
       "      <td>38.88</td>\n",
       "      <td>-23376.00</td>\n",
       "      <td>23376.0</td>\n",
       "      <td>2.9</td>\n",
       "      <td>United States</td>\n",
       "      <td>Endometrioid adenoma, NOS</td>\n",
       "      <td>Stage IA</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>female</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Endometrium</td>\n",
       "      <td>39.76</td>\n",
       "      <td>-21184.50</td>\n",
       "      <td>21185.0</td>\n",
       "      <td>3.5</td>\n",
       "      <td>United States</td>\n",
       "      <td>Endometrioid adenoma, NOS</td>\n",
       "      <td>Stage IA</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>female</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Endometrium</td>\n",
       "      <td>51.19</td>\n",
       "      <td>-18262.50</td>\n",
       "      <td>18263.0</td>\n",
       "      <td>4.5</td>\n",
       "      <td>United States</td>\n",
       "      <td>Endometrioid adenoma, NOS</td>\n",
       "      <td>Stage IA</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>female</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Endometrium</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Carcinosarcoma, NOS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Endometrium</td>\n",
       "      <td>32.69</td>\n",
       "      <td>-27393.75</td>\n",
       "      <td>27394.0</td>\n",
       "      <td>3.5</td>\n",
       "      <td>United States</td>\n",
       "      <td>Endometrioid adenoma, NOS</td>\n",
       "      <td>Stage IA</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>female</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>Endometrium</td>\n",
       "      <td>29.40</td>\n",
       "      <td>-27393.75</td>\n",
       "      <td>27394.0</td>\n",
       "      <td>4.2</td>\n",
       "      <td>Ukraine</td>\n",
       "      <td>Endometrioid adenoma, NOS</td>\n",
       "      <td>Stage IA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>female</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>Endometrium</td>\n",
       "      <td>35.42</td>\n",
       "      <td>-27028.50</td>\n",
       "      <td>27029.0</td>\n",
       "      <td>1.5</td>\n",
       "      <td>Ukraine</td>\n",
       "      <td>Endometrioid adenoma, NOS</td>\n",
       "      <td>Stage III</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>female</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>Endometrium</td>\n",
       "      <td>24.32</td>\n",
       "      <td>-31046.25</td>\n",
       "      <td>31047.0</td>\n",
       "      <td>3.8</td>\n",
       "      <td>United States</td>\n",
       "      <td>Serous carcinoma, NOS</td>\n",
       "      <td>Stage III</td>\n",
       "      <td>black or african american</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>female</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>Endometrium</td>\n",
       "      <td>34.06</td>\n",
       "      <td>-25567.50</td>\n",
       "      <td>25568.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>Ukraine</td>\n",
       "      <td>Serous carcinoma, NOS</td>\n",
       "      <td>Stage IA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>female</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>Endometrium</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ukraine</td>\n",
       "      <td>Serous carcinoma, NOS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>104 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    tissue_or_organ_of_origin    bmi  days_to_birth  age_at_diagnosis  \\\n",
       "0                 Endometrium  38.88      -23376.00           23376.0   \n",
       "1                 Endometrium  39.76      -21184.50           21185.0   \n",
       "2                 Endometrium  51.19      -18262.50           18263.0   \n",
       "3                 Endometrium    NaN            NaN               NaN   \n",
       "4                 Endometrium  32.69      -27393.75           27394.0   \n",
       "..                        ...    ...            ...               ...   \n",
       "99                Endometrium  29.40      -27393.75           27394.0   \n",
       "100               Endometrium  35.42      -27028.50           27029.0   \n",
       "101               Endometrium  24.32      -31046.25           31047.0   \n",
       "102               Endometrium  34.06      -25567.50           25568.0   \n",
       "103               Endometrium    NaN            NaN               NaN   \n",
       "\n",
       "     tumor_largest_dimension_diameter country_of_birth  \\\n",
       "0                                 2.9    United States   \n",
       "1                                 3.5    United States   \n",
       "2                                 4.5    United States   \n",
       "3                                 NaN              NaN   \n",
       "4                                 3.5    United States   \n",
       "..                                ...              ...   \n",
       "99                                4.2          Ukraine   \n",
       "100                               1.5          Ukraine   \n",
       "101                               3.8    United States   \n",
       "102                               5.0          Ukraine   \n",
       "103                               NaN          Ukraine   \n",
       "\n",
       "             primary_diagnosis figo_stage                       race  \\\n",
       "0    Endometrioid adenoma, NOS   Stage IA                      white   \n",
       "1    Endometrioid adenoma, NOS   Stage IA                      white   \n",
       "2    Endometrioid adenoma, NOS   Stage IA                      white   \n",
       "3          Carcinosarcoma, NOS        NaN                        NaN   \n",
       "4    Endometrioid adenoma, NOS   Stage IA                      white   \n",
       "..                         ...        ...                        ...   \n",
       "99   Endometrioid adenoma, NOS   Stage IA                        NaN   \n",
       "100  Endometrioid adenoma, NOS  Stage III                        NaN   \n",
       "101      Serous carcinoma, NOS  Stage III  black or african american   \n",
       "102      Serous carcinoma, NOS   Stage IA                        NaN   \n",
       "103      Serous carcinoma, NOS        NaN                        NaN   \n",
       "\n",
       "                  ethnicity  gender tumor_focality  \n",
       "0    not hispanic or latino  female       Unifocal  \n",
       "1    not hispanic or latino  female       Unifocal  \n",
       "2    not hispanic or latino  female       Unifocal  \n",
       "3                       NaN     NaN            NaN  \n",
       "4    not hispanic or latino  female       Unifocal  \n",
       "..                      ...     ...            ...  \n",
       "99                      NaN  female       Unifocal  \n",
       "100                     NaN  female       Unifocal  \n",
       "101  not hispanic or latino  female       Unifocal  \n",
       "102                     NaN  female       Unifocal  \n",
       "103                     NaN     NaN            NaN  \n",
       "\n",
       "[104 rows x 12 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "harmonized_dataset = bdi.materialize_mapping(dataset, harmonization_spec)\n",
    "harmonized_dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For comparison, here is how our original data looked like:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Tumor_Site</th>\n",
       "      <th>BMI</th>\n",
       "      <th>Age</th>\n",
       "      <th>Age</th>\n",
       "      <th>Tumor_Size_cm</th>\n",
       "      <th>Country</th>\n",
       "      <th>Histologic_type</th>\n",
       "      <th>FIGO_stage</th>\n",
       "      <th>Race</th>\n",
       "      <th>Ethnicity</th>\n",
       "      <th>Gender</th>\n",
       "      <th>Tumor_Focality</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Anterior endometrium</td>\n",
       "      <td>38.88</td>\n",
       "      <td>64.0</td>\n",
       "      <td>64.0</td>\n",
       "      <td>2.9</td>\n",
       "      <td>United States</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>IA</td>\n",
       "      <td>White</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>Female</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Posterior endometrium</td>\n",
       "      <td>39.76</td>\n",
       "      <td>58.0</td>\n",
       "      <td>58.0</td>\n",
       "      <td>3.5</td>\n",
       "      <td>United States</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>IA</td>\n",
       "      <td>White</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>Female</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Other, specify</td>\n",
       "      <td>51.19</td>\n",
       "      <td>50.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>4.5</td>\n",
       "      <td>United States</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>IA</td>\n",
       "      <td>White</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>Female</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Other, specify</td>\n",
       "      <td>32.69</td>\n",
       "      <td>75.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>3.5</td>\n",
       "      <td>United States</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>IA</td>\n",
       "      <td>White</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>Female</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>Other, specify</td>\n",
       "      <td>29.40</td>\n",
       "      <td>75.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>4.2</td>\n",
       "      <td>Ukraine</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>IA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Female</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>Other, specify</td>\n",
       "      <td>35.42</td>\n",
       "      <td>74.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>1.5</td>\n",
       "      <td>Ukraine</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>II</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Female</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>Other, specify</td>\n",
       "      <td>24.32</td>\n",
       "      <td>85.0</td>\n",
       "      <td>85.0</td>\n",
       "      <td>3.8</td>\n",
       "      <td>United States</td>\n",
       "      <td>Serous</td>\n",
       "      <td>II</td>\n",
       "      <td>Black or African American</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>Female</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>Other, specify</td>\n",
       "      <td>34.06</td>\n",
       "      <td>70.0</td>\n",
       "      <td>70.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>Ukraine</td>\n",
       "      <td>Serous</td>\n",
       "      <td>IA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Female</td>\n",
       "      <td>Unifocal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ukraine</td>\n",
       "      <td>Serous</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>104 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                Tumor_Site    BMI   Age   Age  Tumor_Size_cm        Country  \\\n",
       "0     Anterior endometrium  38.88  64.0  64.0            2.9  United States   \n",
       "1    Posterior endometrium  39.76  58.0  58.0            3.5  United States   \n",
       "2           Other, specify  51.19  50.0  50.0            4.5  United States   \n",
       "3                      NaN    NaN   NaN   NaN            NaN            NaN   \n",
       "4           Other, specify  32.69  75.0  75.0            3.5  United States   \n",
       "..                     ...    ...   ...   ...            ...            ...   \n",
       "99          Other, specify  29.40  75.0  75.0            4.2        Ukraine   \n",
       "100         Other, specify  35.42  74.0  74.0            1.5        Ukraine   \n",
       "101         Other, specify  24.32  85.0  85.0            3.8  United States   \n",
       "102         Other, specify  34.06  70.0  70.0            5.0        Ukraine   \n",
       "103                    NaN    NaN   NaN   NaN            NaN        Ukraine   \n",
       "\n",
       "    Histologic_type FIGO_stage                       Race  \\\n",
       "0      Endometrioid         IA                      White   \n",
       "1      Endometrioid         IA                      White   \n",
       "2      Endometrioid         IA                      White   \n",
       "3    Carcinosarcoma        NaN                        NaN   \n",
       "4      Endometrioid         IA                      White   \n",
       "..              ...        ...                        ...   \n",
       "99     Endometrioid         IA                        NaN   \n",
       "100    Endometrioid         II                        NaN   \n",
       "101          Serous         II  Black or African American   \n",
       "102          Serous         IA                        NaN   \n",
       "103          Serous        NaN                        NaN   \n",
       "\n",
       "                  Ethnicity  Gender Tumor_Focality  \n",
       "0    Not-Hispanic or Latino  Female       Unifocal  \n",
       "1    Not-Hispanic or Latino  Female       Unifocal  \n",
       "2    Not-Hispanic or Latino  Female       Unifocal  \n",
       "3                       NaN     NaN            NaN  \n",
       "4    Not-Hispanic or Latino  Female       Unifocal  \n",
       "..                      ...     ...            ...  \n",
       "99                      NaN  Female       Unifocal  \n",
       "100                     NaN  Female       Unifocal  \n",
       "101  Not-Hispanic or Latino  Female       Unifocal  \n",
       "102                     NaN  Female       Unifocal  \n",
       "103                     NaN     NaN            NaN  \n",
       "\n",
       "[104 rows x 12 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "original_columns = map(lambda m: m[\"source\"], harmonization_spec)\n",
    "dataset[original_columns]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
