{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Getting Started"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Harmonization with `bdikit`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Data harmonization is the process of integrating and aligning data from different sources into a consistent format to ensure compatibility and interoperability across data analyses and systems. `bdikit` is a library the helps with key data harmonization steps:\n",
    "- *Schema Mapping*: In this step, data from various sources are mapped to a unified schema or model. This involves identifying equivalent table columns and establishing relationships between disparate datasets.\n",
    "- *Value Mapping (Data Standardization)*: This step involves converting data into a common format or structure, using consistent naming conventions, units, and coding systems to ensure uniformity.\n",
    "\n",
    "In this example, we describe how `bdikit` can be used to map a dataset from Dou et al. (https://pubmed.ncbi.nlm.nih.gov/37567170/) to the [GDC (Genomic Data Commons)](https://portal.gdc.cancer.gov/) standard data format.\n",
    "\n",
    "First, import the `bdikit` library."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import bdikit as bdi\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we load the data using Pandas."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Country</th>\n",
       "      <th>Histologic_Grade_FIGO</th>\n",
       "      <th>Histologic_type</th>\n",
       "      <th>Path_Stage_Primary_Tumor-pT</th>\n",
       "      <th>Path_Stage_Reg_Lymph_Nodes-pN</th>\n",
       "      <th>Clin_Stage_Dist_Mets-cM</th>\n",
       "      <th>Path_Stage_Dist_Mets-pM</th>\n",
       "      <th>tumor_Stage-Pathological</th>\n",
       "      <th>FIGO_stage</th>\n",
       "      <th>BMI</th>\n",
       "      <th>Age</th>\n",
       "      <th>Race</th>\n",
       "      <th>Ethnicity</th>\n",
       "      <th>Gender</th>\n",
       "      <th>Tumor_Site</th>\n",
       "      <th>Tumor_Focality</th>\n",
       "      <th>Tumor_Size_cm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>United States</td>\n",
       "      <td>FIGO grade 1</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>pN0</td>\n",
       "      <td>cM0</td>\n",
       "      <td>Staging Incomplete</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>IA</td>\n",
       "      <td>38.88</td>\n",
       "      <td>64.0</td>\n",
       "      <td>White</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>Female</td>\n",
       "      <td>Anterior endometrium</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>2.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>United States</td>\n",
       "      <td>FIGO grade 1</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>pNX</td>\n",
       "      <td>cM0</td>\n",
       "      <td>Staging Incomplete</td>\n",
       "      <td>Stage IV</td>\n",
       "      <td>IA</td>\n",
       "      <td>39.76</td>\n",
       "      <td>58.0</td>\n",
       "      <td>White</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>Female</td>\n",
       "      <td>Posterior endometrium</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>3.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>United States</td>\n",
       "      <td>FIGO grade 2</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>pN0</td>\n",
       "      <td>cM0</td>\n",
       "      <td>Staging Incomplete</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>IA</td>\n",
       "      <td>51.19</td>\n",
       "      <td>50.0</td>\n",
       "      <td>White</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>Female</td>\n",
       "      <td>Other, specify</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>4.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>United States</td>\n",
       "      <td>FIGO grade 2</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>pNX</td>\n",
       "      <td>cM0</td>\n",
       "      <td>No pathologic evidence of distant metastasis</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>IA</td>\n",
       "      <td>32.69</td>\n",
       "      <td>75.0</td>\n",
       "      <td>White</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>Female</td>\n",
       "      <td>Other, specify</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>3.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>United States</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Serous</td>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>pNX</td>\n",
       "      <td>cM0</td>\n",
       "      <td>Staging Incomplete</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>IA</td>\n",
       "      <td>20.28</td>\n",
       "      <td>63.0</td>\n",
       "      <td>White</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>Female</td>\n",
       "      <td>Other, specify</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>United States</td>\n",
       "      <td>FIGO grade 1</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>pNX</td>\n",
       "      <td>cM0</td>\n",
       "      <td>Staging Incomplete</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>IA</td>\n",
       "      <td>55.67</td>\n",
       "      <td>50.0</td>\n",
       "      <td>White</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>Female</td>\n",
       "      <td>Other, specify</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>4.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Other_specify</td>\n",
       "      <td>FIGO grade 2</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>pNX</td>\n",
       "      <td>cM0</td>\n",
       "      <td>Staging Incomplete</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>IA</td>\n",
       "      <td>25.68</td>\n",
       "      <td>60.0</td>\n",
       "      <td>White</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>Female</td>\n",
       "      <td>Other, specify</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>United States</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Serous</td>\n",
       "      <td>pT3a (FIGO IIIA)</td>\n",
       "      <td>pNX</td>\n",
       "      <td>cM0</td>\n",
       "      <td>Staging Incomplete</td>\n",
       "      <td>Stage III</td>\n",
       "      <td>IIIA</td>\n",
       "      <td>21.57</td>\n",
       "      <td>83.0</td>\n",
       "      <td>White</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>Female</td>\n",
       "      <td>Other, specify</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>United States</td>\n",
       "      <td>FIGO grade 1</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>pT1 (FIGO I)</td>\n",
       "      <td>pN0</td>\n",
       "      <td>cM0</td>\n",
       "      <td>Staging Incomplete</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>IA</td>\n",
       "      <td>34.26</td>\n",
       "      <td>69.0</td>\n",
       "      <td>White</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>Female</td>\n",
       "      <td>Other, specify</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>5.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Country Histologic_Grade_FIGO Histologic_type  \\\n",
       "0  United States          FIGO grade 1    Endometrioid   \n",
       "1  United States          FIGO grade 1    Endometrioid   \n",
       "2  United States          FIGO grade 2    Endometrioid   \n",
       "3            NaN                   NaN  Carcinosarcoma   \n",
       "4  United States          FIGO grade 2    Endometrioid   \n",
       "5  United States                   NaN          Serous   \n",
       "6  United States          FIGO grade 1    Endometrioid   \n",
       "7  Other_specify          FIGO grade 2    Endometrioid   \n",
       "8  United States                   NaN          Serous   \n",
       "9  United States          FIGO grade 1    Endometrioid   \n",
       "\n",
       "  Path_Stage_Primary_Tumor-pT Path_Stage_Reg_Lymph_Nodes-pN  \\\n",
       "0              pT1a (FIGO IA)                           pN0   \n",
       "1              pT1a (FIGO IA)                           pNX   \n",
       "2              pT1a (FIGO IA)                           pN0   \n",
       "3                         NaN                           NaN   \n",
       "4              pT1a (FIGO IA)                           pNX   \n",
       "5              pT1a (FIGO IA)                           pNX   \n",
       "6              pT1a (FIGO IA)                           pNX   \n",
       "7              pT1a (FIGO IA)                           pNX   \n",
       "8            pT3a (FIGO IIIA)                           pNX   \n",
       "9                pT1 (FIGO I)                           pN0   \n",
       "\n",
       "  Clin_Stage_Dist_Mets-cM                       Path_Stage_Dist_Mets-pM  \\\n",
       "0                     cM0                            Staging Incomplete   \n",
       "1                     cM0                            Staging Incomplete   \n",
       "2                     cM0                            Staging Incomplete   \n",
       "3                     NaN                                           NaN   \n",
       "4                     cM0  No pathologic evidence of distant metastasis   \n",
       "5                     cM0                            Staging Incomplete   \n",
       "6                     cM0                            Staging Incomplete   \n",
       "7                     cM0                            Staging Incomplete   \n",
       "8                     cM0                            Staging Incomplete   \n",
       "9                     cM0                            Staging Incomplete   \n",
       "\n",
       "  tumor_Stage-Pathological FIGO_stage    BMI   Age   Race  \\\n",
       "0                  Stage I         IA  38.88  64.0  White   \n",
       "1                 Stage IV         IA  39.76  58.0  White   \n",
       "2                  Stage I         IA  51.19  50.0  White   \n",
       "3                      NaN        NaN    NaN   NaN    NaN   \n",
       "4                  Stage I         IA  32.69  75.0  White   \n",
       "5                  Stage I         IA  20.28  63.0  White   \n",
       "6                  Stage I         IA  55.67  50.0  White   \n",
       "7                  Stage I         IA  25.68  60.0  White   \n",
       "8                Stage III       IIIA  21.57  83.0  White   \n",
       "9                  Stage I         IA  34.26  69.0  White   \n",
       "\n",
       "                Ethnicity  Gender             Tumor_Site Tumor_Focality  \\\n",
       "0  Not-Hispanic or Latino  Female   Anterior endometrium       Unifocal   \n",
       "1  Not-Hispanic or Latino  Female  Posterior endometrium       Unifocal   \n",
       "2  Not-Hispanic or Latino  Female         Other, specify       Unifocal   \n",
       "3                     NaN     NaN                    NaN            NaN   \n",
       "4  Not-Hispanic or Latino  Female         Other, specify       Unifocal   \n",
       "5  Not-Hispanic or Latino  Female         Other, specify       Unifocal   \n",
       "6  Not-Hispanic or Latino  Female         Other, specify       Unifocal   \n",
       "7  Not-Hispanic or Latino  Female         Other, specify       Unifocal   \n",
       "8  Not-Hispanic or Latino  Female         Other, specify       Unifocal   \n",
       "9  Not-Hispanic or Latino  Female         Other, specify       Unifocal   \n",
       "\n",
       "   Tumor_Size_cm  \n",
       "0            2.9  \n",
       "1            3.5  \n",
       "2            4.5  \n",
       "3            NaN  \n",
       "4            3.5  \n",
       "5            6.0  \n",
       "6            4.5  \n",
       "7            5.0  \n",
       "8            4.0  \n",
       "9            5.2  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset = pd.read_csv(\"./datasets/dou.csv\")\n",
    "\n",
    "# columns = [\n",
    "#     \"Country\",\n",
    "#     \"Path_Stage_Primary_Tumor-pT\",\n",
    "#     \"FIGO_stage\",\n",
    "#     \"Race\",\n",
    "#     \"Ethnicity\",\n",
    "#     \"Gender\",\n",
    "#     \"Tumor_Focality\",\n",
    "#     \"Tumor_Site\",\n",
    "# ]\n",
    "columns = [\n",
    "    \"Country\",\n",
    "    \"Path_Stage_Primary_Tumor-pT\",\n",
    "    \"Histologic_type\",\n",
    "    \"FIGO_stage\",\n",
    "    \"BMI\",\n",
    "    \"Age\",\n",
    "    \"Race\",\n",
    "    \"Ethnicity\",\n",
    "    \"Gender\",\n",
    "    \"Tumor_Focality\",\n",
    "    \"Tumor_Size_cm\",\n",
    "]\n",
    "# dataset = dataset[columns]\n",
    "dataset.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Country</th>\n",
       "      <th>Histologic_Grade_FIGO</th>\n",
       "      <th>Histologic_type</th>\n",
       "      <th>Path_Stage_Primary_Tumor-pT</th>\n",
       "      <th>Path_Stage_Reg_Lymph_Nodes-pN</th>\n",
       "      <th>Clin_Stage_Dist_Mets-cM</th>\n",
       "      <th>Path_Stage_Dist_Mets-pM</th>\n",
       "      <th>tumor_Stage-Pathological</th>\n",
       "      <th>FIGO_stage</th>\n",
       "      <th>BMI</th>\n",
       "      <th>Age</th>\n",
       "      <th>Race</th>\n",
       "      <th>Ethnicity</th>\n",
       "      <th>Gender</th>\n",
       "      <th>Tumor_Site</th>\n",
       "      <th>Tumor_Focality</th>\n",
       "      <th>Tumor_Size_cm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>United States</td>\n",
       "      <td>FIGO grade 1</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>pN0</td>\n",
       "      <td>cM0</td>\n",
       "      <td>Staging Incomplete</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>IA</td>\n",
       "      <td>38.88</td>\n",
       "      <td>64.0</td>\n",
       "      <td>White</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>Female</td>\n",
       "      <td>Anterior endometrium</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>2.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>United States</td>\n",
       "      <td>FIGO grade 1</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>pNX</td>\n",
       "      <td>cM0</td>\n",
       "      <td>Staging Incomplete</td>\n",
       "      <td>Stage IV</td>\n",
       "      <td>IA</td>\n",
       "      <td>39.76</td>\n",
       "      <td>58.0</td>\n",
       "      <td>White</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>Female</td>\n",
       "      <td>Posterior endometrium</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>3.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>United States</td>\n",
       "      <td>FIGO grade 2</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>pN0</td>\n",
       "      <td>cM0</td>\n",
       "      <td>Staging Incomplete</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>IA</td>\n",
       "      <td>51.19</td>\n",
       "      <td>50.0</td>\n",
       "      <td>White</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>Female</td>\n",
       "      <td>Other, specify</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>4.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>United States</td>\n",
       "      <td>FIGO grade 2</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>pNX</td>\n",
       "      <td>cM0</td>\n",
       "      <td>No pathologic evidence of distant metastasis</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>IA</td>\n",
       "      <td>32.69</td>\n",
       "      <td>75.0</td>\n",
       "      <td>White</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>Female</td>\n",
       "      <td>Other, specify</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>3.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>Ukraine</td>\n",
       "      <td>FIGO grade 3</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>pNX</td>\n",
       "      <td>cM0</td>\n",
       "      <td>Staging Incomplete</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>IA</td>\n",
       "      <td>29.40</td>\n",
       "      <td>75.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Female</td>\n",
       "      <td>Other, specify</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>4.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>Ukraine</td>\n",
       "      <td>FIGO grade 2</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>pT2 (FIGO II)</td>\n",
       "      <td>pN0</td>\n",
       "      <td>cM0</td>\n",
       "      <td>Staging Incomplete</td>\n",
       "      <td>Stage II</td>\n",
       "      <td>II</td>\n",
       "      <td>35.42</td>\n",
       "      <td>74.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Female</td>\n",
       "      <td>Other, specify</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>1.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>United States</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Serous</td>\n",
       "      <td>pT2 (FIGO II)</td>\n",
       "      <td>pN0</td>\n",
       "      <td>Staging Incomplete</td>\n",
       "      <td>Staging Incomplete</td>\n",
       "      <td>Stage II</td>\n",
       "      <td>II</td>\n",
       "      <td>24.32</td>\n",
       "      <td>85.0</td>\n",
       "      <td>Black or African American</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>Female</td>\n",
       "      <td>Other, specify</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>3.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>Ukraine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Serous</td>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>pN0</td>\n",
       "      <td>cM0</td>\n",
       "      <td>Staging Incomplete</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>IA</td>\n",
       "      <td>34.06</td>\n",
       "      <td>70.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Female</td>\n",
       "      <td>Other, specify</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>Ukraine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Serous</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>104 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Country Histologic_Grade_FIGO Histologic_type  \\\n",
       "0    United States          FIGO grade 1    Endometrioid   \n",
       "1    United States          FIGO grade 1    Endometrioid   \n",
       "2    United States          FIGO grade 2    Endometrioid   \n",
       "3              NaN                   NaN  Carcinosarcoma   \n",
       "4    United States          FIGO grade 2    Endometrioid   \n",
       "..             ...                   ...             ...   \n",
       "99         Ukraine          FIGO grade 3    Endometrioid   \n",
       "100        Ukraine          FIGO grade 2    Endometrioid   \n",
       "101  United States                   NaN          Serous   \n",
       "102        Ukraine                   NaN          Serous   \n",
       "103        Ukraine                   NaN          Serous   \n",
       "\n",
       "    Path_Stage_Primary_Tumor-pT Path_Stage_Reg_Lymph_Nodes-pN  \\\n",
       "0                pT1a (FIGO IA)                           pN0   \n",
       "1                pT1a (FIGO IA)                           pNX   \n",
       "2                pT1a (FIGO IA)                           pN0   \n",
       "3                           NaN                           NaN   \n",
       "4                pT1a (FIGO IA)                           pNX   \n",
       "..                          ...                           ...   \n",
       "99               pT1a (FIGO IA)                           pNX   \n",
       "100               pT2 (FIGO II)                           pN0   \n",
       "101               pT2 (FIGO II)                           pN0   \n",
       "102              pT1a (FIGO IA)                           pN0   \n",
       "103                         NaN                           NaN   \n",
       "\n",
       "    Clin_Stage_Dist_Mets-cM                       Path_Stage_Dist_Mets-pM  \\\n",
       "0                       cM0                            Staging Incomplete   \n",
       "1                       cM0                            Staging Incomplete   \n",
       "2                       cM0                            Staging Incomplete   \n",
       "3                       NaN                                           NaN   \n",
       "4                       cM0  No pathologic evidence of distant metastasis   \n",
       "..                      ...                                           ...   \n",
       "99                      cM0                            Staging Incomplete   \n",
       "100                     cM0                            Staging Incomplete   \n",
       "101      Staging Incomplete                            Staging Incomplete   \n",
       "102                     cM0                            Staging Incomplete   \n",
       "103                     NaN                                           NaN   \n",
       "\n",
       "    tumor_Stage-Pathological FIGO_stage    BMI   Age  \\\n",
       "0                    Stage I         IA  38.88  64.0   \n",
       "1                   Stage IV         IA  39.76  58.0   \n",
       "2                    Stage I         IA  51.19  50.0   \n",
       "3                        NaN        NaN    NaN   NaN   \n",
       "4                    Stage I         IA  32.69  75.0   \n",
       "..                       ...        ...    ...   ...   \n",
       "99                   Stage I         IA  29.40  75.0   \n",
       "100                 Stage II         II  35.42  74.0   \n",
       "101                 Stage II         II  24.32  85.0   \n",
       "102                  Stage I         IA  34.06  70.0   \n",
       "103                      NaN        NaN    NaN   NaN   \n",
       "\n",
       "                          Race               Ethnicity  Gender  \\\n",
       "0                        White  Not-Hispanic or Latino  Female   \n",
       "1                        White  Not-Hispanic or Latino  Female   \n",
       "2                        White  Not-Hispanic or Latino  Female   \n",
       "3                          NaN                     NaN     NaN   \n",
       "4                        White  Not-Hispanic or Latino  Female   \n",
       "..                         ...                     ...     ...   \n",
       "99                         NaN                     NaN  Female   \n",
       "100                        NaN                     NaN  Female   \n",
       "101  Black or African American  Not-Hispanic or Latino  Female   \n",
       "102                        NaN                     NaN  Female   \n",
       "103                        NaN                     NaN     NaN   \n",
       "\n",
       "                Tumor_Site Tumor_Focality  Tumor_Size_cm  \n",
       "0     Anterior endometrium       Unifocal            2.9  \n",
       "1    Posterior endometrium       Unifocal            3.5  \n",
       "2           Other, specify       Unifocal            4.5  \n",
       "3                      NaN            NaN            NaN  \n",
       "4           Other, specify       Unifocal            3.5  \n",
       "..                     ...            ...            ...  \n",
       "99          Other, specify       Unifocal            4.2  \n",
       "100         Other, specify       Unifocal            1.5  \n",
       "101         Other, specify       Unifocal            3.8  \n",
       "102         Other, specify       Unifocal            5.0  \n",
       "103                    NaN            NaN            NaN  \n",
       "\n",
       "[104 rows x 17 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Matching the table schema to GDC standard vocabulary\n",
    "\n",
    "`bdi-kit` offers a suite of functions to help with data harmonization tasks.\n",
    "For instance, it can help with automatic discovery of one-to-one mappings between the columns in the input (source) dataset and a target dataset schema. The target schema can be either another table or a standard data vocabulary such as the GDC (Genomic Data Commons).\n",
    "\n",
    "To achieve this using `bdi-kit`, we can use the `match_schema()` function to match columns to the GDC vocabulary schema as follows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Some weights of RobertaModel were not initialized from the model checkpoint at roberta-base and are newly initialized: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']\n",
      "You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.\n",
      "  0%|          | 0/11 [00:00<?, ?it/s]We strongly recommend passing in an `attention_mask` since your input_ids may be padded. See https://huggingface.co/docs/transformers/troubleshooting#incorrect-output-when-padding-tokens-arent-masked.\n",
      "100%|██████████| 11/11 [00:01<00:00,  9.78it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table features extracted from 11 columns\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 734/734 [00:55<00:00, 13.32it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table features extracted from 734 columns\n",
      "\n",
      "Source column:  Country\n",
      "value_matches:  [ValueMatch(current_value='United States', target_value='Federated States of Micronesia', similarity=0.541)]\n",
      "source: Country target: country_of_birth score: 0.541\n",
      "value_matches:  [ValueMatch(current_value='Other_specify', target_value='other', similarity=0.508), ValueMatch(current_value='nan', target_value='american indian or alaska native', similarity=0.332), ValueMatch(current_value='United States', target_value='white', similarity=0.313), ValueMatch(current_value='Poland', target_value='native hawaiian or other pacific islander', similarity=0.27)]\n",
      "source: Country target: race score: 1.423\n",
      "value_matches:  [ValueMatch(current_value='Poland', target_value='Andorra', similarity=0.317), ValueMatch(current_value='United States', target_value='Guatemala', similarity=0.261), ValueMatch(current_value='Other_specify', target_value='Jersey', similarity=0.253)]\n",
      "source: Country target: country_of_residence_at_enrollment score: 0.8310000000000001\n",
      "value_matches:  [ValueMatch(current_value='Other_specify', target_value='Peripheral zone', similarity=0.288)]\n",
      "source: Country target: zone_of_origin_prostate score: 0.288\n",
      "value_matches:  [ValueMatch(current_value='United States', target_value='Indeterminate', similarity=0.325)]\n",
      "source: Country target: non_nodal_regional_disease score: 0.325\n",
      "value_matches:  []\n",
      "source: Country target: submission_enabled score: 0\n",
      "value_matches:  []\n",
      "source: Country target: is_legacy score: 0\n",
      "value_matches:  []\n",
      "source: Country target: oct_embedded score: 0\n",
      "value_matches:  [ValueMatch(current_value='United States', target_value='Not Reported', similarity=0.325)]\n",
      "source: Country target: perineural_invasion_present score: 0.325\n",
      "value_matches:  [ValueMatch(current_value='United States', target_value='Indeterminate', similarity=0.325)]\n",
      "source: Country target: ovarian_surface_involvement score: 0.325\n",
      "value_matches:  [ValueMatch(current_value='United States', target_value='not reported', similarity=0.311)]\n",
      "source: Country target: ethnicity score: 0.311\n",
      "value_matches:  [ValueMatch(current_value='United States', target_value='Not Reported', similarity=0.338)]\n",
      "source: Country target: overrepresented_sequences score: 0.338\n",
      "value_matches:  []\n",
      "source: Country target: released score: 0\n",
      "value_matches:  [ValueMatch(current_value='United States', target_value='Not Reported', similarity=0.325)]\n",
      "source: Country target: chemo_concurrent_to_radiation score: 0.325\n",
      "value_matches:  []\n",
      "source: Country target: status score: 0\n",
      "value_matches:  [ValueMatch(current_value='United States', target_value='Indeterminate', similarity=0.325)]\n",
      "source: Country target: satellite_nodule_present score: 0.325\n",
      "value_matches:  [ValueMatch(current_value='United States', target_value='Not Reported', similarity=0.315)]\n",
      "source: Country target: ulceration_indicator score: 0.315\n",
      "value_matches:  []\n",
      "source: Country target: request_submission score: 0\n",
      "value_matches:  []\n",
      "source: Country target: consent_type score: 0\n",
      "value_matches:  [ValueMatch(current_value='United States', target_value='Not Reported', similarity=0.338)]\n",
      "source: Country target: kmer_content score: 0.338\n",
      "Top k reranked columns:  [('Country', 'race', 1.423), ('Country', 'country_of_residence_at_enrollment', 0.8310000000000001), ('Country', 'country_of_birth', 0.541), ('Country', 'overrepresented_sequences', 0.338), ('Country', 'kmer_content', 0.338), ('Country', 'non_nodal_regional_disease', 0.325), ('Country', 'perineural_invasion_present', 0.325), ('Country', 'ovarian_surface_involvement', 0.325), ('Country', 'chemo_concurrent_to_radiation', 0.325), ('Country', 'satellite_nodule_present', 0.325), ('Country', 'ulceration_indicator', 0.315), ('Country', 'ethnicity', 0.311), ('Country', 'zone_of_origin_prostate', 0.288), ('Country', 'submission_enabled', 0), ('Country', 'is_legacy', 0), ('Country', 'oct_embedded', 0), ('Country', 'released', 0), ('Country', 'status', 0), ('Country', 'request_submission', 0), ('Country', 'consent_type', 0)]\n",
      "\n",
      "Source column:  Path_Stage_Primary_Tumor-pT\n",
      "value_matches:  [ValueMatch(current_value='pT1a (FIGO IA)', target_value='T1a', similarity=0.563), ValueMatch(current_value='pT1 (FIGO I)', target_value='T1', similarity=0.341), ValueMatch(current_value='pT1b (FIGO IB)', target_value='T1', similarity=0.254)]\n",
      "source: Path_Stage_Primary_Tumor-pT target: uicc_pathologic_t score: 1.158\n",
      "value_matches:  [ValueMatch(current_value='pT3b (FIGO IIIB)', target_value='T3b', similarity=0.478), ValueMatch(current_value='pT3a (FIGO IIIA)', target_value='T3a', similarity=0.468), ValueMatch(current_value='pT1b (FIGO IB)', target_value='T1b2', similarity=0.445), ValueMatch(current_value='pT2 (FIGO II)', target_value='T2', similarity=0.369), ValueMatch(current_value='pT1 (FIGO I)', target_value='T1', similarity=0.358), ValueMatch(current_value='pT1a (FIGO IA)', target_value='T1', similarity=0.273)]\n",
      "source: Path_Stage_Primary_Tumor-pT target: ajcc_pathologic_t score: 2.391\n",
      "value_matches:  [ValueMatch(current_value='pT1 (FIGO I)', target_value='T1', similarity=0.424), ValueMatch(current_value='pT2 (FIGO II)', target_value='T2', similarity=0.424), ValueMatch(current_value='pT1a (FIGO IA)', target_value='T1', similarity=0.311), ValueMatch(current_value='pT1b (FIGO IB)', target_value='T1', similarity=0.302), ValueMatch(current_value='pT3a (FIGO IIIA)', target_value='T3', similarity=0.26), ValueMatch(current_value='pT3b (FIGO IIIB)', target_value='T3', similarity=0.256)]\n",
      "source: Path_Stage_Primary_Tumor-pT target: ensat_pathologic_t score: 1.977\n",
      "value_matches:  [ValueMatch(current_value='pT1a (FIGO IA)', target_value='T1a', similarity=0.603), ValueMatch(current_value='pT3a (FIGO IIIA)', target_value='T3a', similarity=0.46)]\n",
      "source: Path_Stage_Primary_Tumor-pT target: uicc_clinical_t score: 1.063\n",
      "value_matches:  [ValueMatch(current_value='pT1b (FIGO IB)', target_value='T1b', similarity=0.622), ValueMatch(current_value='pT1 (FIGO I)', target_value='T1b', similarity=0.255)]\n",
      "source: Path_Stage_Primary_Tumor-pT target: ajcc_clinical_t score: 0.877\n",
      "value_matches:  [ValueMatch(current_value='nan', target_value='nan', similarity=1.0)]\n",
      "source: Path_Stage_Primary_Tumor-pT target: extrathyroid_extension score: 1.0\n",
      "value_matches:  [ValueMatch(current_value='pT3a (FIGO IIIA)', target_value='Stage IIIA1', similarity=0.489), ValueMatch(current_value='pT3b (FIGO IIIB)', target_value='Stage IIIA1', similarity=0.318), ValueMatch(current_value='pT1b (FIGO IB)', target_value='Stage IB', similarity=0.315), ValueMatch(current_value='pT2 (FIGO II)', target_value='Stage IIIC2', similarity=0.297), ValueMatch(current_value='pT1a (FIGO IA)', target_value='Stage IA1', similarity=0.272)]\n",
      "source: Path_Stage_Primary_Tumor-pT target: uicc_pathologic_stage score: 1.6909999999999998\n",
      "value_matches:  []\n",
      "source: Path_Stage_Primary_Tumor-pT target: somatic_mutation_indexes score: 0\n",
      "value_matches:  [ValueMatch(current_value='pT3a (FIGO IIIA)', target_value='Stage IIIA', similarity=0.549), ValueMatch(current_value='pT3b (FIGO IIIB)', target_value='Stage IIB', similarity=0.487), ValueMatch(current_value='pT1b (FIGO IB)', target_value='Stage IIB', similarity=0.324), ValueMatch(current_value='pT2 (FIGO II)', target_value='Stage III', similarity=0.314), ValueMatch(current_value='pT1a (FIGO IA)', target_value='Stage IIA1', similarity=0.292)]\n",
      "source: Path_Stage_Primary_Tumor-pT target: ajcc_pathologic_stage score: 1.9660000000000002\n",
      "value_matches:  []\n",
      "source: Path_Stage_Primary_Tumor-pT target: annotated_somatic_mutations score: 0\n",
      "value_matches:  []\n",
      "source: Path_Stage_Primary_Tumor-pT target: masked_somatic_mutations score: 0\n",
      "value_matches:  [ValueMatch(current_value='pT3b (FIGO IIIB)', target_value='Stage IIIB', similarity=0.554), ValueMatch(current_value='pT3a (FIGO IIIA)', target_value='Stage IIIAii', similarity=0.524), ValueMatch(current_value='nan', target_value='Unknown', similarity=0.398), ValueMatch(current_value='pT2 (FIGO II)', target_value='Stage IIIAii', similarity=0.303), ValueMatch(current_value='pT1b (FIGO IB)', target_value='Stage IB1', similarity=0.294), ValueMatch(current_value='pT1a (FIGO IA)', target_value='Stage IA', similarity=0.291)]\n",
      "source: Path_Stage_Primary_Tumor-pT target: figo_stage score: 2.364\n",
      "value_matches:  [ValueMatch(current_value='nan', target_value='Transitional zone', similarity=0.297)]\n",
      "source: Path_Stage_Primary_Tumor-pT target: tumor_level_prostate score: 0.297\n",
      "value_matches:  [ValueMatch(current_value='pT3a (FIGO IIIA)', target_value='Stage III', similarity=0.434), ValueMatch(current_value='pT3b (FIGO IIIB)', target_value='Stage III', similarity=0.363), ValueMatch(current_value='pT2 (FIGO II)', target_value='Stage III', similarity=0.304)]\n",
      "source: Path_Stage_Primary_Tumor-pT target: ensat_pathologic_stage score: 1.101\n",
      "value_matches:  [ValueMatch(current_value='pT3b (FIGO IIIB)', target_value='Stage IIIB', similarity=0.555), ValueMatch(current_value='pT3a (FIGO IIIA)', target_value='Stage IIIA', similarity=0.523), ValueMatch(current_value='pT1b (FIGO IB)', target_value='Stage IIIB', similarity=0.301), ValueMatch(current_value='pT1a (FIGO IA)', target_value='Stage IA', similarity=0.261), ValueMatch(current_value='pT2 (FIGO II)', target_value='Stage IIIA', similarity=0.252)]\n",
      "source: Path_Stage_Primary_Tumor-pT target: uicc_clinical_stage score: 1.8920000000000001\n",
      "value_matches:  []\n",
      "source: Path_Stage_Primary_Tumor-pT target: tumor_grade_category score: 0\n",
      "value_matches:  [ValueMatch(current_value='pT3a (FIGO IIIA)', target_value='Stage IIIA', similarity=0.549), ValueMatch(current_value='pT3b (FIGO IIIB)', target_value='Stage III', similarity=0.397), ValueMatch(current_value='pT2 (FIGO II)', target_value='Stage III', similarity=0.328), ValueMatch(current_value='pT1a (FIGO IA)', target_value='Stage IIA', similarity=0.278), ValueMatch(current_value='pT1b (FIGO IB)', target_value='Stage IB2', similarity=0.262)]\n",
      "source: Path_Stage_Primary_Tumor-pT target: ajcc_clinical_stage score: 1.814\n",
      "value_matches:  [ValueMatch(current_value='nan', target_value='Unknown', similarity=0.345)]\n",
      "source: Path_Stage_Primary_Tumor-pT target: inss_stage score: 0.345\n",
      "value_matches:  []\n",
      "source: Path_Stage_Primary_Tumor-pT target: adapter_sequence score: 0\n",
      "value_matches:  []\n",
      "source: Path_Stage_Primary_Tumor-pT target: margin_distance score: 0\n",
      "Top k reranked columns:  [('Path_Stage_Primary_Tumor-pT', 'ajcc_pathologic_t', 2.391), ('Path_Stage_Primary_Tumor-pT', 'figo_stage', 2.364), ('Path_Stage_Primary_Tumor-pT', 'ensat_pathologic_t', 1.977), ('Path_Stage_Primary_Tumor-pT', 'ajcc_pathologic_stage', 1.9660000000000002), ('Path_Stage_Primary_Tumor-pT', 'uicc_clinical_stage', 1.8920000000000001), ('Path_Stage_Primary_Tumor-pT', 'ajcc_clinical_stage', 1.814), ('Path_Stage_Primary_Tumor-pT', 'uicc_pathologic_stage', 1.6909999999999998), ('Path_Stage_Primary_Tumor-pT', 'uicc_pathologic_t', 1.158), ('Path_Stage_Primary_Tumor-pT', 'ensat_pathologic_stage', 1.101), ('Path_Stage_Primary_Tumor-pT', 'uicc_clinical_t', 1.063), ('Path_Stage_Primary_Tumor-pT', 'extrathyroid_extension', 1.0), ('Path_Stage_Primary_Tumor-pT', 'ajcc_clinical_t', 0.877), ('Path_Stage_Primary_Tumor-pT', 'inss_stage', 0.345), ('Path_Stage_Primary_Tumor-pT', 'tumor_level_prostate', 0.297), ('Path_Stage_Primary_Tumor-pT', 'somatic_mutation_indexes', 0), ('Path_Stage_Primary_Tumor-pT', 'annotated_somatic_mutations', 0), ('Path_Stage_Primary_Tumor-pT', 'masked_somatic_mutations', 0), ('Path_Stage_Primary_Tumor-pT', 'tumor_grade_category', 0), ('Path_Stage_Primary_Tumor-pT', 'adapter_sequence', 0), ('Path_Stage_Primary_Tumor-pT', 'margin_distance', 0)]\n",
      "\n",
      "Source column:  Histologic_type\n",
      "value_matches:  [ValueMatch(current_value='Clear cell', target_value='Colorectal Cancer', similarity=0.313), ValueMatch(current_value='Carcinosarcoma', target_value='Phenochromocytoma or Paraganglioma', similarity=0.309)]\n",
      "source: Histologic_type target: history_of_tumor_type score: 0.622\n",
      "value_matches:  [ValueMatch(current_value='Carcinosarcoma', target_value='Carcinomatous', similarity=0.695), ValueMatch(current_value='Serous', target_value='Fibrous', similarity=0.393), ValueMatch(current_value='Endometrioid', target_value='Myeloid', similarity=0.252)]\n",
      "source: Histologic_type target: roots score: 1.34\n",
      "value_matches:  [ValueMatch(current_value='Endometrioid', target_value='Intermediate Risk', similarity=0.277)]\n",
      "source: Histologic_type target: cog_rhabdomyosarcoma_risk_group score: 0.277\n",
      "value_matches:  [ValueMatch(current_value='Serous', target_value='Mucinous and Serous Neoplasms', similarity=0.615), ValueMatch(current_value='Clear cell', target_value='Germ Cell Neoplasms', similarity=0.371), ValueMatch(current_value='Carcinosarcoma', target_value='NOS', similarity=0.304)]\n",
      "source: Histologic_type target: disease_type score: 1.29\n",
      "value_matches:  [ValueMatch(current_value='Serous', target_value='Ovarian serous carcinoma', similarity=0.576), ValueMatch(current_value='Carcinosarcoma', target_value='Basal cell carcinoma', similarity=0.472), ValueMatch(current_value='Clear cell', target_value='Renal cell carcinoma', similarity=0.44)]\n",
      "source: Histologic_type target: described_cases score: 1.488\n",
      "value_matches:  [ValueMatch(current_value='Carcinosarcoma', target_value='Chromium scATAC v1 Library', similarity=0.27)]\n",
      "source: Histologic_type target: single_cell_library score: 0.27\n",
      "value_matches:  []\n",
      "source: Histologic_type target: sarcomatoid_present score: 0\n",
      "value_matches:  [ValueMatch(current_value='Endometrioid', target_value='Bilateral ovaries with endometriotic cyst and surface adhesions', similarity=0.476), ValueMatch(current_value='Carcinosarcoma', target_value='Keratinizing dysplasia; severe (carcinoma in situ)', similarity=0.43), ValueMatch(current_value='Clear cell', target_value='Diffuse and early nodular diabetic glomerulosclerosis', similarity=0.288), ValueMatch(current_value='Serous', target_value='Diffuse and early nodular diabetic glomerulosclerosis', similarity=0.258)]\n",
      "source: Histologic_type target: additional_pathology_findings score: 1.452\n",
      "value_matches:  [ValueMatch(current_value='Carcinosarcoma', target_value='Osteosarcoma', similarity=0.584), ValueMatch(current_value='Clear cell', target_value='Squamous cell carcinoma', similarity=0.353), ValueMatch(current_value='Serous', target_value='Squamous cell carcinoma', similarity=0.259)]\n",
      "source: Histologic_type target: pathology_details score: 1.196\n",
      "value_matches:  [ValueMatch(current_value='Carcinosarcoma', target_value='Osteosarcoma', similarity=0.611), ValueMatch(current_value='Clear cell', target_value='Basal cell carcinoma', similarity=0.441), ValueMatch(current_value='Serous', target_value='Squamous cell carcinoma', similarity=0.259)]\n",
      "source: Histologic_type target: pathology_reports score: 1.311\n",
      "value_matches:  [ValueMatch(current_value='Carcinosarcoma', target_value='Kaposi Sarcoma', similarity=0.582), ValueMatch(current_value='Clear cell', target_value='Gallbladder Cancer', similarity=0.263)]\n",
      "source: Histologic_type target: relationship_primary_diagnosis score: 0.845\n",
      "value_matches:  [ValueMatch(current_value='Endometrioid', target_value='Not Reported', similarity=0.257)]\n",
      "source: Histologic_type target: chromosome score: 0.257\n",
      "value_matches:  []\n",
      "source: Histologic_type target: histone_variant score: 0\n",
      "value_matches:  [ValueMatch(current_value='Carcinosarcoma', target_value='Carcinoma', similarity=0.783), ValueMatch(current_value='Endometrioid', target_value='Endometrial stromal nodule', similarity=0.564), ValueMatch(current_value='Serous', target_value='Pseudomucinous cystadenoma', similarity=0.256)]\n",
      "source: Histologic_type target: primary_diagnosis score: 1.603\n",
      "value_matches:  [ValueMatch(current_value='Clear cell', target_value='Buccal Cells', similarity=0.412), ValueMatch(current_value='Endometrioid', target_value='2D Modified Conditionally Reprogrammed Cells', similarity=0.292)]\n",
      "source: Histologic_type target: composition score: 0.704\n",
      "value_matches:  [ValueMatch(current_value='Clear cell', target_value='Mononuclear Cells from Bone Marrow', similarity=0.491), ValueMatch(current_value='Serous', target_value='Serum', similarity=0.371), ValueMatch(current_value='Carcinosarcoma', target_value='Mononuclear Cells from Bone Marrow', similarity=0.262)]\n",
      "source: Histologic_type target: specimen_type score: 1.124\n",
      "value_matches:  []\n",
      "source: Histologic_type target: icd_10_code score: 0\n",
      "value_matches:  [ValueMatch(current_value='Clear cell', target_value='Likely Pathogenic', similarity=0.252)]\n",
      "source: Histologic_type target: pathogenicity score: 0.252\n",
      "value_matches:  []\n",
      "source: Histologic_type target: papillary_renal_cell_type score: 0\n",
      "value_matches:  []\n",
      "source: Histologic_type target: antigen score: 0\n",
      "Top k reranked columns:  [('Histologic_type', 'primary_diagnosis', 1.603), ('Histologic_type', 'described_cases', 1.488), ('Histologic_type', 'additional_pathology_findings', 1.452), ('Histologic_type', 'roots', 1.34), ('Histologic_type', 'pathology_reports', 1.311), ('Histologic_type', 'disease_type', 1.29), ('Histologic_type', 'pathology_details', 1.196), ('Histologic_type', 'specimen_type', 1.124), ('Histologic_type', 'relationship_primary_diagnosis', 0.845), ('Histologic_type', 'composition', 0.704), ('Histologic_type', 'history_of_tumor_type', 0.622), ('Histologic_type', 'cog_rhabdomyosarcoma_risk_group', 0.277), ('Histologic_type', 'single_cell_library', 0.27), ('Histologic_type', 'chromosome', 0.257), ('Histologic_type', 'pathogenicity', 0.252), ('Histologic_type', 'sarcomatoid_present', 0), ('Histologic_type', 'histone_variant', 0), ('Histologic_type', 'icd_10_code', 0), ('Histologic_type', 'papillary_renal_cell_type', 0), ('Histologic_type', 'antigen', 0)]\n",
      "\n",
      "Source column:  FIGO_stage\n",
      "value_matches:  [ValueMatch(current_value='IIIC1', target_value='Stage IIIC1', similarity=0.818), ValueMatch(current_value='IIIB', target_value='Stage IIIB', similarity=0.76), ValueMatch(current_value='IIIA', target_value='Stage IIIAii', similarity=0.736), ValueMatch(current_value='IVB', target_value='Stage IVB', similarity=0.718), ValueMatch(current_value='II', target_value='Stage IIIAii', similarity=0.628), ValueMatch(current_value='IIIC2', target_value='Stage IC2', similarity=0.555), ValueMatch(current_value='IA', target_value='Stage IA', similarity=0.539), ValueMatch(current_value='IB', target_value='Stage IIIB', similarity=0.456), ValueMatch(current_value='nan', target_value='Unknown', similarity=0.394)]\n",
      "source: FIGO_stage target: figo_stage score: 5.604\n",
      "value_matches:  [ValueMatch(current_value='IIIA', target_value='Stage IIIA', similarity=0.774), ValueMatch(current_value='IVB', target_value='Stage IVB', similarity=0.742), ValueMatch(current_value='II', target_value='Stage III', similarity=0.636), ValueMatch(current_value='IIIC2', target_value='Stage IIIC', similarity=0.62), ValueMatch(current_value='IIIC1', target_value='Stage IIIC', similarity=0.604), ValueMatch(current_value='IA', target_value='Stage IIA', similarity=0.532), ValueMatch(current_value='IIIB', target_value='Stage III', similarity=0.511), ValueMatch(current_value='IB', target_value='Stage IB2', similarity=0.427)]\n",
      "source: FIGO_stage target: ajcc_clinical_stage score: 4.846\n",
      "value_matches:  [ValueMatch(current_value='IIIC2', target_value='Stage IIIC2', similarity=0.84), ValueMatch(current_value='IVB', target_value='Stage IVB', similarity=0.727), ValueMatch(current_value='IIIA', target_value='Stage IIIA1', similarity=0.689), ValueMatch(current_value='IIIC1', target_value='Stage IIIC2', similarity=0.539), ValueMatch(current_value='IA', target_value='Stage IA', similarity=0.527), ValueMatch(current_value='IB', target_value='Stage IB', similarity=0.518), ValueMatch(current_value='II', target_value='Stage IIIA1', similarity=0.494), ValueMatch(current_value='IIIB', target_value='Stage IIIA1', similarity=0.408)]\n",
      "source: FIGO_stage target: uicc_pathologic_stage score: 4.742\n",
      "value_matches:  [ValueMatch(current_value='IIIB', target_value='Stage IIIB', similarity=0.771), ValueMatch(current_value='IIIA', target_value='Stage IIIA', similarity=0.757), ValueMatch(current_value='IVB', target_value='Stage IVB', similarity=0.741), ValueMatch(current_value='IIIC1', target_value='Stage IIC1', similarity=0.738), ValueMatch(current_value='IIIC2', target_value='Stage IIIC', similarity=0.561), ValueMatch(current_value='II', target_value='Stage IIIA', similarity=0.532), ValueMatch(current_value='IA', target_value='Stage IA', similarity=0.518), ValueMatch(current_value='IB', target_value='Stage IIIB', similarity=0.488)]\n",
      "source: FIGO_stage target: uicc_clinical_stage score: 5.106\n",
      "value_matches:  [ValueMatch(current_value='IIIA', target_value='Group IIIa', similarity=0.739), ValueMatch(current_value='IIIB', target_value='Group IIIb', similarity=0.739), ValueMatch(current_value='II', target_value='Group IIIb', similarity=0.518), ValueMatch(current_value='IA', target_value='Group Ia', similarity=0.485), ValueMatch(current_value='IB', target_value='Group Ib', similarity=0.485), ValueMatch(current_value='IIIC1', target_value='Group IIc', similarity=0.47), ValueMatch(current_value='IIIC2', target_value='Group IIc', similarity=0.47), ValueMatch(current_value='IVB', target_value='Group IV', similarity=0.378)]\n",
      "source: FIGO_stage target: irs_group score: 4.284\n",
      "value_matches:  [ValueMatch(current_value='IIIA', target_value='Stage IIIA', similarity=0.761), ValueMatch(current_value='IVB', target_value='Stage IVB', similarity=0.725), ValueMatch(current_value='IIIB', target_value='Stage IIB', similarity=0.626), ValueMatch(current_value='II', target_value='Stage III', similarity=0.617), ValueMatch(current_value='IB', target_value='Stage IIB', similarity=0.509), ValueMatch(current_value='IA', target_value='Stage IIIA', similarity=0.507), ValueMatch(current_value='IIIC1', target_value='Stage III', similarity=0.399), ValueMatch(current_value='IIIC2', target_value='Stage III', similarity=0.389)]\n",
      "source: FIGO_stage target: ajcc_pathologic_stage score: 4.533\n",
      "value_matches:  [ValueMatch(current_value='II', target_value='II', similarity=1.0), ValueMatch(current_value='IIIB', target_value='III', similarity=0.752), ValueMatch(current_value='IIIA', target_value='III', similarity=0.752), ValueMatch(current_value='IIIC1', target_value='III', similarity=0.606), ValueMatch(current_value='IIIC2', target_value='III', similarity=0.606), ValueMatch(current_value='IA', target_value='I', similarity=0.331), ValueMatch(current_value='IB', target_value='I', similarity=0.331), ValueMatch(current_value='nan', target_value='Unknown', similarity=0.327)]\n",
      "source: FIGO_stage target: iss_stage score: 4.705\n",
      "value_matches:  [ValueMatch(current_value='nan', target_value='Unknown', similarity=0.344)]\n",
      "source: FIGO_stage target: inss_stage score: 0.344\n",
      "value_matches:  [ValueMatch(current_value='nan', target_value='Unknown', similarity=0.33), ValueMatch(current_value='IIIC1', target_value='1', similarity=0.281), ValueMatch(current_value='IIIC2', target_value='2', similarity=0.281)]\n",
      "source: FIGO_stage target: irs_stage score: 0.892\n",
      "value_matches:  [ValueMatch(current_value='II', target_value='Stage III', similarity=0.544), ValueMatch(current_value='IIIA', target_value='Stage III', similarity=0.502), ValueMatch(current_value='IIIB', target_value='Stage III', similarity=0.443), ValueMatch(current_value='IIIC1', target_value='Stage III', similarity=0.357), ValueMatch(current_value='IIIC2', target_value='Stage III', similarity=0.357), ValueMatch(current_value='nan', target_value='Unknown', similarity=0.338), ValueMatch(current_value='IVB', target_value='Stage IV', similarity=0.299)]\n",
      "source: FIGO_stage target: cog_liver_stage score: 2.8400000000000003\n",
      "value_matches:  []\n",
      "source: FIGO_stage target: shortest_dimension score: 0\n",
      "value_matches:  [ValueMatch(current_value='IB', target_value='GB', similarity=0.313), ValueMatch(current_value='nan', target_value='Unknown', similarity=0.31)]\n",
      "source: FIGO_stage target: tumor_grade score: 0.623\n",
      "value_matches:  []\n",
      "source: FIGO_stage target: fragment_minimum_length score: 0\n",
      "value_matches:  [ValueMatch(current_value='IVB', target_value='Stage IVb', similarity=0.665), ValueMatch(current_value='IIIA', target_value='Stage IIa', similarity=0.573), ValueMatch(current_value='IIIB', target_value='Stage IIb', similarity=0.564), ValueMatch(current_value='II', target_value='Stage III', similarity=0.56), ValueMatch(current_value='IA', target_value='Stage IIa', similarity=0.507), ValueMatch(current_value='IB', target_value='Stage IIb', similarity=0.458), ValueMatch(current_value='IIIC2', target_value='Stage III', similarity=0.353), ValueMatch(current_value='IIIC1', target_value='Stage III', similarity=0.353)]\n",
      "source: FIGO_stage target: masaoka_stage score: 4.033\n",
      "value_matches:  [ValueMatch(current_value='II', target_value='Stage III', similarity=0.544), ValueMatch(current_value='IIIA', target_value='Stage III', similarity=0.502), ValueMatch(current_value='IIIB', target_value='Stage III', similarity=0.443), ValueMatch(current_value='IIIC1', target_value='Stage III', similarity=0.357), ValueMatch(current_value='IIIC2', target_value='Stage III', similarity=0.357), ValueMatch(current_value='nan', target_value='Unknown', similarity=0.338), ValueMatch(current_value='IVB', target_value='Stage IV', similarity=0.299)]\n",
      "source: FIGO_stage target: cog_renal_stage score: 2.8400000000000003\n",
      "value_matches:  []\n",
      "source: FIGO_stage target: fragment_maximum_length score: 0\n",
      "value_matches:  [ValueMatch(current_value='II', target_value='Stage III', similarity=0.544), ValueMatch(current_value='IIIA', target_value='Stage III', similarity=0.502), ValueMatch(current_value='IIIB', target_value='Stage III', similarity=0.443), ValueMatch(current_value='IIIC1', target_value='Stage III', similarity=0.357), ValueMatch(current_value='IIIC2', target_value='Stage III', similarity=0.357), ValueMatch(current_value='nan', target_value='Unknown', similarity=0.338), ValueMatch(current_value='IVB', target_value='Stage IV', similarity=0.299)]\n",
      "source: FIGO_stage target: ann_arbor_clinical_stage score: 2.8400000000000003\n",
      "value_matches:  []\n",
      "source: FIGO_stage target: fragment_mean_length score: 0\n",
      "value_matches:  [ValueMatch(current_value='II', target_value='Class III', similarity=0.541), ValueMatch(current_value='IIIA', target_value='Class III', similarity=0.492), ValueMatch(current_value='IIIB', target_value='Class III', similarity=0.437), ValueMatch(current_value='IIIC1', target_value='Class III', similarity=0.397), ValueMatch(current_value='IIIC2', target_value='Class III', similarity=0.397), ValueMatch(current_value='IVB', target_value='Class IV', similarity=0.281)]\n",
      "source: FIGO_stage target: myasthenia_gravis_classification score: 2.5450000000000004\n",
      "value_matches:  [ValueMatch(current_value='nan', target_value='Unknown', similarity=0.33)]\n",
      "source: FIGO_stage target: who_nte_grade score: 0.33\n",
      "Top k reranked columns:  [('FIGO_stage', 'figo_stage', 5.604), ('FIGO_stage', 'uicc_clinical_stage', 5.106), ('FIGO_stage', 'ajcc_clinical_stage', 4.846), ('FIGO_stage', 'uicc_pathologic_stage', 4.742), ('FIGO_stage', 'iss_stage', 4.705), ('FIGO_stage', 'ajcc_pathologic_stage', 4.533), ('FIGO_stage', 'irs_group', 4.284), ('FIGO_stage', 'masaoka_stage', 4.033), ('FIGO_stage', 'cog_liver_stage', 2.8400000000000003), ('FIGO_stage', 'cog_renal_stage', 2.8400000000000003), ('FIGO_stage', 'ann_arbor_clinical_stage', 2.8400000000000003), ('FIGO_stage', 'myasthenia_gravis_classification', 2.5450000000000004), ('FIGO_stage', 'irs_stage', 0.892), ('FIGO_stage', 'tumor_grade', 0.623), ('FIGO_stage', 'inss_stage', 0.344), ('FIGO_stage', 'who_nte_grade', 0.33), ('FIGO_stage', 'shortest_dimension', 0), ('FIGO_stage', 'fragment_minimum_length', 0), ('FIGO_stage', 'fragment_maximum_length', 0), ('FIGO_stage', 'fragment_mean_length', 0)]\n",
      "\n",
      "Source column:  BMI\n",
      "value_matches:  [ValueMatch(current_value='28.84', target_value='28.8', similarity=0.789), ValueMatch(current_value='35.42', target_value='35.4', similarity=0.78), ValueMatch(current_value='38.88', target_value='28.8', similarity=0.53), ValueMatch(current_value='32.32', target_value='33.2', similarity=0.43), ValueMatch(current_value='23.88', target_value='28.8', similarity=0.398), ValueMatch(current_value='44.0', target_value='40.2', similarity=0.383), ValueMatch(current_value='28.5', target_value='28.8', similarity=0.377), ValueMatch(current_value='17.11', target_value='31.7', similarity=0.377), ValueMatch(current_value='34.0', target_value='40.2', similarity=0.377), ValueMatch(current_value='20.55', target_value='30.5', similarity=0.372), ValueMatch(current_value='26.22', target_value='22.3', similarity=0.357), ValueMatch(current_value='38.89', target_value='28.8', similarity=0.356), ValueMatch(current_value='36.2', target_value='23.6', similarity=0.353), ValueMatch(current_value='35.0', target_value='35.4', similarity=0.353), ValueMatch(current_value='20.28', target_value='40.2', similarity=0.347), ValueMatch(current_value='39.14', target_value='29.1', similarity=0.344), ValueMatch(current_value='46.64', target_value='26.4', similarity=0.34), ValueMatch(current_value='38.54', target_value='35.4', similarity=0.34), ValueMatch(current_value='33.65', target_value='33.2', similarity=0.336), ValueMatch(current_value='43.0', target_value='30.5', similarity=0.331), ValueMatch(current_value='31.22', target_value='22.3', similarity=0.33), ValueMatch(current_value='55.86', target_value='25.8', similarity=0.318), ValueMatch(current_value='26.0', target_value='26.4', similarity=0.316), ValueMatch(current_value='25.68', target_value='25.8', similarity=0.315), ValueMatch(current_value='31.83', target_value='18.5', similarity=0.314), ValueMatch(current_value='21.83', target_value='18.5', similarity=0.314), ValueMatch(current_value='46.45', target_value='26.4', similarity=0.314), ValueMatch(current_value='17.85', target_value='18.5', similarity=0.313), ValueMatch(current_value='46.41', target_value='26.4', similarity=0.309), ValueMatch(current_value='22.86', target_value='28.8', similarity=0.308), ValueMatch(current_value='17.64', target_value='31.7', similarity=0.306), ValueMatch(current_value='30.85', target_value='30.5', similarity=0.305), ValueMatch(current_value='24.32', target_value='24.9', similarity=0.305), ValueMatch(current_value='32.83', target_value='33.2', similarity=0.305), ValueMatch(current_value='34.26', target_value='26.4', similarity=0.304), ValueMatch(current_value='36.0', target_value='23.6', similarity=0.299), ValueMatch(current_value='29.4', target_value='29.1', similarity=0.299), ValueMatch(current_value='27.1', target_value='27.5', similarity=0.298), ValueMatch(current_value='40.72', target_value='40.2', similarity=0.295), ValueMatch(current_value='45.83', target_value='25.8', similarity=0.292), ValueMatch(current_value='29.37', target_value='37.9', similarity=0.291), ValueMatch(current_value='37.69', target_value='37.9', similarity=0.29), ValueMatch(current_value='31.0', target_value='31.7', similarity=0.288), ValueMatch(current_value='27.31', target_value='31.7', similarity=0.286), ValueMatch(current_value='31.58', target_value='25.8', similarity=0.285), ValueMatch(current_value='26.14', target_value='26.4', similarity=0.284), ValueMatch(current_value='34.06', target_value='40.2', similarity=0.279), ValueMatch(current_value='27.0', target_value='27.5', similarity=0.276), ValueMatch(current_value='25.37', target_value='25.8', similarity=0.274), ValueMatch(current_value='29.62', target_value='29.1', similarity=0.265), ValueMatch(current_value='27.82', target_value='27.5', similarity=0.262), ValueMatch(current_value='30.48', target_value='30.5', similarity=0.262), ValueMatch(current_value='46.85', target_value='18.5', similarity=0.26), ValueMatch(current_value='25.03', target_value='25.8', similarity=0.258), ValueMatch(current_value='34.37', target_value='37.9', similarity=0.258), ValueMatch(current_value='30.47', target_value='30.5', similarity=0.257), ValueMatch(current_value='29.52', target_value='29.1', similarity=0.254)]\n",
      "source: BMI target: bmi score: 19.058\n",
      "value_matches:  [ValueMatch(current_value='31.22', target_value='31.2', similarity=0.785), ValueMatch(current_value='25.37', target_value='25.3', similarity=0.773), ValueMatch(current_value='30.48', target_value='30.4', similarity=0.771), ValueMatch(current_value='27.82', target_value='27.8', similarity=0.767), ValueMatch(current_value='34.53', target_value='34.5', similarity=0.767), ValueMatch(current_value='27.83', target_value='27.8', similarity=0.763), ValueMatch(current_value='38.44', target_value='38.4', similarity=0.76), ValueMatch(current_value='30.47', target_value='30.4', similarity=0.757), ValueMatch(current_value='38.41', target_value='38.4', similarity=0.732), ValueMatch(current_value='32.69', target_value='32.6', similarity=0.728), ValueMatch(current_value='32.32', target_value='32.6', similarity=0.416), ValueMatch(current_value='43.0', target_value='30.4', similarity=0.41), ValueMatch(current_value='44.0', target_value='40.1', similarity=0.367), ValueMatch(current_value='34.0', target_value='40.1', similarity=0.364), ValueMatch(current_value='34.84', target_value='38.4', similarity=0.362), ValueMatch(current_value='45.53', target_value='34.5', similarity=0.359), ValueMatch(current_value='26.22', target_value='22.5', similarity=0.357), ValueMatch(current_value='29.4', target_value='29.9', similarity=0.349), ValueMatch(current_value='27.1', target_value='27.8', similarity=0.34), ValueMatch(current_value='35.0', target_value='35.2', similarity=0.339), ValueMatch(current_value='26.0', target_value='32.6', similarity=0.334), ValueMatch(current_value='33.65', target_value='33.7', similarity=0.329), ValueMatch(current_value='46.45', target_value='34.5', similarity=0.324), ValueMatch(current_value='24.32', target_value='24.1', similarity=0.324), ValueMatch(current_value='29.52', target_value='35.2', similarity=0.324), ValueMatch(current_value='41.44', target_value='24.1', similarity=0.324), ValueMatch(current_value='47.82', target_value='27.8', similarity=0.322), ValueMatch(current_value='48.46', target_value='38.4', similarity=0.316), ValueMatch(current_value='28.84', target_value='38.4', similarity=0.316), ValueMatch(current_value='27.0', target_value='27.8', similarity=0.315), ValueMatch(current_value='45.83', target_value='34.5', similarity=0.314), ValueMatch(current_value='36.84', target_value='38.4', similarity=0.313), ValueMatch(current_value='46.41', target_value='24.1', similarity=0.311), ValueMatch(current_value='35.42', target_value='35.2', similarity=0.307), ValueMatch(current_value='34.37', target_value='33.7', similarity=0.307), ValueMatch(current_value='34.26', target_value='32.6', similarity=0.306), ValueMatch(current_value='22.86', target_value='22.5', similarity=0.304), ValueMatch(current_value='28.5', target_value='28.7', similarity=0.3), ValueMatch(current_value='29.62', target_value='29.9', similarity=0.3), ValueMatch(current_value='38.88', target_value='38.4', similarity=0.292), ValueMatch(current_value='23.88', target_value='38.4', similarity=0.292), ValueMatch(current_value='38.89', target_value='38.4', similarity=0.291), ValueMatch(current_value='31.0', target_value='31.2', similarity=0.29), ValueMatch(current_value='25.03', target_value='25.3', similarity=0.287), ValueMatch(current_value='38.54', target_value='38.4', similarity=0.287), ValueMatch(current_value='32.06', target_value='32.6', similarity=0.287), ValueMatch(current_value='29.37', target_value='37.9', similarity=0.284), ValueMatch(current_value='25.68', target_value='22.5', similarity=0.283), ValueMatch(current_value='17.85', target_value='27.8', similarity=0.282), ValueMatch(current_value='37.69', target_value='37.9', similarity=0.282), ValueMatch(current_value='21.45', target_value='34.5', similarity=0.28), ValueMatch(current_value='42.98', target_value='29.9', similarity=0.276), ValueMatch(current_value='32.83', target_value='32.6', similarity=0.274), ValueMatch(current_value='30.85', target_value='30.4', similarity=0.27), ValueMatch(current_value='59.78', target_value='27.8', similarity=0.27), ValueMatch(current_value='34.06', target_value='40.1', similarity=0.268), ValueMatch(current_value='29.07', target_value='29.9', similarity=0.267), ValueMatch(current_value='27.31', target_value='31.2', similarity=0.265), ValueMatch(current_value='31.83', target_value='31.2', similarity=0.261), ValueMatch(current_value='26.14', target_value='32.6', similarity=0.256), ValueMatch(current_value='40.72', target_value='40.1', similarity=0.253)]\n",
      "source: BMI target: average_base_quality score: 23.353\n",
      "value_matches:  [ValueMatch(current_value='29.4', target_value='2.94', similarity=1.0), ValueMatch(current_value='31.63', target_value='1.63', similarity=0.816), ValueMatch(current_value='32.06', target_value='2.06', similarity=0.788), ValueMatch(current_value='26.14', target_value='2.61', similarity=0.775), ValueMatch(current_value='32.83', target_value='2.83', similarity=0.771), ValueMatch(current_value='22.86', target_value='2.28', similarity=0.756), ValueMatch(current_value='25.37', target_value='2.5', similarity=0.461), ValueMatch(current_value='25.68', target_value='2.5', similarity=0.454), ValueMatch(current_value='25.03', target_value='2.5', similarity=0.436), ValueMatch(current_value='20.55', target_value='3.05', similarity=0.374), ValueMatch(current_value='26.22', target_value='2.28', similarity=0.372), ValueMatch(current_value='39.2', target_value='2.39', similarity=0.369), ValueMatch(current_value='51.19', target_value='1.95', similarity=0.356), ValueMatch(current_value='28.5', target_value='2.28', similarity=0.356), ValueMatch(current_value='17.11', target_value='2.17', similarity=0.349), ValueMatch(current_value='27.1', target_value='2.72', similarity=0.347), ValueMatch(current_value='34.06', target_value='2.06', similarity=0.334), ValueMatch(current_value='43.0', target_value='3.05', similarity=0.332), ValueMatch(current_value='27.82', target_value='2.72', similarity=0.332), ValueMatch(current_value='21.57', target_value='1.52', similarity=0.324), ValueMatch(current_value='34.72', target_value='2.72', similarity=0.324), ValueMatch(current_value='29.52', target_value='1.52', similarity=0.322), ValueMatch(current_value='27.0', target_value='2.72', similarity=0.32), ValueMatch(current_value='20.28', target_value='2.28', similarity=0.318), ValueMatch(current_value='31.22', target_value='2.28', similarity=0.316), ValueMatch(current_value='21.83', target_value='1.85', similarity=0.313), ValueMatch(current_value='17.85', target_value='1.85', similarity=0.311), ValueMatch(current_value='28.84', target_value='2.28', similarity=0.311), ValueMatch(current_value='40.72', target_value='2.72', similarity=0.31), ValueMatch(current_value='31.83', target_value='1.85', similarity=0.309), ValueMatch(current_value='26.0', target_value='2.61', similarity=0.308), ValueMatch(current_value='30.85', target_value='3.05', similarity=0.306), ValueMatch(current_value='17.64', target_value='1.74', similarity=0.302), ValueMatch(current_value='27.83', target_value='2.83', similarity=0.301), ValueMatch(current_value='31.58', target_value='1.52', similarity=0.297), ValueMatch(current_value='32.32', target_value='2.39', similarity=0.29), ValueMatch(current_value='42.19', target_value='1.95', similarity=0.289), ValueMatch(current_value='42.98', target_value='2.94', similarity=0.288), ValueMatch(current_value='31.96', target_value='1.95', similarity=0.286), ValueMatch(current_value='29.62', target_value='2.94', similarity=0.275), ValueMatch(current_value='68.39', target_value='2.39', similarity=0.274), ValueMatch(current_value='23.88', target_value='2.39', similarity=0.27), ValueMatch(current_value='39.76', target_value='2.39', similarity=0.265), ValueMatch(current_value='30.48', target_value='3.05', similarity=0.263), ValueMatch(current_value='27.31', target_value='2.72', similarity=0.261), ValueMatch(current_value='46.85', target_value='1.85', similarity=0.261), ValueMatch(current_value='21.45', target_value='2.17', similarity=0.261), ValueMatch(current_value='39.14', target_value='2.39', similarity=0.259), ValueMatch(current_value='38.54', target_value='1.85', similarity=0.258), ValueMatch(current_value='30.47', target_value='3.05', similarity=0.258), ValueMatch(current_value='34.26', target_value='2.61', similarity=0.252), ValueMatch(current_value='27.66', target_value='2.72', similarity=0.251), ValueMatch(current_value='45.83', target_value='2.83', similarity=0.25)]\n",
      "source: BMI target: body_surface_area score: 19.580999999999996\n",
      "value_matches:  [ValueMatch(current_value='34.06', target_value='40.6', similarity=0.814), ValueMatch(current_value='42.19', target_value='21.9', similarity=0.78), ValueMatch(current_value='30.85', target_value='30.8', similarity=0.78), ValueMatch(current_value='27.1', target_value='7.1', similarity=0.606), ValueMatch(current_value='17.11', target_value='7.1', similarity=0.569), ValueMatch(current_value='37.11', target_value='7.1', similarity=0.538), ValueMatch(current_value='34.84', target_value='4.8', similarity=0.535), ValueMatch(current_value='48.46', target_value='4.8', similarity=0.515), ValueMatch(current_value='29.37', target_value='9.3', similarity=0.509), ValueMatch(current_value='44.81', target_value='4.8', similarity=0.486), ValueMatch(current_value='25.37', target_value='2.5', similarity=0.468), ValueMatch(current_value='25.68', target_value='2.5', similarity=0.459), ValueMatch(current_value='65.71', target_value='7.1', similarity=0.453), ValueMatch(current_value='34.89', target_value='4.8', similarity=0.451), ValueMatch(current_value='25.03', target_value='2.5', similarity=0.441), ValueMatch(current_value='30.48', target_value='4.8', similarity=0.438), ValueMatch(current_value='71.09', target_value='7.1', similarity=0.425), ValueMatch(current_value='44.0', target_value='40.6', similarity=0.392), ValueMatch(current_value='24.32', target_value='24.2', similarity=0.39), ValueMatch(current_value='34.0', target_value='40.6', similarity=0.388), ValueMatch(current_value='43.0', target_value='30.8', similarity=0.346), ValueMatch(current_value='41.44', target_value='34.1', similarity=0.344), ValueMatch(current_value='27.31', target_value='37.3', similarity=0.34), ValueMatch(current_value='38.41', target_value='34.1', similarity=0.339), ValueMatch(current_value='51.19', target_value='21.9', similarity=0.336), ValueMatch(current_value='31.22', target_value='12.0', similarity=0.335), ValueMatch(current_value='46.41', target_value='34.1', similarity=0.33), ValueMatch(current_value='32.06', target_value='40.6', similarity=0.327), ValueMatch(current_value='34.37', target_value='37.3', similarity=0.325), ValueMatch(current_value='31.96', target_value='21.9', similarity=0.312), ValueMatch(current_value='38.44', target_value='43.8', similarity=0.311), ValueMatch(current_value='31.58', target_value='15.4', similarity=0.304), ValueMatch(current_value='21.83', target_value='18.6', similarity=0.303), ValueMatch(current_value='31.83', target_value='18.6', similarity=0.3), ValueMatch(current_value='21.57', target_value='15.4', similarity=0.299), ValueMatch(current_value='38.54', target_value='15.4', similarity=0.298), ValueMatch(current_value='20.28', target_value='12.0', similarity=0.296), ValueMatch(current_value='35.42', target_value='15.4', similarity=0.288), ValueMatch(current_value='22.86', target_value='18.6', similarity=0.281), ValueMatch(current_value='34.26', target_value='24.2', similarity=0.281), ValueMatch(current_value='27.0', target_value='27.5', similarity=0.279), ValueMatch(current_value='55.86', target_value='18.6', similarity=0.271), ValueMatch(current_value='40.72', target_value='40.6', similarity=0.269), ValueMatch(current_value='23.88', target_value='43.8', similarity=0.268), ValueMatch(current_value='38.89', target_value='43.8', similarity=0.268), ValueMatch(current_value='21.45', target_value='21.9', similarity=0.267), ValueMatch(current_value='38.88', target_value='43.8', similarity=0.267), ValueMatch(current_value='27.82', target_value='27.5', similarity=0.267), ValueMatch(current_value='30.47', target_value='30.8', similarity=0.265), ValueMatch(current_value='42.98', target_value='24.2', similarity=0.262), ValueMatch(current_value='37.69', target_value='37.3', similarity=0.256), ValueMatch(current_value='20.55', target_value='12.0', similarity=0.255)]\n",
      "source: BMI target: intermediate_dimension score: 19.926\n",
      "value_matches:  [ValueMatch(current_value='24.32', target_value='43.2', similarity=0.759), ValueMatch(current_value='21.57', target_value='21.5', similarity=0.755), ValueMatch(current_value='27.1', target_value='7.1', similarity=0.603), ValueMatch(current_value='17.11', target_value='7.1', similarity=0.57), ValueMatch(current_value='37.11', target_value='7.1', similarity=0.536), ValueMatch(current_value='34.84', target_value='4.8', similarity=0.535), ValueMatch(current_value='48.46', target_value='4.8', similarity=0.519), ValueMatch(current_value='29.37', target_value='9.3', similarity=0.506), ValueMatch(current_value='44.81', target_value='4.8', similarity=0.487), ValueMatch(current_value='25.37', target_value='2.5', similarity=0.462), ValueMatch(current_value='25.68', target_value='2.5', similarity=0.46), ValueMatch(current_value='65.71', target_value='7.1', similarity=0.458), ValueMatch(current_value='34.89', target_value='4.8', similarity=0.454), ValueMatch(current_value='25.03', target_value='2.5', similarity=0.437), ValueMatch(current_value='30.48', target_value='4.8', similarity=0.435), ValueMatch(current_value='71.09', target_value='7.1', similarity=0.426), ValueMatch(current_value='32.32', target_value='43.2', similarity=0.402), ValueMatch(current_value='38.89', target_value='28.9', similarity=0.35), ValueMatch(current_value='43.0', target_value='43.2', similarity=0.35), ValueMatch(current_value='21.83', target_value='18.2', similarity=0.344), ValueMatch(current_value='35.0', target_value='35.6', similarity=0.342), ValueMatch(current_value='46.45', target_value='46.5', similarity=0.338), ValueMatch(current_value='42.98', target_value='39.8', similarity=0.335), ValueMatch(current_value='31.22', target_value='12.0', similarity=0.331), ValueMatch(current_value='55.67', target_value='35.6', similarity=0.324), ValueMatch(current_value='39.2', target_value='39.8', similarity=0.319), ValueMatch(current_value='27.82', target_value='18.2', similarity=0.318), ValueMatch(current_value='21.45', target_value='21.5', similarity=0.317), ValueMatch(current_value='28.5', target_value='28.9', similarity=0.314), ValueMatch(current_value='47.82', target_value='24.7', similarity=0.313), ValueMatch(current_value='34.72', target_value='24.7', similarity=0.311), ValueMatch(current_value='46.64', target_value='46.5', similarity=0.31), ValueMatch(current_value='34.37', target_value='43.2', similarity=0.31), ValueMatch(current_value='68.39', target_value='39.8', similarity=0.31), ValueMatch(current_value='31.58', target_value='21.5', similarity=0.307), ValueMatch(current_value='46.0', target_value='46.5', similarity=0.305), ValueMatch(current_value='36.57', target_value='46.5', similarity=0.305), ValueMatch(current_value='31.83', target_value='18.2', similarity=0.303), ValueMatch(current_value='46.85', target_value='46.5', similarity=0.303), ValueMatch(current_value='38.97', target_value='28.9', similarity=0.302), ValueMatch(current_value='38.54', target_value='15.4', similarity=0.299), ValueMatch(current_value='20.28', target_value='12.0', similarity=0.298), ValueMatch(current_value='35.42', target_value='15.4', similarity=0.291), ValueMatch(current_value='33.65', target_value='46.5', similarity=0.288), ValueMatch(current_value='28.84', target_value='28.9', similarity=0.282), ValueMatch(current_value='46.41', target_value='46.5', similarity=0.278), ValueMatch(current_value='32.06', target_value='12.0', similarity=0.278), ValueMatch(current_value='30.47', target_value='24.7', similarity=0.271), ValueMatch(current_value='32.83', target_value='43.2', similarity=0.266), ValueMatch(current_value='22.86', target_value='28.9', similarity=0.264), ValueMatch(current_value='39.76', target_value='39.8', similarity=0.257), ValueMatch(current_value='20.55', target_value='12.0', similarity=0.257), ValueMatch(current_value='42.19', target_value='32.1', similarity=0.257), ValueMatch(current_value='39.14', target_value='39.8', similarity=0.252)]\n",
      "source: BMI target: recist_targeted_regions_sum score: 20.003000000000004\n",
      "value_matches:  [ValueMatch(current_value='36.2', target_value='36.2', similarity=1.0), ValueMatch(current_value='42.19', target_value='21.9', similarity=0.772), ValueMatch(current_value='44.0', target_value='4.4', similarity=0.629), ValueMatch(current_value='41.44', target_value='4.4', similarity=0.585), ValueMatch(current_value='38.44', target_value='4.4', similarity=0.525), ValueMatch(current_value='29.52', target_value='5.2', similarity=0.523), ValueMatch(current_value='44.81', target_value='4.4', similarity=0.493), ValueMatch(current_value='38.97', target_value='9.7', similarity=0.48), ValueMatch(current_value='39.76', target_value='9.7', similarity=0.476), ValueMatch(current_value='59.78', target_value='9.7', similarity=0.458), ValueMatch(current_value='31.0', target_value='11.0', similarity=0.448), ValueMatch(current_value='34.0', target_value='40.0', similarity=0.422), ValueMatch(current_value='17.11', target_value='11.0', similarity=0.409), ValueMatch(current_value='37.11', target_value='11.0', similarity=0.354), ValueMatch(current_value='31.22', target_value='12.8', similarity=0.344), ValueMatch(current_value='28.5', target_value='18.5', similarity=0.343), ValueMatch(current_value='26.22', target_value='36.2', similarity=0.342), ValueMatch(current_value='33.65', target_value='33.1', similarity=0.341), ValueMatch(current_value='51.19', target_value='21.9', similarity=0.336), ValueMatch(current_value='71.09', target_value='11.0', similarity=0.327), ValueMatch(current_value='31.63', target_value='26.3', similarity=0.326), ValueMatch(current_value='55.67', target_value='15.6', similarity=0.323), ValueMatch(current_value='29.62', target_value='36.2', similarity=0.317), ValueMatch(current_value='21.83', target_value='18.5', similarity=0.313), ValueMatch(current_value='34.06', target_value='40.0', similarity=0.312), ValueMatch(current_value='29.4', target_value='29.7', similarity=0.312), ValueMatch(current_value='17.85', target_value='18.5', similarity=0.31), ValueMatch(current_value='31.96', target_value='21.9', similarity=0.31), ValueMatch(current_value='31.83', target_value='18.5', similarity=0.31), ValueMatch(current_value='36.0', target_value='36.2', similarity=0.309), ValueMatch(current_value='32.32', target_value='23.4', similarity=0.306), ValueMatch(current_value='26.0', target_value='26.3', similarity=0.306), ValueMatch(current_value='29.37', target_value='29.7', similarity=0.305), ValueMatch(current_value='31.58', target_value='15.6', similarity=0.303), ValueMatch(current_value='21.57', target_value='15.6', similarity=0.298), ValueMatch(current_value='25.68', target_value='15.6', similarity=0.295), ValueMatch(current_value='29.07', target_value='29.7', similarity=0.294), ValueMatch(current_value='40.72', target_value='40.0', similarity=0.293), ValueMatch(current_value='23.88', target_value='23.4', similarity=0.284), ValueMatch(current_value='34.26', target_value='26.3', similarity=0.283), ValueMatch(current_value='28.84', target_value='12.8', similarity=0.28), ValueMatch(current_value='34.84', target_value='4.4', similarity=0.27), ValueMatch(current_value='32.69', target_value='26.3', similarity=0.27), ValueMatch(current_value='21.45', target_value='21.9', similarity=0.266), ValueMatch(current_value='46.85', target_value='18.5', similarity=0.264), ValueMatch(current_value='48.46', target_value='4.4', similarity=0.261), ValueMatch(current_value='38.54', target_value='18.5', similarity=0.261), ValueMatch(current_value='22.86', target_value='12.8', similarity=0.261), ValueMatch(current_value='46.45', target_value='4.4', similarity=0.257), ValueMatch(current_value='20.28', target_value='12.8', similarity=0.255), ValueMatch(current_value='36.84', target_value='36.2', similarity=0.254), ValueMatch(current_value='46.41', target_value='4.4', similarity=0.253), ValueMatch(current_value='30.85', target_value='18.5', similarity=0.25)]\n",
      "source: BMI target: percent_stromal_cells score: 19.118\n",
      "value_matches:  [ValueMatch(current_value='31.22', target_value='31.2', similarity=0.772), ValueMatch(current_value='17.85', target_value='17.8', similarity=0.77), ValueMatch(current_value='33.65', target_value='36.5', similarity=0.759), ValueMatch(current_value='36.57', target_value='36.5', similarity=0.758), ValueMatch(current_value='34.84', target_value='4.8', similarity=0.541), ValueMatch(current_value='48.46', target_value='4.8', similarity=0.522), ValueMatch(current_value='27.31', target_value='7.3', similarity=0.506), ValueMatch(current_value='44.81', target_value='4.8', similarity=0.492), ValueMatch(current_value='34.89', target_value='4.8', similarity=0.456), ValueMatch(current_value='25.37', target_value='2.5', similarity=0.455), ValueMatch(current_value='25.68', target_value='2.5', similarity=0.446), ValueMatch(current_value='30.48', target_value='4.8', similarity=0.444), ValueMatch(current_value='25.03', target_value='2.5', similarity=0.433), ValueMatch(current_value='31.0', target_value='10.1', similarity=0.429), ValueMatch(current_value='35.0', target_value='15.0', similarity=0.382), ValueMatch(current_value='20.55', target_value='25.5', similarity=0.374), ValueMatch(current_value='17.11', target_value='17.8', similarity=0.371), ValueMatch(current_value='45.53', target_value='25.5', similarity=0.357), ValueMatch(current_value='38.89', target_value='28.9', similarity=0.351), ValueMatch(current_value='55.86', target_value='25.5', similarity=0.347), ValueMatch(current_value='55.67', target_value='25.5', similarity=0.337), ValueMatch(current_value='42.98', target_value='39.8', similarity=0.336), ValueMatch(current_value='39.2', target_value='39.8', similarity=0.321), ValueMatch(current_value='34.37', target_value='33.7', similarity=0.319), ValueMatch(current_value='36.2', target_value='36.5', similarity=0.318), ValueMatch(current_value='28.5', target_value='28.9', similarity=0.314), ValueMatch(current_value='65.71', target_value='36.5', similarity=0.314), ValueMatch(current_value='68.39', target_value='39.8', similarity=0.312), ValueMatch(current_value='36.0', target_value='36.5', similarity=0.311), ValueMatch(current_value='71.09', target_value='10.1', similarity=0.31), ValueMatch(current_value='30.47', target_value='20.4', similarity=0.309), ValueMatch(current_value='26.0', target_value='12.6', similarity=0.308), ValueMatch(current_value='38.97', target_value='28.9', similarity=0.304), ValueMatch(current_value='20.28', target_value='20.4', similarity=0.303), ValueMatch(current_value='32.32', target_value='23.1', similarity=0.302), ValueMatch(current_value='17.64', target_value='17.8', similarity=0.301), ValueMatch(current_value='31.58', target_value='15.0', similarity=0.3), ValueMatch(current_value='26.22', target_value='12.6', similarity=0.292), ValueMatch(current_value='21.57', target_value='15.0', similarity=0.291), ValueMatch(current_value='26.14', target_value='12.6', similarity=0.288), ValueMatch(current_value='27.83', target_value='17.8', similarity=0.282), ValueMatch(current_value='28.84', target_value='28.9', similarity=0.282), ValueMatch(current_value='23.88', target_value='23.1', similarity=0.282), ValueMatch(current_value='32.06', target_value='20.4', similarity=0.28), ValueMatch(current_value='27.82', target_value='17.8', similarity=0.269), ValueMatch(current_value='47.82', target_value='17.8', similarity=0.266), ValueMatch(current_value='22.86', target_value='28.9', similarity=0.264), ValueMatch(current_value='39.76', target_value='39.8', similarity=0.259), ValueMatch(current_value='31.83', target_value='31.2', similarity=0.259), ValueMatch(current_value='36.84', target_value='36.5', similarity=0.256), ValueMatch(current_value='37.11', target_value='33.7', similarity=0.255), ValueMatch(current_value='29.37', target_value='33.7', similarity=0.255), ValueMatch(current_value='37.69', target_value='33.7', similarity=0.255), ValueMatch(current_value='59.78', target_value='17.8', similarity=0.253), ValueMatch(current_value='39.14', target_value='39.8', similarity=0.253), ValueMatch(current_value='34.26', target_value='12.6', similarity=0.252)]\n",
      "source: BMI target: longest_dimension score: 20.377\n",
      "value_matches:  [ValueMatch(current_value='20.55', target_value='2.05', similarity=0.8), ValueMatch(current_value='21.57', target_value='2.15', similarity=0.758), ValueMatch(current_value='17.11', target_value='1.7', similarity=0.57), ValueMatch(current_value='26.22', target_value='2.2', similarity=0.568), ValueMatch(current_value='51.19', target_value='1.9', similarity=0.509), ValueMatch(current_value='32.32', target_value='2.3', similarity=0.501), ValueMatch(current_value='20.28', target_value='2.0', similarity=0.497), ValueMatch(current_value='21.83', target_value='1.8', similarity=0.49), ValueMatch(current_value='31.83', target_value='1.8', similarity=0.48), ValueMatch(current_value='22.86', target_value='2.2', similarity=0.478), ValueMatch(current_value='31.22', target_value='2.2', similarity=0.478), ValueMatch(current_value='42.19', target_value='1.9', similarity=0.478), ValueMatch(current_value='31.96', target_value='1.9', similarity=0.471), ValueMatch(current_value='17.85', target_value='1.7', similarity=0.469), ValueMatch(current_value='31.63', target_value='1.6', similarity=0.465), ValueMatch(current_value='23.88', target_value='2.3', similarity=0.463), ValueMatch(current_value='17.64', target_value='1.7', similarity=0.461), ValueMatch(current_value='32.06', target_value='2.0', similarity=0.459), ValueMatch(current_value='21.45', target_value='2.1', similarity=0.439), ValueMatch(current_value='28.5', target_value='1.85', similarity=0.341), ValueMatch(current_value='65.71', target_value='1.65', similarity=0.328), ValueMatch(current_value='29.52', target_value='1.95', similarity=0.318), ValueMatch(current_value='31.58', target_value='2.15', similarity=0.315), ValueMatch(current_value='25.37', target_value='2.25', similarity=0.299), ValueMatch(current_value='25.68', target_value='2.25', similarity=0.295), ValueMatch(current_value='36.57', target_value='1.65', similarity=0.29), ValueMatch(current_value='25.03', target_value='2.25', similarity=0.283), ValueMatch(current_value='33.65', target_value='1.65', similarity=0.273), ValueMatch(current_value='46.85', target_value='1.85', similarity=0.264), ValueMatch(current_value='38.54', target_value='1.85', similarity=0.26), ValueMatch(current_value='37.11', target_value='1.7', similarity=0.255)]\n",
      "source: BMI target: a260_a280_ratio score: 13.354999999999995\n",
      "value_matches:  [ValueMatch(current_value='45.53', target_value='45.5', similarity=0.786), ValueMatch(current_value='29.37', target_value='29.3', similarity=0.773), ValueMatch(current_value='29.4', target_value='9.4', similarity=0.654), ValueMatch(current_value='29.52', target_value='5.2', similarity=0.523), ValueMatch(current_value='36.0', target_value='60.8', similarity=0.379), ValueMatch(current_value='46.0', target_value='60.8', similarity=0.376), ValueMatch(current_value='26.0', target_value='60.8', similarity=0.374), ValueMatch(current_value='55.86', target_value='45.5', similarity=0.361), ValueMatch(current_value='55.67', target_value='45.5', similarity=0.353), ValueMatch(current_value='27.1', target_value='27.7', similarity=0.35), ValueMatch(current_value='20.55', target_value='45.5', similarity=0.35), ValueMatch(current_value='41.44', target_value='34.1', similarity=0.348), ValueMatch(current_value='51.19', target_value='51.2', similarity=0.347), ValueMatch(current_value='38.41', target_value='34.1', similarity=0.343), ValueMatch(current_value='46.45', target_value='45.5', similarity=0.341), ValueMatch(current_value='28.5', target_value='18.5', similarity=0.339), ValueMatch(current_value='46.41', target_value='34.1', similarity=0.334), ValueMatch(current_value='30.85', target_value='60.8', similarity=0.33), ValueMatch(current_value='31.22', target_value='12.8', similarity=0.328), ValueMatch(current_value='39.2', target_value='43.9', similarity=0.328), ValueMatch(current_value='43.0', target_value='43.9', similarity=0.327), ValueMatch(current_value='27.0', target_value='27.7', similarity=0.323), ValueMatch(current_value='21.83', target_value='18.5', similarity=0.311), ValueMatch(current_value='31.83', target_value='18.5', similarity=0.309), ValueMatch(current_value='17.85', target_value='18.5', similarity=0.308), ValueMatch(current_value='32.32', target_value='23.4', similarity=0.308), ValueMatch(current_value='45.83', target_value='45.5', similarity=0.306), ValueMatch(current_value='34.53', target_value='45.5', similarity=0.305), ValueMatch(current_value='21.45', target_value='45.5', similarity=0.303), ValueMatch(current_value='27.82', target_value='27.7', similarity=0.301), ValueMatch(current_value='22.86', target_value='38.6', similarity=0.299), ValueMatch(current_value='39.14', target_value='43.9', similarity=0.298), ValueMatch(current_value='31.63', target_value='16.7', similarity=0.297), ValueMatch(current_value='34.37', target_value='43.9', similarity=0.291), ValueMatch(current_value='23.88', target_value='23.4', similarity=0.288), ValueMatch(current_value='28.84', target_value='12.8', similarity=0.282), ValueMatch(current_value='27.83', target_value='27.7', similarity=0.281), ValueMatch(current_value='29.62', target_value='29.3', similarity=0.28), ValueMatch(current_value='34.0', target_value='34.1', similarity=0.274), ValueMatch(current_value='68.39', target_value='43.9', similarity=0.273), ValueMatch(current_value='39.76', target_value='43.9', similarity=0.265), ValueMatch(current_value='27.31', target_value='27.7', similarity=0.264), ValueMatch(current_value='38.88', target_value='38.6', similarity=0.262), ValueMatch(current_value='38.89', target_value='38.6', similarity=0.262), ValueMatch(current_value='46.85', target_value='18.5', similarity=0.261), ValueMatch(current_value='38.54', target_value='18.5', similarity=0.259), ValueMatch(current_value='20.28', target_value='12.8', similarity=0.259), ValueMatch(current_value='34.84', target_value='34.1', similarity=0.256), ValueMatch(current_value='27.66', target_value='27.7', similarity=0.254)]\n",
      "source: BMI target: spindle_cell_percent score: 16.623\n",
      "value_matches:  [ValueMatch(current_value='29.52', target_value='95.2', similarity=0.817), ValueMatch(current_value='55.86', target_value='55.8', similarity=0.777), ValueMatch(current_value='34.84', target_value='4.8', similarity=0.535), ValueMatch(current_value='48.46', target_value='4.8', similarity=0.517), ValueMatch(current_value='44.81', target_value='4.8', similarity=0.487), ValueMatch(current_value='34.72', target_value='7.2', similarity=0.486), ValueMatch(current_value='40.72', target_value='7.2', similarity=0.464), ValueMatch(current_value='34.89', target_value='4.8', similarity=0.454), ValueMatch(current_value='30.48', target_value='4.8', similarity=0.434), ValueMatch(current_value='27.0', target_value='87.0', similarity=0.407), ValueMatch(current_value='29.4', target_value='69.4', similarity=0.405), ValueMatch(current_value='45.53', target_value='55.8', similarity=0.358), ValueMatch(current_value='31.22', target_value='12.5', similarity=0.343), ValueMatch(current_value='20.55', target_value='55.8', similarity=0.34), ValueMatch(current_value='55.67', target_value='55.8', similarity=0.338), ValueMatch(current_value='33.65', target_value='33.7', similarity=0.337), ValueMatch(current_value='35.42', target_value='25.4', similarity=0.331), ValueMatch(current_value='45.83', target_value='55.8', similarity=0.33), ValueMatch(current_value='38.89', target_value='18.9', similarity=0.329), ValueMatch(current_value='42.98', target_value='98.5', similarity=0.328), ValueMatch(current_value='27.82', target_value='82.1', similarity=0.327), ValueMatch(current_value='28.5', target_value='98.5', similarity=0.325), ValueMatch(current_value='34.37', target_value='33.7', similarity=0.323), ValueMatch(current_value='31.58', target_value='55.8', similarity=0.321), ValueMatch(current_value='31.63', target_value='76.3', similarity=0.319), ValueMatch(current_value='21.83', target_value='82.1', similarity=0.314), ValueMatch(current_value='38.54', target_value='25.4', similarity=0.307), ValueMatch(current_value='37.69', target_value='69.4', similarity=0.303), ValueMatch(current_value='32.69', target_value='69.4', similarity=0.303), ValueMatch(current_value='27.66', target_value='76.3', similarity=0.302), ValueMatch(current_value='31.83', target_value='18.9', similarity=0.301), ValueMatch(current_value='39.76', target_value='76.3', similarity=0.294), ValueMatch(current_value='31.96', target_value='49.6', similarity=0.293), ValueMatch(current_value='47.82', target_value='82.1', similarity=0.287), ValueMatch(current_value='29.62', target_value='49.6', similarity=0.285), ValueMatch(current_value='38.97', target_value='18.9', similarity=0.284), ValueMatch(current_value='17.64', target_value='76.3', similarity=0.264), ValueMatch(current_value='25.37', target_value='33.7', similarity=0.264), ValueMatch(current_value='21.45', target_value='82.1', similarity=0.263), ValueMatch(current_value='25.68', target_value='25.4', similarity=0.258), ValueMatch(current_value='42.19', target_value='82.1', similarity=0.258), ValueMatch(current_value='29.37', target_value='33.7', similarity=0.257), ValueMatch(current_value='37.11', target_value='33.7', similarity=0.255), ValueMatch(current_value='21.57', target_value='82.1', similarity=0.253), ValueMatch(current_value='46.85', target_value='98.5', similarity=0.252)]\n",
      "source: BMI target: percent_normal_cells score: 16.029\n",
      "value_matches:  [ValueMatch(current_value='27.1', target_value='7.1', similarity=0.602), ValueMatch(current_value='17.11', target_value='7.1', similarity=0.569), ValueMatch(current_value='20.55', target_value='0.5', similarity=0.566), ValueMatch(current_value='37.11', target_value='7.1', similarity=0.535), ValueMatch(current_value='42.98', target_value='9.8', similarity=0.518), ValueMatch(current_value='32.32', target_value='2.3', similarity=0.488), ValueMatch(current_value='34.72', target_value='4.7', similarity=0.475), ValueMatch(current_value='47.82', target_value='4.7', similarity=0.471), ValueMatch(current_value='30.47', target_value='4.7', similarity=0.465), ValueMatch(current_value='33.65', target_value='33.3', similarity=0.462), ValueMatch(current_value='65.71', target_value='7.1', similarity=0.453), ValueMatch(current_value='23.88', target_value='2.3', similarity=0.447), ValueMatch(current_value='71.09', target_value='7.1', similarity=0.429), ValueMatch(current_value='31.0', target_value='21.0', similarity=0.416), ValueMatch(current_value='29.4', target_value='29.9', similarity=0.346), ValueMatch(current_value='31.22', target_value='12.5', similarity=0.335), ValueMatch(current_value='26.14', target_value='36.1', similarity=0.331), ValueMatch(current_value='29.52', target_value='39.5', similarity=0.329), ValueMatch(current_value='21.57', target_value='15.2', similarity=0.328), ValueMatch(current_value='39.2', target_value='39.5', similarity=0.317), ValueMatch(current_value='21.83', target_value='18.6', similarity=0.305), ValueMatch(current_value='55.67', target_value='26.7', similarity=0.305), ValueMatch(current_value='31.83', target_value='18.6', similarity=0.302), ValueMatch(current_value='26.0', target_value='26.7', similarity=0.301), ValueMatch(current_value='29.62', target_value='29.9', similarity=0.3), ValueMatch(current_value='31.58', target_value='15.2', similarity=0.299), ValueMatch(current_value='36.2', target_value='36.1', similarity=0.293), ValueMatch(current_value='34.0', target_value='23.4', similarity=0.289), ValueMatch(current_value='36.0', target_value='36.1', similarity=0.286), ValueMatch(current_value='26.22', target_value='26.7', similarity=0.285), ValueMatch(current_value='22.86', target_value='18.6', similarity=0.282), ValueMatch(current_value='36.81', target_value='36.1', similarity=0.281), ValueMatch(current_value='29.37', target_value='29.9', similarity=0.276), ValueMatch(current_value='55.86', target_value='18.6', similarity=0.273), ValueMatch(current_value='34.84', target_value='23.4', similarity=0.27), ValueMatch(current_value='25.37', target_value='12.5', similarity=0.266), ValueMatch(current_value='29.07', target_value='29.9', similarity=0.265), ValueMatch(current_value='68.39', target_value='39.5', similarity=0.263), ValueMatch(current_value='25.68', target_value='12.5', similarity=0.262), ValueMatch(current_value='34.26', target_value='23.4', similarity=0.261), ValueMatch(current_value='39.76', target_value='39.5', similarity=0.256), ValueMatch(current_value='21.45', target_value='21.0', similarity=0.255), ValueMatch(current_value='25.03', target_value='12.5', similarity=0.252), ValueMatch(current_value='39.14', target_value='39.5', similarity=0.25), ValueMatch(current_value='42.19', target_value='21.0', similarity=0.25)]\n",
      "source: BMI target: necrosis_percent score: 15.809000000000001\n",
      "value_matches:  [ValueMatch(current_value='17.11', target_value='1.1', similarity=0.621), ValueMatch(current_value='26.22', target_value='2.2', similarity=0.573), ValueMatch(current_value='41.44', target_value='1.4', similarity=0.547), ValueMatch(current_value='31.22', target_value='1.2', similarity=0.542), ValueMatch(current_value='37.11', target_value='1.1', similarity=0.535), ValueMatch(current_value='24.32', target_value='2.4', similarity=0.531), ValueMatch(current_value='51.19', target_value='1.9', similarity=0.519), ValueMatch(current_value='20.28', target_value='2.0', similarity=0.506), ValueMatch(current_value='21.83', target_value='1.8', similarity=0.501), ValueMatch(current_value='32.32', target_value='2.3', similarity=0.5), ValueMatch(current_value='31.83', target_value='1.8', similarity=0.494), ValueMatch(current_value='31.58', target_value='1.5', similarity=0.492), ValueMatch(current_value='42.19', target_value='1.9', similarity=0.488), ValueMatch(current_value='22.86', target_value='2.2', similarity=0.487), ValueMatch(current_value='21.57', target_value='1.5', similarity=0.484), ValueMatch(current_value='31.96', target_value='1.9', similarity=0.482), ValueMatch(current_value='31.63', target_value='1.6', similarity=0.482), ValueMatch(current_value='17.85', target_value='1.7', similarity=0.476), ValueMatch(current_value='17.64', target_value='1.7', similarity=0.473), ValueMatch(current_value='32.06', target_value='2.0', similarity=0.471), ValueMatch(current_value='25.37', target_value='2.5', similarity=0.465), ValueMatch(current_value='23.88', target_value='2.3', similarity=0.461), ValueMatch(current_value='25.68', target_value='2.5', similarity=0.458), ValueMatch(current_value='21.45', target_value='1.4', similarity=0.457), ValueMatch(current_value='25.03', target_value='2.5', similarity=0.44), ValueMatch(current_value='26.14', target_value='1.4', similarity=0.44), ValueMatch(current_value='20.55', target_value='2.0', similarity=0.437), ValueMatch(current_value='39.14', target_value='1.4', similarity=0.434)]\n",
      "source: BMI target: ribosomal_rna_28s_16s_ratio score: 13.795999999999998\n",
      "value_matches:  [ValueMatch(current_value='28.5', target_value='32850', similarity=0.653), ValueMatch(current_value='36.57', target_value='6570', similarity=0.602), ValueMatch(current_value='65.71', target_value='6570', similarity=0.571), ValueMatch(current_value='42.19', target_value='21915', similarity=0.516), ValueMatch(current_value='32.83', target_value='32850', similarity=0.503), ValueMatch(current_value='25.03', target_value='18250', similarity=0.462), ValueMatch(current_value='33.65', target_value='36500', similarity=0.442), ValueMatch(current_value='20.55', target_value='25550', similarity=0.393), ValueMatch(current_value='26.14', target_value='14610', similarity=0.378), ValueMatch(current_value='29.4', target_value='9490', similarity=0.359), ValueMatch(current_value='21.57', target_value='21915', similarity=0.355), ValueMatch(current_value='45.53', target_value='25550', similarity=0.349), ValueMatch(current_value='27.0', target_value='6570', similarity=0.341), ValueMatch(current_value='55.86', target_value='25550', similarity=0.338), ValueMatch(current_value='25.37', target_value='23725', similarity=0.336), ValueMatch(current_value='55.67', target_value='25550', similarity=0.329), ValueMatch(current_value='35.0', target_value='8035', similarity=0.319), ValueMatch(current_value='39.2', target_value='29220', similarity=0.305), ValueMatch(current_value='31.0', target_value='14610', similarity=0.303), ValueMatch(current_value='51.19', target_value='21915', similarity=0.299), ValueMatch(current_value='26.22', target_value='29220', similarity=0.295), ValueMatch(current_value='36.0', target_value='36500', similarity=0.284), ValueMatch(current_value='29.07', target_value='9490', similarity=0.281), ValueMatch(current_value='27.1', target_value='12775', similarity=0.276), ValueMatch(current_value='32.32', target_value='32850', similarity=0.273), ValueMatch(current_value='20.28', target_value='29220', similarity=0.264), ValueMatch(current_value='39.14', target_value='21915', similarity=0.255), ValueMatch(current_value='29.62', target_value='29220', similarity=0.252)]\n",
      "source: BMI target: age_at_diagnosis score: 10.333000000000002\n",
      "value_matches:  [ValueMatch(current_value='25.03', target_value='25.0', similarity=0.737), ValueMatch(current_value='31.0', target_value='1.0', similarity=0.626), ValueMatch(current_value='35.0', target_value='5.0', similarity=0.577), ValueMatch(current_value='43.0', target_value='3.0', similarity=0.569), ValueMatch(current_value='20.55', target_value='0.5', similarity=0.544), ValueMatch(current_value='20.28', target_value='2.0', similarity=0.496), ValueMatch(current_value='31.58', target_value='1.5', similarity=0.476), ValueMatch(current_value='21.57', target_value='1.5', similarity=0.466), ValueMatch(current_value='71.09', target_value='1.0', similarity=0.452), ValueMatch(current_value='30.48', target_value='3.0', similarity=0.448), ValueMatch(current_value='32.06', target_value='2.0', similarity=0.446), ValueMatch(current_value='30.47', target_value='3.0', similarity=0.44), ValueMatch(current_value='25.37', target_value='2.5', similarity=0.439), ValueMatch(current_value='30.85', target_value='3.0', similarity=0.433), ValueMatch(current_value='25.68', target_value='2.5', similarity=0.432), ValueMatch(current_value='40.72', target_value='0.75', similarity=0.283), ValueMatch(current_value='29.07', target_value='0.75', similarity=0.278)]\n",
      "source: BMI target: concentration score: 8.142000000000001\n",
      "value_matches:  [ValueMatch(current_value='41.44', target_value='14.4', similarity=0.809), ValueMatch(current_value='35.0', target_value='5.0', similarity=0.615), ValueMatch(current_value='27.1', target_value='7.1', similarity=0.601), ValueMatch(current_value='28.5', target_value='8.5', similarity=0.587), ValueMatch(current_value='17.11', target_value='7.1', similarity=0.571), ValueMatch(current_value='20.55', target_value='0.5', similarity=0.568), ValueMatch(current_value='37.11', target_value='7.1', similarity=0.537), ValueMatch(current_value='24.32', target_value='2.4', similarity=0.534), ValueMatch(current_value='31.63', target_value='6.3', similarity=0.524), ValueMatch(current_value='31.22', target_value='1.2', similarity=0.518), ValueMatch(current_value='23.88', target_value='3.8', similarity=0.494), ValueMatch(current_value='38.88', target_value='3.8', similarity=0.492), ValueMatch(current_value='38.89', target_value='3.8', similarity=0.492), ValueMatch(current_value='25.03', target_value='5.0', similarity=0.48), ValueMatch(current_value='65.71', target_value='7.1', similarity=0.453), ValueMatch(current_value='46.85', target_value='8.5', similarity=0.45), ValueMatch(current_value='38.54', target_value='8.5', similarity=0.446), ValueMatch(current_value='17.85', target_value='8.5', similarity=0.439), ValueMatch(current_value='30.85', target_value='8.5', similarity=0.426), ValueMatch(current_value='38.41', target_value='3.8', similarity=0.425), ValueMatch(current_value='71.09', target_value='7.1', similarity=0.425), ValueMatch(current_value='38.44', target_value='3.8', similarity=0.417), ValueMatch(current_value='44.0', target_value='14.4', similarity=0.411), ValueMatch(current_value='38.97', target_value='3.8', similarity=0.403), ValueMatch(current_value='44.81', target_value='14.4', similarity=0.374), ValueMatch(current_value='36.0', target_value='16.0', similarity=0.372), ValueMatch(current_value='46.0', target_value='16.0', similarity=0.369), ValueMatch(current_value='26.0', target_value='16.0', similarity=0.368), ValueMatch(current_value='40.72', target_value='20.7', similarity=0.344), ValueMatch(current_value='29.07', target_value='20.7', similarity=0.339), ValueMatch(current_value='51.19', target_value='11.2', similarity=0.333), ValueMatch(current_value='21.83', target_value='18.5', similarity=0.31), ValueMatch(current_value='31.83', target_value='18.5', similarity=0.308), ValueMatch(current_value='21.45', target_value='14.4', similarity=0.307), ValueMatch(current_value='20.28', target_value='20.7', similarity=0.301), ValueMatch(current_value='26.14', target_value='14.4', similarity=0.297), ValueMatch(current_value='39.14', target_value='14.4', similarity=0.293), ValueMatch(current_value='28.84', target_value='12.8', similarity=0.287), ValueMatch(current_value='32.06', target_value='20.7', similarity=0.276), ValueMatch(current_value='22.86', target_value='12.8', similarity=0.27), ValueMatch(current_value='55.86', target_value='8.5', similarity=0.261)]\n",
      "source: BMI target: tumor_burden score: 17.525999999999996\n",
      "value_matches:  [ValueMatch(current_value='36.0', target_value='6.0', similarity=0.631), ValueMatch(current_value='46.0', target_value='6.0', similarity=0.627), ValueMatch(current_value='26.0', target_value='6.0', similarity=0.623), ValueMatch(current_value='28.5', target_value='8.5', similarity=0.592), ValueMatch(current_value='43.0', target_value='4.3', similarity=0.571), ValueMatch(current_value='31.22', target_value='1.2', similarity=0.546), ValueMatch(current_value='34.84', target_value='4.8', similarity=0.537), ValueMatch(current_value='48.46', target_value='4.8', similarity=0.519), ValueMatch(current_value='33.65', target_value='3.3', similarity=0.517), ValueMatch(current_value='71.09', target_value='0.9', similarity=0.515), ValueMatch(current_value='34.37', target_value='4.3', similarity=0.508), ValueMatch(current_value='21.83', target_value='1.8', similarity=0.505), ValueMatch(current_value='31.83', target_value='1.8', similarity=0.499), ValueMatch(current_value='38.54', target_value='5.4', similarity=0.497), ValueMatch(current_value='38.97', target_value='9.7', similarity=0.492), ValueMatch(current_value='44.81', target_value='4.8', similarity=0.489), ValueMatch(current_value='39.76', target_value='9.7', similarity=0.488), ValueMatch(current_value='34.72', target_value='7.2', similarity=0.487), ValueMatch(current_value='35.42', target_value='5.4', similarity=0.476), ValueMatch(current_value='59.78', target_value='9.7', similarity=0.471), ValueMatch(current_value='25.37', target_value='2.5', similarity=0.468), ValueMatch(current_value='40.72', target_value='7.2', similarity=0.466), ValueMatch(current_value='46.85', target_value='8.5', similarity=0.46), ValueMatch(current_value='25.68', target_value='2.5', similarity=0.459), ValueMatch(current_value='21.45', target_value='2.1', similarity=0.456), ValueMatch(current_value='34.89', target_value='4.8', similarity=0.451), ValueMatch(current_value='17.85', target_value='8.5', similarity=0.448), ValueMatch(current_value='42.19', target_value='2.1', similarity=0.447), ValueMatch(current_value='30.48', target_value='4.8', similarity=0.443), ValueMatch(current_value='25.03', target_value='2.5', similarity=0.441), ValueMatch(current_value='21.57', target_value='2.1', similarity=0.439), ValueMatch(current_value='30.85', target_value='8.5', similarity=0.435), ValueMatch(current_value='24.32', target_value='4.3', similarity=0.427), ValueMatch(current_value='29.37', target_value='3.7', similarity=0.426), ValueMatch(current_value='37.69', target_value='3.7', similarity=0.426), ValueMatch(current_value='37.11', target_value='3.7', similarity=0.424), ValueMatch(current_value='31.0', target_value='10.4', similarity=0.405), ValueMatch(current_value='30.47', target_value='10.4', similarity=0.3), ValueMatch(current_value='55.86', target_value='8.5', similarity=0.261), ValueMatch(current_value='17.11', target_value='2.1', similarity=0.257)]\n",
      "source: BMI target: tumor_width_measurement score: 18.929\n",
      "value_matches:  [ValueMatch(current_value='27.83', target_value='27830973', similarity=0.595), ValueMatch(current_value='29.52', target_value='29567829', similarity=0.468), ValueMatch(current_value='46.41', target_value='26786412', similarity=0.408), ValueMatch(current_value='39.76', target_value='33289765', similarity=0.405), ValueMatch(current_value='38.97', target_value='33289765', similarity=0.401), ValueMatch(current_value='46.85', target_value='25468954', similarity=0.393), ValueMatch(current_value='27.82', target_value='29567829', similarity=0.389), ValueMatch(current_value='47.82', target_value='28954789', similarity=0.385), ValueMatch(current_value='32.83', target_value='33289765', similarity=0.384), ValueMatch(current_value='55.67', target_value='35678901', similarity=0.381), ValueMatch(current_value='34.53', target_value='39012345', similarity=0.372), ValueMatch(current_value='32.06', target_value='32098765', similarity=0.365), ValueMatch(current_value='42.98', target_value='31298756', similarity=0.36), ValueMatch(current_value='31.22', target_value='31298756', similarity=0.336), ValueMatch(current_value='33.65', target_value='33289765', similarity=0.312), ValueMatch(current_value='30.48', target_value='30541042', similarity=0.284), ValueMatch(current_value='30.47', target_value='30541042', similarity=0.281), ValueMatch(current_value='27.31', target_value='27830973', similarity=0.281), ValueMatch(current_value='29.4', target_value='29567829', similarity=0.271), ValueMatch(current_value='38.54', target_value='25468954', similarity=0.27), ValueMatch(current_value='25.68', target_value='25468954', similarity=0.266), ValueMatch(current_value='23.88', target_value='38901234', similarity=0.265), ValueMatch(current_value='29.62', target_value='29567829', similarity=0.261), ValueMatch(current_value='35.42', target_value='25468954', similarity=0.26), ValueMatch(current_value='38.89', target_value='28954789', similarity=0.253)]\n",
      "source: BMI target: pmid score: 8.645999999999999\n",
      "value_matches:  [ValueMatch(current_value='38.89', target_value='88.9', similarity=0.831), ValueMatch(current_value='25.68', target_value='25.6', similarity=0.779), ValueMatch(current_value='40.72', target_value='40.7', similarity=0.763), ValueMatch(current_value='20.55', target_value='205.1', similarity=0.608), ValueMatch(current_value='38.88', target_value='88.9', similarity=0.511), ValueMatch(current_value='44.0', target_value='40.7', similarity=0.384), ValueMatch(current_value='45.53', target_value='55.4', similarity=0.378), ValueMatch(current_value='34.0', target_value='40.7', similarity=0.376), ValueMatch(current_value='36.0', target_value='60.8', similarity=0.366), ValueMatch(current_value='46.0', target_value='60.8', similarity=0.363), ValueMatch(current_value='26.0', target_value='60.8', similarity=0.361), ValueMatch(current_value='43.0', target_value='30.0', similarity=0.359), ValueMatch(current_value='29.07', target_value='90.0', similarity=0.351), ValueMatch(current_value='31.0', target_value='110.4', similarity=0.351), ValueMatch(current_value='34.89', target_value='88.9', similarity=0.349), ValueMatch(current_value='23.88', target_value='88.9', similarity=0.345), ValueMatch(current_value='28.84', target_value='88.9', similarity=0.341), ValueMatch(current_value='29.52', target_value='45.2', similarity=0.336), ValueMatch(current_value='17.11', target_value='110.4', similarity=0.336), ValueMatch(current_value='38.97', target_value='88.9', similarity=0.335), ValueMatch(current_value='55.67', target_value='25.6', similarity=0.333), ValueMatch(current_value='38.54', target_value='55.4', similarity=0.33), ValueMatch(current_value='55.86', target_value='55.4', similarity=0.325), ValueMatch(current_value='30.85', target_value='60.8', similarity=0.323), ValueMatch(current_value='35.42', target_value='55.4', similarity=0.318), ValueMatch(current_value='25.37', target_value='75.3', similarity=0.313), ValueMatch(current_value='35.0', target_value='150.5', similarity=0.311), ValueMatch(current_value='34.53', target_value='75.3', similarity=0.292), ValueMatch(current_value='37.11', target_value='110.4', similarity=0.289), ValueMatch(current_value='46.45', target_value='45.2', similarity=0.287), ValueMatch(current_value='21.45', target_value='45.2', similarity=0.285), ValueMatch(current_value='30.48', target_value='30.0', similarity=0.285), ValueMatch(current_value='36.57', target_value='95.7', similarity=0.284), ValueMatch(current_value='30.47', target_value='30.0', similarity=0.28), ValueMatch(current_value='34.06', target_value='40.7', similarity=0.277), ValueMatch(current_value='51.19', target_value='205.1', similarity=0.277), ValueMatch(current_value='65.71', target_value='95.7', similarity=0.275), ValueMatch(current_value='21.57', target_value='95.7', similarity=0.27), ValueMatch(current_value='25.03', target_value='25.6', similarity=0.267), ValueMatch(current_value='31.58', target_value='150.5', similarity=0.263), ValueMatch(current_value='20.28', target_value='130.2', similarity=0.263), ValueMatch(current_value='17.85', target_value='175.0', similarity=0.26), ValueMatch(current_value='71.09', target_value='110.4', similarity=0.257)]\n",
      "source: BMI target: gross_tumor_weight score: 15.486999999999998\n",
      "value_matches:  [ValueMatch(current_value='20.28', target_value='0.028', similarity=0.65), ValueMatch(current_value='35.0', target_value='0.035', similarity=0.384), ValueMatch(current_value='20.55', target_value='0.005', similarity=0.321), ValueMatch(current_value='25.03', target_value='0.03', similarity=0.316), ValueMatch(current_value='40.72', target_value='0.007', similarity=0.282), ValueMatch(current_value='29.07', target_value='0.007', similarity=0.278), ValueMatch(current_value='31.22', target_value='0.012', similarity=0.276), ValueMatch(current_value='33.65', target_value='0.033', similarity=0.27), ValueMatch(current_value='28.5', target_value='0.028', similarity=0.266), ValueMatch(current_value='21.83', target_value='0.018', similarity=0.25)]\n",
      "source: BMI target: proportion_base_mismatch score: 3.2929999999999997\n",
      "value_matches:  [ValueMatch(current_value='23.88', target_value='23.8', similarity=0.802), ValueMatch(current_value='34.53', target_value='45.3', similarity=0.798), ValueMatch(current_value='35.0', target_value='3.5', similarity=0.575), ValueMatch(current_value='45.53', target_value='45.3', similarity=0.531), ValueMatch(current_value='55.67', target_value='5.6', similarity=0.516), ValueMatch(current_value='39.14', target_value='9.1', similarity=0.515), ValueMatch(current_value='34.72', target_value='7.2', similarity=0.491), ValueMatch(current_value='25.68', target_value='5.6', similarity=0.475), ValueMatch(current_value='35.42', target_value='3.5', similarity=0.47), ValueMatch(current_value='40.72', target_value='7.2', similarity=0.463), ValueMatch(current_value='27.0', target_value='70.4', similarity=0.423), ValueMatch(current_value='27.1', target_value='47.1', similarity=0.354), ValueMatch(current_value='36.2', target_value='62.8', similarity=0.354), ValueMatch(current_value='30.47', target_value='70.4', similarity=0.352), ValueMatch(current_value='31.22', target_value='12.5', similarity=0.338), ValueMatch(current_value='38.89', target_value='18.9', similarity=0.336), ValueMatch(current_value='17.11', target_value='47.1', similarity=0.336), ValueMatch(current_value='26.22', target_value='62.8', similarity=0.318), ValueMatch(current_value='37.11', target_value='47.1', similarity=0.316), ValueMatch(current_value='28.5', target_value='62.8', similarity=0.314), ValueMatch(current_value='22.86', target_value='62.8', similarity=0.314), ValueMatch(current_value='32.32', target_value='23.8', similarity=0.314), ValueMatch(current_value='46.45', target_value='45.3', similarity=0.308), ValueMatch(current_value='30.48', target_value='70.4', similarity=0.304), ValueMatch(current_value='34.89', target_value='18.9', similarity=0.303), ValueMatch(current_value='31.58', target_value='15.4', similarity=0.302), ValueMatch(current_value='21.83', target_value='18.9', similarity=0.302), ValueMatch(current_value='45.83', target_value='45.3', similarity=0.301), ValueMatch(current_value='31.83', target_value='18.9', similarity=0.301), ValueMatch(current_value='21.57', target_value='15.4', similarity=0.296), ValueMatch(current_value='38.54', target_value='15.4', similarity=0.295), ValueMatch(current_value='29.62', target_value='62.8', similarity=0.293), ValueMatch(current_value='38.97', target_value='18.9', similarity=0.29), ValueMatch(current_value='25.37', target_value='45.3', similarity=0.286), ValueMatch(current_value='47.82', target_value='47.1', similarity=0.283), ValueMatch(current_value='28.84', target_value='62.8', similarity=0.282), ValueMatch(current_value='38.88', target_value='23.8', similarity=0.273), ValueMatch(current_value='65.71', target_value='47.1', similarity=0.267), ValueMatch(current_value='21.45', target_value='45.3', similarity=0.266), ValueMatch(current_value='34.0', target_value='34.9', similarity=0.263), ValueMatch(current_value='20.28', target_value='62.8', similarity=0.258), ValueMatch(current_value='25.03', target_value='12.5', similarity=0.25), ValueMatch(current_value='71.09', target_value='47.1', similarity=0.25)]\n",
      "source: BMI target: percent_inflam_infiltration score: 15.677999999999997\n",
      "Top k reranked columns:  [('BMI', 'average_base_quality', 23.353), ('BMI', 'longest_dimension', 20.377), ('BMI', 'recist_targeted_regions_sum', 20.003000000000004), ('BMI', 'intermediate_dimension', 19.926), ('BMI', 'body_surface_area', 19.580999999999996), ('BMI', 'percent_stromal_cells', 19.118), ('BMI', 'bmi', 19.058), ('BMI', 'tumor_width_measurement', 18.929), ('BMI', 'tumor_burden', 17.525999999999996), ('BMI', 'spindle_cell_percent', 16.623), ('BMI', 'percent_normal_cells', 16.029), ('BMI', 'necrosis_percent', 15.809000000000001), ('BMI', 'percent_inflam_infiltration', 15.677999999999997), ('BMI', 'gross_tumor_weight', 15.486999999999998), ('BMI', 'ribosomal_rna_28s_16s_ratio', 13.795999999999998), ('BMI', 'a260_a280_ratio', 13.354999999999995), ('BMI', 'age_at_diagnosis', 10.333000000000002), ('BMI', 'pmid', 8.645999999999999), ('BMI', 'concentration', 8.142000000000001), ('BMI', 'proportion_base_mismatch', 3.2929999999999997)]\n",
      "\n",
      "Source column:  Age\n",
      "value_matches:  [ValueMatch(current_value='55.0', target_value='45.5', similarity=0.488), ValueMatch(current_value='45.0', target_value='45.5', similarity=0.467), ValueMatch(current_value='58.0', target_value='58.6', similarity=0.425), ValueMatch(current_value='78.0', target_value='67.8', similarity=0.425), ValueMatch(current_value='86.0', target_value='58.6', similarity=0.42), ValueMatch(current_value='63.0', target_value='63.7', similarity=0.42), ValueMatch(current_value='76.0', target_value='76.8', similarity=0.42), ValueMatch(current_value='73.0', target_value='73.2', similarity=0.415), ValueMatch(current_value='67.0', target_value='67.8', similarity=0.409), ValueMatch(current_value='90.0', target_value='89.0', similarity=0.394), ValueMatch(current_value='68.0', target_value='76.8', similarity=0.394), ValueMatch(current_value='69.0', target_value='89.0', similarity=0.349), ValueMatch(current_value='59.0', target_value='89.0', similarity=0.347), ValueMatch(current_value='61.0', target_value='110.5', similarity=0.341), ValueMatch(current_value='71.0', target_value='110.5', similarity=0.34), ValueMatch(current_value='41.0', target_value='110.5', similarity=0.336), ValueMatch(current_value='77.0', target_value='70.1', similarity=0.325), ValueMatch(current_value='70.0', target_value='70.1', similarity=0.319), ValueMatch(current_value='57.0', target_value='70.1', similarity=0.273)]\n",
      "source: Age target: weight score: 7.3069999999999995\n",
      "value_matches:  [ValueMatch(current_value='55.0', target_value='25550', similarity=0.79), ValueMatch(current_value='57.0', target_value='6570', similarity=0.764), ValueMatch(current_value='73.0', target_value='7305', similarity=0.755), ValueMatch(current_value='61.0', target_value='14610', similarity=0.679), ValueMatch(current_value='65.0', target_value='36500', similarity=0.632), ValueMatch(current_value='85.0', target_value='32850', similarity=0.597), ValueMatch(current_value='50.0', target_value='36500', similarity=0.584), ValueMatch(current_value='90.0', target_value='9490', similarity=0.378), ValueMatch(current_value='41.0', target_value='14610', similarity=0.347), ValueMatch(current_value='77.0', target_value='12775', similarity=0.343), ValueMatch(current_value='72.0', target_value='23725', similarity=0.339), ValueMatch(current_value='59.0', target_value='9490', similarity=0.335), ValueMatch(current_value='69.0', target_value='9490', similarity=0.333), ValueMatch(current_value='75.0', target_value='12775', similarity=0.319), ValueMatch(current_value='42.0', target_value='16425', similarity=0.302), ValueMatch(current_value='38.0', target_value='8035', similarity=0.301), ValueMatch(current_value='67.0', target_value='6570', similarity=0.293), ValueMatch(current_value='53.0', target_value='7305', similarity=0.292), ValueMatch(current_value='46.0', target_value='14610', similarity=0.291), ValueMatch(current_value='62.0', target_value='29220', similarity=0.289), ValueMatch(current_value='71.0', target_value='14610', similarity=0.287), ValueMatch(current_value='64.0', target_value='16425', similarity=0.28), ValueMatch(current_value='70.0', target_value='6570', similarity=0.275), ValueMatch(current_value='58.0', target_value='8035', similarity=0.266), ValueMatch(current_value='60.0', target_value='36500', similarity=0.251)]\n",
      "source: Age target: age_at_diagnosis score: 10.322\n",
      "value_matches:  [ValueMatch(current_value='55.0', target_value='55', similarity=0.726), ValueMatch(current_value='48.0', target_value='48', similarity=0.693), ValueMatch(current_value='42.0', target_value='42', similarity=0.689), ValueMatch(current_value='45.0', target_value='45', similarity=0.687), ValueMatch(current_value='58.0', target_value='58', similarity=0.679), ValueMatch(current_value='62.0', target_value='62', similarity=0.678), ValueMatch(current_value='53.0', target_value='53', similarity=0.676), ValueMatch(current_value='65.0', target_value='65', similarity=0.675), ValueMatch(current_value='77.0', target_value='70', similarity=0.653), ValueMatch(current_value='70.0', target_value='70', similarity=0.633), ValueMatch(current_value='50.0', target_value='50', similarity=0.603), ValueMatch(current_value='57.0', target_value='70', similarity=0.546), ValueMatch(current_value='67.0', target_value='70', similarity=0.544), ValueMatch(current_value='75.0', target_value='50', similarity=0.503), ValueMatch(current_value='85.0', target_value='50', similarity=0.497), ValueMatch(current_value='72.0', target_value='27', similarity=0.303), ValueMatch(current_value='44.0', target_value='45', similarity=0.283)]\n",
      "source: Age target: relationship_age_at_diagnosis score: 10.068000000000001\n",
      "value_matches:  [ValueMatch(current_value='55.0', target_value='55', similarity=0.727), ValueMatch(current_value='45.0', target_value='45', similarity=0.689), ValueMatch(current_value='75.0', target_value='75', similarity=0.683), ValueMatch(current_value='65.0', target_value='65', similarity=0.677), ValueMatch(current_value='44.0', target_value='40', similarity=0.671), ValueMatch(current_value='77.0', target_value='70', similarity=0.652), ValueMatch(current_value='70.0', target_value='70', similarity=0.627), ValueMatch(current_value='66.0', target_value='60', similarity=0.622), ValueMatch(current_value='60.0', target_value='60', similarity=0.601), ValueMatch(current_value='50.0', target_value='50', similarity=0.596), ValueMatch(current_value='53.0', target_value='30', similarity=0.574), ValueMatch(current_value='63.0', target_value='30', similarity=0.571), ValueMatch(current_value='73.0', target_value='30', similarity=0.567), ValueMatch(current_value='64.0', target_value='40', similarity=0.562), ValueMatch(current_value='83.0', target_value='30', similarity=0.562), ValueMatch(current_value='74.0', target_value='40', similarity=0.558), ValueMatch(current_value='58.0', target_value='80', similarity=0.546), ValueMatch(current_value='68.0', target_value='80', similarity=0.544), ValueMatch(current_value='57.0', target_value='70', similarity=0.543), ValueMatch(current_value='67.0', target_value='70', similarity=0.541), ValueMatch(current_value='78.0', target_value='80', similarity=0.54), ValueMatch(current_value='48.0', target_value='80', similarity=0.536), ValueMatch(current_value='38.0', target_value='80', similarity=0.527), ValueMatch(current_value='56.0', target_value='60', similarity=0.512), ValueMatch(current_value='76.0', target_value='60', similarity=0.506), ValueMatch(current_value='46.0', target_value='60', similarity=0.502), ValueMatch(current_value='86.0', target_value='60', similarity=0.5), ValueMatch(current_value='85.0', target_value='50', similarity=0.494), ValueMatch(current_value='42.0', target_value='24', similarity=0.339)]\n",
      "source: Age target: age_at_onset score: 16.569\n",
      "value_matches:  [ValueMatch(current_value='55.0', target_value='55', similarity=0.728), ValueMatch(current_value='45.0', target_value='45', similarity=0.689), ValueMatch(current_value='85.0', target_value='85', similarity=0.689), ValueMatch(current_value='75.0', target_value='75', similarity=0.684), ValueMatch(current_value='90.0', target_value='90', similarity=0.682), ValueMatch(current_value='65.0', target_value='65', similarity=0.677), ValueMatch(current_value='44.0', target_value='40', similarity=0.671), ValueMatch(current_value='77.0', target_value='70', similarity=0.651), ValueMatch(current_value='70.0', target_value='70', similarity=0.626), ValueMatch(current_value='66.0', target_value='60', similarity=0.622), ValueMatch(current_value='59.0', target_value='90', similarity=0.609), ValueMatch(current_value='69.0', target_value='90', similarity=0.607), ValueMatch(current_value='60.0', target_value='60', similarity=0.599), ValueMatch(current_value='50.0', target_value='50', similarity=0.594), ValueMatch(current_value='53.0', target_value='30', similarity=0.573), ValueMatch(current_value='63.0', target_value='30', similarity=0.571), ValueMatch(current_value='73.0', target_value='30', similarity=0.567), ValueMatch(current_value='83.0', target_value='30', similarity=0.563), ValueMatch(current_value='64.0', target_value='40', similarity=0.561), ValueMatch(current_value='74.0', target_value='40', similarity=0.557), ValueMatch(current_value='57.0', target_value='70', similarity=0.543), ValueMatch(current_value='58.0', target_value='80', similarity=0.54), ValueMatch(current_value='67.0', target_value='70', similarity=0.54), ValueMatch(current_value='68.0', target_value='80', similarity=0.538), ValueMatch(current_value='78.0', target_value='80', similarity=0.534), ValueMatch(current_value='48.0', target_value='80', similarity=0.531), ValueMatch(current_value='38.0', target_value='80', similarity=0.521), ValueMatch(current_value='56.0', target_value='60', similarity=0.511), ValueMatch(current_value='76.0', target_value='60', similarity=0.505), ValueMatch(current_value='46.0', target_value='60', similarity=0.502), ValueMatch(current_value='86.0', target_value='60', similarity=0.502), ValueMatch(current_value='42.0', target_value='24', similarity=0.354)]\n",
      "source: Age target: age_at_last_exposure score: 18.641\n",
      "value_matches:  [ValueMatch(current_value='70.0', target_value='70.0', similarity=1.0), ValueMatch(current_value='58.0', target_value='55.8', similarity=0.467), ValueMatch(current_value='55.0', target_value='55.8', similarity=0.466), ValueMatch(current_value='59.0', target_value='65.9', similarity=0.448), ValueMatch(current_value='53.0', target_value='75.3', similarity=0.432), ValueMatch(current_value='75.0', target_value='75.3', similarity=0.409), ValueMatch(current_value='77.0', target_value='70.0', similarity=0.408), ValueMatch(current_value='85.0', target_value='85.1', similarity=0.403), ValueMatch(current_value='45.0', target_value='45.2', similarity=0.399), ValueMatch(current_value='65.0', target_value='65.9', similarity=0.395), ValueMatch(current_value='90.0', target_value='90.4', similarity=0.393), ValueMatch(current_value='68.0', target_value='80.6', similarity=0.375), ValueMatch(current_value='57.0', target_value='35.7', similarity=0.369), ValueMatch(current_value='44.0', target_value='40.3', similarity=0.362), ValueMatch(current_value='67.0', target_value='70.0', similarity=0.348), ValueMatch(current_value='69.0', target_value='90.4', similarity=0.345), ValueMatch(current_value='50.0', target_value='70.0', similarity=0.342), ValueMatch(current_value='60.0', target_value='70.0', similarity=0.34), ValueMatch(current_value='63.0', target_value='30.5', similarity=0.316), ValueMatch(current_value='73.0', target_value='30.5', similarity=0.314), ValueMatch(current_value='83.0', target_value='30.5', similarity=0.31), ValueMatch(current_value='64.0', target_value='40.3', similarity=0.305), ValueMatch(current_value='74.0', target_value='40.3', similarity=0.303), ValueMatch(current_value='78.0', target_value='80.6', similarity=0.299), ValueMatch(current_value='48.0', target_value='80.6', similarity=0.296), ValueMatch(current_value='38.0', target_value='80.6', similarity=0.293), ValueMatch(current_value='66.0', target_value='60.1', similarity=0.292)]\n",
      "source: Age target: initial_weight score: 10.428999999999998\n",
      "value_matches:  [ValueMatch(current_value='59.0', target_value='59', similarity=0.689), ValueMatch(current_value='41.0', target_value='41', similarity=0.688), ValueMatch(current_value='85.0', target_value='85', similarity=0.685), ValueMatch(current_value='78.0', target_value='78', similarity=0.683), ValueMatch(current_value='53.0', target_value='53', similarity=0.682), ValueMatch(current_value='72.0', target_value='72', similarity=0.681), ValueMatch(current_value='65.0', target_value='65', similarity=0.675), ValueMatch(current_value='55.0', target_value='5', similarity=0.528), ValueMatch(current_value='61.0', target_value='1', similarity=0.361), ValueMatch(current_value='71.0', target_value='1', similarity=0.36), ValueMatch(current_value='50.0', target_value='5', similarity=0.302), ValueMatch(current_value='56.0', target_value='5', similarity=0.284), ValueMatch(current_value='75.0', target_value='5', similarity=0.283), ValueMatch(current_value='45.0', target_value='5', similarity=0.281), ValueMatch(current_value='58.0', target_value='5', similarity=0.279), ValueMatch(current_value='57.0', target_value='5', similarity=0.278), ValueMatch(current_value='74.0', target_value='47', similarity=0.264), ValueMatch(current_value='44.0', target_value='41', similarity=0.262)]\n",
      "source: Age target: age_at_index score: 8.264999999999999\n",
      "value_matches:  [ValueMatch(current_value='73.0', target_value='-7300', similarity=0.81), ValueMatch(current_value='57.0', target_value='-6570', similarity=0.776), ValueMatch(current_value='38.0', target_value='-4380', similarity=0.739), ValueMatch(current_value='41.0', target_value='-12410', similarity=0.662), ValueMatch(current_value='46.0', target_value='-14600', similarity=0.658), ValueMatch(current_value='60.0', target_value='-14600', similarity=0.585), ValueMatch(current_value='68.0', target_value='-11680', similarity=0.578), ValueMatch(current_value='55.0', target_value='-9855', similarity=0.388), ValueMatch(current_value='75.0', target_value='-6575', similarity=0.375), ValueMatch(current_value='65.0', target_value='-6575', similarity=0.367), ValueMatch(current_value='85.0', target_value='-9855', similarity=0.353), ValueMatch(current_value='83.0', target_value='-8030', similarity=0.331), ValueMatch(current_value='90.0', target_value='-2190', similarity=0.302), ValueMatch(current_value='67.0', target_value='-6570', similarity=0.3), ValueMatch(current_value='70.0', target_value='-7300', similarity=0.298), ValueMatch(current_value='61.0', target_value='-5110', similarity=0.295), ValueMatch(current_value='71.0', target_value='-5110', similarity=0.293), ValueMatch(current_value='62.0', target_value='-2920', similarity=0.293), ValueMatch(current_value='72.0', target_value='-2920', similarity=0.291), ValueMatch(current_value='42.0', target_value='-2920', similarity=0.288), ValueMatch(current_value='77.0', target_value='-6570', similarity=0.286), ValueMatch(current_value='48.0', target_value='-4380', similarity=0.28), ValueMatch(current_value='63.0', target_value='-7300', similarity=0.268), ValueMatch(current_value='53.0', target_value='-7300', similarity=0.267), ValueMatch(current_value='69.0', target_value='-2190', similarity=0.263), ValueMatch(current_value='59.0', target_value='-2190', similarity=0.263)]\n",
      "source: Age target: days_to_birth score: 10.608999999999998\n",
      "value_matches:  [ValueMatch(current_value='66.0', target_value='66.7', similarity=0.476), ValueMatch(current_value='78.0', target_value='78.8', similarity=0.458), ValueMatch(current_value='77.0', target_value='77.9', similarity=0.457), ValueMatch(current_value='67.0', target_value='66.7', similarity=0.442), ValueMatch(current_value='83.0', target_value='83.7', similarity=0.439), ValueMatch(current_value='72.0', target_value='72.5', similarity=0.43), ValueMatch(current_value='69.0', target_value='69.5', similarity=0.423), ValueMatch(current_value='65.0', target_value='65.4', similarity=0.399), ValueMatch(current_value='68.0', target_value='68.9', similarity=0.395), ValueMatch(current_value='74.0', target_value='74.3', similarity=0.395), ValueMatch(current_value='75.0', target_value='75.2', similarity=0.39), ValueMatch(current_value='70.0', target_value='70.3', similarity=0.314), ValueMatch(current_value='58.0', target_value='80.1', similarity=0.267), ValueMatch(current_value='57.0', target_value='70.3', similarity=0.266), ValueMatch(current_value='48.0', target_value='80.1', similarity=0.264), ValueMatch(current_value='38.0', target_value='80.1', similarity=0.26)]\n",
      "source: Age target: fev1_fvc_pre_bronch_percent score: 6.075\n",
      "value_matches:  [ValueMatch(current_value='85.0', target_value='85.0', similarity=1.0), ValueMatch(current_value='77.0', target_value='77.4', similarity=0.477), ValueMatch(current_value='74.0', target_value='77.4', similarity=0.459), ValueMatch(current_value='83.0', target_value='83.6', similarity=0.437), ValueMatch(current_value='72.0', target_value='72.6', similarity=0.431), ValueMatch(current_value='69.0', target_value='69.7', similarity=0.43), ValueMatch(current_value='62.0', target_value='62.8', similarity=0.422), ValueMatch(current_value='67.0', target_value='67.5', similarity=0.408), ValueMatch(current_value='75.0', target_value='67.5', similarity=0.403), ValueMatch(current_value='64.0', target_value='64.3', similarity=0.393), ValueMatch(current_value='68.0', target_value='68.9', similarity=0.388), ValueMatch(current_value='90.0', target_value='90.3', similarity=0.379), ValueMatch(current_value='55.0', target_value='85.0', similarity=0.344), ValueMatch(current_value='50.0', target_value='85.0', similarity=0.333), ValueMatch(current_value='59.0', target_value='90.3', similarity=0.332), ValueMatch(current_value='70.0', target_value='70.1', similarity=0.309), ValueMatch(current_value='65.0', target_value='85.0', similarity=0.284), ValueMatch(current_value='45.0', target_value='85.0', similarity=0.278), ValueMatch(current_value='57.0', target_value='70.1', similarity=0.263), ValueMatch(current_value='71.0', target_value='70.1', similarity=0.262)]\n",
      "source: Age target: fev1_fvc_post_bronch_percent score: 8.032000000000002\n",
      "value_matches:  [ValueMatch(current_value='90.0', target_value='90.0', similarity=1.0), ValueMatch(current_value='75.0', target_value='175.0', similarity=0.712), ValueMatch(current_value='55.0', target_value='55.4', similarity=0.472), ValueMatch(current_value='53.0', target_value='75.3', similarity=0.452), ValueMatch(current_value='59.0', target_value='90.0', similarity=0.414), ValueMatch(current_value='69.0', target_value='90.0', similarity=0.412), ValueMatch(current_value='45.0', target_value='45.2', similarity=0.411), ValueMatch(current_value='56.0', target_value='25.6', similarity=0.396), ValueMatch(current_value='74.0', target_value='40.7', similarity=0.389), ValueMatch(current_value='57.0', target_value='95.7', similarity=0.386), ValueMatch(current_value='41.0', target_value='110.4', similarity=0.386), ValueMatch(current_value='44.0', target_value='40.7', similarity=0.368), ValueMatch(current_value='86.0', target_value='60.8', similarity=0.364), ValueMatch(current_value='63.0', target_value='30.0', similarity=0.359), ValueMatch(current_value='73.0', target_value='30.0', similarity=0.358), ValueMatch(current_value='83.0', target_value='30.0', similarity=0.353), ValueMatch(current_value='71.0', target_value='110.4', similarity=0.323), ValueMatch(current_value='61.0', target_value='110.4', similarity=0.323), ValueMatch(current_value='66.0', target_value='60.8', similarity=0.31), ValueMatch(current_value='64.0', target_value='40.7', similarity=0.308), ValueMatch(current_value='50.0', target_value='30.0', similarity=0.306), ValueMatch(current_value='60.0', target_value='60.8', similarity=0.304), ValueMatch(current_value='70.0', target_value='30.0', similarity=0.291), ValueMatch(current_value='62.0', target_value='205.1', similarity=0.263), ValueMatch(current_value='72.0', target_value='205.1', similarity=0.262), ValueMatch(current_value='42.0', target_value='205.1', similarity=0.26), ValueMatch(current_value='76.0', target_value='60.8', similarity=0.252), ValueMatch(current_value='46.0', target_value='60.8', similarity=0.25)]\n",
      "source: Age target: gross_tumor_weight score: 10.684000000000001\n",
      "value_matches:  []\n",
      "source: Age target: demographics score: 0\n",
      "value_matches:  [ValueMatch(current_value='nan', target_value='Unknown', similarity=0.289)]\n",
      "source: Age target: menopause_status score: 0.289\n",
      "value_matches:  [ValueMatch(current_value='nan', target_value='Anemia', similarity=0.342)]\n",
      "source: Age target: comorbidities score: 0.342\n",
      "value_matches:  [ValueMatch(current_value='50.0', target_value='500', similarity=1.0), ValueMatch(current_value='45.0', target_value='450', similarity=1.0), ValueMatch(current_value='55.0', target_value='550', similarity=1.0), ValueMatch(current_value='75.0', target_value='75', similarity=0.707), ValueMatch(current_value='90.0', target_value='90', similarity=0.699), ValueMatch(current_value='59.0', target_value='90', similarity=0.608), ValueMatch(current_value='69.0', target_value='90', similarity=0.602), ValueMatch(current_value='65.0', target_value='50', similarity=0.448), ValueMatch(current_value='44.0', target_value='400', similarity=0.443), ValueMatch(current_value='85.0', target_value='50', similarity=0.44), ValueMatch(current_value='62.0', target_value='220', similarity=0.388), ValueMatch(current_value='72.0', target_value='220', similarity=0.386), ValueMatch(current_value='42.0', target_value='220', similarity=0.383), ValueMatch(current_value='53.0', target_value='300', similarity=0.381), ValueMatch(current_value='63.0', target_value='300', similarity=0.378), ValueMatch(current_value='73.0', target_value='300', similarity=0.376), ValueMatch(current_value='83.0', target_value='300', similarity=0.372), ValueMatch(current_value='64.0', target_value='400', similarity=0.371), ValueMatch(current_value='74.0', target_value='400', similarity=0.369), ValueMatch(current_value='60.0', target_value='500', similarity=0.298), ValueMatch(current_value='70.0', target_value='500', similarity=0.289), ValueMatch(current_value='58.0', target_value='180', similarity=0.288), ValueMatch(current_value='68.0', target_value='180', similarity=0.285), ValueMatch(current_value='78.0', target_value='180', similarity=0.283), ValueMatch(current_value='48.0', target_value='180', similarity=0.281), ValueMatch(current_value='38.0', target_value='180', similarity=0.277), ValueMatch(current_value='77.0', target_value='75', similarity=0.275), ValueMatch(current_value='57.0', target_value='75', similarity=0.256)]\n",
      "source: Age target: current_weight score: 12.883\n",
      "value_matches:  [ValueMatch(current_value='50.0', target_value='50.0', similarity=1.0), ValueMatch(current_value='44.0', target_value='34.4', similarity=0.502), ValueMatch(current_value='55.0', target_value='55.6', similarity=0.497), ValueMatch(current_value='56.0', target_value='55.6', similarity=0.462), ValueMatch(current_value='41.0', target_value='41.8', similarity=0.45), ValueMatch(current_value='83.0', target_value='68.3', similarity=0.446), ValueMatch(current_value='53.0', target_value='45.3', similarity=0.431), ValueMatch(current_value='45.0', target_value='45.3', similarity=0.425), ValueMatch(current_value='72.0', target_value='87.2', similarity=0.423), ValueMatch(current_value='68.0', target_value='68.3', similarity=0.419), ValueMatch(current_value='62.0', target_value='62.1', similarity=0.407), ValueMatch(current_value='74.0', target_value='74.9', similarity=0.4), ValueMatch(current_value='90.0', target_value='29.0', similarity=0.393), ValueMatch(current_value='58.0', target_value='80.5', similarity=0.361), ValueMatch(current_value='60.0', target_value='50.0', similarity=0.358), ValueMatch(current_value='70.0', target_value='50.0', similarity=0.348), ValueMatch(current_value='59.0', target_value='29.0', similarity=0.343), ValueMatch(current_value='69.0', target_value='29.0', similarity=0.343), ValueMatch(current_value='65.0', target_value='50.0', similarity=0.323), ValueMatch(current_value='75.0', target_value='50.0', similarity=0.32), ValueMatch(current_value='85.0', target_value='50.0', similarity=0.319), ValueMatch(current_value='78.0', target_value='80.5', similarity=0.282), ValueMatch(current_value='48.0', target_value='80.5', similarity=0.281), ValueMatch(current_value='38.0', target_value='80.5', similarity=0.278)]\n",
      "source: Age target: epithelioid_cell_percent score: 9.811000000000003\n",
      "value_matches:  [ValueMatch(current_value='85.0', target_value='85.0', similarity=1.0), ValueMatch(current_value='58.0', target_value='58.8', similarity=0.473), ValueMatch(current_value='74.0', target_value='74.5', similarity=0.433), ValueMatch(current_value='63.0', target_value='63.4', similarity=0.427), ValueMatch(current_value='44.0', target_value='40.0', similarity=0.42), ValueMatch(current_value='56.0', target_value='45.6', similarity=0.418), ValueMatch(current_value='69.0', target_value='69.1', similarity=0.416), ValueMatch(current_value='73.0', target_value='97.3', similarity=0.415), ValueMatch(current_value='45.0', target_value='45.6', similarity=0.407), ValueMatch(current_value='90.0', target_value='90.8', similarity=0.381), ValueMatch(current_value='64.0', target_value='40.0', similarity=0.361), ValueMatch(current_value='50.0', target_value='40.0', similarity=0.335), ValueMatch(current_value='60.0', target_value='40.0', similarity=0.332), ValueMatch(current_value='59.0', target_value='90.8', similarity=0.331), ValueMatch(current_value='55.0', target_value='85.0', similarity=0.324), ValueMatch(current_value='70.0', target_value='40.0', similarity=0.323), ValueMatch(current_value='65.0', target_value='85.0', similarity=0.269), ValueMatch(current_value='75.0', target_value='85.0', similarity=0.267)]\n",
      "source: Age target: percent_tumor_invasion score: 7.332000000000001\n",
      "value_matches:  [ValueMatch(current_value='42.0', target_value='2024', similarity=0.319), ValueMatch(current_value='62.0', target_value='2020', similarity=0.316), ValueMatch(current_value='72.0', target_value='2020', similarity=0.313)]\n",
      "source: Age target: year_of_death score: 0.948\n",
      "value_matches:  [ValueMatch(current_value='73.0', target_value='730', similarity=1.0), ValueMatch(current_value='65.0', target_value='-3650', similarity=0.766), ValueMatch(current_value='90.0', target_value='90', similarity=0.676), ValueMatch(current_value='45.0', target_value='-14500', similarity=0.656), ValueMatch(current_value='69.0', target_value='90', similarity=0.587), ValueMatch(current_value='59.0', target_value='90', similarity=0.586), ValueMatch(current_value='50.0', target_value='-14500', similarity=0.579), ValueMatch(current_value='72.0', target_value='32872', similarity=0.374), ValueMatch(current_value='60.0', target_value='0', similarity=0.328), ValueMatch(current_value='70.0', target_value='0', similarity=0.32), ValueMatch(current_value='63.0', target_value='730', similarity=0.31), ValueMatch(current_value='53.0', target_value='730', similarity=0.309), ValueMatch(current_value='83.0', target_value='730', similarity=0.307), ValueMatch(current_value='58.0', target_value='180', similarity=0.281), ValueMatch(current_value='68.0', target_value='180', similarity=0.281), ValueMatch(current_value='78.0', target_value='180', similarity=0.281), ValueMatch(current_value='38.0', target_value='180', similarity=0.279), ValueMatch(current_value='48.0', target_value='180', similarity=0.275), ValueMatch(current_value='55.0', target_value='-3650', similarity=0.262)]\n",
      "source: Age target: days_to_last_known_disease_status score: 8.457\n",
      "value_matches:  [ValueMatch(current_value='55.0', target_value='550X', similarity=0.859), ValueMatch(current_value='45.0', target_value='450X', similarity=0.845), ValueMatch(current_value='65.0', target_value='650X', similarity=0.839), ValueMatch(current_value='70.0', target_value='700X', similarity=0.834), ValueMatch(current_value='60.0', target_value='600X', similarity=0.821), ValueMatch(current_value='50.0', target_value='500X', similarity=0.811), ValueMatch(current_value='44.0', target_value='400X', similarity=0.382), ValueMatch(current_value='77.0', target_value='700X', similarity=0.367), ValueMatch(current_value='61.0', target_value='100X', similarity=0.357), ValueMatch(current_value='62.0', target_value='200X', similarity=0.357), ValueMatch(current_value='72.0', target_value='200X', similarity=0.353), ValueMatch(current_value='71.0', target_value='100X', similarity=0.353), ValueMatch(current_value='42.0', target_value='200X', similarity=0.351), ValueMatch(current_value='41.0', target_value='100X', similarity=0.351), ValueMatch(current_value='63.0', target_value='30X', similarity=0.345), ValueMatch(current_value='53.0', target_value='30X', similarity=0.345), ValueMatch(current_value='66.0', target_value='60X', similarity=0.344), ValueMatch(current_value='73.0', target_value='30X', similarity=0.341), ValueMatch(current_value='83.0', target_value='30X', similarity=0.335), ValueMatch(current_value='64.0', target_value='400X', similarity=0.322), ValueMatch(current_value='74.0', target_value='400X', similarity=0.318), ValueMatch(current_value='57.0', target_value='700X', similarity=0.308), ValueMatch(current_value='67.0', target_value='700X', similarity=0.307), ValueMatch(current_value='56.0', target_value='60X', similarity=0.279), ValueMatch(current_value='76.0', target_value='60X', similarity=0.274), ValueMatch(current_value='46.0', target_value='60X', similarity=0.272), ValueMatch(current_value='86.0', target_value='60X', similarity=0.27)]\n",
      "source: Age target: mean_coverage score: 11.94\n",
      "Top k reranked columns:  [('Age', 'age_at_last_exposure', 18.641), ('Age', 'age_at_onset', 16.569), ('Age', 'current_weight', 12.883), ('Age', 'mean_coverage', 11.94), ('Age', 'gross_tumor_weight', 10.684000000000001), ('Age', 'days_to_birth', 10.608999999999998), ('Age', 'initial_weight', 10.428999999999998), ('Age', 'age_at_diagnosis', 10.322), ('Age', 'relationship_age_at_diagnosis', 10.068000000000001), ('Age', 'epithelioid_cell_percent', 9.811000000000003), ('Age', 'days_to_last_known_disease_status', 8.457), ('Age', 'age_at_index', 8.264999999999999), ('Age', 'fev1_fvc_post_bronch_percent', 8.032000000000002), ('Age', 'percent_tumor_invasion', 7.332000000000001), ('Age', 'weight', 7.3069999999999995), ('Age', 'fev1_fvc_pre_bronch_percent', 6.075), ('Age', 'year_of_death', 0.948), ('Age', 'comorbidities', 0.342), ('Age', 'menopause_status', 0.289), ('Age', 'demographics', 0)]\n",
      "\n",
      "Source column:  Race\n",
      "value_matches:  [ValueMatch(current_value='White', target_value='white', similarity=1.0), ValueMatch(current_value=' White', target_value='white', similarity=1.0), ValueMatch(current_value='Not Reported', target_value='not reported', similarity=1.0), ValueMatch(current_value='Asian', target_value='asian', similarity=1.0), ValueMatch(current_value='Black or African American', target_value='black or african american', similarity=1.0), ValueMatch(current_value='nan', target_value='american indian or alaska native', similarity=0.353)]\n",
      "source: Race target: race score: 5.353\n",
      "value_matches:  [ValueMatch(current_value='Not Reported', target_value='not reported', similarity=1.0), ValueMatch(current_value='Black or African American', target_value='hispanic or latino', similarity=0.311)]\n",
      "source: Race target: ethnicity score: 1.311\n",
      "value_matches:  [ValueMatch(current_value='Not Reported', target_value='Not Reported', similarity=1.0), ValueMatch(current_value='White', target_value='White Matter', similarity=0.66), ValueMatch(current_value=' White', target_value='White Matter', similarity=0.66)]\n",
      "source: Race target: supratentorial_localization score: 2.3200000000000003\n",
      "value_matches:  [ValueMatch(current_value='Not Reported', target_value='Unrelated', similarity=0.291)]\n",
      "source: Race target: relationship_type score: 0.291\n",
      "value_matches:  [ValueMatch(current_value='Asian', target_value='Malaysia', similarity=0.334), ValueMatch(current_value='Not Reported', target_value='Federated States of Micronesia', similarity=0.3), ValueMatch(current_value='nan', target_value='China', similarity=0.264)]\n",
      "source: Race target: country_of_birth score: 0.898\n",
      "value_matches:  [ValueMatch(current_value='Black or African American', target_value='Family History of Cancer', similarity=0.335)]\n",
      "source: Race target: demographics score: 0.335\n",
      "value_matches:  []\n",
      "source: Race target: well_number score: 0\n",
      "value_matches:  [ValueMatch(current_value='Not Reported', target_value='Not Reported', similarity=1.0)]\n",
      "source: Race target: eye_color score: 1.0\n",
      "value_matches:  [ValueMatch(current_value='Asian', target_value='Not Classified', similarity=0.317), ValueMatch(current_value='Not Reported', target_value='Not Classified', similarity=0.267)]\n",
      "source: Race target: fab_morphology_code score: 0.5840000000000001\n",
      "value_matches:  []\n",
      "source: Race target: plate_well score: 0\n",
      "value_matches:  [ValueMatch(current_value='Not Reported', target_value='Not Reported', similarity=1.0)]\n",
      "source: Race target: education_level score: 1.0\n",
      "value_matches:  [ValueMatch(current_value='Not Reported', target_value='not reported', similarity=1.0)]\n",
      "source: Race target: gender score: 1.0\n",
      "value_matches:  [ValueMatch(current_value='Black or African American', target_value='France', similarity=0.281), ValueMatch(current_value='Asian', target_value='Saudi Arabia', similarity=0.262)]\n",
      "source: Race target: country_of_residence_at_enrollment score: 0.543\n",
      "value_matches:  [ValueMatch(current_value='Not Reported', target_value='Not Reported', similarity=1.0)]\n",
      "source: Race target: child_pugh_classification score: 1.0\n",
      "value_matches:  [ValueMatch(current_value='White', target_value='BWH', similarity=0.279), ValueMatch(current_value=' White', target_value='BWH', similarity=0.279)]\n",
      "source: Race target: short_name score: 0.558\n",
      "value_matches:  [ValueMatch(current_value='Black or African American', target_value='Respirable Crystalline Silica', similarity=0.307), ValueMatch(current_value='nan', target_value='Marijuana', similarity=0.292)]\n",
      "source: Race target: exposure_type score: 0.599\n",
      "value_matches:  [ValueMatch(current_value='Not Reported', target_value='Not Reported', similarity=1.0)]\n",
      "source: Race target: alcohol_type score: 1.0\n",
      "value_matches:  []\n",
      "source: Race target: base_caller_version score: 0\n",
      "value_matches:  [ValueMatch(current_value='Black or African American', target_value='Melanocytic', similarity=0.261)]\n",
      "source: Race target: roots score: 0.261\n",
      "value_matches:  [ValueMatch(current_value='Black or African American', target_value='Technical Variation Minimization', similarity=0.253)]\n",
      "source: Race target: methylation_array_harmonization_workflows score: 0.253\n",
      "Top k reranked columns:  [('Race', 'race', 5.353), ('Race', 'supratentorial_localization', 2.3200000000000003), ('Race', 'ethnicity', 1.311), ('Race', 'eye_color', 1.0), ('Race', 'education_level', 1.0), ('Race', 'gender', 1.0), ('Race', 'child_pugh_classification', 1.0), ('Race', 'alcohol_type', 1.0), ('Race', 'country_of_birth', 0.898), ('Race', 'exposure_type', 0.599), ('Race', 'fab_morphology_code', 0.5840000000000001), ('Race', 'short_name', 0.558), ('Race', 'country_of_residence_at_enrollment', 0.543), ('Race', 'demographics', 0.335), ('Race', 'relationship_type', 0.291), ('Race', 'roots', 0.261), ('Race', 'methylation_array_harmonization_workflows', 0.253), ('Race', 'well_number', 0), ('Race', 'plate_well', 0), ('Race', 'base_caller_version', 0)]\n",
      "\n",
      "Source column:  Ethnicity\n",
      "value_matches:  [ValueMatch(current_value='Not reported', target_value='not reported', similarity=1.0), ValueMatch(current_value='Hispanic or Latino', target_value='hispanic or latino', similarity=1.0), ValueMatch(current_value='Not-Hispanic or Latino', target_value='not hispanic or latino', similarity=0.937)]\n",
      "source: Ethnicity target: ethnicity score: 2.9370000000000003\n",
      "value_matches:  [ValueMatch(current_value='Not reported', target_value='not reported', similarity=1.0), ValueMatch(current_value='Not-Hispanic or Latino', target_value='native hawaiian or other pacific islander', similarity=0.453), ValueMatch(current_value='Hispanic or Latino', target_value='native hawaiian or other pacific islander', similarity=0.446), ValueMatch(current_value='nan', target_value='american indian or alaska native', similarity=0.334)]\n",
      "source: Ethnicity target: race score: 2.233\n",
      "value_matches:  [ValueMatch(current_value='Not reported', target_value='not reported', similarity=1.0), ValueMatch(current_value='Not-Hispanic or Latino', target_value='not reported', similarity=0.307)]\n",
      "source: Ethnicity target: gender score: 1.307\n",
      "value_matches:  [ValueMatch(current_value='Not reported', target_value='not reported', similarity=1.0), ValueMatch(current_value='Not-Hispanic or Latino', target_value='not reported', similarity=0.307)]\n",
      "source: Ethnicity target: relationship_gender score: 1.307\n",
      "value_matches:  [ValueMatch(current_value='Not reported', target_value='Federated States of Micronesia', similarity=0.314), ValueMatch(current_value='Not-Hispanic or Latino', target_value='State of Palestine', similarity=0.289), ValueMatch(current_value='Hispanic or Latino', target_value='State of Palestine', similarity=0.289), ValueMatch(current_value='nan', target_value='China', similarity=0.263)]\n",
      "source: Ethnicity target: country_of_birth score: 1.1549999999999998\n",
      "value_matches:  [ValueMatch(current_value='Hispanic or Latino', target_value='Geographic Location', similarity=0.278), ValueMatch(current_value='Not-Hispanic or Latino', target_value='Geographic Location', similarity=0.265)]\n",
      "source: Ethnicity target: demographics score: 0.543\n",
      "value_matches:  [ValueMatch(current_value='Not reported', target_value='Not Reported', similarity=1.0), ValueMatch(current_value='Hispanic or Latino', target_value='Spinal Cord', similarity=0.284), ValueMatch(current_value='Not-Hispanic or Latino', target_value='Spinal Cord', similarity=0.265)]\n",
      "source: Ethnicity target: supratentorial_localization score: 1.549\n",
      "value_matches:  []\n",
      "source: Ethnicity target: well_number score: 0\n",
      "value_matches:  [ValueMatch(current_value='Not reported', target_value='Not Reported', similarity=1.0), ValueMatch(current_value='Not-Hispanic or Latino', target_value='Not Reported', similarity=0.293)]\n",
      "source: Ethnicity target: eye_color score: 1.293\n",
      "value_matches:  [ValueMatch(current_value='Not reported', target_value='Not Reported', similarity=1.0), ValueMatch(current_value='Not-Hispanic or Latino', target_value='Not Reported', similarity=0.31)]\n",
      "source: Ethnicity target: child_pugh_classification score: 1.31\n",
      "value_matches:  []\n",
      "source: Ethnicity target: ajcc_pathologic_n score: 0\n",
      "value_matches:  [ValueMatch(current_value='Not-Hispanic or Latino', target_value='Not Classified', similarity=0.378), ValueMatch(current_value='Hispanic or Latino', target_value='Not Classified', similarity=0.34), ValueMatch(current_value='Not reported', target_value='Not Classified', similarity=0.262)]\n",
      "source: Ethnicity target: fab_morphology_code score: 0.98\n",
      "value_matches:  [ValueMatch(current_value='Not reported', target_value='Unrelated', similarity=0.317)]\n",
      "source: Ethnicity target: relationship_type score: 0.317\n",
      "value_matches:  []\n",
      "source: Ethnicity target: plate_well score: 0\n",
      "value_matches:  []\n",
      "source: Ethnicity target: ajcc_clinical_n score: 0\n",
      "value_matches:  [ValueMatch(current_value='Not reported', target_value='Separated', similarity=0.311), ValueMatch(current_value='Hispanic or Latino', target_value='Domestic Partnership', similarity=0.295), ValueMatch(current_value='Not-Hispanic or Latino', target_value='Domestic Partnership', similarity=0.283)]\n",
      "source: Ethnicity target: marital_status score: 0.889\n",
      "value_matches:  [ValueMatch(current_value='Not reported', target_value='not reported', similarity=1.0), ValueMatch(current_value='Not-Hispanic or Latino', target_value='no', similarity=0.361), ValueMatch(current_value='Hispanic or Latino', target_value='no', similarity=0.254)]\n",
      "source: Ethnicity target: relative_with_cancer_history score: 1.615\n",
      "value_matches:  [ValueMatch(current_value='Not reported', target_value='Not Reported', similarity=1.0), ValueMatch(current_value='Not-Hispanic or Latino', target_value='Not Reported', similarity=0.299)]\n",
      "source: Ethnicity target: histone_family score: 1.299\n",
      "value_matches:  [ValueMatch(current_value='Hispanic or Latino', target_value='Occupational', similarity=0.32), ValueMatch(current_value='Not-Hispanic or Latino', target_value='Occupational', similarity=0.309)]\n",
      "source: Ethnicity target: exposure_source score: 0.629\n",
      "value_matches:  [ValueMatch(current_value='Hispanic or Latino', target_value='Kiribati', similarity=0.291), ValueMatch(current_value='Not-Hispanic or Latino', target_value='Kiribati', similarity=0.259)]\n",
      "source: Ethnicity target: country_of_residence_at_enrollment score: 0.55\n",
      "Top k reranked columns:  [('Ethnicity', 'ethnicity', 2.9370000000000003), ('Ethnicity', 'race', 2.233), ('Ethnicity', 'relative_with_cancer_history', 1.615), ('Ethnicity', 'supratentorial_localization', 1.549), ('Ethnicity', 'child_pugh_classification', 1.31), ('Ethnicity', 'gender', 1.307), ('Ethnicity', 'relationship_gender', 1.307), ('Ethnicity', 'histone_family', 1.299), ('Ethnicity', 'eye_color', 1.293), ('Ethnicity', 'country_of_birth', 1.1549999999999998), ('Ethnicity', 'fab_morphology_code', 0.98), ('Ethnicity', 'marital_status', 0.889), ('Ethnicity', 'exposure_source', 0.629), ('Ethnicity', 'country_of_residence_at_enrollment', 0.55), ('Ethnicity', 'demographics', 0.543), ('Ethnicity', 'relationship_type', 0.317), ('Ethnicity', 'well_number', 0), ('Ethnicity', 'ajcc_pathologic_n', 0), ('Ethnicity', 'plate_well', 0), ('Ethnicity', 'ajcc_clinical_n', 0)]\n",
      "\n",
      "Source column:  Gender\n",
      "value_matches:  [ValueMatch(current_value='Female', target_value='female', similarity=1.0)]\n",
      "source: Gender target: gender score: 1.0\n",
      "value_matches:  [ValueMatch(current_value='Female', target_value='female', similarity=1.0)]\n",
      "source: Gender target: relationship_gender score: 1.0\n",
      "value_matches:  [ValueMatch(current_value='nan', target_value='No', similarity=0.273)]\n",
      "source: Gender target: premature_at_birth score: 0.273\n",
      "value_matches:  [ValueMatch(current_value='nan', target_value='No', similarity=0.273)]\n",
      "source: Gender target: pregnant_at_diagnosis score: 0.273\n",
      "value_matches:  [ValueMatch(current_value='nan', target_value='Ectopic Pregnancy', similarity=0.304)]\n",
      "source: Gender target: pregnancy_outcome score: 0.304\n",
      "value_matches:  []\n",
      "source: Gender target: marital_status score: 0\n",
      "value_matches:  []\n",
      "source: Gender target: menopause_status score: 0\n",
      "value_matches:  [ValueMatch(current_value='Female', target_value='Maternal Half Sister', similarity=0.27)]\n",
      "source: Gender target: relationship_type score: 0.27\n",
      "value_matches:  [ValueMatch(current_value='Female', target_value='False', similarity=0.276)]\n",
      "source: Gender target: age_is_obfuscated score: 0.276\n",
      "value_matches:  [ValueMatch(current_value='nan', target_value='No', similarity=0.273)]\n",
      "source: Gender target: tumor_confined_to_organ_of_origin score: 0.273\n",
      "value_matches:  [ValueMatch(current_value='Female', target_value='Guatemala', similarity=0.413), ValueMatch(current_value='nan', target_value='Antigua and Barbuda', similarity=0.273)]\n",
      "source: Gender target: country_of_birth score: 0.6859999999999999\n",
      "value_matches:  []\n",
      "source: Gender target: hysterectomy_type score: 0\n",
      "value_matches:  [ValueMatch(current_value='Female', target_value='Guatemala', similarity=0.42)]\n",
      "source: Gender target: country_of_residence_at_enrollment score: 0.42\n",
      "value_matches:  [ValueMatch(current_value='nan', target_value='Anus and anal canal', similarity=0.549), ValueMatch(current_value='Female', target_value='Other and unspecified female genital organs', similarity=0.457)]\n",
      "source: Gender target: primary_site score: 1.006\n",
      "value_matches:  [ValueMatch(current_value='Female', target_value='False', similarity=0.276)]\n",
      "source: Gender target: selected_normal_wxs score: 0.276\n",
      "value_matches:  []\n",
      "source: Gender target: weeks_gestation_at_birth score: 0\n",
      "value_matches:  [ValueMatch(current_value='Female', target_value='False', similarity=0.276)]\n",
      "source: Gender target: selected_normal_wgs score: 0.276\n",
      "value_matches:  []\n",
      "source: Gender target: ethnicity score: 0\n",
      "value_matches:  [ValueMatch(current_value='Female', target_value='Alleles', similarity=0.306)]\n",
      "source: Gender target: variant_type score: 0.306\n",
      "value_matches:  [ValueMatch(current_value='Female', target_value='Left', similarity=0.269)]\n",
      "source: Gender target: laterality score: 0.269\n",
      "Top k reranked columns:  [('Gender', 'primary_site', 1.006), ('Gender', 'gender', 1.0), ('Gender', 'relationship_gender', 1.0), ('Gender', 'country_of_birth', 0.6859999999999999), ('Gender', 'country_of_residence_at_enrollment', 0.42), ('Gender', 'variant_type', 0.306), ('Gender', 'pregnancy_outcome', 0.304), ('Gender', 'age_is_obfuscated', 0.276), ('Gender', 'selected_normal_wxs', 0.276), ('Gender', 'selected_normal_wgs', 0.276), ('Gender', 'premature_at_birth', 0.273), ('Gender', 'pregnant_at_diagnosis', 0.273), ('Gender', 'tumor_confined_to_organ_of_origin', 0.273), ('Gender', 'relationship_type', 0.27), ('Gender', 'laterality', 0.269), ('Gender', 'marital_status', 0), ('Gender', 'menopause_status', 0), ('Gender', 'hysterectomy_type', 0), ('Gender', 'weeks_gestation_at_birth', 0), ('Gender', 'ethnicity', 0)]\n",
      "\n",
      "Source column:  Tumor_Focality\n",
      "value_matches:  [ValueMatch(current_value='Unifocal', target_value='Unifocal', similarity=1.0), ValueMatch(current_value='Multifocal', target_value='Multifocal', similarity=1.0)]\n",
      "source: Tumor_Focality target: tumor_focality score: 2.0\n",
      "value_matches:  [ValueMatch(current_value='nan', target_value='Unknown', similarity=0.305)]\n",
      "source: Tumor_Focality target: tumor_shape score: 0.305\n",
      "value_matches:  []\n",
      "source: Tumor_Focality target: tumor_depth_descriptor score: 0\n",
      "value_matches:  [ValueMatch(current_value='Unifocal', target_value='Focal', similarity=0.667), ValueMatch(current_value='Multifocal', target_value='Focal', similarity=0.586)]\n",
      "source: Tumor_Focality target: extracapsular_extension score: 1.2530000000000001\n",
      "value_matches:  [ValueMatch(current_value='nan', target_value='Many', similarity=0.327)]\n",
      "source: Tumor_Focality target: tumor_infiltrating_macrophages score: 0.327\n",
      "value_matches:  []\n",
      "source: Tumor_Focality target: enneking_msts_tumor_site score: 0\n",
      "value_matches:  []\n",
      "source: Tumor_Focality target: tissue_type score: 0\n",
      "value_matches:  [ValueMatch(current_value='nan', target_value='No', similarity=0.271)]\n",
      "source: Tumor_Focality target: history_of_tumor score: 0.271\n",
      "value_matches:  []\n",
      "source: Tumor_Focality target: distance_normal_to_tumor score: 0\n",
      "value_matches:  []\n",
      "source: Tumor_Focality target: residual_tumor_measurement score: 0\n",
      "value_matches:  []\n",
      "source: Tumor_Focality target: wilms_tumor_histologic_subtype score: 0\n",
      "value_matches:  []\n",
      "source: Tumor_Focality target: tumor_width_measurement score: 0\n",
      "value_matches:  [ValueMatch(current_value='nan', target_value='No', similarity=0.259)]\n",
      "source: Tumor_Focality target: tumor_confined_to_organ_of_origin score: 0.259\n",
      "value_matches:  [ValueMatch(current_value='nan', target_value='No', similarity=0.259)]\n",
      "source: Tumor_Focality target: non_nodal_tumor_deposits score: 0.259\n",
      "value_matches:  []\n",
      "source: Tumor_Focality target: tumor_thickness score: 0\n",
      "value_matches:  [ValueMatch(current_value='nan', target_value='No Extranodal Extension', similarity=0.254)]\n",
      "source: Tumor_Focality target: extranodal_extension score: 0.254\n",
      "value_matches:  [ValueMatch(current_value='nan', target_value='Many', similarity=0.327)]\n",
      "source: Tumor_Focality target: tumor_infiltrating_lymphocytes score: 0.327\n",
      "value_matches:  []\n",
      "source: Tumor_Focality target: margin_distance score: 0\n",
      "value_matches:  []\n",
      "source: Tumor_Focality target: tumor_grade_category score: 0\n",
      "value_matches:  [ValueMatch(current_value='nan', target_value='Manta', similarity=0.269)]\n",
      "source: Tumor_Focality target: somatic_mutation_calling_workflows score: 0.269\n",
      "Top k reranked columns:  [('Tumor_Focality', 'tumor_focality', 2.0), ('Tumor_Focality', 'extracapsular_extension', 1.2530000000000001), ('Tumor_Focality', 'tumor_infiltrating_macrophages', 0.327), ('Tumor_Focality', 'tumor_infiltrating_lymphocytes', 0.327), ('Tumor_Focality', 'tumor_shape', 0.305), ('Tumor_Focality', 'history_of_tumor', 0.271), ('Tumor_Focality', 'somatic_mutation_calling_workflows', 0.269), ('Tumor_Focality', 'tumor_confined_to_organ_of_origin', 0.259), ('Tumor_Focality', 'non_nodal_tumor_deposits', 0.259), ('Tumor_Focality', 'extranodal_extension', 0.254), ('Tumor_Focality', 'tumor_depth_descriptor', 0), ('Tumor_Focality', 'enneking_msts_tumor_site', 0), ('Tumor_Focality', 'tissue_type', 0), ('Tumor_Focality', 'distance_normal_to_tumor', 0), ('Tumor_Focality', 'residual_tumor_measurement', 0), ('Tumor_Focality', 'wilms_tumor_histologic_subtype', 0), ('Tumor_Focality', 'tumor_width_measurement', 0), ('Tumor_Focality', 'tumor_thickness', 0), ('Tumor_Focality', 'margin_distance', 0), ('Tumor_Focality', 'tumor_grade_category', 0)]\n",
      "\n",
      "Source column:  Tumor_Size_cm\n",
      "value_matches:  [ValueMatch(current_value='6.0', target_value='6.0', similarity=1.0), ValueMatch(current_value='0.9', target_value='0.9', similarity=1.0), ValueMatch(current_value='3.7', target_value='3.7', similarity=1.0), ValueMatch(current_value='8.5', target_value='8.5', similarity=1.0), ValueMatch(current_value='1.2', target_value='1.2', similarity=1.0), ValueMatch(current_value='4.3', target_value='4.3', similarity=1.0), ValueMatch(current_value='2.5', target_value='2.5', similarity=1.0), ValueMatch(current_value='1.8', target_value='1.8', similarity=1.0), ValueMatch(current_value='1.0', target_value='10.4', similarity=0.595), ValueMatch(current_value='9.0', target_value='0.9', similarity=0.452), ValueMatch(current_value='2.7', target_value='7.2', similarity=0.416), ValueMatch(current_value='11.0', target_value='10.4', similarity=0.407), ValueMatch(current_value='4.5', target_value='5.4', similarity=0.395), ValueMatch(current_value='5.2', target_value='2.5', similarity=0.387), ValueMatch(current_value='3.5', target_value='3.3', similarity=0.331), ValueMatch(current_value='5.5', target_value='2.5', similarity=0.331), ValueMatch(current_value='2.2', target_value='2.5', similarity=0.331), ValueMatch(current_value='1.3', target_value='3.3', similarity=0.329), ValueMatch(current_value='2.6', target_value='6.0', similarity=0.318), ValueMatch(current_value='6.3', target_value='6.0', similarity=0.318), ValueMatch(current_value='6.5', target_value='6.0', similarity=0.318), ValueMatch(current_value='3.2', target_value='3.3', similarity=0.312), ValueMatch(current_value='3.0', target_value='3.3', similarity=0.312), ValueMatch(current_value='2.3', target_value='3.3', similarity=0.312), ValueMatch(current_value='3.8', target_value='3.3', similarity=0.299), ValueMatch(current_value='3.9', target_value='3.3', similarity=0.296), ValueMatch(current_value='9.5', target_value='0.9', similarity=0.279), ValueMatch(current_value='2.9', target_value='0.9', similarity=0.279), ValueMatch(current_value='1.4', target_value='10.4', similarity=0.271), ValueMatch(current_value='4.0', target_value='10.4', similarity=0.266), ValueMatch(current_value='7.0', target_value='3.7', similarity=0.263), ValueMatch(current_value='8.0', target_value='8.5', similarity=0.263), ValueMatch(current_value='0.8', target_value='8.5', similarity=0.263), ValueMatch(current_value='1.7', target_value='3.7', similarity=0.262), ValueMatch(current_value='4.7', target_value='3.7', similarity=0.257)]\n",
      "source: Tumor_Size_cm target: tumor_width_measurement score: 16.862000000000002\n",
      "value_matches:  [ValueMatch(current_value='3.5', target_value='3.5', similarity=1.0), ValueMatch(current_value='6.0', target_value='6.0', similarity=1.0), ValueMatch(current_value='1.5', target_value='1.5', similarity=1.0), ValueMatch(current_value='1.2', target_value='1.2', similarity=1.0), ValueMatch(current_value='0.8', target_value='0.8', similarity=1.0), ValueMatch(current_value='3.0', target_value='3.0', similarity=1.0), ValueMatch(current_value='4.2', target_value='4.2', similarity=1.0), ValueMatch(current_value='4.7', target_value='4.7', similarity=1.0), ValueMatch(current_value='13.5', target_value='3.5', similarity=0.61), ValueMatch(current_value='6.5', target_value='5.6', similarity=0.437), ValueMatch(current_value='8.0', target_value='0.8', similarity=0.429), ValueMatch(current_value='5.5', target_value='3.5', similarity=0.334), ValueMatch(current_value='2.2', target_value='1.2', similarity=0.329), ValueMatch(current_value='6.3', target_value='6.0', similarity=0.297), ValueMatch(current_value='2.6', target_value='6.0', similarity=0.297), ValueMatch(current_value='9.5', target_value='9.1', similarity=0.282), ValueMatch(current_value='9.0', target_value='9.1', similarity=0.281), ValueMatch(current_value='2.9', target_value='9.1', similarity=0.281), ValueMatch(current_value='0.9', target_value='9.1', similarity=0.281), ValueMatch(current_value='3.9', target_value='9.1', similarity=0.281), ValueMatch(current_value='2.7', target_value='4.7', similarity=0.27), ValueMatch(current_value='3.7', target_value='4.7', similarity=0.27), ValueMatch(current_value='7.0', target_value='4.7', similarity=0.27), ValueMatch(current_value='1.7', target_value='4.7', similarity=0.269), ValueMatch(current_value='4.5', target_value='4.2', similarity=0.25), ValueMatch(current_value='8.5', target_value='0.8', similarity=0.25)]\n",
      "source: Tumor_Size_cm target: tumor_depth_measurement score: 13.718000000000002\n",
      "value_matches:  [ValueMatch(current_value='5.2', target_value='5.2', similarity=1.0), ValueMatch(current_value='6.0', target_value='6.0', similarity=1.0), ValueMatch(current_value='1.4', target_value='1.4', similarity=1.0), ValueMatch(current_value='3.7', target_value='3.7', similarity=1.0), ValueMatch(current_value='2.5', target_value='2.5', similarity=1.0), ValueMatch(current_value='1.8', target_value='1.8', similarity=1.0), ValueMatch(current_value='7.0', target_value='7.0', similarity=1.0), ValueMatch(current_value='2.2', target_value='2.2', similarity=1.0), ValueMatch(current_value='0.9', target_value='0.9', similarity=1.0), ValueMatch(current_value='9.0', target_value='0.9', similarity=0.452), ValueMatch(current_value='9.5', target_value='5.9', similarity=0.428), ValueMatch(current_value='8.0', target_value='8.8', similarity=0.414), ValueMatch(current_value='3.8', target_value='8.8', similarity=0.414), ValueMatch(current_value='0.8', target_value='8.8', similarity=0.414), ValueMatch(current_value='8.5', target_value='8.8', similarity=0.414), ValueMatch(current_value='3.2', target_value='2.2', similarity=0.337), ValueMatch(current_value='2.3', target_value='2.2', similarity=0.337), ValueMatch(current_value='1.2', target_value='2.2', similarity=0.334), ValueMatch(current_value='3.5', target_value='3.3', similarity=0.331), ValueMatch(current_value='5.5', target_value='5.2', similarity=0.329), ValueMatch(current_value='1.3', target_value='3.3', similarity=0.327), ValueMatch(current_value='4.2', target_value='2.2', similarity=0.326), ValueMatch(current_value='2.7', target_value='2.2', similarity=0.323), ValueMatch(current_value='2.9', target_value='2.2', similarity=0.32), ValueMatch(current_value='2.6', target_value='2.2', similarity=0.316), ValueMatch(current_value='3.0', target_value='3.3', similarity=0.312), ValueMatch(current_value='4.3', target_value='3.3', similarity=0.301), ValueMatch(current_value='6.3', target_value='6.0', similarity=0.297), ValueMatch(current_value='6.5', target_value='6.0', similarity=0.297), ValueMatch(current_value='3.9', target_value='3.3', similarity=0.296), ValueMatch(current_value='1.7', target_value='7.0', similarity=0.26), ValueMatch(current_value='4.7', target_value='7.0', similarity=0.255)]\n",
      "source: Tumor_Size_cm target: size_extraocular_nodule score: 16.834\n",
      "value_matches:  [ValueMatch(current_value='2.5', target_value='2.5', similarity=1.0), ValueMatch(current_value='2.7', target_value='2.72', similarity=0.683), ValueMatch(current_value='3.9', target_value='2.39', similarity=0.654), ValueMatch(current_value='2.2', target_value='2.28', similarity=0.652), ValueMatch(current_value='8.5', target_value='1.85', similarity=0.648), ValueMatch(current_value='1.8', target_value='1.85', similarity=0.648), ValueMatch(current_value='6.3', target_value='1.63', similarity=0.639), ValueMatch(current_value='9.5', target_value='1.95', similarity=0.632), ValueMatch(current_value='2.6', target_value='2.61', similarity=0.631), ValueMatch(current_value='1.5', target_value='1.52', similarity=0.63), ValueMatch(current_value='1.7', target_value='2.17', similarity=0.621), ValueMatch(current_value='3.0', target_value='3.05', similarity=0.618), ValueMatch(current_value='5.2', target_value='1.52', similarity=0.618), ValueMatch(current_value='2.9', target_value='2.94', similarity=0.613), ValueMatch(current_value='2.3', target_value='2.39', similarity=0.605), ValueMatch(current_value='4.7', target_value='1.74', similarity=0.321), ValueMatch(current_value='5.5', target_value='2.5', similarity=0.317), ValueMatch(current_value='6.0', target_value='2.06', similarity=0.289), ValueMatch(current_value='3.8', target_value='2.83', similarity=0.277), ValueMatch(current_value='1.4', target_value='1.74', similarity=0.269), ValueMatch(current_value='4.2', target_value='2.94', similarity=0.252)]\n",
      "source: Tumor_Size_cm target: body_surface_area score: 11.617\n",
      "value_matches:  [ValueMatch(current_value='3.5', target_value='3.5', similarity=1.0), ValueMatch(current_value='4.5', target_value='4.5', similarity=1.0), ValueMatch(current_value='4.0', target_value='4.0', similarity=1.0), ValueMatch(current_value='6.0', target_value='6.0', similarity=1.0), ValueMatch(current_value='5.2', target_value='5.2', similarity=1.0), ValueMatch(current_value='0.8', target_value='0.8', similarity=1.0), ValueMatch(current_value='1.5', target_value='1.5', similarity=1.0), ValueMatch(current_value='1.2', target_value='1.2', similarity=1.0), ValueMatch(current_value='13.5', target_value='3.5', similarity=0.608), ValueMatch(current_value='3.7', target_value='7.3', similarity=0.442), ValueMatch(current_value='8.0', target_value='0.8', similarity=0.429), ValueMatch(current_value='1.8', target_value='8.1', similarity=0.422), ValueMatch(current_value='4.2', target_value='2.4', similarity=0.4), ValueMatch(current_value='1.3', target_value='3.1', similarity=0.394), ValueMatch(current_value='2.5', target_value='5.2', similarity=0.367), ValueMatch(current_value='5.0', target_value='0.5', similarity=0.339), ValueMatch(current_value='6.5', target_value='6.0', similarity=0.324), ValueMatch(current_value='2.2', target_value='5.2', similarity=0.323), ValueMatch(current_value='2.6', target_value='6.0', similarity=0.322), ValueMatch(current_value='5.5', target_value='3.5', similarity=0.322), ValueMatch(current_value='6.3', target_value='6.0', similarity=0.321), ValueMatch(current_value='7.0', target_value='7.3', similarity=0.284), ValueMatch(current_value='2.7', target_value='7.3', similarity=0.283), ValueMatch(current_value='1.7', target_value='7.3', similarity=0.281), ValueMatch(current_value='4.7', target_value='7.3', similarity=0.274), ValueMatch(current_value='8.5', target_value='0.8', similarity=0.268), ValueMatch(current_value='3.8', target_value='0.8', similarity=0.266), ValueMatch(current_value='4.3', target_value='4.5', similarity=0.251), ValueMatch(current_value='1.4', target_value='4.5', similarity=0.25)]\n",
      "source: Tumor_Size_cm target: tumor_thickness score: 15.170000000000002\n",
      "value_matches:  [ValueMatch(current_value='6.5', target_value='6.5', similarity=1.0), ValueMatch(current_value='1.5', target_value='1.5', similarity=1.0), ValueMatch(current_value='9.5', target_value='9.5', similarity=1.0), ValueMatch(current_value='4.2', target_value='4.2', similarity=1.0), ValueMatch(current_value='1.0', target_value='10.0', similarity=0.691), ValueMatch(current_value='11.0', target_value='10.0', similarity=0.458), ValueMatch(current_value='6.3', target_value='3.6', similarity=0.447), ValueMatch(current_value='3.7', target_value='7.3', similarity=0.422), ValueMatch(current_value='2.9', target_value='9.2', similarity=0.405), ValueMatch(current_value='8.5', target_value='5.8', similarity=0.4), ValueMatch(current_value='1.2', target_value='2.1', similarity=0.371), ValueMatch(current_value='5.5', target_value='1.5', similarity=0.317), ValueMatch(current_value='2.2', target_value='4.2', similarity=0.316), ValueMatch(current_value='2.6', target_value='6.5', similarity=0.299), ValueMatch(current_value='6.0', target_value='6.5', similarity=0.296), ValueMatch(current_value='4.5', target_value='4.2', similarity=0.264), ValueMatch(current_value='4.3', target_value='4.2', similarity=0.262), ValueMatch(current_value='1.4', target_value='4.2', similarity=0.262), ValueMatch(current_value='4.0', target_value='4.2', similarity=0.26), ValueMatch(current_value='13.5', target_value='1.5', similarity=0.259), ValueMatch(current_value='4.7', target_value='4.2', similarity=0.253)]\n",
      "source: Tumor_Size_cm target: rin score: 9.982000000000001\n",
      "value_matches:  [ValueMatch(current_value='4.5', target_value='4.5', similarity=1.0), ValueMatch(current_value='1.2', target_value='1.2', similarity=1.0), ValueMatch(current_value='6.3', target_value='6.3', similarity=1.0), ValueMatch(current_value='7.0', target_value='7.0', similarity=1.0), ValueMatch(current_value='1.5', target_value='11.5', similarity=0.684), ValueMatch(current_value='1.0', target_value='10.1', similarity=0.672), ValueMatch(current_value='2.7', target_value='12.7', similarity=0.667), ValueMatch(current_value='3.9', target_value='13.9', similarity=0.665), ValueMatch(current_value='1.3', target_value='13.9', similarity=0.592), ValueMatch(current_value='11.0', target_value='10.1', similarity=0.514), ValueMatch(current_value='8.5', target_value='5.8', similarity=0.418), ValueMatch(current_value='4.2', target_value='2.4', similarity=0.405), ValueMatch(current_value='13.5', target_value='13.9', similarity=0.361), ValueMatch(current_value='5.0', target_value='0.5', similarity=0.352), ValueMatch(current_value='2.2', target_value='1.2', similarity=0.35), ValueMatch(current_value='6.5', target_value='6.3', similarity=0.322), ValueMatch(current_value='2.6', target_value='6.3', similarity=0.32), ValueMatch(current_value='6.0', target_value='6.3', similarity=0.32), ValueMatch(current_value='5.5', target_value='4.5', similarity=0.297), ValueMatch(current_value='1.7', target_value='12.7', similarity=0.292), ValueMatch(current_value='3.7', target_value='7.0', similarity=0.28), ValueMatch(current_value='4.7', target_value='7.0', similarity=0.271), ValueMatch(current_value='1.8', target_value='5.8', similarity=0.269), ValueMatch(current_value='3.8', target_value='5.8', similarity=0.268), ValueMatch(current_value='8.0', target_value='5.8', similarity=0.267), ValueMatch(current_value='0.8', target_value='5.8', similarity=0.267), ValueMatch(current_value='9.5', target_value='9.4', similarity=0.259), ValueMatch(current_value='0.9', target_value='9.4', similarity=0.257), ValueMatch(current_value='2.9', target_value='9.4', similarity=0.257), ValueMatch(current_value='9.0', target_value='9.4', similarity=0.257), ValueMatch(current_value='1.4', target_value='4.5', similarity=0.253), ValueMatch(current_value='4.3', target_value='4.5', similarity=0.252), ValueMatch(current_value='4.0', target_value='4.5', similarity=0.251)]\n",
      "source: Tumor_Size_cm target: greatest_tumor_dimension score: 14.639\n",
      "value_matches:  [ValueMatch(current_value='3.5', target_value='3.5', similarity=1.0), ValueMatch(current_value='6.0', target_value='6.0', similarity=1.0), ValueMatch(current_value='0.8', target_value='0.8', similarity=1.0), ValueMatch(current_value='4.7', target_value='4.7', similarity=1.0), ValueMatch(current_value='1.2', target_value='1.2', similarity=1.0), ValueMatch(current_value='2.6', target_value='12.6', similarity=0.672), ValueMatch(current_value='3.8', target_value='13.8', similarity=0.662), ValueMatch(current_value='11.0', target_value='11.1', similarity=0.634), ValueMatch(current_value='5.0', target_value='15.0', similarity=0.629), ValueMatch(current_value='13.5', target_value='3.5', similarity=0.626), ValueMatch(current_value='1.5', target_value='15.0', similarity=0.626), ValueMatch(current_value='1.3', target_value='13.8', similarity=0.597), ValueMatch(current_value='1.0', target_value='10.5', similarity=0.593), ValueMatch(current_value='8.0', target_value='0.8', similarity=0.434), ValueMatch(current_value='2.7', target_value='7.2', similarity=0.428), ValueMatch(current_value='1.8', target_value='8.8', similarity=0.4), ValueMatch(current_value='8.5', target_value='8.8', similarity=0.398), ValueMatch(current_value='2.2', target_value='1.2', similarity=0.35), ValueMatch(current_value='5.5', target_value='3.5', similarity=0.334), ValueMatch(current_value='6.5', target_value='6.0', similarity=0.3), ValueMatch(current_value='6.3', target_value='6.0', similarity=0.299), ValueMatch(current_value='9.5', target_value='9.4', similarity=0.273), ValueMatch(current_value='9.0', target_value='9.4', similarity=0.273), ValueMatch(current_value='0.9', target_value='9.4', similarity=0.273), ValueMatch(current_value='2.9', target_value='9.4', similarity=0.272), ValueMatch(current_value='3.9', target_value='9.4', similarity=0.272), ValueMatch(current_value='1.7', target_value='4.7', similarity=0.27), ValueMatch(current_value='7.0', target_value='4.7', similarity=0.269), ValueMatch(current_value='3.7', target_value='4.7', similarity=0.267), ValueMatch(current_value='1.4', target_value='4.7', similarity=0.251), ValueMatch(current_value='4.5', target_value='4.7', similarity=0.25), ValueMatch(current_value='4.0', target_value='4.7', similarity=0.25)]\n",
      "source: Tumor_Size_cm target: tumor_largest_dimension_diameter score: 15.901999999999996\n",
      "value_matches:  [ValueMatch(current_value='2.9', target_value='2.9', similarity=1.0), ValueMatch(current_value='6.0', target_value='6.0', similarity=1.0), ValueMatch(current_value='1.2', target_value='1.2', similarity=1.0), ValueMatch(current_value='1.4', target_value='1.4', similarity=1.0), ValueMatch(current_value='1.7', target_value='1.7', similarity=1.0), ValueMatch(current_value='2.5', target_value='2.5', similarity=1.0), ValueMatch(current_value='2.2', target_value='2.2', similarity=1.0), ValueMatch(current_value='4.7', target_value='4.7', similarity=1.0), ValueMatch(current_value='3.8', target_value='3.8', similarity=1.0), ValueMatch(current_value='0.9', target_value='0.9', similarity=1.0), ValueMatch(current_value='9.0', target_value='0.9', similarity=0.447), ValueMatch(current_value='7.0', target_value='0.7', similarity=0.411), ValueMatch(current_value='5.2', target_value='2.5', similarity=0.387), ValueMatch(current_value='4.0', target_value='4.4', similarity=0.381), ValueMatch(current_value='4.2', target_value='4.4', similarity=0.379), ValueMatch(current_value='4.5', target_value='4.4', similarity=0.379), ValueMatch(current_value='4.3', target_value='4.4', similarity=0.377), ValueMatch(current_value='1.5', target_value='5.1', similarity=0.371), ValueMatch(current_value='5.0', target_value='0.5', similarity=0.358), ValueMatch(current_value='3.5', target_value='3.3', similarity=0.343), ValueMatch(current_value='1.3', target_value='3.3', similarity=0.341), ValueMatch(current_value='5.5', target_value='2.5', similarity=0.331), ValueMatch(current_value='3.0', target_value='3.3', similarity=0.325), ValueMatch(current_value='3.2', target_value='2.2', similarity=0.324), ValueMatch(current_value='2.3', target_value='2.2', similarity=0.324), ValueMatch(current_value='6.5', target_value='6.0', similarity=0.32), ValueMatch(current_value='2.6', target_value='6.0', similarity=0.32), ValueMatch(current_value='6.3', target_value='6.0', similarity=0.318), ValueMatch(current_value='2.7', target_value='2.2', similarity=0.312), ValueMatch(current_value='3.7', target_value='3.3', similarity=0.31), ValueMatch(current_value='3.9', target_value='3.3', similarity=0.307), ValueMatch(current_value='0.8', target_value='3.8', similarity=0.297), ValueMatch(current_value='8.0', target_value='3.8', similarity=0.297), ValueMatch(current_value='8.5', target_value='3.8', similarity=0.296), ValueMatch(current_value='1.8', target_value='3.8', similarity=0.294), ValueMatch(current_value='9.5', target_value='0.9', similarity=0.28)]\n",
      "source: Tumor_Size_cm target: shortest_dimension score: 18.829\n",
      "value_matches:  [ValueMatch(current_value='4.5', target_value='4.5', similarity=1.0), ValueMatch(current_value='6.0', target_value='6.0', similarity=1.0), ValueMatch(current_value='4.0', target_value='4.0', similarity=1.0), ValueMatch(current_value='1.2', target_value='1.2', similarity=1.0), ValueMatch(current_value='5.5', target_value='5.5', similarity=1.0), ValueMatch(current_value='1.8', target_value='1.8', similarity=1.0), ValueMatch(current_value='2.5', target_value='2.5', similarity=1.0), ValueMatch(current_value='3.2', target_value='3.2', similarity=1.0), ValueMatch(current_value='0.8', target_value='0.8', similarity=1.0), ValueMatch(current_value='6.3', target_value='3.6', similarity=0.442), ValueMatch(current_value='8.0', target_value='0.8', similarity=0.437), ValueMatch(current_value='2.7', target_value='7.2', similarity=0.437), ValueMatch(current_value='2.3', target_value='3.2', similarity=0.38), ValueMatch(current_value='5.2', target_value='2.5', similarity=0.367), ValueMatch(current_value='1.5', target_value='5.1', similarity=0.367), ValueMatch(current_value='3.0', target_value='0.3', similarity=0.346), ValueMatch(current_value='5.0', target_value='0.5', similarity=0.333), ValueMatch(current_value='2.2', target_value='2.5', similarity=0.323), ValueMatch(current_value='3.5', target_value='5.5', similarity=0.323), ValueMatch(current_value='6.5', target_value='6.0', similarity=0.304), ValueMatch(current_value='2.6', target_value='6.0', similarity=0.303), ValueMatch(current_value='8.5', target_value='5.5', similarity=0.288), ValueMatch(current_value='7.0', target_value='7.2', similarity=0.287), ValueMatch(current_value='3.7', target_value='7.2', similarity=0.283), ValueMatch(current_value='3.8', target_value='0.8', similarity=0.283), ValueMatch(current_value='9.5', target_value='5.5', similarity=0.28), ValueMatch(current_value='1.7', target_value='7.2', similarity=0.28), ValueMatch(current_value='4.7', target_value='7.2', similarity=0.273), ValueMatch(current_value='4.2', target_value='4.0', similarity=0.268), ValueMatch(current_value='4.3', target_value='4.0', similarity=0.267), ValueMatch(current_value='1.4', target_value='4.0', similarity=0.264), ValueMatch(current_value='11.0', target_value='1.2', similarity=0.26)]\n",
      "source: Tumor_Size_cm target: analyte_quantity score: 16.395000000000003\n",
      "value_matches:  [ValueMatch(current_value='3.5', target_value='3.5', similarity=1.0), ValueMatch(current_value='9.0', target_value='9.0', similarity=1.0), ValueMatch(current_value='0.8', target_value='0.8', similarity=1.0), ValueMatch(current_value='1.2', target_value='1.2', similarity=1.0), ValueMatch(current_value='4.3', target_value='14.3', similarity=0.648), ValueMatch(current_value='1.4', target_value='14.3', similarity=0.641), ValueMatch(current_value='13.5', target_value='3.5', similarity=0.615), ValueMatch(current_value='1.5', target_value='15.7', similarity=0.598), ValueMatch(current_value='1.0', target_value='10.5', similarity=0.596), ValueMatch(current_value='11.0', target_value='11.2', similarity=0.486), ValueMatch(current_value='0.9', target_value='9.0', similarity=0.452), ValueMatch(current_value='8.0', target_value='0.8', similarity=0.44), ValueMatch(current_value='3.8', target_value='8.3', similarity=0.416), ValueMatch(current_value='4.2', target_value='2.4', similarity=0.405), ValueMatch(current_value='2.2', target_value='1.2', similarity=0.362), ValueMatch(current_value='5.5', target_value='3.5', similarity=0.334), ValueMatch(current_value='9.5', target_value='9.0', similarity=0.28), ValueMatch(current_value='3.9', target_value='9.0', similarity=0.279), ValueMatch(current_value='2.9', target_value='9.0', similarity=0.279), ValueMatch(current_value='1.8', target_value='12.8', similarity=0.278), ValueMatch(current_value='6.5', target_value='7.6', similarity=0.273), ValueMatch(current_value='6.0', target_value='7.6', similarity=0.272), ValueMatch(current_value='6.3', target_value='7.6', similarity=0.272), ValueMatch(current_value='2.6', target_value='7.6', similarity=0.272), ValueMatch(current_value='1.7', target_value='15.7', similarity=0.271), ValueMatch(current_value='8.5', target_value='0.8', similarity=0.264)]\n",
      "source: Tumor_Size_cm target: tumor_depth score: 12.732999999999999\n",
      "value_matches:  [ValueMatch(current_value='5.2', target_value='5.2', similarity=1.0), ValueMatch(current_value='2.5', target_value='2.5', similarity=1.0), ValueMatch(current_value='1.5', target_value='1.5', similarity=1.0), ValueMatch(current_value='3.7', target_value='3.7', similarity=1.0), ValueMatch(current_value='6.3', target_value='6.3', similarity=1.0), ValueMatch(current_value='2.2', target_value='2.2', similarity=1.0), ValueMatch(current_value='3.9', target_value='3.9', similarity=1.0), ValueMatch(current_value='4.7', target_value='7.4', similarity=0.451), ValueMatch(current_value='6.5', target_value='5.6', similarity=0.423), ValueMatch(current_value='1.7', target_value='7.1', similarity=0.422), ValueMatch(current_value='1.4', target_value='4.1', similarity=0.411), ValueMatch(current_value='3.2', target_value='2.2', similarity=0.339), ValueMatch(current_value='2.3', target_value='2.2', similarity=0.339), ValueMatch(current_value='1.2', target_value='2.2', similarity=0.336), ValueMatch(current_value='4.2', target_value='2.2', similarity=0.326), ValueMatch(current_value='2.9', target_value='2.2', similarity=0.323), ValueMatch(current_value='2.7', target_value='2.2', similarity=0.323), ValueMatch(current_value='3.5', target_value='3.3', similarity=0.321), ValueMatch(current_value='2.6', target_value='2.2', similarity=0.32), ValueMatch(current_value='1.3', target_value='3.3', similarity=0.317), ValueMatch(current_value='5.5', target_value='1.5', similarity=0.317), ValueMatch(current_value='3.0', target_value='3.3', similarity=0.295), ValueMatch(current_value='4.3', target_value='3.3', similarity=0.29), ValueMatch(current_value='3.8', target_value='3.3', similarity=0.281), ValueMatch(current_value='6.0', target_value='6.3', similarity=0.276), ValueMatch(current_value='8.5', target_value='6.8', similarity=0.27), ValueMatch(current_value='1.8', target_value='6.8', similarity=0.268), ValueMatch(current_value='9.5', target_value='3.9', similarity=0.266), ValueMatch(current_value='0.8', target_value='6.8', similarity=0.265), ValueMatch(current_value='8.0', target_value='6.8', similarity=0.265), ValueMatch(current_value='7.0', target_value='3.7', similarity=0.26), ValueMatch(current_value='13.5', target_value='1.5', similarity=0.26), ValueMatch(current_value='9.0', target_value='3.9', similarity=0.26), ValueMatch(current_value='0.9', target_value='3.9', similarity=0.26)]\n",
      "source: Tumor_Size_cm target: imaging_suv score: 15.484000000000002\n",
      "value_matches:  [ValueMatch(current_value='0.8', target_value='0.8', similarity=1.0), ValueMatch(current_value='1.2', target_value='1.2', similarity=1.0), ValueMatch(current_value='2.5', target_value='2.5', similarity=1.0), ValueMatch(current_value='1.3', target_value='11.3', similarity=0.689), ValueMatch(current_value='1.0', target_value='10.1', similarity=0.673), ValueMatch(current_value='2.6', target_value='12.6', similarity=0.667), ValueMatch(current_value='4.2', target_value='14.2', similarity=0.648), ValueMatch(current_value='1.4', target_value='14.2', similarity=0.641), ValueMatch(current_value='1.5', target_value='15.8', similarity=0.602), ValueMatch(current_value='11.0', target_value='10.1', similarity=0.516), ValueMatch(current_value='8.0', target_value='0.8', similarity=0.435), ValueMatch(current_value='6.5', target_value='5.6', similarity=0.428), ValueMatch(current_value='13.5', target_value='11.3', similarity=0.422), ValueMatch(current_value='1.7', target_value='7.1', similarity=0.418), ValueMatch(current_value='5.2', target_value='2.5', similarity=0.387), ValueMatch(current_value='2.2', target_value='1.2', similarity=0.35), ValueMatch(current_value='3.5', target_value='3.3', similarity=0.343), ValueMatch(current_value='5.5', target_value='2.5', similarity=0.331), ValueMatch(current_value='2.3', target_value='3.3', similarity=0.323), ValueMatch(current_value='3.2', target_value='3.3', similarity=0.323), ValueMatch(current_value='3.0', target_value='3.3', similarity=0.322), ValueMatch(current_value='4.3', target_value='3.3', similarity=0.313), ValueMatch(current_value='3.8', target_value='3.3', similarity=0.313), ValueMatch(current_value='3.9', target_value='3.3', similarity=0.307), ValueMatch(current_value='6.3', target_value='3.3', similarity=0.307), ValueMatch(current_value='3.7', target_value='3.3', similarity=0.307), ValueMatch(current_value='8.5', target_value='15.8', similarity=0.271), ValueMatch(current_value='2.7', target_value='7.1', similarity=0.267), ValueMatch(current_value='7.0', target_value='7.1', similarity=0.265), ValueMatch(current_value='6.0', target_value='5.6', similarity=0.263), ValueMatch(current_value='1.8', target_value='15.8', similarity=0.261), ValueMatch(current_value='4.7', target_value='7.1', similarity=0.259), ValueMatch(current_value='2.9', target_value='8.9', similarity=0.257), ValueMatch(current_value='9.5', target_value='8.9', similarity=0.257), ValueMatch(current_value='9.0', target_value='8.9', similarity=0.255), ValueMatch(current_value='0.9', target_value='8.9', similarity=0.255)]\n",
      "source: Tumor_Size_cm target: tumor_length_measurement score: 15.675000000000002\n",
      "value_matches:  [ValueMatch(current_value='6.0', target_value='600', similarity=0.729), ValueMatch(current_value='7.0', target_value='700', similarity=0.725), ValueMatch(current_value='8.0', target_value='800', similarity=0.725), ValueMatch(current_value='4.0', target_value='400', similarity=0.722), ValueMatch(current_value='5.5', target_value='550', similarity=0.721), ValueMatch(current_value='3.0', target_value='300', similarity=0.709), ValueMatch(current_value='6.5', target_value='650', similarity=0.708), ValueMatch(current_value='8.5', target_value='850', similarity=0.703), ValueMatch(current_value='4.5', target_value='450', similarity=0.699), ValueMatch(current_value='1.5', target_value='150', similarity=0.691), ValueMatch(current_value='2.5', target_value='250', similarity=0.685), ValueMatch(current_value='3.5', target_value='350', similarity=0.666), ValueMatch(current_value='5.0', target_value='500', similarity=0.638), ValueMatch(current_value='13.5', target_value='350', similarity=0.398), ValueMatch(current_value='0.8', target_value='800', similarity=0.333), ValueMatch(current_value='1.0', target_value='150', similarity=0.265), ValueMatch(current_value='5.2', target_value='250', similarity=0.253), ValueMatch(current_value='2.2', target_value='250', similarity=0.251)]\n",
      "source: Tumor_Size_cm target: average_insert_size score: 10.621\n",
      "value_matches:  [ValueMatch(current_value='3.5', target_value='3.5', similarity=1.0), ValueMatch(current_value='4.5', target_value='4.5', similarity=1.0), ValueMatch(current_value='5.0', target_value='5.0', similarity=1.0), ValueMatch(current_value='6.0', target_value='6.0', similarity=1.0), ValueMatch(current_value='4.0', target_value='4.0', similarity=1.0), ValueMatch(current_value='5.5', target_value='5.5', similarity=1.0), ValueMatch(current_value='2.5', target_value='2.5', similarity=1.0), ValueMatch(current_value='7.0', target_value='7.0', similarity=1.0), ValueMatch(current_value='1.0', target_value='1.0', similarity=1.0), ValueMatch(current_value='1.5', target_value='1.5', similarity=1.0), ValueMatch(current_value='8.0', target_value='8.0', similarity=1.0), ValueMatch(current_value='3.0', target_value='3.0', similarity=1.0), ValueMatch(current_value='6.5', target_value='6.5', similarity=1.0), ValueMatch(current_value='11.0', target_value='1.0', similarity=0.69), ValueMatch(current_value='13.5', target_value='3.5', similarity=0.602), ValueMatch(current_value='0.8', target_value='8.0', similarity=0.444), ValueMatch(current_value='5.2', target_value='2.5', similarity=0.374), ValueMatch(current_value='2.2', target_value='2.5', similarity=0.365), ValueMatch(current_value='8.5', target_value='8.0', similarity=0.309), ValueMatch(current_value='3.8', target_value='8.0', similarity=0.302), ValueMatch(current_value='6.3', target_value='6.5', similarity=0.302), ValueMatch(current_value='2.6', target_value='6.5', similarity=0.301), ValueMatch(current_value='1.8', target_value='8.0', similarity=0.299), ValueMatch(current_value='3.7', target_value='7.0', similarity=0.284), ValueMatch(current_value='2.7', target_value='7.0', similarity=0.282), ValueMatch(current_value='1.7', target_value='7.0', similarity=0.281), ValueMatch(current_value='4.7', target_value='7.0', similarity=0.274), ValueMatch(current_value='4.3', target_value='4.5', similarity=0.268), ValueMatch(current_value='4.2', target_value='4.5', similarity=0.266), ValueMatch(current_value='1.4', target_value='4.5', similarity=0.264), ValueMatch(current_value='9.5', target_value='5.5', similarity=0.253), ValueMatch(current_value='1.3', target_value='1.0', similarity=0.252)]\n",
      "source: Tumor_Size_cm target: circumferential_resection_margin score: 19.412\n",
      "value_matches:  [ValueMatch(current_value='5.5', target_value='v5.5.5', similarity=0.801), ValueMatch(current_value='6.0', target_value='v6.0.0', similarity=0.584), ValueMatch(current_value='4.0', target_value='v4.0.0', similarity=0.576), ValueMatch(current_value='8.0', target_value='v8.0.0', similarity=0.575), ValueMatch(current_value='1.0', target_value='v1.0.0', similarity=0.54), ValueMatch(current_value='7.0', target_value='v7.0.3', similarity=0.505), ValueMatch(current_value='2.3', target_value='v1.2.3', similarity=0.502), ValueMatch(current_value='1.2', target_value='v1.2.3', similarity=0.484), ValueMatch(current_value='5.0', target_value='v5.0.1', similarity=0.481), ValueMatch(current_value='3.0', target_value='v3.0.5', similarity=0.479), ValueMatch(current_value='11.0', target_value='v1.0.0', similarity=0.348), ValueMatch(current_value='6.3', target_value='v6.6.6', similarity=0.333), ValueMatch(current_value='6.5', target_value='v6.6.6', similarity=0.332), ValueMatch(current_value='2.6', target_value='v6.6.6', similarity=0.332), ValueMatch(current_value='3.7', target_value='v7.7.7', similarity=0.314), ValueMatch(current_value='1.7', target_value='v7.7.7', similarity=0.313), ValueMatch(current_value='2.7', target_value='v7.7.7', similarity=0.313), ValueMatch(current_value='4.7', target_value='v7.7.7', similarity=0.301), ValueMatch(current_value='4.3', target_value='v4.4.4', similarity=0.297), ValueMatch(current_value='4.2', target_value='v4.4.4', similarity=0.296), ValueMatch(current_value='4.5', target_value='v4.4.4', similarity=0.296), ValueMatch(current_value='1.4', target_value='v4.4.4', similarity=0.296), ValueMatch(current_value='0.8', target_value='v8.0.0', similarity=0.288)]\n",
      "source: Tumor_Size_cm target: schema_version score: 9.585999999999997\n",
      "value_matches:  [ValueMatch(current_value='6.5', target_value='6.5', similarity=1.0), ValueMatch(current_value='1.5', target_value='1.5', similarity=1.0), ValueMatch(current_value='1.0', target_value='1.0', similarity=1.0), ValueMatch(current_value='9.5', target_value='9.5', similarity=1.0), ValueMatch(current_value='4.2', target_value='4.2', similarity=1.0), ValueMatch(current_value='11.0', target_value='1.0', similarity=0.677), ValueMatch(current_value='6.3', target_value='3.6', similarity=0.447), ValueMatch(current_value='3.7', target_value='7.3', similarity=0.434), ValueMatch(current_value='8.5', target_value='5.8', similarity=0.406), ValueMatch(current_value='1.2', target_value='2.1', similarity=0.371), ValueMatch(current_value='5.0', target_value='0.5', similarity=0.345), ValueMatch(current_value='2.2', target_value='4.2', similarity=0.316), ValueMatch(current_value='5.5', target_value='1.5', similarity=0.306), ValueMatch(current_value='6.0', target_value='6.5', similarity=0.301), ValueMatch(current_value='2.6', target_value='6.5', similarity=0.3), ValueMatch(current_value='9.0', target_value='9.5', similarity=0.283), ValueMatch(current_value='0.9', target_value='9.5', similarity=0.283), ValueMatch(current_value='2.9', target_value='9.5', similarity=0.282), ValueMatch(current_value='3.9', target_value='9.5', similarity=0.28), ValueMatch(current_value='4.5', target_value='4.2', similarity=0.265), ValueMatch(current_value='7.0', target_value='7.3', similarity=0.264), ValueMatch(current_value='4.0', target_value='4.2', similarity=0.264), ValueMatch(current_value='2.7', target_value='7.3', similarity=0.263), ValueMatch(current_value='1.4', target_value='4.2', similarity=0.262), ValueMatch(current_value='1.7', target_value='7.3', similarity=0.262), ValueMatch(current_value='4.3', target_value='4.2', similarity=0.262), ValueMatch(current_value='13.5', target_value='1.5', similarity=0.254), ValueMatch(current_value='4.7', target_value='7.3', similarity=0.253), ValueMatch(current_value='0.8', target_value='5.8', similarity=0.252), ValueMatch(current_value='8.0', target_value='5.8', similarity=0.252), ValueMatch(current_value='3.8', target_value='5.8', similarity=0.25), ValueMatch(current_value='1.8', target_value='5.8', similarity=0.25)]\n",
      "source: Tumor_Size_cm target: rna_integrity_number score: 13.383999999999999\n",
      "value_matches:  [ValueMatch(current_value='6.0', target_value='6.0.0', similarity=0.743), ValueMatch(current_value='8.0', target_value='8.0.0', similarity=0.743), ValueMatch(current_value='7.0', target_value='7.0.0', similarity=0.74), ValueMatch(current_value='4.0', target_value='4.0.0', similarity=0.732), ValueMatch(current_value='5.0', target_value='5.0.0', similarity=0.723), ValueMatch(current_value='3.0', target_value='3.0.0', similarity=0.723), ValueMatch(current_value='1.0', target_value='1.0.0', similarity=0.715), ValueMatch(current_value='1.2', target_value='1.2.1', similarity=0.693), ValueMatch(current_value='4.2', target_value='4.2.0', similarity=0.647), ValueMatch(current_value='1.4', target_value='3.1.4', similarity=0.63), ValueMatch(current_value='2.5', target_value='2.5.3', similarity=0.628), ValueMatch(current_value='3.2', target_value='5.3.2', similarity=0.628), ValueMatch(current_value='4.5', target_value='7.4.5', similarity=0.61), ValueMatch(current_value='11.0', target_value='1.0.0', similarity=0.481), ValueMatch(current_value='0.8', target_value='8.0.0', similarity=0.372), ValueMatch(current_value='4.7', target_value='7.4.5', similarity=0.31), ValueMatch(current_value='4.3', target_value='3.1.4', similarity=0.257), ValueMatch(current_value='1.3', target_value='3.1.4', similarity=0.25)]\n",
      "source: Tumor_Size_cm target: workflow_version score: 10.624999999999998\n",
      "value_matches:  [ValueMatch(current_value='5.0', target_value='5.0', similarity=1.0), ValueMatch(current_value='1.2', target_value='1.2', similarity=1.0), ValueMatch(current_value='3.8', target_value='3.8', similarity=1.0), ValueMatch(current_value='8.5', target_value='8.5', similarity=1.0), ValueMatch(current_value='6.3', target_value='6.3', similarity=1.0), ValueMatch(current_value='1.4', target_value='14.4', similarity=0.72), ValueMatch(current_value='1.8', target_value='18.5', similarity=0.653), ValueMatch(current_value='6.0', target_value='16.0', similarity=0.644), ValueMatch(current_value='11.0', target_value='11.2', similarity=0.481), ValueMatch(current_value='9.5', target_value='9.9', similarity=0.436), ValueMatch(current_value='9.0', target_value='9.9', similarity=0.436), ValueMatch(current_value='0.9', target_value='9.9', similarity=0.436), ValueMatch(current_value='2.9', target_value='9.9', similarity=0.436), ValueMatch(current_value='3.9', target_value='9.9', similarity=0.432), ValueMatch(current_value='1.7', target_value='7.1', similarity=0.418), ValueMatch(current_value='4.2', target_value='2.4', similarity=0.411), ValueMatch(current_value='2.2', target_value='1.2', similarity=0.35), ValueMatch(current_value='5.5', target_value='5.0', similarity=0.321), ValueMatch(current_value='4.5', target_value='14.4', similarity=0.299), ValueMatch(current_value='4.0', target_value='14.4', similarity=0.299), ValueMatch(current_value='6.5', target_value='6.3', similarity=0.297), ValueMatch(current_value='2.6', target_value='6.3', similarity=0.297), ValueMatch(current_value='4.3', target_value='14.4', similarity=0.296), ValueMatch(current_value='4.7', target_value='14.4', similarity=0.284), ValueMatch(current_value='7.0', target_value='20.7', similarity=0.278), ValueMatch(current_value='2.7', target_value='20.7', similarity=0.278), ValueMatch(current_value='3.7', target_value='7.1', similarity=0.265), ValueMatch(current_value='8.0', target_value='8.5', similarity=0.261), ValueMatch(current_value='0.8', target_value='8.5', similarity=0.261)]\n",
      "source: Tumor_Size_cm target: tumor_burden score: 14.288999999999998\n",
      "value_matches:  [ValueMatch(current_value='5.0', target_value='5.0', similarity=1.0), ValueMatch(current_value='4.0', target_value='4.0', similarity=1.0), ValueMatch(current_value='6.0', target_value='6.0', similarity=1.0), ValueMatch(current_value='8.0', target_value='8.0', similarity=1.0), ValueMatch(current_value='3.0', target_value='3.0', similarity=1.0), ValueMatch(current_value='1.0', target_value='1.0', similarity=1.0), ValueMatch(current_value='2.5', target_value='2.5', similarity=1.0), ValueMatch(current_value='7.0', target_value='7.0', similarity=1.0), ValueMatch(current_value='9.0', target_value='9.0', similarity=1.0), ValueMatch(current_value='1.5', target_value='1.5', similarity=1.0), ValueMatch(current_value='11.0', target_value='1.0', similarity=0.663), ValueMatch(current_value='0.9', target_value='9.0', similarity=0.428), ValueMatch(current_value='0.8', target_value='8.0', similarity=0.428), ValueMatch(current_value='5.2', target_value='2.5', similarity=0.389), ValueMatch(current_value='2.2', target_value='2.5', similarity=0.36), ValueMatch(current_value='5.5', target_value='5.0', similarity=0.338), ValueMatch(current_value='6.5', target_value='6.0', similarity=0.332), ValueMatch(current_value='6.3', target_value='6.0', similarity=0.326), ValueMatch(current_value='2.6', target_value='6.0', similarity=0.326), ValueMatch(current_value='8.5', target_value='8.0', similarity=0.311), ValueMatch(current_value='9.5', target_value='9.0', similarity=0.311), ValueMatch(current_value='1.7', target_value='7.0', similarity=0.308), ValueMatch(current_value='1.8', target_value='8.0', similarity=0.308), ValueMatch(current_value='2.9', target_value='9.0', similarity=0.305), ValueMatch(current_value='2.7', target_value='7.0', similarity=0.305), ValueMatch(current_value='3.9', target_value='9.0', similarity=0.305), ValueMatch(current_value='3.7', target_value='7.0', similarity=0.305), ValueMatch(current_value='3.8', target_value='8.0', similarity=0.305), ValueMatch(current_value='4.7', target_value='7.0', similarity=0.294), ValueMatch(current_value='4.5', target_value='4.0', similarity=0.292), ValueMatch(current_value='1.4', target_value='4.0', similarity=0.29), ValueMatch(current_value='4.3', target_value='4.0', similarity=0.287), ValueMatch(current_value='4.2', target_value='4.0', similarity=0.287), ValueMatch(current_value='13.5', target_value='1.5', similarity=0.259)]\n",
      "source: Tumor_Size_cm target: margin_distance score: 18.062\n",
      "Top k reranked columns:  [('Tumor_Size_cm', 'circumferential_resection_margin', 19.412), ('Tumor_Size_cm', 'shortest_dimension', 18.829), ('Tumor_Size_cm', 'margin_distance', 18.062), ('Tumor_Size_cm', 'tumor_width_measurement', 16.862000000000002), ('Tumor_Size_cm', 'size_extraocular_nodule', 16.834), ('Tumor_Size_cm', 'analyte_quantity', 16.395000000000003), ('Tumor_Size_cm', 'tumor_largest_dimension_diameter', 15.901999999999996), ('Tumor_Size_cm', 'tumor_length_measurement', 15.675000000000002), ('Tumor_Size_cm', 'imaging_suv', 15.484000000000002), ('Tumor_Size_cm', 'tumor_thickness', 15.170000000000002), ('Tumor_Size_cm', 'greatest_tumor_dimension', 14.639), ('Tumor_Size_cm', 'tumor_burden', 14.288999999999998), ('Tumor_Size_cm', 'tumor_depth_measurement', 13.718000000000002), ('Tumor_Size_cm', 'rna_integrity_number', 13.383999999999999), ('Tumor_Size_cm', 'tumor_depth', 12.732999999999999), ('Tumor_Size_cm', 'body_surface_area', 11.617), ('Tumor_Size_cm', 'workflow_version', 10.624999999999998), ('Tumor_Size_cm', 'average_insert_size', 10.621), ('Tumor_Size_cm', 'rin', 9.982000000000001), ('Tumor_Size_cm', 'schema_version', 9.585999999999997)]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Country</td>\n",
       "      <td>race</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>ajcc_pathologic_t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>primary_diagnosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FIGO_stage</td>\n",
       "      <td>figo_stage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BMI</td>\n",
       "      <td>average_base_quality</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Age</td>\n",
       "      <td>age_at_last_exposure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Race</td>\n",
       "      <td>race</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Ethnicity</td>\n",
       "      <td>ethnicity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Gender</td>\n",
       "      <td>primary_site</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Tumor_Focality</td>\n",
       "      <td>tumor_focality</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Tumor_Size_cm</td>\n",
       "      <td>circumferential_resection_margin</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         source                            target\n",
       "0                       Country                              race\n",
       "1   Path_Stage_Primary_Tumor-pT                 ajcc_pathologic_t\n",
       "2               Histologic_type                 primary_diagnosis\n",
       "3                    FIGO_stage                        figo_stage\n",
       "4                           BMI              average_base_quality\n",
       "5                           Age              age_at_last_exposure\n",
       "6                          Race                              race\n",
       "7                     Ethnicity                         ethnicity\n",
       "8                        Gender                      primary_site\n",
       "9                Tumor_Focality                    tumor_focality\n",
       "10                Tumor_Size_cm  circumferential_resection_margin"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "column_mappings = bdi.match_schema(dataset[columns], target=\"gdc\", method=\"two_phase\")\n",
    "column_mappings"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generating a harmonized table\n",
    "\n",
    "After discovering a schema mapping, we can generate a new table (DataFrame) using the new column names from the GDC standard vocabulary.\n",
    "\n",
    "To do so using `bdi-kit`, we can use the function `materialize_mapping()` as follows. Note that the column headers have been renamed to the target schema."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>country_of_birth</th>\n",
       "      <th>history_of_tumor_type</th>\n",
       "      <th>figo_stage</th>\n",
       "      <th>average_base_quality</th>\n",
       "      <th>age_at_diagnosis</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>gender</th>\n",
       "      <th>tumor_focality</th>\n",
       "      <th>tumor_width_measurement</th>\n",
       "      <th>tumor_level_prostate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>United States</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>IA</td>\n",
       "      <td>38.88</td>\n",
       "      <td>64.0</td>\n",
       "      <td>White</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>Female</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>2.9</td>\n",
       "      <td>Anterior endometrium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>United States</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>IA</td>\n",
       "      <td>39.76</td>\n",
       "      <td>58.0</td>\n",
       "      <td>White</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>Female</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>3.5</td>\n",
       "      <td>Posterior endometrium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>United States</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>IA</td>\n",
       "      <td>51.19</td>\n",
       "      <td>50.0</td>\n",
       "      <td>White</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>Female</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>4.5</td>\n",
       "      <td>Other, specify</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>United States</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>IA</td>\n",
       "      <td>32.69</td>\n",
       "      <td>75.0</td>\n",
       "      <td>White</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>Female</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>3.5</td>\n",
       "      <td>Other, specify</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>Ukraine</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>IA</td>\n",
       "      <td>29.40</td>\n",
       "      <td>75.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Female</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>4.2</td>\n",
       "      <td>Other, specify</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>Ukraine</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>II</td>\n",
       "      <td>35.42</td>\n",
       "      <td>74.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Female</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>1.5</td>\n",
       "      <td>Other, specify</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>United States</td>\n",
       "      <td>Serous</td>\n",
       "      <td>II</td>\n",
       "      <td>24.32</td>\n",
       "      <td>85.0</td>\n",
       "      <td>Black or African American</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>Female</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>3.8</td>\n",
       "      <td>Other, specify</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>Ukraine</td>\n",
       "      <td>Serous</td>\n",
       "      <td>IA</td>\n",
       "      <td>34.06</td>\n",
       "      <td>70.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Female</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>5.0</td>\n",
       "      <td>Other, specify</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>Ukraine</td>\n",
       "      <td>Serous</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>104 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    country_of_birth history_of_tumor_type figo_stage  average_base_quality  \\\n",
       "0      United States          Endometrioid         IA                 38.88   \n",
       "1      United States          Endometrioid         IA                 39.76   \n",
       "2      United States          Endometrioid         IA                 51.19   \n",
       "3                NaN        Carcinosarcoma        NaN                   NaN   \n",
       "4      United States          Endometrioid         IA                 32.69   \n",
       "..               ...                   ...        ...                   ...   \n",
       "99           Ukraine          Endometrioid         IA                 29.40   \n",
       "100          Ukraine          Endometrioid         II                 35.42   \n",
       "101    United States                Serous         II                 24.32   \n",
       "102          Ukraine                Serous         IA                 34.06   \n",
       "103          Ukraine                Serous        NaN                   NaN   \n",
       "\n",
       "     age_at_diagnosis                       race               ethnicity  \\\n",
       "0                64.0                      White  Not-Hispanic or Latino   \n",
       "1                58.0                      White  Not-Hispanic or Latino   \n",
       "2                50.0                      White  Not-Hispanic or Latino   \n",
       "3                 NaN                        NaN                     NaN   \n",
       "4                75.0                      White  Not-Hispanic or Latino   \n",
       "..                ...                        ...                     ...   \n",
       "99               75.0                        NaN                     NaN   \n",
       "100              74.0                        NaN                     NaN   \n",
       "101              85.0  Black or African American  Not-Hispanic or Latino   \n",
       "102              70.0                        NaN                     NaN   \n",
       "103               NaN                        NaN                     NaN   \n",
       "\n",
       "     gender tumor_focality  tumor_width_measurement   tumor_level_prostate  \n",
       "0    Female       Unifocal                      2.9   Anterior endometrium  \n",
       "1    Female       Unifocal                      3.5  Posterior endometrium  \n",
       "2    Female       Unifocal                      4.5         Other, specify  \n",
       "3       NaN            NaN                      NaN                    NaN  \n",
       "4    Female       Unifocal                      3.5         Other, specify  \n",
       "..      ...            ...                      ...                    ...  \n",
       "99   Female       Unifocal                      4.2         Other, specify  \n",
       "100  Female       Unifocal                      1.5         Other, specify  \n",
       "101  Female       Unifocal                      3.8         Other, specify  \n",
       "102  Female       Unifocal                      5.0         Other, specify  \n",
       "103     NaN            NaN                      NaN                    NaN  \n",
       "\n",
       "[104 rows x 11 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bdi.materialize_mapping(dataset, column_mappings)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generating a harmonized table with value mappings\n",
    "\n",
    "`bdi-kit` can also help with translation of the values from the source table to the target standard format.\n",
    "\n",
    "To this end, `bdi-kit` provides the function `match_values()` that automatically creates value mappings for each string column.\n",
    "The output of `match_values()` can be fed to `materialize_mapping()` which materialized the final target using both schema and value mappings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "value_mappings = bdi.match_values(dataset, column_mapping=column_mappings, target=\"gdc\", method=\"tfidf\")\n",
    "bdi.materialize_mapping(dataset, value_mappings)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Verifying and Correcting Automatic Mappings"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Verifying the schema mappings\n",
    "\n",
    "Sometimes the mappings generated automatically may be incorrect or you may to want verify them individually.\n",
    "To verify the suggested column mappings, `bdi-kit` offers additional APIs to visualize the data and make any modifications when necessary. \n",
    "\n",
    "For this example, we will use the column `Histologic_type`. We can start by exploring the columns most similar to `Histologic_type`. \n",
    "\n",
    "For this, we can use the `top_matches()` function. Here, we notice that `primary_diagnosis` could be a potential target column.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Some weights of RobertaModel were not initialized from the model checkpoint at roberta-base and are newly initialized: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']\n",
      "You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.\n",
      "100%|██████████| 1/1 [00:00<00:00, 17.62it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table features extracted from 1 columns\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 734/734 [00:55<00:00, 13.25it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table features extracted from 734 columns\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>disease_type</td>\n",
       "      <td>0.539168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>sample_type</td>\n",
       "      <td>0.530217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>roots</td>\n",
       "      <td>0.525866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>history_of_tumor_type</td>\n",
       "      <td>0.524959</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>additional_pathology_findings</td>\n",
       "      <td>0.517125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>specimen_type</td>\n",
       "      <td>0.511611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>morphologic_architectural_pattern</td>\n",
       "      <td>0.482257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>histone_variant</td>\n",
       "      <td>0.478656</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>viral_hepatitis_serologies</td>\n",
       "      <td>0.471346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>0.471309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>analyte_type_id</td>\n",
       "      <td>0.450211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>composition</td>\n",
       "      <td>0.446170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>described_cases</td>\n",
       "      <td>0.445148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>pathology_details</td>\n",
       "      <td>0.434320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>antigen</td>\n",
       "      <td>0.432611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>pathology_reports</td>\n",
       "      <td>0.429676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>cog_rhabdomyosarcoma_risk_group</td>\n",
       "      <td>0.420320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>biospecimen_type</td>\n",
       "      <td>0.420141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>analyte_type</td>\n",
       "      <td>0.403565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>single_cell_library</td>\n",
       "      <td>0.403199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>laboratory_test</td>\n",
       "      <td>0.399581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>slide_images</td>\n",
       "      <td>0.398477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>tumor_descriptor</td>\n",
       "      <td>0.396314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>sample_type_id</td>\n",
       "      <td>0.391940</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>relationship_primary_diagnosis</td>\n",
       "      <td>0.391895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>icd_10_code</td>\n",
       "      <td>0.388311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>stain_type</td>\n",
       "      <td>0.385822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>histone_family</td>\n",
       "      <td>0.385273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>pathogenicity</td>\n",
       "      <td>0.383606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>contiguous_organ_invaded</td>\n",
       "      <td>0.383036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>aids_risk_factors</td>\n",
       "      <td>0.382602</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>icd_10</td>\n",
       "      <td>0.375303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>biospecimen_anatomic_site</td>\n",
       "      <td>0.372609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>primary_diagnosis</td>\n",
       "      <td>0.371001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>dysplasia_type</td>\n",
       "      <td>0.369152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>tissue_type</td>\n",
       "      <td>0.367687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>gene_symbol</td>\n",
       "      <td>0.366179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>sarcomatoid_present</td>\n",
       "      <td>0.354912</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>papillary_renal_cell_type</td>\n",
       "      <td>0.351789</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>hysterectomy_type</td>\n",
       "      <td>0.348222</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             source                             target  similarity\n",
       "0   Histologic_type                       disease_type    0.539168\n",
       "1   Histologic_type                        sample_type    0.530217\n",
       "2   Histologic_type                              roots    0.525866\n",
       "3   Histologic_type              history_of_tumor_type    0.524959\n",
       "4   Histologic_type      additional_pathology_findings    0.517125\n",
       "5   Histologic_type                      specimen_type    0.511611\n",
       "6   Histologic_type  morphologic_architectural_pattern    0.482257\n",
       "7   Histologic_type                    histone_variant    0.478656\n",
       "8   Histologic_type         viral_hepatitis_serologies    0.471346\n",
       "9   Histologic_type                         chromosome    0.471309\n",
       "10  Histologic_type                    analyte_type_id    0.450211\n",
       "11  Histologic_type                        composition    0.446170\n",
       "12  Histologic_type                    described_cases    0.445148\n",
       "13  Histologic_type                  pathology_details    0.434320\n",
       "14  Histologic_type                            antigen    0.432611\n",
       "15  Histologic_type                  pathology_reports    0.429676\n",
       "16  Histologic_type    cog_rhabdomyosarcoma_risk_group    0.420320\n",
       "17  Histologic_type                   biospecimen_type    0.420141\n",
       "18  Histologic_type                       analyte_type    0.403565\n",
       "19  Histologic_type                single_cell_library    0.403199\n",
       "20  Histologic_type                    laboratory_test    0.399581\n",
       "21  Histologic_type                       slide_images    0.398477\n",
       "22  Histologic_type                   tumor_descriptor    0.396314\n",
       "23  Histologic_type                     sample_type_id    0.391940\n",
       "24  Histologic_type     relationship_primary_diagnosis    0.391895\n",
       "25  Histologic_type                        icd_10_code    0.388311\n",
       "26  Histologic_type                         stain_type    0.385822\n",
       "27  Histologic_type                     histone_family    0.385273\n",
       "28  Histologic_type                      pathogenicity    0.383606\n",
       "29  Histologic_type           contiguous_organ_invaded    0.383036\n",
       "30  Histologic_type                  aids_risk_factors    0.382602\n",
       "31  Histologic_type                             icd_10    0.375303\n",
       "32  Histologic_type          biospecimen_anatomic_site    0.372609\n",
       "33  Histologic_type                  primary_diagnosis    0.371001\n",
       "34  Histologic_type                     dysplasia_type    0.369152\n",
       "35  Histologic_type                        tissue_type    0.367687\n",
       "36  Histologic_type                        gene_symbol    0.366179\n",
       "37  Histologic_type                sarcomatoid_present    0.354912\n",
       "38  Histologic_type          papillary_renal_cell_type    0.351789\n",
       "39  Histologic_type                  hysterectomy_type    0.348222"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hist_type_matches = bdi.top_matches(dataset, columns=[\"Histologic_type\"], target=\"gdc\", top_k=40)\n",
    "hist_type_matches"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Viewing the column domains\n",
    "\n",
    "To verify that `primary_diagnosis` is a good target column, we view and compare the domains of each column using the `preview_domain()` function. For the source table, it returns the list of unique values in the source column. For the GDC target, it returns the list of unique valid values that a column can have.\n",
    "\n",
    "Here we see that the values seem to be related."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>value_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Endometrioid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Carcinosarcoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Serous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Clear cell</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       value_name\n",
       "0    Endometrioid\n",
       "1  Carcinosarcoma\n",
       "2          Serous\n",
       "3      Clear cell"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bdi.preview_domain(dataset, \"Histologic_type\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>value_name</th>\n",
       "      <th>value_description</th>\n",
       "      <th>column_description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Abdominal desmoid</td>\n",
       "      <td>An insidious poorly circumscribed neoplasm ari...</td>\n",
       "      <td>Text term used to describe the patient's histo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Abdominal fibromatosis</td>\n",
       "      <td>An insidious poorly circumscribed neoplasm ari...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Achromic nevus</td>\n",
       "      <td>A benign nevus characterized by the absence of...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Acidophil adenocarcinoma</td>\n",
       "      <td>A malignant epithelial neoplasm of the anterio...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Acidophil adenoma</td>\n",
       "      <td>An epithelial neoplasm of the anterior pituita...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2620</th>\n",
       "      <td>Wolffian duct tumor</td>\n",
       "      <td>An epithelial neoplasm of the female reproduct...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2621</th>\n",
       "      <td>Xanthofibroma</td>\n",
       "      <td>A benign neoplasm composed of fibroblastic spi...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2622</th>\n",
       "      <td>Yolk sac tumor</td>\n",
       "      <td>A non-seminomatous malignant germ cell tumor c...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2623</th>\n",
       "      <td>Unknown</td>\n",
       "      <td>Not known, not observed, not recorded, or refu...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2624</th>\n",
       "      <td>Not Reported</td>\n",
       "      <td>Not provided or available.</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2625 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    value_name  \\\n",
       "0            Abdominal desmoid   \n",
       "1       Abdominal fibromatosis   \n",
       "2               Achromic nevus   \n",
       "3     Acidophil adenocarcinoma   \n",
       "4            Acidophil adenoma   \n",
       "...                        ...   \n",
       "2620       Wolffian duct tumor   \n",
       "2621             Xanthofibroma   \n",
       "2622            Yolk sac tumor   \n",
       "2623                   Unknown   \n",
       "2624              Not Reported   \n",
       "\n",
       "                                      value_description  \\\n",
       "0     An insidious poorly circumscribed neoplasm ari...   \n",
       "1     An insidious poorly circumscribed neoplasm ari...   \n",
       "2     A benign nevus characterized by the absence of...   \n",
       "3     A malignant epithelial neoplasm of the anterio...   \n",
       "4     An epithelial neoplasm of the anterior pituita...   \n",
       "...                                                 ...   \n",
       "2620  An epithelial neoplasm of the female reproduct...   \n",
       "2621  A benign neoplasm composed of fibroblastic spi...   \n",
       "2622  A non-seminomatous malignant germ cell tumor c...   \n",
       "2623  Not known, not observed, not recorded, or refu...   \n",
       "2624                         Not provided or available.   \n",
       "\n",
       "                                     column_description  \n",
       "0     Text term used to describe the patient's histo...  \n",
       "1                                                        \n",
       "2                                                        \n",
       "3                                                        \n",
       "4                                                        \n",
       "...                                                 ...  \n",
       "2620                                                     \n",
       "2621                                                     \n",
       "2622                                                     \n",
       "2623                                                     \n",
       "2624                                                     \n",
       "\n",
       "[2625 rows x 3 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bdi.preview_domain(\"gdc\", \"primary_diagnosis\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since `primary_diagnosis` looks like a correct match for `Histologic_type`, we can modify the `column_mappings` variable directly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Country</td>\n",
       "      <td>country_of_birth</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>primary_diagnosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FIGO_stage</td>\n",
       "      <td>irs_stage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BMI</td>\n",
       "      <td>age_at_diagnosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Age</td>\n",
       "      <td>weight</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Race</td>\n",
       "      <td>race</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Ethnicity</td>\n",
       "      <td>ethnicity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Gender</td>\n",
       "      <td>gender</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Tumor_Focality</td>\n",
       "      <td>tumor_focality</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Tumor_Size_cm</td>\n",
       "      <td>tumor_depth</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Tumor_Site</td>\n",
       "      <td>tumor_shape</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             source             target\n",
       "0           Country   country_of_birth\n",
       "1   Histologic_type  primary_diagnosis\n",
       "2        FIGO_stage          irs_stage\n",
       "3               BMI   age_at_diagnosis\n",
       "4               Age             weight\n",
       "5              Race               race\n",
       "6         Ethnicity          ethnicity\n",
       "7            Gender             gender\n",
       "8    Tumor_Focality     tumor_focality\n",
       "9     Tumor_Size_cm        tumor_depth\n",
       "10       Tumor_Site        tumor_shape"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "column_mappings.loc[column_mappings[\"source\"] == \"Histologic_type\", \"target\"] = \"primary_diagnosis\"\n",
    "column_mappings"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Finding correct value mappings\n",
    "\n",
    "After finding the correct column, we need to find appropriate value mappings. \n",
    "Using `match_values()`, we can inspect what the possible value mappings for this would look like after the harmonization.\n",
    "\n",
    "`bdi-kit` implements multiple methods for value mapping discovery, including:\n",
    "\n",
    " - `edit_distance` - Computes value similarities using Levenstein's edit distance measure.\n",
    " - `tfidf` - A method based on tf-idf importance weighting computed over charcter n-grams.\n",
    " - `embeddings` - Uses BERT word embeddings to compute \"semantic similarity\" between the values.\n",
    "\n",
    "To specify a value mapping approach, we can pass the `method` parameter."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>Carcinosarcoma, NOS</td>\n",
       "      <td>0.848485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Clear cell</td>\n",
       "      <td>Clear cell adenoma</td>\n",
       "      <td>0.714286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>Stromal endometriosis</td>\n",
       "      <td>0.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Serous</td>\n",
       "      <td>Neuronevus</td>\n",
       "      <td>0.625000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source                 target  similarity\n",
       "0  Carcinosarcoma    Carcinosarcoma, NOS    0.848485\n",
       "1      Clear cell     Clear cell adenoma    0.714286\n",
       "2    Endometrioid  Stromal endometriosis    0.666667\n",
       "3          Serous             Neuronevus    0.625000"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bdi.match_values(\n",
    "    dataset, column_mapping=(\"Histologic_type\", \"primary_diagnosis\"), target=\"gdc\", method=\"edit_distance\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>Carcinosarcoma, NOS</td>\n",
       "      <td>0.969</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>Endometrioid adenoma, NOS</td>\n",
       "      <td>0.897</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Clear cell</td>\n",
       "      <td>Clear cell adenoma</td>\n",
       "      <td>0.853</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Serous</td>\n",
       "      <td>Serous carcinoma, NOS</td>\n",
       "      <td>0.755</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source                     target  similarity\n",
       "0  Carcinosarcoma        Carcinosarcoma, NOS       0.969\n",
       "1    Endometrioid  Endometrioid adenoma, NOS       0.897\n",
       "2      Clear cell         Clear cell adenoma       0.853\n",
       "3          Serous      Serous carcinoma, NOS       0.755"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bdi.match_values(\n",
    "    dataset, column_mapping=(\"Histologic_type\", \"primary_diagnosis\"), target=\"gdc\", method=\"tfidf\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>Carcinofibroma</td>\n",
       "      <td>0.919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>Endometrioid cystadenocarcinoma</td>\n",
       "      <td>0.810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Clear cell</td>\n",
       "      <td>Clear cell carcinoma</td>\n",
       "      <td>0.760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Serous</td>\n",
       "      <td>Serous cystoma</td>\n",
       "      <td>0.661</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source                           target  similarity\n",
       "0  Carcinosarcoma                   Carcinofibroma       0.919\n",
       "1    Endometrioid  Endometrioid cystadenocarcinoma       0.810\n",
       "2      Clear cell             Clear cell carcinoma       0.760\n",
       "3          Serous                   Serous cystoma       0.661"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bdi.match_values(\n",
    "    dataset, column_mapping=(\"Histologic_type\", \"primary_diagnosis\"), target=\"gdc\", method=\"embedding\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>Carcinosarcoma, NOS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Clear cell</td>\n",
       "      <td>Clear cell adenocarcinoma, NOS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>Endometrioid carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Serous</td>\n",
       "      <td>Serous cystadenocarcinoma</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source                          target\n",
       "0  Carcinosarcoma             Carcinosarcoma, NOS\n",
       "1      Clear cell  Clear cell adenocarcinoma, NOS\n",
       "2    Endometrioid          Endometrioid carcinoma\n",
       "3          Serous       Serous cystadenocarcinoma"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hist_type_vmap = pd.DataFrame(\n",
    "    columns=[\"source\", \"target\"],\n",
    "    data=[\n",
    "        (\"Carcinosarcoma\", \"Carcinosarcoma, NOS\"),\n",
    "        (\"Clear cell\", \"Clear cell adenocarcinoma, NOS\"),\n",
    "        (\"Endometrioid\", \"Endometrioid carcinoma\"),\n",
    "        (\"Serous\", \"Serous cystadenocarcinoma\"),\n",
    "    ],\n",
    ")\n",
    "hist_type_vmap"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Verifying multiple value mappings at once\n",
    "\n",
    "Besides verifying value mappings individually, you can also do it for all column mappings at once."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>disease_type</td>\n",
       "      <td>0.539168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>sample_type</td>\n",
       "      <td>0.530217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>roots</td>\n",
       "      <td>0.525866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>history_of_tumor_type</td>\n",
       "      <td>0.524959</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>additional_pathology_findings</td>\n",
       "      <td>0.517125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>specimen_type</td>\n",
       "      <td>0.511611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>morphologic_architectural_pattern</td>\n",
       "      <td>0.482257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>histone_variant</td>\n",
       "      <td>0.478656</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>viral_hepatitis_serologies</td>\n",
       "      <td>0.471346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Histologic_type</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>0.471309</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            source                             target  similarity\n",
       "0  Histologic_type                       disease_type    0.539168\n",
       "1  Histologic_type                        sample_type    0.530217\n",
       "2  Histologic_type                              roots    0.525866\n",
       "3  Histologic_type              history_of_tumor_type    0.524959\n",
       "4  Histologic_type      additional_pathology_findings    0.517125\n",
       "5  Histologic_type                      specimen_type    0.511611\n",
       "6  Histologic_type  morphologic_architectural_pattern    0.482257\n",
       "7  Histologic_type                    histone_variant    0.478656\n",
       "8  Histologic_type         viral_hepatitis_serologies    0.471346\n",
       "9  Histologic_type                         chromosome    0.471309"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hist_type_matches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Histologic_type => disease_type (coverage: 75.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Serous</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>0.563</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>Soft Tissue Tumors and Sarcomas, NOS</td>\n",
       "      <td>0.450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Clear cell</td>\n",
       "      <td>Acinar Cell Neoplasms</td>\n",
       "      <td>0.319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source                                 target  similarity\n",
       "0          Serous  Cystic, Mucinous and Serous Neoplasms       0.563\n",
       "1  Carcinosarcoma   Soft Tissue Tumors and Sarcomas, NOS       0.450\n",
       "2      Clear cell                  Acinar Cell Neoplasms       0.319\n",
       "3    Endometrioid                                   None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Histologic_type => sample_type (coverage: 25.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Clear cell</td>\n",
       "      <td>Mononuclear Cells from Bone Marrow Normal</td>\n",
       "      <td>0.502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Serous</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source                                     target  similarity\n",
       "0      Clear cell  Mononuclear Cells from Bone Marrow Normal       0.502\n",
       "1  Carcinosarcoma                                       None         NaN\n",
       "2    Endometrioid                                       None         NaN\n",
       "3          Serous                                       None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Histologic_type => history_of_tumor_type (coverage: 50.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Clear cell</td>\n",
       "      <td>Colorectal Cancer</td>\n",
       "      <td>0.313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>Phenochromocytoma or Paraganglioma</td>\n",
       "      <td>0.309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Serous</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source                              target  similarity\n",
       "0      Clear cell                   Colorectal Cancer       0.313\n",
       "1  Carcinosarcoma  Phenochromocytoma or Paraganglioma       0.309\n",
       "2    Endometrioid                                None         NaN\n",
       "3          Serous                                None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Histologic_type => additional_pathology_findings (coverage: 75.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>Endometriosis</td>\n",
       "      <td>0.746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>Carcinoma in situ</td>\n",
       "      <td>0.643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Clear cell</td>\n",
       "      <td>Diffuse and early nodular diabetic glomerulosc...</td>\n",
       "      <td>0.273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Serous</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source                                             target  \\\n",
       "0    Endometrioid                                      Endometriosis   \n",
       "1  Carcinosarcoma                                  Carcinoma in situ   \n",
       "2      Clear cell  Diffuse and early nodular diabetic glomerulosc...   \n",
       "3          Serous                                               None   \n",
       "\n",
       "   similarity  \n",
       "0       0.746  \n",
       "1       0.643  \n",
       "2       0.273  \n",
       "3         NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Histologic_type => specimen_type (coverage: 75.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Clear cell</td>\n",
       "      <td>Cell</td>\n",
       "      <td>0.615</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Serous</td>\n",
       "      <td>Serum</td>\n",
       "      <td>0.369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>Bone Marrow Components NOS</td>\n",
       "      <td>0.272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source                      target  similarity\n",
       "0      Clear cell                        Cell       0.615\n",
       "1          Serous                       Serum       0.369\n",
       "2  Carcinosarcoma  Bone Marrow Components NOS       0.272\n",
       "3    Endometrioid                        None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Histologic_type => morphologic_architectural_pattern (coverage: 50.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Clear cell</td>\n",
       "      <td>Papillary Renal Cell</td>\n",
       "      <td>0.516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>Papillary, NOS</td>\n",
       "      <td>0.259</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Serous</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source                target  similarity\n",
       "0      Clear cell  Papillary Renal Cell       0.516\n",
       "1  Carcinosarcoma        Papillary, NOS       0.259\n",
       "2    Endometrioid                  None         NaN\n",
       "3          Serous                  None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Histologic_type => histone_variant (coverage: 25.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>0.269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Clear cell</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Serous</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source        target  similarity\n",
       "0    Endometrioid  Not Reported       0.269\n",
       "1  Carcinosarcoma          None         NaN\n",
       "2      Clear cell          None         NaN\n",
       "3          Serous          None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Histologic_type => viral_hepatitis_serologies (coverage: 0.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Serous</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Clear cell</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source target similarity\n",
       "0          Serous   None       None\n",
       "1  Carcinosarcoma   None       None\n",
       "2      Clear cell   None       None\n",
       "3    Endometrioid   None       None"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Histologic_type => chromosome (coverage: 25.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>0.28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Clear cell</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Serous</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source        target  similarity\n",
       "0    Endometrioid  Not Reported        0.28\n",
       "1  Carcinosarcoma          None         NaN\n",
       "2      Clear cell          None         NaN\n",
       "3          Serous          None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Histologic_type => analyte_type_id (coverage: 75.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Serous</td>\n",
       "      <td>S</td>\n",
       "      <td>0.414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>D</td>\n",
       "      <td>0.298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Clear cell</td>\n",
       "      <td>E</td>\n",
       "      <td>0.269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source target  similarity\n",
       "0          Serous      S       0.414\n",
       "1    Endometrioid      D       0.298\n",
       "2      Clear cell      E       0.269\n",
       "3  Carcinosarcoma   None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Histologic_type => composition (coverage: 75.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Clear cell</td>\n",
       "      <td>Cell</td>\n",
       "      <td>0.621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Serous</td>\n",
       "      <td>Serum</td>\n",
       "      <td>0.374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>Bone Marrow Components NOS</td>\n",
       "      <td>0.279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source                      target  similarity\n",
       "0      Clear cell                        Cell       0.621\n",
       "1          Serous                       Serum       0.374\n",
       "2  Carcinosarcoma  Bone Marrow Components NOS       0.279\n",
       "3    Endometrioid                        None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Histologic_type => antigen (coverage: 75.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>Squamous Cell Carcinoma Antigen (SCCA)</td>\n",
       "      <td>0.514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Clear cell</td>\n",
       "      <td>CEA</td>\n",
       "      <td>0.342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Serous</td>\n",
       "      <td>NSE</td>\n",
       "      <td>0.280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source                                  target  similarity\n",
       "0  Carcinosarcoma  Squamous Cell Carcinoma Antigen (SCCA)       0.514\n",
       "1      Clear cell                                     CEA       0.342\n",
       "2          Serous                                     NSE       0.280\n",
       "3    Endometrioid                                    None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Histologic_type => cog_rhabdomyosarcoma_risk_group (coverage: 25.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>Intermediate Risk</td>\n",
       "      <td>0.277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Clear cell</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Serous</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source             target  similarity\n",
       "0    Endometrioid  Intermediate Risk       0.277\n",
       "1  Carcinosarcoma               None         NaN\n",
       "2      Clear cell               None         NaN\n",
       "3          Serous               None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Histologic_type => biospecimen_type (coverage: 75.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Serous</td>\n",
       "      <td>Serum</td>\n",
       "      <td>0.365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>Buccal Mucosa</td>\n",
       "      <td>0.295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Clear cell</td>\n",
       "      <td>Muscle Tissue</td>\n",
       "      <td>0.271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source         target  similarity\n",
       "0          Serous          Serum       0.365\n",
       "1  Carcinosarcoma  Buccal Mucosa       0.295\n",
       "2      Clear cell  Muscle Tissue       0.271\n",
       "3    Endometrioid           None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Histologic_type => analyte_type (coverage: 25.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Clear cell</td>\n",
       "      <td>Nuclei RNA</td>\n",
       "      <td>0.313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Serous</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source      target  similarity\n",
       "0      Clear cell  Nuclei RNA       0.313\n",
       "1  Carcinosarcoma        None         NaN\n",
       "2    Endometrioid        None         NaN\n",
       "3          Serous        None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Histologic_type => single_cell_library (coverage: 25.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>Chromium scATAC v1 Library</td>\n",
       "      <td>0.27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Clear cell</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Serous</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source                      target  similarity\n",
       "0  Carcinosarcoma  Chromium scATAC v1 Library        0.27\n",
       "1    Endometrioid                        None         NaN\n",
       "2      Clear cell                        None         NaN\n",
       "3          Serous                        None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Histologic_type => laboratory_test (coverage: 25.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Clear cell</td>\n",
       "      <td>Cellularity</td>\n",
       "      <td>0.392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Serous</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source       target  similarity\n",
       "0      Clear cell  Cellularity       0.392\n",
       "1  Carcinosarcoma         None         NaN\n",
       "2    Endometrioid         None         NaN\n",
       "3          Serous         None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Histologic_type => tumor_descriptor (coverage: 50.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Clear cell</td>\n",
       "      <td>Not Allowed To Collect</td>\n",
       "      <td>0.382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>NOS</td>\n",
       "      <td>0.302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Serous</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source                  target  similarity\n",
       "0      Clear cell  Not Allowed To Collect       0.382\n",
       "1  Carcinosarcoma                     NOS       0.302\n",
       "2    Endometrioid                    None         NaN\n",
       "3          Serous                    None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Histologic_type => sample_type_id (coverage: 0.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Serous</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Clear cell</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source target similarity\n",
       "0          Serous   None       None\n",
       "1  Carcinosarcoma   None       None\n",
       "2      Clear cell   None       None\n",
       "3    Endometrioid   None       None"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Histologic_type => relationship_primary_diagnosis (coverage: 75.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>Sarcoma</td>\n",
       "      <td>0.697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Clear cell</td>\n",
       "      <td>Basal Cell Cancer</td>\n",
       "      <td>0.461</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>Thyroid Cancer</td>\n",
       "      <td>0.253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Serous</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source             target  similarity\n",
       "0  Carcinosarcoma            Sarcoma       0.697\n",
       "1      Clear cell  Basal Cell Cancer       0.461\n",
       "2    Endometrioid     Thyroid Cancer       0.253\n",
       "3          Serous               None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Histologic_type => stain_type (coverage: 0.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Serous</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Clear cell</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source target similarity\n",
       "0          Serous   None       None\n",
       "1  Carcinosarcoma   None       None\n",
       "2      Clear cell   None       None\n",
       "3    Endometrioid   None       None"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Histologic_type => histone_family (coverage: 25.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>0.258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Clear cell</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Serous</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source        target  similarity\n",
       "0    Endometrioid  Not Reported       0.258\n",
       "1  Carcinosarcoma          None         NaN\n",
       "2      Clear cell          None         NaN\n",
       "3          Serous          None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Histologic_type => pathogenicity (coverage: 25.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Clear cell</td>\n",
       "      <td>Likely Pathogenic</td>\n",
       "      <td>0.252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Serous</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source             target  similarity\n",
       "0      Clear cell  Likely Pathogenic       0.252\n",
       "1  Carcinosarcoma               None         NaN\n",
       "2    Endometrioid               None         NaN\n",
       "3          Serous               None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Histologic_type => contiguous_organ_invaded (coverage: 25.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Clear cell</td>\n",
       "      <td>Small Bowel</td>\n",
       "      <td>0.262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Serous</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source       target  similarity\n",
       "0      Clear cell  Small Bowel       0.262\n",
       "1  Carcinosarcoma         None         NaN\n",
       "2    Endometrioid         None         NaN\n",
       "3          Serous         None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Histologic_type => aids_risk_factors (coverage: 75.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>Coccidioidomycosis</td>\n",
       "      <td>0.377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Clear cell</td>\n",
       "      <td>Salmonella Septicemia</td>\n",
       "      <td>0.305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>Nocardiosis</td>\n",
       "      <td>0.283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Serous</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source                 target  similarity\n",
       "0    Endometrioid     Coccidioidomycosis       0.377\n",
       "1      Clear cell  Salmonella Septicemia       0.305\n",
       "2  Carcinosarcoma            Nocardiosis       0.283\n",
       "3          Serous                   None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Histologic_type => biospecimen_anatomic_site (coverage: 100.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Clear cell</td>\n",
       "      <td>Cell-Line</td>\n",
       "      <td>0.474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>Abdomen</td>\n",
       "      <td>0.376</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Serous</td>\n",
       "      <td>Venous</td>\n",
       "      <td>0.318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>Stomach - Mucosa Only</td>\n",
       "      <td>0.270</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source                 target  similarity\n",
       "0      Clear cell              Cell-Line       0.474\n",
       "1    Endometrioid                Abdomen       0.376\n",
       "2          Serous                 Venous       0.318\n",
       "3  Carcinosarcoma  Stomach - Mucosa Only       0.270"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Histologic_type => primary_diagnosis (coverage: 100.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>Carcinosarcoma, NOS</td>\n",
       "      <td>0.969</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>Endometrioid adenoma, NOS</td>\n",
       "      <td>0.897</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Clear cell</td>\n",
       "      <td>Clear cell adenoma</td>\n",
       "      <td>0.853</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Serous</td>\n",
       "      <td>Serous carcinoma, NOS</td>\n",
       "      <td>0.755</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source                     target  similarity\n",
       "0  Carcinosarcoma        Carcinosarcoma, NOS       0.969\n",
       "1    Endometrioid  Endometrioid adenoma, NOS       0.897\n",
       "2      Clear cell         Clear cell adenoma       0.853\n",
       "3          Serous      Serous carcinoma, NOS       0.755"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mappings = bdi.match_values(\n",
    "    dataset,\n",
    "    column_mapping=hist_type_matches.head(34),\n",
    "    target=\"gdc\",\n",
    "    method=\"tfidf\",\n",
    "    # default_missing=None\n",
    ")\n",
    "\n",
    "for mapping in mappings:\n",
    "    mapping.attrs['default_missing'] = \"___\"\n",
    "    print(f\"{mapping.attrs['source']} => {mapping.attrs['target']} (coverage: {mapping.attrs['coverage']:.2%})\")\n",
    "    display(mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>Carcinosarcoma, NOS</td>\n",
       "      <td>0.969</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>Endometrioid adenoma, NOS</td>\n",
       "      <td>0.897</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Clear cell</td>\n",
       "      <td>Clear cell adenoma</td>\n",
       "      <td>0.853</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Serous</td>\n",
       "      <td>Serous carcinoma, NOS</td>\n",
       "      <td>0.755</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source                     target  similarity\n",
       "0  Carcinosarcoma        Carcinosarcoma, NOS       0.969\n",
       "1    Endometrioid  Endometrioid adenoma, NOS       0.897\n",
       "2      Clear cell         Clear cell adenoma       0.853\n",
       "3          Serous      Serous carcinoma, NOS       0.755"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "scores = [\n",
    "    np.sum(m[\"similarity\"])\n",
    "    for m in mappings\n",
    "]\n",
    "scores\n",
    "sorted_mappings = [m for _, m in sorted(zip(scores, mappings), key=lambda it: it[0], reverse=True)]\n",
    "sorted_mappings[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Some weights of RobertaModel were not initialized from the model checkpoint at roberta-base and are newly initialized: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']\n",
      "You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.\n",
      "100%|██████████| 1/1 [00:00<00:00, 11.98it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table features extracted from 1 columns\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 734/734 [00:52<00:00, 13.92it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table features extracted from 734 columns\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>uicc_pathologic_t</td>\n",
       "      <td>0.677620</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>ajcc_pathologic_t</td>\n",
       "      <td>0.660834</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>ensat_pathologic_t</td>\n",
       "      <td>0.660739</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>uicc_clinical_t</td>\n",
       "      <td>0.645708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>ajcc_clinical_t</td>\n",
       "      <td>0.592968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>extrathyroid_extension</td>\n",
       "      <td>0.589977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>uicc_pathologic_stage</td>\n",
       "      <td>0.584754</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>somatic_mutation_indexes</td>\n",
       "      <td>0.571665</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>ajcc_pathologic_stage</td>\n",
       "      <td>0.563254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>annotated_somatic_mutations</td>\n",
       "      <td>0.563235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>masked_somatic_mutations</td>\n",
       "      <td>0.536071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>figo_stage</td>\n",
       "      <td>0.535757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>tumor_level_prostate</td>\n",
       "      <td>0.525700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>ensat_pathologic_stage</td>\n",
       "      <td>0.524734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>uicc_clinical_stage</td>\n",
       "      <td>0.512399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>tumor_grade_category</td>\n",
       "      <td>0.510672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>ajcc_clinical_stage</td>\n",
       "      <td>0.501770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>inss_stage</td>\n",
       "      <td>0.477118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>adapter_sequence</td>\n",
       "      <td>0.473948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>margin_distance</td>\n",
       "      <td>0.438898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>masaoka_stage</td>\n",
       "      <td>0.434978</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>enneking_msts_tumor_site</td>\n",
       "      <td>0.433908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>tumor_regression_grade</td>\n",
       "      <td>0.421245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>tumor_grade</td>\n",
       "      <td>0.401092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>aggregated_somatic_mutations</td>\n",
       "      <td>0.397975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>fastq_name</td>\n",
       "      <td>0.391604</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>route_of_administration</td>\n",
       "      <td>0.388670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>distance_normal_to_tumor</td>\n",
       "      <td>0.381779</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>target_capture_kit_catalog_number</td>\n",
       "      <td>0.378923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>aa_change</td>\n",
       "      <td>0.376404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>submitted_aligned_reads_files</td>\n",
       "      <td>0.375468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>shortest_dimension</td>\n",
       "      <td>0.371787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>submitted_genomic_profiles</td>\n",
       "      <td>0.371026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>best_overall_response</td>\n",
       "      <td>0.366447</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>data_subtypes</td>\n",
       "      <td>0.363905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>ann_arbor_pathologic_stage</td>\n",
       "      <td>0.361731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>zone_of_origin_prostate</td>\n",
       "      <td>0.360281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>simple_germline_variations</td>\n",
       "      <td>0.358423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>adverse_event_grade</td>\n",
       "      <td>0.357383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>Path_Stage_Primary_Tumor-pT</td>\n",
       "      <td>simple_somatic_mutations</td>\n",
       "      <td>0.353099</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         source                             target  similarity\n",
       "0   Path_Stage_Primary_Tumor-pT                  uicc_pathologic_t    0.677620\n",
       "1   Path_Stage_Primary_Tumor-pT                  ajcc_pathologic_t    0.660834\n",
       "2   Path_Stage_Primary_Tumor-pT                 ensat_pathologic_t    0.660739\n",
       "3   Path_Stage_Primary_Tumor-pT                    uicc_clinical_t    0.645708\n",
       "4   Path_Stage_Primary_Tumor-pT                    ajcc_clinical_t    0.592968\n",
       "5   Path_Stage_Primary_Tumor-pT             extrathyroid_extension    0.589977\n",
       "6   Path_Stage_Primary_Tumor-pT              uicc_pathologic_stage    0.584754\n",
       "7   Path_Stage_Primary_Tumor-pT           somatic_mutation_indexes    0.571665\n",
       "8   Path_Stage_Primary_Tumor-pT              ajcc_pathologic_stage    0.563254\n",
       "9   Path_Stage_Primary_Tumor-pT        annotated_somatic_mutations    0.563235\n",
       "10  Path_Stage_Primary_Tumor-pT           masked_somatic_mutations    0.536071\n",
       "11  Path_Stage_Primary_Tumor-pT                         figo_stage    0.535757\n",
       "12  Path_Stage_Primary_Tumor-pT               tumor_level_prostate    0.525700\n",
       "13  Path_Stage_Primary_Tumor-pT             ensat_pathologic_stage    0.524734\n",
       "14  Path_Stage_Primary_Tumor-pT                uicc_clinical_stage    0.512399\n",
       "15  Path_Stage_Primary_Tumor-pT               tumor_grade_category    0.510672\n",
       "16  Path_Stage_Primary_Tumor-pT                ajcc_clinical_stage    0.501770\n",
       "17  Path_Stage_Primary_Tumor-pT                         inss_stage    0.477118\n",
       "18  Path_Stage_Primary_Tumor-pT                   adapter_sequence    0.473948\n",
       "19  Path_Stage_Primary_Tumor-pT                    margin_distance    0.438898\n",
       "20  Path_Stage_Primary_Tumor-pT                      masaoka_stage    0.434978\n",
       "21  Path_Stage_Primary_Tumor-pT           enneking_msts_tumor_site    0.433908\n",
       "22  Path_Stage_Primary_Tumor-pT             tumor_regression_grade    0.421245\n",
       "23  Path_Stage_Primary_Tumor-pT                        tumor_grade    0.401092\n",
       "24  Path_Stage_Primary_Tumor-pT       aggregated_somatic_mutations    0.397975\n",
       "25  Path_Stage_Primary_Tumor-pT                         fastq_name    0.391604\n",
       "26  Path_Stage_Primary_Tumor-pT            route_of_administration    0.388670\n",
       "27  Path_Stage_Primary_Tumor-pT           distance_normal_to_tumor    0.381779\n",
       "28  Path_Stage_Primary_Tumor-pT  target_capture_kit_catalog_number    0.378923\n",
       "29  Path_Stage_Primary_Tumor-pT                          aa_change    0.376404\n",
       "30  Path_Stage_Primary_Tumor-pT      submitted_aligned_reads_files    0.375468\n",
       "31  Path_Stage_Primary_Tumor-pT                 shortest_dimension    0.371787\n",
       "32  Path_Stage_Primary_Tumor-pT         submitted_genomic_profiles    0.371026\n",
       "33  Path_Stage_Primary_Tumor-pT              best_overall_response    0.366447\n",
       "34  Path_Stage_Primary_Tumor-pT                      data_subtypes    0.363905\n",
       "35  Path_Stage_Primary_Tumor-pT         ann_arbor_pathologic_stage    0.361731\n",
       "36  Path_Stage_Primary_Tumor-pT            zone_of_origin_prostate    0.360281\n",
       "37  Path_Stage_Primary_Tumor-pT         simple_germline_variations    0.358423\n",
       "38  Path_Stage_Primary_Tumor-pT                adverse_event_grade    0.357383\n",
       "39  Path_Stage_Primary_Tumor-pT           simple_somatic_mutations    0.353099"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path_stage_matches = bdi.top_matches(dataset, columns=[\"Path_Stage_Primary_Tumor-pT\"], target=\"gdc\", top_k=40)\n",
    "path_stage_matches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Path_Stage_Primary_Tumor-pT => uicc_pathologic_t (coverage: 100.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pT1b (FIGO IB)</td>\n",
       "      <td>T1b</td>\n",
       "      <td>0.534</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>T1a</td>\n",
       "      <td>0.509</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pT3b (FIGO IIIB)</td>\n",
       "      <td>T3b</td>\n",
       "      <td>0.436</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pT3a (FIGO IIIA)</td>\n",
       "      <td>T3a</td>\n",
       "      <td>0.418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>nan</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>pT1 (FIGO I)</td>\n",
       "      <td>T1</td>\n",
       "      <td>0.289</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>pT2 (FIGO II)</td>\n",
       "      <td>T2</td>\n",
       "      <td>0.281</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             source   target  similarity\n",
       "0    pT1b (FIGO IB)      T1b       0.534\n",
       "1    pT1a (FIGO IA)      T1a       0.509\n",
       "2  pT3b (FIGO IIIB)      T3b       0.436\n",
       "3  pT3a (FIGO IIIA)      T3a       0.418\n",
       "4               nan  Unknown       0.373\n",
       "5      pT1 (FIGO I)       T1       0.289\n",
       "6     pT2 (FIGO II)       T2       0.281"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Path_Stage_Primary_Tumor-pT => ajcc_pathologic_t (coverage: 100.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pT1b (FIGO IB)</td>\n",
       "      <td>T1b</td>\n",
       "      <td>0.535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>T1a</td>\n",
       "      <td>0.507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pT3b (FIGO IIIB)</td>\n",
       "      <td>T3b</td>\n",
       "      <td>0.437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pT3a (FIGO IIIA)</td>\n",
       "      <td>T3a</td>\n",
       "      <td>0.416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>nan</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>pT1 (FIGO I)</td>\n",
       "      <td>T1</td>\n",
       "      <td>0.290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>pT2 (FIGO II)</td>\n",
       "      <td>T2</td>\n",
       "      <td>0.281</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             source   target  similarity\n",
       "0    pT1b (FIGO IB)      T1b       0.535\n",
       "1    pT1a (FIGO IA)      T1a       0.507\n",
       "2  pT3b (FIGO IIIB)      T3b       0.437\n",
       "3  pT3a (FIGO IIIA)      T3a       0.416\n",
       "4               nan  Unknown       0.348\n",
       "5      pT1 (FIGO I)       T1       0.290\n",
       "6     pT2 (FIGO II)       T2       0.281"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Path_Stage_Primary_Tumor-pT => ensat_pathologic_t (coverage: 85.71%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pT1 (FIGO I)</td>\n",
       "      <td>T1</td>\n",
       "      <td>0.424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pT2 (FIGO II)</td>\n",
       "      <td>T2</td>\n",
       "      <td>0.424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>T1</td>\n",
       "      <td>0.311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pT1b (FIGO IB)</td>\n",
       "      <td>T1</td>\n",
       "      <td>0.302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pT3a (FIGO IIIA)</td>\n",
       "      <td>T3</td>\n",
       "      <td>0.260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>pT3b (FIGO IIIB)</td>\n",
       "      <td>T3</td>\n",
       "      <td>0.256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>nan</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             source target  similarity\n",
       "0      pT1 (FIGO I)     T1       0.424\n",
       "1     pT2 (FIGO II)     T2       0.424\n",
       "2    pT1a (FIGO IA)     T1       0.311\n",
       "3    pT1b (FIGO IB)     T1       0.302\n",
       "4  pT3a (FIGO IIIA)     T3       0.260\n",
       "5  pT3b (FIGO IIIB)     T3       0.256\n",
       "6               nan   None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Path_Stage_Primary_Tumor-pT => uicc_clinical_t (coverage: 100.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pT1b (FIGO IB)</td>\n",
       "      <td>T1b</td>\n",
       "      <td>0.536</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>T1a</td>\n",
       "      <td>0.512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pT3b (FIGO IIIB)</td>\n",
       "      <td>T3b</td>\n",
       "      <td>0.437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pT3a (FIGO IIIA)</td>\n",
       "      <td>T3a</td>\n",
       "      <td>0.418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>nan</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>pT1 (FIGO I)</td>\n",
       "      <td>T1</td>\n",
       "      <td>0.297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>pT2 (FIGO II)</td>\n",
       "      <td>T2</td>\n",
       "      <td>0.285</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             source   target  similarity\n",
       "0    pT1b (FIGO IB)      T1b       0.536\n",
       "1    pT1a (FIGO IA)      T1a       0.512\n",
       "2  pT3b (FIGO IIIB)      T3b       0.437\n",
       "3  pT3a (FIGO IIIA)      T3a       0.418\n",
       "4               nan  Unknown       0.373\n",
       "5      pT1 (FIGO I)       T1       0.297\n",
       "6     pT2 (FIGO II)       T2       0.285"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Path_Stage_Primary_Tumor-pT => ajcc_clinical_t (coverage: 100.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pT1b (FIGO IB)</td>\n",
       "      <td>T1b</td>\n",
       "      <td>0.537</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>T1a</td>\n",
       "      <td>0.509</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pT3b (FIGO IIIB)</td>\n",
       "      <td>T3b</td>\n",
       "      <td>0.437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pT3a (FIGO IIIA)</td>\n",
       "      <td>T3a</td>\n",
       "      <td>0.416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>nan</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>pT1 (FIGO I)</td>\n",
       "      <td>T1</td>\n",
       "      <td>0.298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>pT2 (FIGO II)</td>\n",
       "      <td>T2</td>\n",
       "      <td>0.285</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             source   target  similarity\n",
       "0    pT1b (FIGO IB)      T1b       0.537\n",
       "1    pT1a (FIGO IA)      T1a       0.509\n",
       "2  pT3b (FIGO IIIB)      T3b       0.437\n",
       "3  pT3a (FIGO IIIA)      T3a       0.416\n",
       "4               nan  Unknown       0.348\n",
       "5      pT1 (FIGO I)       T1       0.298\n",
       "6     pT2 (FIGO II)       T2       0.285"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Path_Stage_Primary_Tumor-pT => extrathyroid_extension (coverage: 0.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pT1b (FIGO IB)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pT3b (FIGO IIIB)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>nan</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pT1 (FIGO I)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pT3a (FIGO IIIA)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>pT2 (FIGO II)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             source target similarity\n",
       "0    pT1b (FIGO IB)   None       None\n",
       "1  pT3b (FIGO IIIB)   None       None\n",
       "2               nan   None       None\n",
       "3      pT1 (FIGO I)   None       None\n",
       "4  pT3a (FIGO IIIA)   None       None\n",
       "5    pT1a (FIGO IA)   None       None\n",
       "6     pT2 (FIGO II)   None       None"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Path_Stage_Primary_Tumor-pT => uicc_pathologic_stage (coverage: 71.43%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pT3b (FIGO IIIB)</td>\n",
       "      <td>Stage IIIB</td>\n",
       "      <td>0.532</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pT3a (FIGO IIIA)</td>\n",
       "      <td>Stage IIIA</td>\n",
       "      <td>0.478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>nan</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.379</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pT1b (FIGO IB)</td>\n",
       "      <td>Stage IB</td>\n",
       "      <td>0.339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pT2 (FIGO II)</td>\n",
       "      <td>Stage III</td>\n",
       "      <td>0.265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>pT1 (FIGO I)</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             source      target  similarity\n",
       "0  pT3b (FIGO IIIB)  Stage IIIB       0.532\n",
       "1  pT3a (FIGO IIIA)  Stage IIIA       0.478\n",
       "2               nan     Unknown       0.379\n",
       "3    pT1b (FIGO IB)    Stage IB       0.339\n",
       "4     pT2 (FIGO II)   Stage III       0.265\n",
       "5    pT1a (FIGO IA)        None         NaN\n",
       "6      pT1 (FIGO I)        None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Path_Stage_Primary_Tumor-pT => ajcc_pathologic_stage (coverage: 71.43%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pT3b (FIGO IIIB)</td>\n",
       "      <td>Stage IIIB</td>\n",
       "      <td>0.532</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pT3a (FIGO IIIA)</td>\n",
       "      <td>Stage IIIA</td>\n",
       "      <td>0.478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>nan</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.379</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pT1b (FIGO IB)</td>\n",
       "      <td>Stage IB</td>\n",
       "      <td>0.339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pT2 (FIGO II)</td>\n",
       "      <td>Stage III</td>\n",
       "      <td>0.265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>pT1 (FIGO I)</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             source      target  similarity\n",
       "0  pT3b (FIGO IIIB)  Stage IIIB       0.532\n",
       "1  pT3a (FIGO IIIA)  Stage IIIA       0.478\n",
       "2               nan     Unknown       0.379\n",
       "3    pT1b (FIGO IB)    Stage IB       0.339\n",
       "4     pT2 (FIGO II)   Stage III       0.265\n",
       "5    pT1a (FIGO IA)        None         NaN\n",
       "6      pT1 (FIGO I)        None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mappings = bdi.match_values(\n",
    "    dataset,\n",
    "    column_mapping=path_stage_matches.head(10),\n",
    "    target=\"gdc\",\n",
    "    method=\"tfidf\",\n",
    "    # default_missing=None\n",
    ")\n",
    "\n",
    "for mapping in mappings:\n",
    "    mapping.attrs['default_missing'] = \"___\"\n",
    "    print(f\"{mapping.attrs['source']} => {mapping.attrs['target']} (coverage: {mapping.attrs['coverage']:.2%})\")\n",
    "    display(mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'mappings' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[8], line 4\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mnumpy\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01mnp\u001b[39;00m\n\u001b[1;32m      2\u001b[0m scores \u001b[38;5;241m=\u001b[39m [\n\u001b[1;32m      3\u001b[0m     np\u001b[38;5;241m.\u001b[39msum(m[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msimilarity\u001b[39m\u001b[38;5;124m\"\u001b[39m])\n\u001b[0;32m----> 4\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m m \u001b[38;5;129;01min\u001b[39;00m \u001b[43mmappings\u001b[49m\n\u001b[1;32m      5\u001b[0m ]\n\u001b[1;32m      6\u001b[0m scores\n\u001b[1;32m      7\u001b[0m sorted_mappings \u001b[38;5;241m=\u001b[39m [m \u001b[38;5;28;01mfor\u001b[39;00m _, m \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28msorted\u001b[39m(\u001b[38;5;28mzip\u001b[39m(scores, mappings), key\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mlambda\u001b[39;00m it: it[\u001b[38;5;241m0\u001b[39m], reverse\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)]\n",
      "\u001b[0;31mNameError\u001b[0m: name 'mappings' is not defined"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "scores = [\n",
    "    np.sum(m[\"similarity\"])\n",
    "    for m in mappings\n",
    "]\n",
    "scores\n",
    "sorted_mappings = [m for _, m in sorted(zip(scores, mappings), key=lambda it: it[0], reverse=True)]\n",
    "print(scores[:2])\n",
    "print(sorted_mappings[1].attrs[\"target\"])\n",
    "sorted_mappings[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Some weights of RobertaModel were not initialized from the model checkpoint at roberta-base and are newly initialized: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']\n",
      "You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.\n",
      "100%|██████████| 1/1 [00:00<00:00, 17.54it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table features extracted from 1 columns\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 734/734 [00:48<00:00, 15.13it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table features extracted from 734 columns\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Country</td>\n",
       "      <td>country_of_birth</td>\n",
       "      <td>0.491808</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Country</td>\n",
       "      <td>country_of_residence_at_enrollment</td>\n",
       "      <td>0.419452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Country</td>\n",
       "      <td>oct_embedded</td>\n",
       "      <td>0.371549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Country</td>\n",
       "      <td>race</td>\n",
       "      <td>0.352190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Country</td>\n",
       "      <td>submission_enabled</td>\n",
       "      <td>0.325194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Country</td>\n",
       "      <td>zone_of_origin_prostate</td>\n",
       "      <td>0.281882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Country</td>\n",
       "      <td>is_legacy</td>\n",
       "      <td>0.279952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Country</td>\n",
       "      <td>released</td>\n",
       "      <td>0.271991</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Country</td>\n",
       "      <td>non_nodal_regional_disease</td>\n",
       "      <td>0.268099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Country</td>\n",
       "      <td>perineural_invasion_present</td>\n",
       "      <td>0.225525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Country</td>\n",
       "      <td>ethnicity</td>\n",
       "      <td>0.217912</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Country</td>\n",
       "      <td>consent_type</td>\n",
       "      <td>0.210355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Country</td>\n",
       "      <td>request_submission</td>\n",
       "      <td>0.208036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Country</td>\n",
       "      <td>releasable</td>\n",
       "      <td>0.207206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Country</td>\n",
       "      <td>project_id</td>\n",
       "      <td>0.204526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Country</td>\n",
       "      <td>overrepresented_sequences</td>\n",
       "      <td>0.203413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Country</td>\n",
       "      <td>year_of_diagnosis</td>\n",
       "      <td>0.192407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Country</td>\n",
       "      <td>status</td>\n",
       "      <td>0.189337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Country</td>\n",
       "      <td>ulceration_indicator</td>\n",
       "      <td>0.188119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Country</td>\n",
       "      <td>vascular_invasion_present</td>\n",
       "      <td>0.188106</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     source                              target  similarity\n",
       "0   Country                    country_of_birth    0.491808\n",
       "1   Country  country_of_residence_at_enrollment    0.419452\n",
       "2   Country                        oct_embedded    0.371549\n",
       "3   Country                                race    0.352190\n",
       "4   Country                  submission_enabled    0.325194\n",
       "5   Country             zone_of_origin_prostate    0.281882\n",
       "6   Country                           is_legacy    0.279952\n",
       "7   Country                            released    0.271991\n",
       "8   Country          non_nodal_regional_disease    0.268099\n",
       "9   Country         perineural_invasion_present    0.225525\n",
       "10  Country                           ethnicity    0.217912\n",
       "11  Country                        consent_type    0.210355\n",
       "12  Country                  request_submission    0.208036\n",
       "13  Country                          releasable    0.207206\n",
       "14  Country                          project_id    0.204526\n",
       "15  Country           overrepresented_sequences    0.203413\n",
       "16  Country                   year_of_diagnosis    0.192407\n",
       "17  Country                              status    0.189337\n",
       "18  Country                ulceration_indicator    0.188119\n",
       "19  Country           vascular_invasion_present    0.188106"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "country_matches = bdi.top_matches(dataset, columns=[\"Country\"], target=\"gdc\", top_k=20)\n",
    "country_matches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Country => country_of_birth (coverage: 60.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>United States</td>\n",
       "      <td>United States</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Poland</td>\n",
       "      <td>Poland</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ukraine</td>\n",
       "      <td>Ukraine</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>nan</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Other_specify</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          source         target  similarity\n",
       "0  United States  United States         1.0\n",
       "1         Poland         Poland         1.0\n",
       "2        Ukraine        Ukraine         1.0\n",
       "3            nan           None         NaN\n",
       "4  Other_specify           None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Country => country_of_residence_at_enrollment (coverage: 60.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>United States</td>\n",
       "      <td>United States</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Poland</td>\n",
       "      <td>Poland</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ukraine</td>\n",
       "      <td>Ukraine</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>nan</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Other_specify</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          source         target  similarity\n",
       "0  United States  United States         1.0\n",
       "1         Poland         Poland         1.0\n",
       "2        Ukraine        Ukraine         1.0\n",
       "3            nan           None         NaN\n",
       "4  Other_specify           None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Country => race (coverage: 80.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Other_specify</td>\n",
       "      <td>other</td>\n",
       "      <td>0.502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nan</td>\n",
       "      <td>american indian or alaska native</td>\n",
       "      <td>0.344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>United States</td>\n",
       "      <td>white</td>\n",
       "      <td>0.297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Poland</td>\n",
       "      <td>native hawaiian or other pacific islander</td>\n",
       "      <td>0.274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Ukraine</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          source                                     target  similarity\n",
       "0  Other_specify                                      other       0.502\n",
       "1            nan           american indian or alaska native       0.344\n",
       "2  United States                                      white       0.297\n",
       "3         Poland  native hawaiian or other pacific islander       0.274\n",
       "4        Ukraine                                       None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Country => zone_of_origin_prostate (coverage: 20.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Other_specify</td>\n",
       "      <td>Peripheral zone</td>\n",
       "      <td>0.288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nan</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ukraine</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Poland</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>United States</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          source           target  similarity\n",
       "0  Other_specify  Peripheral zone       0.288\n",
       "1            nan             None         NaN\n",
       "2        Ukraine             None         NaN\n",
       "3         Poland             None         NaN\n",
       "4  United States             None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Country => non_nodal_regional_disease (coverage: 20.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>United States</td>\n",
       "      <td>Indeterminate</td>\n",
       "      <td>0.325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nan</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ukraine</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Poland</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Other_specify</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          source         target  similarity\n",
       "0  United States  Indeterminate       0.325\n",
       "1            nan           None         NaN\n",
       "2        Ukraine           None         NaN\n",
       "3         Poland           None         NaN\n",
       "4  Other_specify           None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Country => perineural_invasion_present (coverage: 20.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>United States</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>0.325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nan</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ukraine</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Poland</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Other_specify</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          source        target  similarity\n",
       "0  United States  Not Reported       0.325\n",
       "1            nan          None         NaN\n",
       "2        Ukraine          None         NaN\n",
       "3         Poland          None         NaN\n",
       "4  Other_specify          None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Country => ethnicity (coverage: 0.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>nan</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>United States</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ukraine</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Poland</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Other_specify</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          source target similarity\n",
       "0            nan   None       None\n",
       "1  United States   None       None\n",
       "2        Ukraine   None       None\n",
       "3         Poland   None       None\n",
       "4  Other_specify   None       None"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Country => consent_type (coverage: 0.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>nan</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>United States</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ukraine</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Poland</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Other_specify</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          source target similarity\n",
       "0            nan   None       None\n",
       "1  United States   None       None\n",
       "2        Ukraine   None       None\n",
       "3         Poland   None       None\n",
       "4  Other_specify   None       None"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Country => overrepresented_sequences (coverage: 20.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>United States</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>0.338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nan</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ukraine</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Poland</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Other_specify</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          source        target  similarity\n",
       "0  United States  Not Reported       0.338\n",
       "1            nan          None         NaN\n",
       "2        Ukraine          None         NaN\n",
       "3         Poland          None         NaN\n",
       "4  Other_specify          None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Country => status (coverage: 0.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>nan</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>United States</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ukraine</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Poland</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Other_specify</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          source target similarity\n",
       "0            nan   None       None\n",
       "1  United States   None       None\n",
       "2        Ukraine   None       None\n",
       "3         Poland   None       None\n",
       "4  Other_specify   None       None"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Country => ulceration_indicator (coverage: 20.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>United States</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>0.315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nan</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ukraine</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Poland</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Other_specify</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          source        target  similarity\n",
       "0  United States  Not Reported       0.315\n",
       "1            nan          None         NaN\n",
       "2        Ukraine          None         NaN\n",
       "3         Poland          None         NaN\n",
       "4  Other_specify          None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Country => vascular_invasion_present (coverage: 20.00%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>United States</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>0.325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nan</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ukraine</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Poland</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Other_specify</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          source        target  similarity\n",
       "0  United States  Not Reported       0.325\n",
       "1            nan          None         NaN\n",
       "2        Ukraine          None         NaN\n",
       "3         Poland          None         NaN\n",
       "4  Other_specify          None         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mappings = bdi.match_values(\n",
    "    dataset,\n",
    "    column_mapping=country_matches,\n",
    "    target=\"gdc\",\n",
    "    method=\"tfidf\",\n",
    "    # default_missing=None\n",
    ")\n",
    "\n",
    "for mapping in mappings:\n",
    "    print(f\"{mapping.attrs['source']} => {mapping.attrs['target']} (coverage: {mapping.attrs['coverage']:.2%})\")\n",
    "    display(mapping)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fixing remaining value mappings\n",
    "\n",
    "We need fix a few value mappings:\n",
    "- Race\n",
    "- Ethnicity\n",
    "- Tumor_Site\n",
    "\n",
    "For race, we need to fix: `nan` -> `american indian or alaska native`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>White</td>\n",
       "      <td>white</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Asian</td>\n",
       "      <td>asian</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Not Reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Black or African American</td>\n",
       "      <td>black or african american</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>nan</td>\n",
       "      <td>american indian or alaska native</td>\n",
       "      <td>0.359</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      source                            target  similarity\n",
       "0                      White                             white       1.000\n",
       "1                      Asian                             asian       1.000\n",
       "2               Not Reported                      not reported       1.000\n",
       "3  Black or African American         black or african american       1.000\n",
       "4                        nan  american indian or alaska native       0.359"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "race_vmap = bdi.match_values(\n",
    "    dataset,\n",
    "    column_mapping=(\"Race\", \"race\"),\n",
    "    target=\"gdc\",\n",
    "    method=\"tfidf\",\n",
    ")\n",
    "race_vmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>White</td>\n",
       "      <td>white</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Asian</td>\n",
       "      <td>asian</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Not Reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Black or African American</td>\n",
       "      <td>black or african american</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      source                     target  similarity\n",
       "0                      White                      white         1.0\n",
       "1                      Asian                      asian         1.0\n",
       "2               Not Reported               not reported         1.0\n",
       "3  Black or African American  black or african american         1.0"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "race_vmap = race_vmap[race_vmap[\"similarity\"] >= 1.0]\n",
    "race_vmap"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For `Ethnicity`, we need to fix: `Not reported` -> `not hispanic or latino`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Hispanic or Latino</td>\n",
       "      <td>hispanic or latino</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>0.935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Not reported</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>0.268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>nan</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   source                  target  similarity\n",
       "0      Hispanic or Latino      hispanic or latino       1.000\n",
       "1  Not-Hispanic or Latino  not hispanic or latino       0.935\n",
       "2            Not reported  not hispanic or latino       0.268\n",
       "3                     nan                    None         NaN"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ethinicity_vmap = bdi.match_values(\n",
    "    dataset,\n",
    "    column_mapping=(\"Ethnicity\", \"ethnicity\"),\n",
    "    target=\"gdc\",\n",
    "    method=\"tfidf\",\n",
    ")\n",
    "ethinicity_vmap\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Hispanic or Latino</td>\n",
       "      <td>hispanic or latino</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>0.935</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   source                  target  similarity\n",
       "0      Hispanic or Latino      hispanic or latino       1.000\n",
       "1  Not-Hispanic or Latino  not hispanic or latino       0.935"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ethinicity_vmap = ethinicity_vmap[ethinicity_vmap[\"similarity\"] > 0.9]\n",
    "ethinicity_vmap"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For `Tumor_Site`, given that this dataset is about endometrial cancer, all values must be mapped to \"Endometrium\". So instead of fixing each mapping individually, we will write a custom function that returns \"Endometrium\" regardless of the input value. Later, we will show how to use this function to transform the dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Anterior endometrium</td>\n",
       "      <td>Endometrium</td>\n",
       "      <td>0.852</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Posterior endometrium</td>\n",
       "      <td>Endometrium</td>\n",
       "      <td>0.823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Other, specify</td>\n",
       "      <td>Other specified parts of pancreas</td>\n",
       "      <td>0.543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>nan</td>\n",
       "      <td>Anal canal</td>\n",
       "      <td>0.301</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  source                             target  similarity\n",
       "0   Anterior endometrium                        Endometrium       0.852\n",
       "1  Posterior endometrium                        Endometrium       0.823\n",
       "2         Other, specify  Other specified parts of pancreas       0.543\n",
       "3                    nan                         Anal canal       0.301"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bdi.match_values(\n",
    "    dataset, column_mapping=(\"Tumor_Site\", \"tissue_or_organ_of_origin\"), target=\"gdc\", method=\"tfidf\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Custom mapping function that will be used to map the values of the 'Tumor_Site' column\n",
    "def map_tumor_site(source_value):\n",
    "    return \"Endometrium\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Combining custom user mappings with suggested mappings\n",
    "\n",
    "Before generating a final harmonized dataset, we can combine the automatically generated value mappings with the fixed mappings provided by the user. To do so, we use `bdi.merge_mappings()` functions, which take a list of mappings (e.g., generated automatically) and a list of \"user-defined mapping overrides\" that will be combined with the first list of mappings and will take precedence whenever they conflict.\n",
    "\n",
    "In our example below, all mappings specified in the variable `user_mappings` will override the mappings in `value_mappings` generated by the `bdi.match_values()` function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from math import ceil\n",
    "\n",
    "user_mappings = [\n",
    "    {\n",
    "        # When no mapping is need, specifying the source and target is enough\n",
    "        \"source\": \"BMI\",\n",
    "        \"target\": \"bmi\",\n",
    "    },\n",
    "    {\n",
    "        \"source\": \"Tumor_Size_cm\",\n",
    "        \"target\": \"tumor_largest_dimension_diameter\",\n",
    "    },\n",
    "    {\n",
    "        # mapper can be a custom Python function\n",
    "        \"source\": \"Tumor_Site\",\n",
    "        \"target\": \"tissue_or_organ_of_origin\",\n",
    "        \"mapper\": map_tumor_site,\n",
    "    },\n",
    "    {\n",
    "        # Lambda functions can also be used as mappers\n",
    "        \"source\": \"Age\",\n",
    "        \"target\": \"days_to_birth\",\n",
    "        \"mapper\": lambda age: -age * 365.25,\n",
    "    },\n",
    "    {\n",
    "        \"source\": \"Age\",\n",
    "        \"target\": \"age_at_diagnosis\",\n",
    "        \"mapper\": lambda age: float(\"nan\") if pd.isnull(age) else ceil(age*365.25),\n",
    "    },\n",
    "    {\n",
    "        # We can also use a data frame to specify value mappings using the `matches` attribute\n",
    "        \"source\": \"Histologic_type\",\n",
    "        \"target\": \"primary_diagnosis\",\n",
    "        \"matches\": hist_type_vmap\n",
    "    },\n",
    "    # For dataframes that contain the 'source' and 'target' columns as attributes,\n",
    "    # such as the ones returned by the match_values() function, we can directly\n",
    "    # use them as mappings\n",
    "    ethinicity_vmap,\n",
    "    race_vmap,\n",
    "]\n",
    "\n",
    "\n",
    "harmonization_spec = bdi.merge_mappings(value_mappings, user_mappings)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, we generate the harmonized dataset, with the user-defined value mappings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bmi</th>\n",
       "      <th>tumor_largest_dimension_diameter</th>\n",
       "      <th>tissue_or_organ_of_origin</th>\n",
       "      <th>days_to_birth</th>\n",
       "      <th>age_at_diagnosis</th>\n",
       "      <th>primary_diagnosis</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>race</th>\n",
       "      <th>country_of_birth</th>\n",
       "      <th>history_of_tumor_type</th>\n",
       "      <th>irs_stage</th>\n",
       "      <th>gender</th>\n",
       "      <th>tumor_focality</th>\n",
       "      <th>tumor_shape</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>38.88</td>\n",
       "      <td>2.9</td>\n",
       "      <td>Endometrium</td>\n",
       "      <td>-23376.00</td>\n",
       "      <td>23376.0</td>\n",
       "      <td>Endometrioid carcinoma</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>white</td>\n",
       "      <td>United States</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>female</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>Dome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>39.76</td>\n",
       "      <td>3.5</td>\n",
       "      <td>Endometrium</td>\n",
       "      <td>-21184.50</td>\n",
       "      <td>21185.0</td>\n",
       "      <td>Endometrioid carcinoma</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>white</td>\n",
       "      <td>United States</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>female</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>Dome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>51.19</td>\n",
       "      <td>4.5</td>\n",
       "      <td>Endometrium</td>\n",
       "      <td>-18262.50</td>\n",
       "      <td>18263.0</td>\n",
       "      <td>Endometrioid carcinoma</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>white</td>\n",
       "      <td>United States</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>female</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Endometrium</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Carcinosarcoma, NOS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Phenochromocytoma or Paraganglioma</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>32.69</td>\n",
       "      <td>3.5</td>\n",
       "      <td>Endometrium</td>\n",
       "      <td>-27393.75</td>\n",
       "      <td>27394.0</td>\n",
       "      <td>Endometrioid carcinoma</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>white</td>\n",
       "      <td>United States</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>female</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>29.40</td>\n",
       "      <td>4.2</td>\n",
       "      <td>Endometrium</td>\n",
       "      <td>-27393.75</td>\n",
       "      <td>27394.0</td>\n",
       "      <td>Endometrioid carcinoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ukraine</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>female</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>35.42</td>\n",
       "      <td>1.5</td>\n",
       "      <td>Endometrium</td>\n",
       "      <td>-27028.50</td>\n",
       "      <td>27029.0</td>\n",
       "      <td>Endometrioid carcinoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ukraine</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>female</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>24.32</td>\n",
       "      <td>3.8</td>\n",
       "      <td>Endometrium</td>\n",
       "      <td>-31046.25</td>\n",
       "      <td>31047.0</td>\n",
       "      <td>Serous cystadenocarcinoma</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>black or african american</td>\n",
       "      <td>United States</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>female</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>34.06</td>\n",
       "      <td>5.0</td>\n",
       "      <td>Endometrium</td>\n",
       "      <td>-25567.50</td>\n",
       "      <td>25568.0</td>\n",
       "      <td>Serous cystadenocarcinoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ukraine</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>female</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Endometrium</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Serous cystadenocarcinoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ukraine</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>104 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       bmi  tumor_largest_dimension_diameter tissue_or_organ_of_origin  \\\n",
       "0    38.88                               2.9               Endometrium   \n",
       "1    39.76                               3.5               Endometrium   \n",
       "2    51.19                               4.5               Endometrium   \n",
       "3      NaN                               NaN               Endometrium   \n",
       "4    32.69                               3.5               Endometrium   \n",
       "..     ...                               ...                       ...   \n",
       "99   29.40                               4.2               Endometrium   \n",
       "100  35.42                               1.5               Endometrium   \n",
       "101  24.32                               3.8               Endometrium   \n",
       "102  34.06                               5.0               Endometrium   \n",
       "103    NaN                               NaN               Endometrium   \n",
       "\n",
       "     days_to_birth  age_at_diagnosis          primary_diagnosis  \\\n",
       "0        -23376.00           23376.0     Endometrioid carcinoma   \n",
       "1        -21184.50           21185.0     Endometrioid carcinoma   \n",
       "2        -18262.50           18263.0     Endometrioid carcinoma   \n",
       "3              NaN               NaN        Carcinosarcoma, NOS   \n",
       "4        -27393.75           27394.0     Endometrioid carcinoma   \n",
       "..             ...               ...                        ...   \n",
       "99       -27393.75           27394.0     Endometrioid carcinoma   \n",
       "100      -27028.50           27029.0     Endometrioid carcinoma   \n",
       "101      -31046.25           31047.0  Serous cystadenocarcinoma   \n",
       "102      -25567.50           25568.0  Serous cystadenocarcinoma   \n",
       "103            NaN               NaN  Serous cystadenocarcinoma   \n",
       "\n",
       "                  ethnicity                       race country_of_birth  \\\n",
       "0    not hispanic or latino                      white    United States   \n",
       "1    not hispanic or latino                      white    United States   \n",
       "2    not hispanic or latino                      white    United States   \n",
       "3                       NaN                        NaN              NaN   \n",
       "4    not hispanic or latino                      white    United States   \n",
       "..                      ...                        ...              ...   \n",
       "99                      NaN                        NaN          Ukraine   \n",
       "100                     NaN                        NaN          Ukraine   \n",
       "101  not hispanic or latino  black or african american    United States   \n",
       "102                     NaN                        NaN          Ukraine   \n",
       "103                     NaN                        NaN          Ukraine   \n",
       "\n",
       "                  history_of_tumor_type irs_stage  gender tumor_focality  \\\n",
       "0                                  None      None  female       Unifocal   \n",
       "1                                  None      None  female       Unifocal   \n",
       "2                                  None      None  female       Unifocal   \n",
       "3    Phenochromocytoma or Paraganglioma      None     NaN            NaN   \n",
       "4                                  None      None  female       Unifocal   \n",
       "..                                  ...       ...     ...            ...   \n",
       "99                                 None      None  female       Unifocal   \n",
       "100                                None      None  female       Unifocal   \n",
       "101                                None      None  female       Unifocal   \n",
       "102                                None      None  female       Unifocal   \n",
       "103                                None      None     NaN            NaN   \n",
       "\n",
       "    tumor_shape  \n",
       "0          Dome  \n",
       "1          Dome  \n",
       "2          None  \n",
       "3           NaN  \n",
       "4          None  \n",
       "..          ...  \n",
       "99         None  \n",
       "100        None  \n",
       "101        None  \n",
       "102        None  \n",
       "103        None  \n",
       "\n",
       "[104 rows x 14 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "harmonized_dataset = bdi.materialize_mapping(dataset, harmonization_spec)\n",
    "harmonized_dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For comparison, here is how our original data looked like:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BMI</th>\n",
       "      <th>Tumor_Size_cm</th>\n",
       "      <th>Tumor_Site</th>\n",
       "      <th>Age</th>\n",
       "      <th>Age</th>\n",
       "      <th>Histologic_type</th>\n",
       "      <th>Ethnicity</th>\n",
       "      <th>Race</th>\n",
       "      <th>Country</th>\n",
       "      <th>Histologic_type</th>\n",
       "      <th>FIGO_stage</th>\n",
       "      <th>Gender</th>\n",
       "      <th>Tumor_Focality</th>\n",
       "      <th>Tumor_Site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>38.88</td>\n",
       "      <td>2.9</td>\n",
       "      <td>Anterior endometrium</td>\n",
       "      <td>64.0</td>\n",
       "      <td>64.0</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>White</td>\n",
       "      <td>United States</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>IA</td>\n",
       "      <td>Female</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>Anterior endometrium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>39.76</td>\n",
       "      <td>3.5</td>\n",
       "      <td>Posterior endometrium</td>\n",
       "      <td>58.0</td>\n",
       "      <td>58.0</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>White</td>\n",
       "      <td>United States</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>IA</td>\n",
       "      <td>Female</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>Posterior endometrium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>51.19</td>\n",
       "      <td>4.5</td>\n",
       "      <td>Other, specify</td>\n",
       "      <td>50.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>White</td>\n",
       "      <td>United States</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>IA</td>\n",
       "      <td>Female</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>Other, specify</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>32.69</td>\n",
       "      <td>3.5</td>\n",
       "      <td>Other, specify</td>\n",
       "      <td>75.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>White</td>\n",
       "      <td>United States</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>IA</td>\n",
       "      <td>Female</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>Other, specify</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>29.40</td>\n",
       "      <td>4.2</td>\n",
       "      <td>Other, specify</td>\n",
       "      <td>75.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ukraine</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>IA</td>\n",
       "      <td>Female</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>Other, specify</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>35.42</td>\n",
       "      <td>1.5</td>\n",
       "      <td>Other, specify</td>\n",
       "      <td>74.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ukraine</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>II</td>\n",
       "      <td>Female</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>Other, specify</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>24.32</td>\n",
       "      <td>3.8</td>\n",
       "      <td>Other, specify</td>\n",
       "      <td>85.0</td>\n",
       "      <td>85.0</td>\n",
       "      <td>Serous</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>Black or African American</td>\n",
       "      <td>United States</td>\n",
       "      <td>Serous</td>\n",
       "      <td>II</td>\n",
       "      <td>Female</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>Other, specify</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>34.06</td>\n",
       "      <td>5.0</td>\n",
       "      <td>Other, specify</td>\n",
       "      <td>70.0</td>\n",
       "      <td>70.0</td>\n",
       "      <td>Serous</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ukraine</td>\n",
       "      <td>Serous</td>\n",
       "      <td>IA</td>\n",
       "      <td>Female</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>Other, specify</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Serous</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ukraine</td>\n",
       "      <td>Serous</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>104 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       BMI  Tumor_Size_cm             Tumor_Site   Age   Age Histologic_type  \\\n",
       "0    38.88            2.9   Anterior endometrium  64.0  64.0    Endometrioid   \n",
       "1    39.76            3.5  Posterior endometrium  58.0  58.0    Endometrioid   \n",
       "2    51.19            4.5         Other, specify  50.0  50.0    Endometrioid   \n",
       "3      NaN            NaN                    NaN   NaN   NaN  Carcinosarcoma   \n",
       "4    32.69            3.5         Other, specify  75.0  75.0    Endometrioid   \n",
       "..     ...            ...                    ...   ...   ...             ...   \n",
       "99   29.40            4.2         Other, specify  75.0  75.0    Endometrioid   \n",
       "100  35.42            1.5         Other, specify  74.0  74.0    Endometrioid   \n",
       "101  24.32            3.8         Other, specify  85.0  85.0          Serous   \n",
       "102  34.06            5.0         Other, specify  70.0  70.0          Serous   \n",
       "103    NaN            NaN                    NaN   NaN   NaN          Serous   \n",
       "\n",
       "                  Ethnicity                       Race        Country  \\\n",
       "0    Not-Hispanic or Latino                      White  United States   \n",
       "1    Not-Hispanic or Latino                      White  United States   \n",
       "2    Not-Hispanic or Latino                      White  United States   \n",
       "3                       NaN                        NaN            NaN   \n",
       "4    Not-Hispanic or Latino                      White  United States   \n",
       "..                      ...                        ...            ...   \n",
       "99                      NaN                        NaN        Ukraine   \n",
       "100                     NaN                        NaN        Ukraine   \n",
       "101  Not-Hispanic or Latino  Black or African American  United States   \n",
       "102                     NaN                        NaN        Ukraine   \n",
       "103                     NaN                        NaN        Ukraine   \n",
       "\n",
       "    Histologic_type FIGO_stage  Gender Tumor_Focality             Tumor_Site  \n",
       "0      Endometrioid         IA  Female       Unifocal   Anterior endometrium  \n",
       "1      Endometrioid         IA  Female       Unifocal  Posterior endometrium  \n",
       "2      Endometrioid         IA  Female       Unifocal         Other, specify  \n",
       "3    Carcinosarcoma        NaN     NaN            NaN                    NaN  \n",
       "4      Endometrioid         IA  Female       Unifocal         Other, specify  \n",
       "..              ...        ...     ...            ...                    ...  \n",
       "99     Endometrioid         IA  Female       Unifocal         Other, specify  \n",
       "100    Endometrioid         II  Female       Unifocal         Other, specify  \n",
       "101          Serous         II  Female       Unifocal         Other, specify  \n",
       "102          Serous         IA  Female       Unifocal         Other, specify  \n",
       "103          Serous        NaN     NaN            NaN                    NaN  \n",
       "\n",
       "[104 rows x 14 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "original_columns = map(lambda m: m[\"source\"], harmonization_spec)\n",
    "dataset[original_columns]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
